// @backstage/cli-module-auth/dist/commands/logout.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
require('../cli-internal/src/InternalCliModule.cjs.js');
require('../cli-internal/src/InternalCommandNode.cjs.js');
var authIdentifiers = require('../cli-internal/src/authIdentifiers.cjs.js');
var secretStore = require('../cli-internal/src/secretStore.cjs.js');
require('../cli-internal/src/knownPluginPackages.cjs.js');
var storage = require('../lib/storage.cjs.js');
var http = require('../lib/http.cjs.js');
var prompt = require('../lib/prompt.cjs.js');

var logout = async ({ args, info }) => {
  const {
    flags: { instance: instanceFlag }
  } = cleye.cli(
    {
      name: info.usage,
      flags: {
        instance: {
          type: String,
          description: "Name of the instance to log out"
        }
      }
    },
    void 0,
    args
  );
  const { name: instanceName } = await prompt.pickInstance(instanceFlag);
  await storage.withMetadataLock(async () => {
    const instance = await storage.getInstanceByName(instanceName);
    const secretStore$1 = await secretStore.getSecretStore();
    const service = authIdentifiers.getAuthInstanceService(instanceName);
    const refreshToken = await secretStore$1.get(service, "refreshToken") ?? "";
    if (refreshToken) {
      try {
        const authBaseUrl = new URL("/api/auth", instance.baseUrl).toString().replace(/\/$/, "");
        await http.httpJson(`${authBaseUrl}/v1/revoke`, {
          method: "POST",
          body: {
            token: refreshToken,
            token_type_hint: "refresh_token"
          },
          signal: AbortSignal.timeout(3e4)
        });
      } catch {
      }
    }
    await secretStore$1.delete(service, "accessToken");
    await secretStore$1.delete(service, "refreshToken");
    await storage.removeInstance(instance.name);
  });
  process.stdout.write("Logged out\n");
};

exports.default = logout;
