// @backstage/cli-module-auth/dist/commands/show.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var cliNode = require('@backstage/cli-node');
var http = require('../lib/http.cjs.js');

var show = async ({ args, info }) => {
  const {
    flags: { instance: instanceFlag }
  } = cleye.cli(
    {
      name: info.usage,
      flags: {
        instance: {
          type: String,
          description: "Name of the instance to show"
        }
      }
    },
    void 0,
    args
  );
  const auth = await cliNode.CliAuth.create({ instanceName: instanceFlag });
  const accessToken = await auth.getAccessToken();
  const authBase = new URL("/api/auth", auth.getBaseUrl()).toString().replace(/\/$/, "");
  const userinfo = await http.httpJson(
    `${authBase}/v1/userinfo`,
    {
      headers: { Authorization: `Bearer ${accessToken}` },
      signal: AbortSignal.timeout(3e4)
    }
  );
  process.stdout.write(`User: ${userinfo.claims.sub}
`);
  process.stdout.write(`
`);
  process.stdout.write(`Ownership:
`);
  for (const ent of userinfo.claims.ent ?? []) {
    process.stdout.write(`  - ${ent}
`);
  }
};

exports.default = show;
