// @backstage/cli-module-build/dist/commands/package/start/startPackage.cjs.js
'use strict';

var startBackend = require('./startBackend.cjs.js');
var startFrontend = require('./startFrontend.cjs.js');
var node_path = require('node:path');
var glob = require('glob');

function resolveEntryPath(entrypoint = "dev", targetDir) {
  const { dir: entryDir, name: entryName } = node_path.parse(entrypoint);
  const [entryFile] = glob.glob.sync(
    `${node_path.resolve(targetDir, entryDir, entryName)}.*`,
    { windowsPathsNoEscape: true }
  );
  if (entryFile) {
    return node_path.join(entryDir, entryName);
  }
  return node_path.join(entryDir, entryName, "index");
}
async function startPackage(options) {
  switch (options.role) {
    case "backend":
      return startBackend.startBackend(options);
    case "backend-plugin":
    case "backend-plugin-module":
    case "node-library":
      return startBackend.startBackendPlugin(options);
    case "frontend":
      return startFrontend.startFrontend({
        ...options,
        entry: "src/index",
        verifyVersions: true
      });
    case "web-library":
    case "frontend-plugin":
    case "frontend-plugin-module":
      return startFrontend.startFrontend({
        ...options,
        entry: resolveEntryPath(options.entrypoint, options.targetDir)
      });
    case "frontend-dynamic-container":
      return startFrontend.startFrontend({
        entry: "src/index",
        ...options,
        skipOpenBrowser: true,
        isModuleFederationRemote: true
      });
    default:
      throw new Error(
        `Start command is not supported for package role '${options.role}'`
      );
  }
}

exports.resolveEntryPath = resolveEntryPath;
exports.startPackage = startPackage;
