// @backstage/ui/dist/hooks/useCollectionAdapter.esm.js
import { useRef } from 'react';
import { isAsyncListSource, getItemKeys } from '../utils/selectableCollection.esm.js';

function useCollectionAdapter({
  items,
  selectedKeys = [],
  search,
  loading,
  retainSelectedItems = true
}) {
  const retainedItems = useRef(/* @__PURE__ */ new Map());
  let asyncSource;
  let sourceItems;
  if (isAsyncListSource(items)) {
    asyncSource = items;
    sourceItems = [...items.items];
  } else {
    sourceItems = Array.from(items);
  }
  const selectedIdSet = new Set(selectedKeys);
  const searchProps = typeof search === "object" ? search : void 0;
  const isServerSearch = searchProps?.mode === "server";
  for (const key of retainedItems.current.keys()) {
    if (!selectedIdSet.has(key)) {
      retainedItems.current.delete(key);
    }
  }
  for (const item of sourceItems) {
    if (selectedIdSet.has(item.id)) {
      retainedItems.current.set(item.id, item);
    }
  }
  const sourceIds = getItemKeys(sourceItems);
  const shouldRetainSelectedItems = isServerSearch && retainSelectedItems;
  let canonicalItems = sourceItems;
  let visibleIds;
  if (shouldRetainSelectedItems) {
    const missingSelectedItems = Array.from(retainedItems.current.entries()).filter(([id]) => !sourceIds.has(id)).map(([, item]) => item);
    canonicalItems = [...sourceItems, ...missingSelectedItems];
    visibleIds = sourceIds;
  }
  let resolvedLoading = loading;
  if (asyncSource) {
    resolvedLoading = {
      state: asyncSource.loadingState,
      onLoadMore: asyncSource.loadMore
    };
  }
  let inputValue = searchProps?.inputValue;
  let onInputChange = searchProps?.onInputChange;
  if (asyncSource && isServerSearch) {
    inputValue = asyncSource.filterText;
    onInputChange = asyncSource.setFilterText;
  }
  return {
    canonicalItems,
    visibleIds,
    loading: resolvedLoading,
    isStale: sourceItems.length > 0 && (resolvedLoading?.state === "filtering" || resolvedLoading?.state === "sorting"),
    inputValue,
    defaultInputValue: searchProps?.defaultInputValue,
    onInputChange
  };
}

export { useCollectionAdapter };
