// @backstage/ui/dist/components/FieldLabel/FieldLabel.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { Label, Text } from 'react-aria-components';
import { forwardRef } from 'react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { FieldLabelDefinition } from './definition.esm.js';

const FieldLabel = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(FieldLabelDefinition, props);
    const {
      classes,
      label,
      secondaryLabel,
      description,
      htmlFor,
      id,
      descriptionId,
      descriptionSlot
    } = ownProps;
    if (!label) return null;
    return /* @__PURE__ */ jsxs("div", { className: classes.root, ...restProps, ref, children: [
      label && /* @__PURE__ */ jsxs(Label, { className: classes.label, htmlFor, id, children: [
        label,
        secondaryLabel && /* @__PURE__ */ jsxs("span", { "aria-hidden": "true", className: classes.secondaryLabel, children: [
          "(",
          secondaryLabel,
          ")"
        ] })
      ] }),
      description && /* @__PURE__ */ jsx(
        Text,
        {
          slot: descriptionSlot,
          className: classes.description,
          elementType: "div",
          id: descriptionId,
          children: description
        }
      )
    ] });
  }
);
FieldLabel.displayName = "FieldLabel";

export { FieldLabel };
