// @backstage/catalog-model/dist/model/modelActions/addRelationPair.cjs.js
'use strict';

var declareRelation = require('../operations/declareRelation.cjs.js');

function opsFromCatalogModelRelationPair(relationPair) {
  const ops = [];
  for (const firstKind of [relationPair.fromKind].flat()) {
    for (const secondKind of [relationPair.toKind].flat()) {
      ops.push(
        declareRelation.createDeclareRelationOp({
          fromKind: firstKind,
          type: relationPair.forward.type,
          toKind: secondKind,
          properties: {
            reverseType: relationPair.reverse.type,
            title: relationPair.forward.title,
            description: relationPair.description
          }
        })
      );
      ops.push(
        declareRelation.createDeclareRelationOp({
          fromKind: secondKind,
          type: relationPair.reverse.type,
          toKind: firstKind,
          properties: {
            reverseType: relationPair.forward.type,
            title: relationPair.reverse.title,
            description: relationPair.description
          }
        })
      );
    }
  }
  return ops;
}

exports.opsFromCatalogModelRelationPair = opsFromCatalogModelRelationPair;
