// @backstage/catalog-model/dist/model/modelActions/updateKind.cjs.js
'use strict';

var reduceKindSchema = require('../jsonSchema/reduceKindSchema.cjs.js');
var validateMetaSchema = require('../jsonSchema/validateMetaSchema.cjs.js');
var updateKind = require('../operations/updateKind.cjs.js');
var updateKindVersion = require('../operations/updateKindVersion.cjs.js');

function opsFromCatalogModelUpdateKind(kind) {
  const ops = [];
  if (kind.names.singular || kind.names.plural || kind.description) {
    ops.push(
      updateKind.createUpdateKindOp({
        kind: kind.names.kind,
        properties: {
          singular: kind.names.singular,
          plural: kind.names.plural,
          description: kind.description
        }
      })
    );
  }
  for (const version of kind.versions ?? []) {
    const jsonSchema = version.schema ? reduceKindSchema.reduceKindSchema(version.schema.jsonSchema) : void 0;
    if (jsonSchema) {
      validateMetaSchema.validateMetaSchema(jsonSchema);
    }
    const names = Array.isArray(version.name) ? version.name : [version.name];
    for (const name of names) {
      const specTypes = version.specType?.length ? [version.specType].flat() : [void 0];
      for (const specType of specTypes) {
        ops.push(
          updateKindVersion.createUpdateKindVersionOp({
            kind: kind.names.kind,
            name,
            specType,
            properties: {
              description: version.description,
              relationFields: version.relationFields,
              schema: jsonSchema ? { jsonSchema } : void 0
            }
          })
        );
      }
    }
  }
  return ops;
}

exports.opsFromCatalogModelUpdateKind = opsFromCatalogModelUpdateKind;
