// @backstage/catalog-model/dist/model/modelActions/updateRelationPair.esm.js
import { createUpdateRelationOp } from '../operations/updateRelation.esm.js';

function opsFromCatalogModelUpdateRelationPair(relationPair) {
  const ops = [];
  for (const firstKind of [relationPair.fromKind].flat()) {
    for (const secondKind of [relationPair.toKind].flat()) {
      ops.push(
        createUpdateRelationOp({
          fromKind: firstKind,
          type: relationPair.forward.type,
          toKind: secondKind,
          properties: {
            reverseType: relationPair.reverse.type,
            title: relationPair.forward.title,
            description: relationPair.description
          }
        })
      );
      if (relationPair.reverse.type) {
        ops.push(
          createUpdateRelationOp({
            fromKind: secondKind,
            type: relationPair.reverse.type,
            toKind: firstKind,
            properties: {
              reverseType: relationPair.forward.type,
              title: relationPair.reverse.title,
              description: relationPair.description
            }
          })
        );
      }
    }
  }
  return ops;
}

export { opsFromCatalogModelUpdateRelationPair };
