// @backstage/frontend-plugin-api/dist/schema/createPortableSchema.esm.js
import { z } from 'zod/v3';
import zodToJsonSchema from 'zod-to-json-schema';

function createDeprecatedConfigSchema(fields) {
  const resolved = {};
  for (const [key, field] of Object.entries(fields)) {
    resolved[key] = resolveZodField(key, field(z));
  }
  return buildPortableSchema(resolved);
}
function createConfigSchema(fields) {
  const resolved = {};
  for (const [key, field] of Object.entries(fields)) {
    resolved[key] = resolveField(key, field);
  }
  return buildPortableSchema(resolved);
}
function mergePortableSchemas(a, b) {
  if (!a && !b) {
    return void 0;
  }
  if (!a) {
    return b;
  }
  if (!b) {
    return a;
  }
  return buildPortableSchema({
    ...a._fields,
    ...b._fields
  });
}
function buildPortableSchema(fields) {
  function parse(input) {
    if (input !== void 0 && input !== null && (typeof input !== "object" || Array.isArray(input))) {
      throw new Error(
        `Invalid config input, expected object but got ${Array.isArray(input) ? "array" : typeof input}`
      );
    }
    const inputObj = input ?? {};
    const result2 = {};
    const errors = [];
    for (const [key, field] of Object.entries(fields)) {
      const validated = field.validate(inputObj[key]);
      if ("errors" in validated) {
        errors.push(...validated.errors);
      } else if (validated.value !== void 0 || key in inputObj) {
        result2[key] = validated.value;
      }
    }
    if (errors.length > 0) {
      throw new Error(errors.join("; "));
    }
    return result2;
  }
  let cached;
  const result = {
    parse,
    schema() {
      if (!cached) {
        cached = { schema: buildObjectJsonSchema(fields) };
      }
      return cached;
    }
  };
  Object.defineProperty(result, "_fields", {
    value: fields,
    enumerable: false
  });
  return result;
}
function resolveField(key, schema) {
  if (isZodV3Type(schema)) {
    throw new Error(
      `Config schema for field '${key}' uses a Zod v3 schema, which is not supported by the \`configSchema\` option. Upgrade to the \`zod\` v4 package (\`zod@^4.0.0\`). Note that the \`zod/v4\` subpath export from the zod v3 package is also not supported, as it does not include JSON Schema conversion.`
    );
  }
  if (isStandardSchema(schema)) {
    if (!hasJsonSchemaConverter(schema)) {
      throw new Error(
        `Config schema for field '${key}' does not support JSON Schema conversion. Use a schema library that implements the Standard JSON Schema interface (like zod v4+).`
      );
    }
    return resolveStandardField(key, schema);
  }
  throw new Error(
    `Config schema for field '${key}' is not a valid Standard Schema`
  );
}
function resolveZodField(key, schema) {
  const wrapper = z.object({ [key]: schema });
  return {
    validate(value) {
      const result = wrapper.safeParse({ [key]: value });
      if (result.success) {
        return { value: result.data[key] };
      }
      return { errors: result.error.issues.map(formatZodIssue) };
    },
    toJsonSchema() {
      const wholeJsonSchema = zodToJsonSchema(wrapper);
      return wholeJsonSchema.properties?.[key] ?? {};
    },
    required: !schema.isOptional()
  };
}
function resolveStandardField(key, schema) {
  const required = isFieldRequired(schema);
  return {
    validate(value) {
      const result = schema["~standard"].validate(value);
      if (result instanceof Promise) {
        throw new Error(
          `Config schema for '${key}' returned a Promise \u2014 async schemas are not supported`
        );
      }
      if (result.issues) {
        return {
          errors: Array.from(result.issues).map(
            (issue) => formatStandardIssue(key, issue)
          )
        };
      }
      return { value: result.value };
    },
    toJsonSchema() {
      const raw = schema["~standard"].jsonSchema.input({ target: "draft-07" });
      const { $schema: _, ...rest } = raw;
      return rest;
    },
    required
  };
}
function buildObjectJsonSchema(fields) {
  const properties = {};
  const required = [];
  for (const [key, field] of Object.entries(fields)) {
    properties[key] = field.toJsonSchema();
    if (field.required) {
      required.push(key);
    }
  }
  const schema = {
    type: "object",
    properties,
    additionalProperties: false
  };
  if (required.length > 0) {
    schema.required = required;
  }
  return schema;
}
function isZodV3Type(value) {
  return typeof value === "object" && value !== null && typeof value._parse === "function" && "_def" in value;
}
function isStandardSchema(value) {
  return typeof value === "object" && value !== null && "~standard" in value && typeof value["~standard"]?.validate === "function";
}
function hasJsonSchemaConverter(schema) {
  const std = schema["~standard"];
  return typeof std?.jsonSchema?.input === "function";
}
function isFieldRequired(schema) {
  const result = schema["~standard"].validate(void 0);
  if (result instanceof Promise) {
    return true;
  }
  return (result.issues?.length ?? 0) > 0;
}
function formatZodIssue(issue) {
  if (issue.code === "invalid_union" && issue.unionErrors?.[0]?.issues?.[0]) {
    return formatZodIssue(issue.unionErrors[0].issues[0]);
  }
  let message = issue.message;
  if (message === "Required") {
    message = "Missing required value";
  }
  if (issue.path.length) {
    message += ` at '${issue.path.join(".")}'`;
  }
  return message;
}
function formatStandardIssue(fieldKey, issue) {
  let message = issue.message;
  if (message === "Required") {
    message = "Missing required value";
  }
  const path = issue.path?.length ? `${fieldKey}.${issue.path.map(
    (p) => typeof p === "object" ? p.key : p
  ).join(".")}` : fieldKey;
  return `${message} at '${path}'`;
}
function warnConfigSchemaPropDeprecation(callSite) {
  console.warn(
    `DEPRECATION WARNING: The \`config.schema\` option for extension config is deprecated. Use the \`configSchema\` option instead with Standard Schema values, for example \`configSchema: { title: z.string() }\` using the \`zod\` v4 package (\`zod@^4.0.0\`). Declared at ${callSite}`
  );
}

export { createConfigSchema, createDeprecatedConfigSchema, mergePortableSchemas, warnConfigSchemaPropDeprecation };
