// @backstage/ui/dist/components/Select/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Select.module.css.esm.js';

const SelectDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Select",
    popover: "bui-SelectPopover"
  },
  propDefs: {
    icon: {},
    size: { dataAttribute: true, default: "small" },
    options: {},
    items: {},
    children: {},
    dependencies: {},
    search: {},
    loading: {},
    searchable: {},
    searchPlaceholder: {},
    label: {},
    secondaryLabel: {},
    description: {},
    isRequired: {},
    className: {}
  }
});
const SelectTriggerDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectTrigger",
      chevron: "bui-SelectTriggerChevron",
      value: "bui-SelectValue"
    },
    bg: "consumer",
    propDefs: {
      icon: {}
    }
  }
);
const SelectContentDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectContent",
      searchWrapper: "bui-SelectSearchWrapper",
      results: "bui-SelectResults",
      search: "bui-SelectSearch",
      searchClear: "bui-SelectSearchClear"
    },
    propDefs: {
      search: {},
      options: {},
      items: {},
      children: {},
      dependencies: {},
      loading: {},
      isStale: {},
      visibleIds: {},
      retainedOptions: {}
    }
  }
);
const SelectListBoxDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectList",
      noResults: "bui-SelectNoResults",
      loading: "bui-SelectLoading",
      loadingRow: "bui-SelectLoadingRow"
    },
    propDefs: {
      options: {},
      items: {},
      children: {},
      dependencies: {},
      loading: {},
      isStale: {},
      retainedOptions: {}
    }
  }
);
const SelectListBoxItemDefinition = defineComponent()({
  styles,
  classNames: {
    label: "bui-SelectItemLabel"
  },
  propDefs: {}
});
const SelectItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-SelectItem",
    indicator: "bui-SelectItemIndicator",
    content: "bui-SelectItemContent"
  },
  propDefs: {
    children: {},
    showSelectionIndicator: {},
    className: {}
  }
});
const SelectItemTextDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-SelectItemText",
    content: "bui-SelectItemContent",
    leadingIcon: "bui-SelectItemLeadingIcon",
    text: "bui-SelectItemTextContent",
    title: "bui-SelectItemTitle",
    description: "bui-SelectItemDescription"
  },
  propDefs: {
    title: {},
    description: {},
    leadingIcon: {},
    className: {}
  }
});
const SelectItemProfileDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-SelectItemProfile",
    content: "bui-SelectItemContent",
    avatar: "bui-SelectItemAvatar",
    name: "bui-SelectItemTitle"
  },
  propDefs: {
    name: {},
    src: {},
    className: {}
  }
});
const SelectSectionDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectSection",
      header: "bui-SelectSectionHeader"
    },
    propDefs: {}
  }
);

export { SelectContentDefinition, SelectDefinition, SelectItemDefinition, SelectItemProfileDefinition, SelectItemTextDefinition, SelectListBoxDefinition, SelectListBoxItemDefinition, SelectSectionDefinition, SelectTriggerDefinition };
