// @backstage/ui/dist/components/Select/SelectContent.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Autocomplete, SearchField, Input, Button } from 'react-aria-components';
import { useFilter } from 'react-aria';
import { RiCloseCircleLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { SelectContentDefinition } from './definition.esm.js';
import { SelectListBox } from './SelectListBox.esm.js';

function getSearchFilter({
  visibleIds,
  filter,
  contains
}) {
  if (visibleIds) {
    return (_textValue, _inputValue, node) => visibleIds.has(node.key);
  }
  if (filter) {
    return (_textValue, inputValue, node) => filter(node.value, inputValue);
  }
  return (textValue, inputValue) => contains(textValue, inputValue);
}
function SelectContent(props) {
  const { contains } = useFilter({ sensitivity: "base" });
  const { ownProps } = useDefinition(SelectContentDefinition, props);
  const {
    classes,
    search,
    options,
    items,
    children,
    dependencies,
    loading,
    isStale,
    visibleIds,
    retainedOptions
  } = ownProps;
  const listBox = /* @__PURE__ */ jsx("div", { className: classes.results, children: /* @__PURE__ */ jsx(
    SelectListBox,
    {
      options,
      items,
      dependencies,
      loading,
      isStale,
      retainedOptions,
      children
    }
  ) });
  if (!search) {
    return /* @__PURE__ */ jsx("div", { className: classes.root, children: listBox });
  }
  const searchProps = typeof search === "object" ? search : void 0;
  const placeholder = searchProps?.placeholder ?? "Search...";
  const filter = getSearchFilter({
    visibleIds,
    filter: searchProps?.filter,
    contains
  });
  return /* @__PURE__ */ jsx(
    Autocomplete,
    {
      inputValue: searchProps?.inputValue,
      defaultInputValue: searchProps?.defaultInputValue,
      onInputChange: searchProps?.onInputChange,
      filter,
      children: /* @__PURE__ */ jsxs("div", { className: classes.root, children: [
        /* @__PURE__ */ jsxs(
          SearchField,
          {
            autoFocus: true,
            className: classes.searchWrapper,
            "aria-label": placeholder,
            children: [
              /* @__PURE__ */ jsx(Input, { placeholder, className: classes.search }),
              /* @__PURE__ */ jsx(Button, { className: classes.searchClear, children: /* @__PURE__ */ jsx(RiCloseCircleLine, {}) })
            ]
          }
        ),
        listBox
      ] })
    }
  );
}

export { SelectContent };
