// @backstage/ui/dist/components/Select/SelectListBox.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { ListBox, Collection, ListBoxLoadMoreItem, Text, ListBoxSection, Header } from 'react-aria-components';
import clsx from 'clsx';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { SelectListBoxDefinition, SelectItemTextDefinition, SelectListBoxItemDefinition, SelectSectionDefinition } from './definition.esm.js';
import { normalizeOptions } from '../../utils/selectableCollection.esm.js';
import { SelectItem } from './SelectItem.esm.js';
import { Skeleton } from '../Skeleton/Skeleton.esm.js';
import '../Skeleton/Skeleton.module.css.esm.js';
import { useDelayedVisibility } from '../../hooks/useDelayedVisibility.esm.js';

const loadingRowWidths = ["70%", "55%", "65%"];
const loadingIndicatorDelayMs = 300;
function LoadingRows({
  count,
  className,
  rowClassName
}) {
  return /* @__PURE__ */ jsx("div", { className, "aria-hidden": "true", children: loadingRowWidths.slice(0, count).map((width) => /* @__PURE__ */ jsx("div", { className: rowClassName, children: /* @__PURE__ */ jsx(Skeleton, { width, height: "0.75rem" }) }, width)) });
}
const NoResults = () => {
  const { ownProps } = useDefinition(SelectListBoxDefinition, {});
  const { classes } = ownProps;
  return /* @__PURE__ */ jsx("div", { className: classes.noResults, children: "No results found." });
};
function PlainOptionItem({ option }) {
  const { ownProps } = useDefinition(SelectItemTextDefinition, {
    title: option.label,
    description: option.description,
    leadingIcon: option.leadingIcon
  });
  const { classes } = ownProps;
  const {
    ownProps: { classes: legacyOptionClasses }
  } = useDefinition(SelectListBoxItemDefinition, {});
  return /* @__PURE__ */ jsx(
    SelectItem,
    {
      id: option.id,
      value: option,
      className: classes.root,
      textValue: option.label,
      isDisabled: option.disabled,
      showSelectionIndicator: true,
      children: /* @__PURE__ */ jsxs("div", { className: classes.content, children: [
        option.leadingIcon && /* @__PURE__ */ jsx("div", { className: classes.leadingIcon, children: option.leadingIcon }),
        /* @__PURE__ */ jsxs("div", { className: classes.text, children: [
          /* @__PURE__ */ jsx(
            Text,
            {
              slot: "label",
              className: clsx(classes.title, legacyOptionClasses.label),
              children: option.label
            }
          ),
          option.description && /* @__PURE__ */ jsx(Text, { slot: "description", className: classes.description, children: option.description })
        ] })
      ] })
    }
  );
}
function SelectSectionItems({ section }) {
  const { ownProps } = useDefinition(SelectSectionDefinition, {});
  const { classes } = ownProps;
  return /* @__PURE__ */ jsxs(ListBoxSection, { className: classes.root, children: [
    /* @__PURE__ */ jsx(Header, { className: classes.header, children: section.title }),
    section.options.map((option) => /* @__PURE__ */ jsx(PlainOptionItem, { option }, option.id))
  ] });
}
function renderSelectOption(item) {
  if ("options" in item) {
    return /* @__PURE__ */ jsx(SelectSectionItems, { section: item }, item.title);
  }
  return /* @__PURE__ */ jsx(PlainOptionItem, { option: item }, item.id);
}
function SelectCollection({
  normalizedOptions,
  retainedOptions,
  items,
  children,
  dependencies
}) {
  if (normalizedOptions) {
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      normalizedOptions.map(renderSelectOption),
      Array.from(retainedOptions ?? [], (option) => /* @__PURE__ */ jsx(PlainOptionItem, { option }, option.id))
    ] });
  }
  if (items) {
    const renderItem = (item) => {
      if (typeof children === "function") {
        return children(item);
      }
      return /* @__PURE__ */ jsx(PlainOptionItem, { option: item });
    };
    return /* @__PURE__ */ jsx(Collection, { items, dependencies, children: renderItem });
  }
  return children;
}
function SelectEmptyState({
  isLoading,
  className,
  rowClassName
}) {
  if (!isLoading) {
    return /* @__PURE__ */ jsx(NoResults, {});
  }
  return /* @__PURE__ */ jsx(LoadingRows, { count: 3, className, rowClassName });
}
function SelectLoadMoreIndicator({
  loading,
  isCollectionLoading,
  isIndicatorVisible,
  className,
  rowClassName
}) {
  if (!loading?.onLoadMore || isCollectionLoading) {
    return null;
  }
  const onLoadMore = loading.state === "loadingMore" ? void 0 : loading.onLoadMore;
  return /* @__PURE__ */ jsx(
    ListBoxLoadMoreItem,
    {
      isLoading: isIndicatorVisible,
      onLoadMore,
      scrollOffset: 0,
      children: /* @__PURE__ */ jsx(
        LoadingRows,
        {
          count: 1,
          className,
          rowClassName
        }
      )
    }
  );
}
function SelectListBox(props) {
  const { ownProps } = useDefinition(SelectListBoxDefinition, props);
  const {
    classes,
    options,
    items,
    children,
    dependencies,
    loading,
    isStale,
    retainedOptions
  } = ownProps;
  const normalizedOptions = options && normalizeOptions(options);
  const isCollectionLoading = loading?.state === "loading" || loading?.state === "filtering" || loading?.state === "sorting";
  const isBusy = isCollectionLoading || loading?.state === "loadingMore";
  const showStale = useDelayedVisibility(
    isStale ?? false,
    loadingIndicatorDelayMs
  );
  const showLoadMoreIndicator = useDelayedVisibility(
    loading?.state === "loadingMore",
    loadingIndicatorDelayMs
  );
  const listBox = /* @__PURE__ */ jsxs(
    ListBox,
    {
      className: classes.root,
      "data-stale": showStale || void 0,
      renderEmptyState: () => /* @__PURE__ */ jsx(
        SelectEmptyState,
        {
          isLoading: isCollectionLoading,
          className: classes.loading,
          rowClassName: classes.loadingRow
        }
      ),
      children: [
        /* @__PURE__ */ jsx(
          SelectCollection,
          {
            normalizedOptions,
            retainedOptions,
            items,
            dependencies,
            children
          }
        ),
        /* @__PURE__ */ jsx(
          SelectLoadMoreIndicator,
          {
            loading,
            isCollectionLoading,
            isIndicatorVisible: showLoadMoreIndicator,
            className: classes.loading,
            rowClassName: classes.loadingRow
          }
        )
      ]
    }
  );
  return /* @__PURE__ */ jsx("div", { "aria-busy": isBusy || void 0, children: listBox });
}

export { SelectListBox };
