// @backstage/ui/dist/components/Select/SelectItem.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { RiCheckLine } from '@remixicon/react';
import { ListBoxItem, Text } from 'react-aria-components';
import { Avatar } from '../Avatar/Avatar.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../Avatar/Avatar.module.css.esm.js';
import { SelectItemDefinition, SelectItemProfileDefinition, SelectItemTextDefinition } from './definition.esm.js';

function SelectItem(props) {
  const { ownProps, restProps } = useDefinition(SelectItemDefinition, props);
  const { classes, children, showSelectionIndicator } = ownProps;
  return /* @__PURE__ */ jsx(ListBoxItem, { className: classes.root, ...restProps, children: (values) => {
    const content = typeof children === "function" ? children(values) : children;
    if (!showSelectionIndicator) {
      return content;
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("div", { className: classes.indicator, children: /* @__PURE__ */ jsx(RiCheckLine, { "aria-hidden": "true" }) }),
      /* @__PURE__ */ jsx("div", { className: classes.content, children: content })
    ] });
  } });
}
function SelectItemText(props) {
  const { ownProps, restProps } = useDefinition(
    SelectItemTextDefinition,
    props
  );
  const { classes, title, description, leadingIcon } = ownProps;
  return /* @__PURE__ */ jsx(
    SelectItem,
    {
      className: classes.root,
      textValue: title,
      showSelectionIndicator: true,
      ...restProps,
      children: /* @__PURE__ */ jsxs("div", { className: classes.content, children: [
        leadingIcon && /* @__PURE__ */ jsx("div", { className: classes.leadingIcon, children: leadingIcon }),
        /* @__PURE__ */ jsxs("div", { className: classes.text, children: [
          /* @__PURE__ */ jsx(Text, { slot: "label", className: classes.title, children: title }),
          description && /* @__PURE__ */ jsx(Text, { slot: "description", className: classes.description, children: description })
        ] })
      ] })
    }
  );
}
function SelectItemProfile(props) {
  const { ownProps, restProps } = useDefinition(
    SelectItemProfileDefinition,
    props
  );
  const { classes, name, src } = ownProps;
  return /* @__PURE__ */ jsx(
    SelectItem,
    {
      className: classes.root,
      textValue: name,
      showSelectionIndicator: true,
      ...restProps,
      children: /* @__PURE__ */ jsxs("div", { className: classes.content, children: [
        /* @__PURE__ */ jsx("div", { className: classes.avatar, children: /* @__PURE__ */ jsx(
          Avatar,
          {
            src: src ?? "",
            name,
            size: "x-small",
            purpose: "decoration"
          }
        ) }),
        /* @__PURE__ */ jsx(Text, { slot: "label", className: classes.name, children: name })
      ] })
    }
  );
}

export { SelectItem, SelectItemProfile, SelectItemText };
