// @backstage/ui/dist/components/Select/Select.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect } from 'react';
import { Select as Select$1 } from 'react-aria-components';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { SelectDefinition } from './definition.esm.js';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { Popover } from '../Popover/Popover.esm.js';
import '../Popover/Popover.module.css.esm.js';
import { SelectTrigger } from './SelectTrigger.esm.js';
import { SelectContent } from './SelectContent.esm.js';
import { useCollectionAdapter } from '../../hooks/useCollectionAdapter.esm.js';
import { useTrackedSelectionKeys } from '../../hooks/useTrackedSelectionKeys.esm.js';
import { resolveCollectionSource, toSelection, getItemKeys } from '../../utils/selectableCollection.esm.js';

function resolveSelectSearch({
  search,
  searchable,
  searchPlaceholder
}) {
  if (search !== void 0) {
    return search;
  }
  if (searchable) {
    return { placeholder: searchPlaceholder };
  }
  return void 0;
}
function resolveContentSearch(search, collection) {
  if (typeof search !== "object") {
    return search;
  }
  return {
    ...search,
    inputValue: collection.inputValue,
    defaultInputValue: collection.defaultInputValue,
    onInputChange: collection.onInputChange
  };
}
function getRetainedOptions(flatOptions, canonicalItems) {
  if (!flatOptions) {
    return void 0;
  }
  const sourceIds = getItemKeys(flatOptions);
  return canonicalItems.filter((item) => !sourceIds.has(item.id));
}
function SelectImpl(props, ref) {
  const { ownProps, restProps, dataAttributes } = useDefinition(
    SelectDefinition,
    {
      placeholder: "Select an option",
      ...props
    }
  );
  const {
    classes,
    label,
    description,
    options,
    icon,
    searchable,
    searchPlaceholder,
    search,
    loading,
    items,
    children,
    dependencies,
    isRequired,
    secondaryLabel
  } = ownProps;
  const ariaLabel = restProps["aria-label"];
  const ariaLabelledBy = restProps["aria-labelledby"];
  useEffect(() => {
    if (!label && !ariaLabel && !ariaLabelledBy) {
      console.warn(
        "Select requires either a visible label, aria-label, or aria-labelledby for accessibility"
      );
    }
  }, [label, ariaLabel, ariaLabelledBy]);
  const secondaryLabelText = secondaryLabel || (isRequired ? "Required" : null);
  const resolvedSearch = resolveSelectSearch({
    search,
    searchable,
    searchPlaceholder
  });
  const collectionSource = resolveCollectionSource({ options, items });
  const controlledValue = restProps.value !== void 0 ? restProps.value : restProps.selectedKey;
  const defaultValue = restProps.defaultValue !== void 0 ? restProps.defaultValue : restProps.defaultSelectedKey;
  const trackedSelection = useTrackedSelectionKeys({
    selectedKeys: controlledValue === void 0 ? void 0 : toSelection(controlledValue),
    defaultSelectedKeys: toSelection(defaultValue)
  });
  const collection = useCollectionAdapter({
    items: collectionSource.source,
    selectedKeys: trackedSelection.selectedKeys,
    search: resolvedSearch,
    loading
  });
  const retainedOptions = getRetainedOptions(
    collectionSource.flatOptions,
    collection.canonicalItems
  );
  const renderedItems = collectionSource.rendersItems ? collection.canonicalItems : void 0;
  const contentSearch = resolveContentSearch(resolvedSearch, collection);
  const handleChange = (value) => {
    trackedSelection.onSelectionChange(toSelection(value));
    restProps.onChange?.(value);
  };
  return /* @__PURE__ */ jsxs(
    Select$1,
    {
      className: classes.root,
      ...dataAttributes,
      ref,
      ...restProps,
      onChange: handleChange,
      allowsEmptyCollection: true,
      children: [
        /* @__PURE__ */ jsx(
          FieldLabel,
          {
            label,
            secondaryLabel: secondaryLabelText,
            description,
            descriptionSlot: "description"
          }
        ),
        /* @__PURE__ */ jsx(SelectTrigger, { icon }),
        /* @__PURE__ */ jsx(FieldError, {}),
        /* @__PURE__ */ jsx(Popover, { className: classes.popover, hideArrow: true, ...dataAttributes, children: /* @__PURE__ */ jsx(
          SelectContent,
          {
            search: contentSearch,
            options: collectionSource.options,
            items: renderedItems,
            dependencies,
            loading: collection.loading,
            isStale: collection.isStale,
            visibleIds: collection.visibleIds,
            retainedOptions,
            children
          }
        ) })
      ]
    }
  );
}
const Select = forwardRef(SelectImpl);
Select.displayName = "Select";

export { Select };
