// @backstage/ui/dist/components/Combobox/ComboboxItem.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { ListBoxItem, Text } from 'react-aria-components';
import { RiCheckLine } from '@remixicon/react';
import { Avatar } from '../Avatar/Avatar.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../Avatar/Avatar.module.css.esm.js';
import { ComboboxItemDefinition, ComboboxItemProfileDefinition, ComboboxItemTextDefinition } from './definition.esm.js';

function ComboboxItem(props) {
  const { ownProps, restProps } = useDefinition(ComboboxItemDefinition, props);
  const { classes, children, textValue, showSelectionIndicator } = ownProps;
  return /* @__PURE__ */ jsx(ListBoxItem, { ...restProps, className: classes.root, textValue, children: (values) => {
    const content = typeof children === "function" ? children(values) : children;
    if (!showSelectionIndicator) {
      return content;
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("div", { className: classes.indicator, children: /* @__PURE__ */ jsx(RiCheckLine, { "aria-hidden": "true" }) }),
      /* @__PURE__ */ jsx("div", { className: classes.content, children: content })
    ] });
  } });
}
function ComboboxItemText(props) {
  const { ownProps, restProps } = useDefinition(
    ComboboxItemTextDefinition,
    props
  );
  const { classes, title, description, leadingIcon, textValue } = ownProps;
  return /* @__PURE__ */ jsxs(
    ComboboxItem,
    {
      ...restProps,
      className: classes.root,
      textValue: textValue ?? title,
      showSelectionIndicator: true,
      children: [
        leadingIcon && /* @__PURE__ */ jsx("div", { className: classes.leadingIcon, children: leadingIcon }),
        /* @__PURE__ */ jsxs("div", { className: classes.text, children: [
          /* @__PURE__ */ jsx(Text, { slot: "label", className: classes.title, children: title }),
          description && /* @__PURE__ */ jsx(Text, { slot: "description", className: classes.description, children: description })
        ] })
      ]
    }
  );
}
function ComboboxItemProfile(props) {
  const { ownProps, restProps } = useDefinition(
    ComboboxItemProfileDefinition,
    props
  );
  const { classes, name, src, textValue } = ownProps;
  return /* @__PURE__ */ jsxs(
    ComboboxItem,
    {
      ...restProps,
      className: classes.root,
      textValue: textValue ?? name,
      showSelectionIndicator: true,
      children: [
        /* @__PURE__ */ jsx("div", { className: classes.avatar, children: /* @__PURE__ */ jsx(
          Avatar,
          {
            src: src ?? "",
            name,
            size: "x-small",
            purpose: "decoration"
          }
        ) }),
        /* @__PURE__ */ jsx(Text, { slot: "label", className: classes.name, children: name })
      ]
    }
  );
}

export { ComboboxItem, ComboboxItemProfile, ComboboxItemText };
