// @backstage/ui/dist/components/Combobox/Combobox.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect } from 'react';
import { ComboBox } from 'react-aria-components';
import { useFilter } from 'react-aria';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { ComboboxDefinition } from './definition.esm.js';
import { Popover } from '../Popover/Popover.esm.js';
import '../Popover/Popover.module.css.esm.js';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { ComboboxInput } from './ComboboxInput.esm.js';
import { ComboboxListBox } from './ComboboxListBox.esm.js';
import { useCollectionAdapter } from '../../hooks/useCollectionAdapter.esm.js';
import { resolveCollectionSource, isAsyncListSource } from '../../utils/selectableCollection.esm.js';
import { useAsyncComboboxState, getAsyncComboboxItemTextValue } from './useAsyncComboboxState.esm.js';

function resolveComboboxStateProps({
  runtimeState,
  asyncState,
  collection,
  hasSearch
}) {
  if (asyncState) {
    return asyncState;
  }
  const selectionProps = {
    value: runtimeState.value,
    defaultValue: runtimeState.defaultValue,
    onChange: runtimeState.onChange
  };
  if (hasSearch) {
    return {
      ...selectionProps,
      inputValue: collection.inputValue,
      defaultInputValue: collection.defaultInputValue,
      onInputChange: collection.onInputChange
    };
  }
  return {
    ...selectionProps,
    inputValue: runtimeState.inputValue,
    defaultInputValue: runtimeState.defaultInputValue,
    onInputChange: runtimeState.onInputChange
  };
}
function ComboboxImpl(props, ref) {
  const { contains } = useFilter({ sensitivity: "base" });
  const { ownProps, restProps, dataAttributes } = useDefinition(
    ComboboxDefinition,
    props
  );
  const {
    classes,
    label,
    description,
    options,
    items,
    children,
    dependencies,
    icon,
    placeholder,
    isRequired,
    secondaryLabel,
    search,
    loading
  } = ownProps;
  const ariaLabel = restProps["aria-label"];
  const ariaLabelledBy = restProps["aria-labelledby"];
  useEffect(() => {
    if (!label && !ariaLabel && !ariaLabelledBy) {
      console.warn(
        "Combobox requires either a visible label, aria-label, or aria-labelledby for accessibility"
      );
    }
  }, [label, ariaLabel, ariaLabelledBy]);
  const secondaryLabelText = secondaryLabel || (isRequired ? "Required" : null);
  const collectionSource = resolveCollectionSource({
    options,
    items
  });
  const collection = useCollectionAdapter({
    items: collectionSource.source,
    search,
    loading,
    retainSelectedItems: false
  });
  const renderedItems = collectionSource.rendersItems ? collection.canonicalItems : void 0;
  const rootItems = renderedItems ?? (collectionSource.options !== void 0 && search !== void 0 ? collection.canonicalItems : void 0);
  const searchProps = typeof search === "object" ? search : void 0;
  const isDirectAsyncServer = searchProps?.mode === "server" && isAsyncListSource(collectionSource.source);
  const shouldDisableRootFilter = searchProps?.mode === "server" || searchProps?.filter !== void 0;
  const {
    value,
    defaultValue,
    onChange,
    inputValue,
    defaultInputValue,
    onInputChange,
    ...ariaProps
  } = restProps;
  const asyncComboboxProps = isDirectAsyncServer ? {
    source: collectionSource.source,
    value,
    defaultValue,
    onChange,
    allowsCustomValue: restProps.allowsCustomValue
  } : void 0;
  const asyncComboboxState = useAsyncComboboxState(asyncComboboxProps);
  const comboboxStateProps = resolveComboboxStateProps({
    runtimeState: {
      value,
      defaultValue,
      onChange,
      inputValue,
      defaultInputValue,
      onInputChange
    },
    asyncState: asyncComboboxState,
    collection,
    hasSearch: search !== void 0
  });
  const defaultFilter = shouldDisableRootFilter ? () => true : contains;
  const getItemTextValue = isDirectAsyncServer && items !== void 0 ? getAsyncComboboxItemTextValue : void 0;
  return /* @__PURE__ */ jsxs(
    ComboBox,
    {
      className: classes.root,
      defaultFilter,
      items: rootItems,
      ...dataAttributes,
      ref,
      ...ariaProps,
      ...comboboxStateProps,
      allowsEmptyCollection: true,
      children: [
        /* @__PURE__ */ jsx(
          FieldLabel,
          {
            label,
            secondaryLabel: secondaryLabelText,
            description,
            descriptionSlot: "description"
          }
        ),
        /* @__PURE__ */ jsx(ComboboxInput, { icon, placeholder }),
        /* @__PURE__ */ jsx(FieldError, {}),
        /* @__PURE__ */ jsx(Popover, { className: classes.popover, hideArrow: true, ...dataAttributes, children: /* @__PURE__ */ jsx(
          ComboboxListBox,
          {
            options: collectionSource.options,
            items: renderedItems,
            dependencies,
            search,
            loading: collection.loading,
            isStale: collection.isStale,
            getItemTextValue,
            children
          }
        ) })
      ]
    }
  );
}
const Combobox = forwardRef(ComboboxImpl);
Combobox.displayName = "Combobox";

export { Combobox };
