// @backstage/ui/dist/components/Combobox/ComboboxListBox.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { useContext, cloneElement } from 'react';
import { ComboBoxStateContext, ListBox, SelectableCollectionContext, Collection, ListBoxLoadMoreItem, ListBoxSection, Header, Text } from 'react-aria-components';
import { useFilter } from 'react-aria';
import clsx from 'clsx';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { normalizeOptions } from '../../utils/selectableCollection.esm.js';
import { ComboboxItem } from './ComboboxItem.esm.js';
import { ComboboxListBoxDefinition, ComboboxSectionDefinition, ComboboxItemTextDefinition, ComboboxListBoxItemDefinition } from './definition.esm.js';
import { Skeleton } from '../Skeleton/Skeleton.esm.js';
import '../Skeleton/Skeleton.module.css.esm.js';
import { useDelayedVisibility } from '../../hooks/useDelayedVisibility.esm.js';

const loadingRowWidths = ["70%", "55%", "65%"];
const loadingIndicatorDelayMs = 300;
function LoadingRows({
  count,
  className,
  rowClassName
}) {
  return /* @__PURE__ */ jsx("div", { className, "aria-hidden": "true", children: loadingRowWidths.slice(0, count).map((width) => /* @__PURE__ */ jsx("div", { className: rowClassName, children: /* @__PURE__ */ jsx(Skeleton, { width, height: "0.75rem" }) }, width)) });
}
const NoResults = () => {
  const { ownProps } = useDefinition(ComboboxListBoxDefinition, {});
  const { classes } = ownProps;
  return /* @__PURE__ */ jsx("div", { className: classes.noResults, children: "No results found." });
};
function ComboboxOptionItem({ option }) {
  const { ownProps } = useDefinition(ComboboxItemTextDefinition, {
    title: option.label,
    description: option.description,
    leadingIcon: option.leadingIcon
  });
  const { classes } = ownProps;
  const {
    ownProps: { classes: legacyOptionClasses }
  } = useDefinition(ComboboxListBoxItemDefinition, {});
  return /* @__PURE__ */ jsxs(
    ComboboxItem,
    {
      id: option.id,
      value: option,
      className: classes.root,
      textValue: option.label,
      isDisabled: option.disabled,
      showSelectionIndicator: true,
      children: [
        option.leadingIcon && /* @__PURE__ */ jsx("div", { className: classes.leadingIcon, children: option.leadingIcon }),
        /* @__PURE__ */ jsxs("div", { className: classes.text, children: [
          /* @__PURE__ */ jsx(
            Text,
            {
              slot: "label",
              className: clsx(classes.title, legacyOptionClasses.label),
              children: option.label
            }
          ),
          option.description && /* @__PURE__ */ jsx(Text, { slot: "description", className: classes.description, children: option.description })
        ] })
      ]
    }
  );
}
function ComboboxSectionItems({
  section
}) {
  const { ownProps } = useDefinition(ComboboxSectionDefinition, {});
  const { classes } = ownProps;
  return /* @__PURE__ */ jsxs(ListBoxSection, { className: classes.root, children: [
    /* @__PURE__ */ jsx(Header, { className: classes.header, children: section.title }),
    section.options.map((option) => /* @__PURE__ */ jsx(ComboboxOptionItem, { option }, option.id))
  ] });
}
function renderComboboxOption(item) {
  if ("options" in item) {
    return /* @__PURE__ */ jsx(ComboboxSectionItems, { section: item }, item.title);
  }
  return /* @__PURE__ */ jsx(ComboboxOptionItem, { option: item }, item.id);
}
function getCollectionFilter({
  search,
  hasItems,
  inputValue,
  contains
}) {
  const searchProps = typeof search === "object" ? search : void 0;
  if (searchProps?.mode === "server") {
    return void 0;
  }
  const customFilter = searchProps?.filter;
  if (customFilter) {
    return (_textValue, node) => customFilter(node.value, inputValue);
  }
  if (search || hasItems) {
    return (textValue) => contains(textValue, inputValue);
  }
  return void 0;
}
function ComboboxCollection({
  normalizedOptions,
  items,
  children,
  dependencies,
  getItemTextValue
}) {
  if (normalizedOptions) {
    return /* @__PURE__ */ jsx(Fragment, { children: normalizedOptions.map(renderComboboxOption) });
  }
  if (items) {
    const renderItemWithTextValue = (item) => {
      let renderedItem;
      if (typeof children === "function") {
        renderedItem = children(item);
      } else {
        renderedItem = /* @__PURE__ */ jsx(ComboboxOptionItem, { option: item });
      }
      if (!getItemTextValue) {
        return renderedItem;
      }
      return cloneElement(renderedItem, {
        textValue: getItemTextValue(item)
      });
    };
    return /* @__PURE__ */ jsx(Collection, { items, dependencies, children: renderItemWithTextValue });
  }
  return children;
}
function ComboboxEmptyState({
  isLoading,
  className,
  rowClassName
}) {
  if (!isLoading) {
    return /* @__PURE__ */ jsx(NoResults, {});
  }
  return /* @__PURE__ */ jsx(LoadingRows, { count: 3, className, rowClassName });
}
function ComboboxLoadMoreIndicator({
  loading,
  isCollectionLoading,
  isIndicatorVisible,
  className,
  rowClassName
}) {
  if (!loading?.onLoadMore || isCollectionLoading) {
    return null;
  }
  const onLoadMore = loading.state === "loadingMore" ? void 0 : loading.onLoadMore;
  return /* @__PURE__ */ jsx(
    ListBoxLoadMoreItem,
    {
      isLoading: isIndicatorVisible,
      onLoadMore,
      scrollOffset: 0,
      children: /* @__PURE__ */ jsx(
        LoadingRows,
        {
          count: 1,
          className,
          rowClassName
        }
      )
    }
  );
}
function ComboboxListBox(props) {
  const { ownProps } = useDefinition(ComboboxListBoxDefinition, props);
  const {
    classes,
    options,
    items,
    children,
    dependencies,
    search,
    loading,
    isStale,
    getItemTextValue
  } = ownProps;
  const normalizedOptions = options && normalizeOptions(options);
  const state = useContext(ComboBoxStateContext);
  const { contains } = useFilter({ sensitivity: "base" });
  const inputValue = state?.inputValue ?? "";
  const filter = getCollectionFilter({
    search,
    hasItems: items !== void 0,
    inputValue,
    contains
  });
  const isCollectionLoading = loading?.state === "loading" || loading?.state === "filtering" || loading?.state === "sorting";
  const isBusy = isCollectionLoading || loading?.state === "loadingMore";
  const showStale = useDelayedVisibility(
    isStale ?? false,
    loadingIndicatorDelayMs
  );
  const showLoadMoreIndicator = useDelayedVisibility(
    loading?.state === "loadingMore",
    loadingIndicatorDelayMs
  );
  const listBox = /* @__PURE__ */ jsxs(
    ListBox,
    {
      className: classes.root,
      "data-stale": showStale || void 0,
      renderEmptyState: () => /* @__PURE__ */ jsx(
        ComboboxEmptyState,
        {
          isLoading: isCollectionLoading,
          className: classes.loading,
          rowClassName: classes.loadingRow
        }
      ),
      children: [
        /* @__PURE__ */ jsx(
          ComboboxCollection,
          {
            normalizedOptions,
            items,
            dependencies,
            getItemTextValue,
            children
          }
        ),
        /* @__PURE__ */ jsx(
          ComboboxLoadMoreIndicator,
          {
            loading,
            isCollectionLoading,
            isIndicatorVisible: showLoadMoreIndicator,
            className: classes.loading,
            rowClassName: classes.loadingRow
          }
        )
      ]
    }
  );
  const busyListBox = /* @__PURE__ */ jsx("div", { "aria-busy": isBusy || void 0, children: listBox });
  if (!filter) {
    return busyListBox;
  }
  return /* @__PURE__ */ jsx(SelectableCollectionContext.Provider, { value: { filter }, children: busyListBox });
}

export { ComboboxListBox };
