// @backstage/ui/dist/components/Combobox/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Combobox.module.css.esm.js';

const ComboboxDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Combobox",
    popover: "bui-ComboboxPopover"
  },
  propDefs: {
    icon: {},
    size: { dataAttribute: true, default: "small" },
    options: {},
    items: {},
    children: {},
    dependencies: {},
    search: {},
    loading: {},
    placeholder: {},
    label: {},
    secondaryLabel: {},
    description: {},
    isRequired: {},
    className: {}
  }
});
const ComboboxInputDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-ComboboxInput",
      icon: "bui-ComboboxInputIcon",
      input: "bui-ComboboxInputField",
      chevron: "bui-ComboboxInputChevron"
    },
    bg: "consumer",
    propDefs: {
      icon: {},
      placeholder: {}
    }
  }
);
const ComboboxListBoxDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxList",
    noResults: "bui-ComboboxNoResults",
    loading: "bui-ComboboxLoading",
    loadingRow: "bui-ComboboxLoadingRow"
  },
  propDefs: {
    options: {},
    items: {},
    children: {},
    dependencies: {},
    search: {},
    loading: {},
    isStale: {},
    getItemTextValue: {}
  }
});
const ComboboxListBoxItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxItem",
    indicator: "bui-ComboboxItemIndicator",
    label: "bui-ComboboxItemLabel"
  },
  propDefs: {}
});
const ComboboxItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxItem",
    indicator: "bui-ComboboxItemIndicator",
    content: "bui-ComboboxItemContent"
  },
  propDefs: {
    children: {},
    textValue: {},
    showSelectionIndicator: {},
    className: {}
  }
});
const ComboboxItemTextDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxItemText",
    leadingIcon: "bui-ComboboxItemTextLeadingIcon",
    text: "bui-ComboboxItemTextContent",
    title: "bui-ComboboxItemTitle",
    description: "bui-ComboboxItemDescription"
  },
  propDefs: {
    title: {},
    description: {},
    leadingIcon: {},
    textValue: {},
    className: {}
  }
});
const ComboboxItemProfileDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxItemProfile",
    avatar: "bui-ComboboxItemAvatar",
    name: "bui-ComboboxItemTitle"
  },
  propDefs: {
    name: {},
    src: {},
    textValue: {},
    className: {}
  }
});
const ComboboxSectionDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-ComboboxSection",
    header: "bui-ComboboxSectionHeader"
  },
  propDefs: {}
});

export { ComboboxDefinition, ComboboxInputDefinition, ComboboxItemDefinition, ComboboxItemProfileDefinition, ComboboxItemTextDefinition, ComboboxListBoxDefinition, ComboboxListBoxItemDefinition, ComboboxSectionDefinition };
