// @backstage/ui/dist/components/Combobox/useAsyncComboboxState.esm.js
import { useState, useRef, useCallback, useEffect } from 'react';

function getItemTextValue(item) {
  return "textValue" in item ? item.textValue : item.label;
}
function resolveNextSelectedItem({
  key,
  currentItem,
  sourceItems,
  isControlled
}) {
  if (key === null) {
    return null;
  }
  if (isControlled && currentItem?.id === key) {
    return currentItem;
  }
  return sourceItems.find((item) => item.id === key) ?? currentItem;
}
function useAsyncComboboxState(props) {
  const isEnabled = props !== void 0;
  const { source, value, defaultValue, onChange, allowsCustomValue } = props ?? {};
  const isControlled = value !== void 0;
  const [uncontrolledValue, setUncontrolledValue] = useState(
    defaultValue ?? null
  );
  const selectedItem = isControlled ? value ?? null : uncontrolledValue;
  const selectedItemRef = useRef(selectedItem);
  const sourceItemsRef = useRef([]);
  const inputValueRef = useRef("");
  const sourceFilterText = source?.filterText ?? "";
  const setSourceFilterText = source?.setFilterText;
  const sourceFilterTextRef = useRef(sourceFilterText);
  const preserveInputOnClearRef = useRef(false);
  const [inputValue, setInputValue] = useState(
    () => selectedItem ? getAsyncComboboxItemTextValue(selectedItem) : sourceFilterText
  );
  selectedItemRef.current = selectedItem;
  sourceItemsRef.current = source ? [...source.items] : [];
  inputValueRef.current = inputValue;
  sourceFilterTextRef.current = sourceFilterText;
  const updateInputValue = useCallback(
    (nextInputValue) => {
      setInputValue(nextInputValue);
      if (isEnabled && setSourceFilterText && sourceFilterTextRef.current !== nextInputValue) {
        sourceFilterTextRef.current = nextInputValue;
        setSourceFilterText(nextInputValue);
      }
    },
    [isEnabled, setSourceFilterText]
  );
  useEffect(() => {
    if (isEnabled) {
      setInputValue(sourceFilterText);
    }
  }, [isEnabled, sourceFilterText]);
  const selectedId = selectedItem?.id ?? null;
  const selectedTextValue = selectedItem ? getAsyncComboboxItemTextValue(selectedItem) : void 0;
  const previousSelectedId = useRef(null);
  const previousSelectedTextValue = useRef();
  useEffect(() => {
    if (!isEnabled) {
      return;
    }
    const selectionChanged = previousSelectedId.current !== selectedId || previousSelectedTextValue.current !== selectedTextValue;
    previousSelectedId.current = selectedId;
    previousSelectedTextValue.current = selectedTextValue;
    if (selectionChanged) {
      if (selectedTextValue !== void 0) {
        updateInputValue(selectedTextValue);
      } else if (preserveInputOnClearRef.current) {
        preserveInputOnClearRef.current = false;
      } else {
        updateInputValue("");
      }
    }
  }, [isEnabled, selectedId, selectedTextValue, updateInputValue]);
  const freshSelectedItem = selectedItem ? sourceItemsRef.current.find((item) => item.id === selectedItem.id) : void 0;
  const isEditing = selectedTextValue !== void 0 && inputValueRef.current !== selectedTextValue;
  useEffect(() => {
    if (isEnabled && !isControlled && !isEditing && freshSelectedItem && freshSelectedItem !== selectedItemRef.current) {
      setUncontrolledValue(freshSelectedItem);
    }
  }, [freshSelectedItem, isControlled, isEditing, isEnabled]);
  const handleChange = useCallback(
    (key) => {
      const currentItem = selectedItemRef.current;
      const currentTextValue = currentItem ? getAsyncComboboxItemTextValue(currentItem) : void 0;
      if (key === null && allowsCustomValue && currentTextValue === inputValueRef.current) {
        updateInputValue(currentTextValue);
        return;
      }
      if (key === null && allowsCustomValue && currentItem) {
        preserveInputOnClearRef.current = true;
      }
      if (key === null && !currentItem && !allowsCustomValue) {
        updateInputValue("");
        return;
      }
      const nextItem = resolveNextSelectedItem({
        key,
        currentItem,
        sourceItems: sourceItemsRef.current,
        isControlled
      });
      if (nextItem) {
        updateInputValue(getAsyncComboboxItemTextValue(nextItem));
      }
      if (!isControlled) {
        setUncontrolledValue(nextItem);
      }
      if (nextItem?.id !== currentItem?.id) {
        onChange?.(nextItem);
      }
    },
    [allowsCustomValue, isControlled, onChange, updateInputValue]
  );
  if (!isEnabled) {
    return void 0;
  }
  return {
    value: selectedItem?.id ?? null,
    inputValue,
    onChange: handleChange,
    onInputChange: updateInputValue
  };
}
function getAsyncComboboxItemTextValue(item) {
  return getItemTextValue(item);
}

export { getAsyncComboboxItemTextValue, useAsyncComboboxState };
