// @backstage/plugin-catalog-react/dist/components/InspectEntityDialog/InspectEntityDialog.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { Dialog, DialogHeader, DialogBody, Tabs, TabList, Tab, TabPanel } from '@backstage/ui';
import { useMemo } from 'react';
import { AncestryPage } from './components/AncestryPage.esm.js';
import { ColocatedPage } from './components/ColocatedPage.esm.js';
import { JsonPage } from './components/JsonPage.esm.js';
import { OverviewPage } from './components/OverviewPage.esm.js';
import { YamlPage } from './components/YamlPage.esm.js';
import { catalogReactTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const TAB_KEYS = [
  "overview",
  "ancestry",
  "colocated",
  "json",
  "yaml"
];
function DialogContents(props) {
  const { entity, initialTab, onSelect } = props;
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const tabNames = useMemo(
    () => ({
      overview: t("inspectEntityDialog.tabNames.overview"),
      ancestry: t("inspectEntityDialog.tabNames.ancestry"),
      colocated: t("inspectEntityDialog.tabNames.colocated"),
      json: t("inspectEntityDialog.tabNames.json"),
      yaml: t("inspectEntityDialog.tabNames.yaml")
    }),
    [t]
  );
  const tabContent = {
    overview: /* @__PURE__ */ jsx(OverviewPage, { entity }),
    ancestry: /* @__PURE__ */ jsx(AncestryPage, { entity }),
    colocated: /* @__PURE__ */ jsx(ColocatedPage, { entity }),
    json: /* @__PURE__ */ jsx(JsonPage, { entity }),
    yaml: /* @__PURE__ */ jsx(YamlPage, { entity })
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(DialogHeader, { children: t("inspectEntityDialog.title") }),
    /* @__PURE__ */ jsx(DialogBody, { children: /* @__PURE__ */ jsxs(
      Tabs,
      {
        defaultSelectedKey: initialTab || "overview",
        onSelectionChange: (key) => onSelect?.(key),
        children: [
          /* @__PURE__ */ jsx(TabList, { "aria-label": t("inspectEntityDialog.tabsAriaLabel"), children: TAB_KEYS.map((tab) => /* @__PURE__ */ jsx(Tab, { id: tab, children: tabNames[tab] }, tab)) }),
          TAB_KEYS.map((tab) => /* @__PURE__ */ jsx(TabPanel, { id: tab, children: tabContent[tab] }, tab))
        ]
      }
    ) })
  ] });
}
function InspectEntityDialog(props) {
  const { open, entity, initialTab, onClose, onSelect } = props;
  if (!entity) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    Dialog,
    {
      isOpen: open,
      onOpenChange: (isOpen) => !isOpen && onClose(),
      width: "940px",
      height: "100vh",
      children: open && /* @__PURE__ */ jsx(
        DialogContents,
        {
          entity,
          initialTab,
          onSelect
        }
      )
    }
  );
}

export { InspectEntityDialog };
