// @backstage/ui/dist/components/NumberField/NumberField.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect, useMemo } from 'react';
import { NumberField as NumberField$1, Group, Input } from 'react-aria-components';
import { RiSubtractLine, RiAddLine } from '@remixicon/react';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { ButtonIcon } from '../ButtonIcon/ButtonIcon.esm.js';
import '../ButtonIcon/ButtonIcon.module.css.esm.js';
import { NumberFieldDefinition } from './definition.esm.js';

const NumberField = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      NumberFieldDefinition,
      props
    );
    const { classes, label, icon, secondaryLabel, placeholder, description } = ownProps;
    useEffect(() => {
      if (!label && !restProps["aria-label"] && !restProps["aria-labelledby"]) {
        console.warn(
          "NumberField requires either a visible label, aria-label, or aria-labelledby for accessibility"
        );
      }
    }, [label, restProps["aria-label"], restProps["aria-labelledby"]]);
    const secondaryLabelText = secondaryLabel || (restProps.isRequired ? "Required" : null);
    const formatOptions = useMemo(
      () => ({
        useGrouping: false,
        ...restProps.formatOptions
      }),
      [restProps.formatOptions]
    );
    return /* @__PURE__ */ jsxs(
      NumberField$1,
      {
        className: classes.root,
        ...dataAttributes,
        ...restProps,
        formatOptions,
        ref,
        children: [
          /* @__PURE__ */ jsx(
            FieldLabel,
            {
              label,
              secondaryLabel: secondaryLabelText,
              description,
              descriptionSlot: "description"
            }
          ),
          /* @__PURE__ */ jsxs(
            Group,
            {
              className: classes.inputWrapper,
              "data-size": dataAttributes["data-size"],
              children: [
                icon && /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: classes.inputIcon,
                    "data-size": dataAttributes["data-size"],
                    "aria-hidden": "true",
                    children: icon
                  }
                ),
                /* @__PURE__ */ jsx(
                  Input,
                  {
                    className: classes.input,
                    ...icon && { "data-icon": true },
                    placeholder
                  }
                ),
                /* @__PURE__ */ jsxs("div", { className: classes.stepperButtons, children: [
                  /* @__PURE__ */ jsx(
                    ButtonIcon,
                    {
                      slot: "decrement",
                      variant: "tertiary",
                      size: ownProps.size,
                      className: classes.stepperButton,
                      icon: /* @__PURE__ */ jsx(RiSubtractLine, {}),
                      "aria-label": "Decrease"
                    }
                  ),
                  /* @__PURE__ */ jsx(
                    ButtonIcon,
                    {
                      slot: "increment",
                      variant: "tertiary",
                      size: ownProps.size,
                      className: classes.stepperButton,
                      icon: /* @__PURE__ */ jsx(RiAddLine, {}),
                      "aria-label": "Increase"
                    }
                  )
                ] })
              ]
            }
          ),
          /* @__PURE__ */ jsx(FieldError, {})
        ]
      }
    );
  }
);
NumberField.displayName = "NumberField";

export { NumberField };
