// @backstage/cli-module-build/dist/lib/bundler/moduleFederation.cjs.js
'use strict';

var entryPoints = require('../entryPoints.cjs.js');
var typeDistProject = require('../typeDistProject.cjs.js');
var moduleFederationCommon = require('@backstage/module-federation-common');
var node_path = require('node:path');
var fs = require('fs-extra');
var chokidar = require('chokidar');
var PQueue = require('p-queue');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var chokidar__default = /*#__PURE__*/_interopDefaultCompat(chokidar);
var PQueue__default = /*#__PURE__*/_interopDefaultCompat(PQueue);

async function getModuleFederationRemoteOptions(packageJson, packageDir) {
  let exposes;
  const packageRole = packageJson.backstage?.role;
  if (packageJson.exports && packageRole) {
    const project = await typeDistProject.createTypeDistProject();
    exposes = Object.fromEntries(
      entryPoints.readEntryPoints(packageJson).filter((ep) => {
        if (ep.mount === "./package.json") {
          return false;
        }
        if (ep.mount === ".") {
          return true;
        }
        return typeDistProject.getEntryPointDefaultFeatureType(
          packageRole,
          packageDir,
          project,
          ep.path
        ) !== null;
      }).map((ep) => [ep.mount, ep.path])
    );
  }
  return {
    // The default output mode requires the name to be a usable as a code
    // symbol, there might be better options here but for now we need to
    // sanitize the name.
    name: packageJson.name.replaceAll("@", "").replaceAll("/", "__").replaceAll("-", "_"),
    exposes,
    sharedDependencies: moduleFederationCommon.defaultRemoteSharedDependencies()
  };
}
function prepareRuntimeSharedDependenciesScript(hostSharedDependencies) {
  const items = Object.entries(hostSharedDependencies).map(
    ([name, sharedDep]) => {
      if (!sharedDep.version) {
        throw new Error(`Version is required for shared dependency '${name}'`);
      }
      return {
        name,
        version: sharedDep.version,
        lib: name,
        // Converted into import below
        shareConfig: {
          singleton: sharedDep.singleton,
          requiredVersion: sharedDep.requiredVersion,
          eager: sharedDep.eager
        }
      };
    }
  );
  return `window['${moduleFederationCommon.BACKSTAGE_RUNTIME_SHARED_DEPENDENCIES_GLOBAL}'] = ${JSON.stringify(
    { items, version: "v1" },
    null,
    2
  ).replace(
    /"lib": ("[^"]+")/gm,
    (_, name) => `"lib": () => import(${name})`
  )};`;
}
const RUNTIME_SHARED_DEPENDENCIES_MODULE_NAME = "__backstage-module-federation-runtime-shared-dependencies__";
const writeQueue = new PQueue__default.default({ concurrency: 1 });
async function writeRuntimeSharedDependenciesModule(targetPath, runtimeSharedDependencies) {
  const script = prepareRuntimeSharedDependenciesScript(
    runtimeSharedDependencies
  );
  await writeQueue.add(async () => {
    const path = node_path.join(
      targetPath,
      "node_modules",
      `${RUNTIME_SHARED_DEPENDENCIES_MODULE_NAME}.js`
    );
    await fs__default.default.ensureDir(node_path.dirname(path));
    await fs__default.default.writeFile(path, script);
  });
}
function resolveSharedDependencyVersions(targetPath, hostSharedDependencies) {
  return Object.fromEntries(
    Object.entries(hostSharedDependencies).filter(([_, sharedDep]) => sharedDep !== void 0).flatMap(([importPath, sharedDep]) => {
      const moduleName = importPath.startsWith("@") ? importPath.split("/").slice(0, 2).join("/") : importPath.split("/")[0];
      let version;
      try {
        const packagePath = require.resolve(`${moduleName}/package.json`, {
          paths: [targetPath]
        });
        version = require(packagePath).version;
      } catch (e) {
        console.log(
          `Skipping module federation shared dependency '${importPath}' because it could not be resolved.`
        );
        return [];
      }
      return [[importPath, { ...sharedDep, version }]];
    })
  );
}
async function createRuntimeSharedDependenciesEntryPoint(options) {
  const { targetPath, watch } = options;
  const doWriteSharedDependenciesModule = async () => {
    const sharedDependencies = moduleFederationCommon.defaultHostSharedDependencies();
    await writeRuntimeSharedDependenciesModule(
      targetPath,
      resolveSharedDependencyVersions(targetPath, sharedDependencies)
    );
  };
  if (watch) {
    const watcher = chokidar__default.default.watch(node_path.resolve(targetPath, "package.json"));
    watcher.on("change", async () => {
      await doWriteSharedDependenciesModule();
      watch();
    });
  }
  await doWriteSharedDependenciesModule();
  return [RUNTIME_SHARED_DEPENDENCIES_MODULE_NAME];
}

exports.createRuntimeSharedDependenciesEntryPoint = createRuntimeSharedDependenciesEntryPoint;
exports.getModuleFederationRemoteOptions = getModuleFederationRemoteOptions;
exports.prepareRuntimeSharedDependenciesScript = prepareRuntimeSharedDependenciesScript;
