// @acarmisc/backstage-plugin-litellm-backend/dist/bridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeycloakJWTVerifier = exports.BridgeAuthError = exports.BridgeConfigError = void 0;
exports.readBridgeConfig = readBridgeConfig;
exports.newDefaultVerifier = newDefaultVerifier;
exports.resolveBridgeUserId = resolveBridgeUserId;
exports.getOrProvisionUserFromClaims = getOrProvisionUserFromClaims;
exports.bridgeListKeys = bridgeListKeys;
exports.bridgeGenerateKey = bridgeGenerateKey;
exports.bridgeRegenerateKey = bridgeRegenerateKey;
/**
 * Bridge — lets CLI clients (Abby) list/mint LiteLLM virtual keys without ever
 * holding the LiteLLM master key.
 *
 * Trust model: the Backstage backend holds the master key (as it already does
 * for the UI). A CLI authenticates with its Keycloak access token (the same
 * Keycloak realm Backstage uses). This module verifies that JWT against the
 * realm JWKS, resolves the caller to a LiteLLM user_id, ensures that user
 * exists (provisioning from claims if enabled), and then lists/generates keys
 * via the existing master-key-authed {@link LiteLLMClient}.
 *
 * Unlike the UI endpoints in router.ts, the bridge routes do NOT call
 * Backstage's `auth.authenticate` (which expects a Backstage-issued token);
 * they verify the raw Keycloak JWT themselves.
 */
const jose_1 = require("jose");
const provisioning_1 = require("./provisioning");
function readBridgeConfig(config) {
    const enabled = config.getOptionalBoolean('litellm.bridge.enabled') ?? false;
    const issuer = config.getOptionalString('litellm.bridge.issuer');
    const clientId = config.getOptionalString('litellm.bridge.clientId') ?? 'abby-cli';
    return { enabled, issuer, clientId };
}
/** Thrown when the bridge is misconfigured (e.g. enabled without an issuer). */
class BridgeConfigError extends Error {
}
exports.BridgeConfigError = BridgeConfigError;
/** Thrown when a presented token fails verification → maps to HTTP 401. */
class BridgeAuthError extends Error {
    constructor() {
        super(...arguments);
        this.status = 401;
    }
}
exports.BridgeAuthError = BridgeAuthError;
/**
 * Verifies a Keycloak access token against the realm JWKS and ensures it was
 * issued for {@link clientId} (via azp, falling back to aud). Uses jose's
 * remote JWKS client (cached, with cooldown on errors).
 */
class KeycloakJWTVerifier {
    constructor(opts) {
        this.issuer = opts.issuer.replace(/\/$/, '');
        this.clientId = opts.clientId;
        this.jwks = (0, jose_1.createRemoteJWKSet)(new URL(`${this.issuer}/protocol/openid-connect/certs`));
    }
    async verify(token) {
        let payload;
        try {
            const result = await (0, jose_1.jwtVerify)(token, this.jwks, {
                issuer: this.issuer,
            });
            payload = result.payload;
        }
        catch (err) {
            // Expired, bad signature, wrong issuer, malformed, JWKS unreachable, etc.
            throw new BridgeAuthError(`invalid keycloak token: ${err.message ?? err}`);
        }
        const azp = payload.azp;
        const aud = payload.aud;
        const audMatches = Array.isArray(aud)
            ? aud.includes(this.clientId)
            : aud === this.clientId;
        if (azp !== this.clientId && !audMatches) {
            throw new BridgeAuthError(`token not issued for client "${this.clientId}" (azp=${azp ?? 'none'})`);
        }
        return {
            sub: payload.sub ?? '',
            email: payload.email,
            preferred_username: payload
                .preferred_username,
            name: payload.name,
            azp,
            aud,
        };
    }
}
exports.KeycloakJWTVerifier = KeycloakJWTVerifier;
/** Builds the default verifier from config, or throws BridgeConfigError. */
function newDefaultVerifier(cfg) {
    if (!cfg.issuer) {
        throw new BridgeConfigError('litellm.bridge.issuer is required when litellm.bridge.enabled is true ' +
            '(e.g. https://auth.ces.abssrv.it/realms/solution-innovation)');
    }
    return new KeycloakJWTVerifier({ issuer: cfg.issuer, clientId: cfg.clientId });
}
/**
 * Resolves the LiteLLM user_id from the verified claims, matching exactly how
 * the UI derives it so the bridge addresses the *same* LiteLLM user (not a
 * duplicate). The UI uses toLiteLLMUserId(entityRef, userIdDomain) where the
 * Backstage entity name has been rewritten by the Keycloak user transformer:
 * because the catalog rejects '@' in entity names, usernames imported as full
 * emails are stored as their local-part. We replicate that here — strip the
 * '@domain' off the username, then apply the same userIdDomain rule.
 */
function resolveBridgeUserId(claims, userIdDomain) {
    const raw = claims.preferred_username ?? claims.email ?? claims.sub;
    const entityName = raw.includes('@') ? raw.split('@')[0] : raw;
    return (0, provisioning_1.toLiteLLMUserId)(entityName, userIdDomain);
}
/**
 * Ensures a LiteLLM user exists for the verified identity. If the user is
 * missing and provisioning is enabled, creates it from the JWT claims (email +
 * name); if provisioning is disabled, throws a 404 telling the caller to log
 * in to Backstage first (the UI is the primary provisioning entry point).
 */
async function getOrProvisionUserFromClaims(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain) {
    const userId = resolveBridgeUserId(claims, userIdDomain);
    const existing = await client.getUserInfo(userId);
    if (existing)
        return existing;
    if (!provisioningEnabled) {
        throw new provisioning_1.ProvisioningError('User not found in LiteLLM', 'No LiteLLM user for this identity. Log in to Backstage once to be provisioned, or enable litellm.provisioning.enabled.', false);
    }
    const profile = {
        email: claims.email ?? claims.preferred_username,
        displayName: claims.name ?? claims.preferred_username,
    };
    const created = await (0, provisioning_1.provisionUser)(client, userId, provisioningDefaults, profile, undefined, logger);
    if (!created) {
        throw new provisioning_1.ProvisioningError('User not found in LiteLLM', 'Provisioning attempted but returned no user — check LiteLLM logs', true, 500);
    }
    return created;
}
/** Lists the caller's virtual keys (provisioning the user first if needed). */
async function bridgeListKeys(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain) {
    await getOrProvisionUserFromClaims(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain);
    return client.listKeys(resolveBridgeUserId(claims, userIdDomain));
}
/** Mints a new virtual key for the caller (provisioning the user first if needed). */
async function bridgeGenerateKey(client, claims, provisioningEnabled, provisioningDefaults, logger, request, userIdDomain) {
    await getOrProvisionUserFromClaims(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain);
    const userId = resolveBridgeUserId(claims, userIdDomain);
    const enriched = {
        ...request,
        user_id: userId,
        metadata: {
            ...(request.metadata ?? {}),
            created_via: 'abby-cli',
            created_by: userId,
            created_at_iso: new Date().toISOString(),
        },
    };
    return client.generateKey(enriched);
}
/**
 * Rotates the caller's key identified by alias, returning a fresh secret. The
 * bridge can't recover a usable secret from a listing (it's masked), so a CLI
 * that lost its cached key but whose stable alias still exists uses this to
 * rotate in place instead of minting a duplicate. Throws 404 when no key with
 * that alias exists for the caller — the CLI then mints one.
 */
async function bridgeRegenerateKey(client, claims, provisioningEnabled, provisioningDefaults, logger, alias, userIdDomain) {
    await getOrProvisionUserFromClaims(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain);
    const userId = resolveBridgeUserId(claims, userIdDomain);
    const keys = await client.listKeys(userId);
    const match = keys.find(k => k.key_alias === alias);
    if (!match) {
        throw new provisioning_1.ProvisioningError('Key not found', `No key with alias "${alias}" for this identity; mint one first.`, false, 404);
    }
    return client.regenerateKey(match.token);
}
