// @acarmisc/backstage-plugin-litellm-backend/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bridgeGenerateKey = exports.bridgeListKeys = exports.getOrProvisionUserFromClaims = exports.resolveBridgeUserId = exports.readBridgeConfig = exports.newDefaultVerifier = exports.KeycloakJWTVerifier = exports.ProvisioningError = exports.resolveUserRole = exports.applyRoleOverrides = exports.readRoleConfigs = exports.readProvisioningDefaults = exports.provisionUser = exports.getOrProvisionUser = exports.toLiteLLMUserId = exports.resolveUserProfile = exports.resolveUserId = exports.LiteLLMClient = exports.createRouter = exports.default = exports.litellmPlugin = void 0;
var plugin_1 = require("./plugin");
Object.defineProperty(exports, "litellmPlugin", { enumerable: true, get: function () { return plugin_1.litellmPlugin; } });
var plugin_2 = require("./plugin");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return plugin_2.litellmPlugin; } });
var router_1 = require("./router");
Object.defineProperty(exports, "createRouter", { enumerable: true, get: function () { return router_1.createRouter; } });
__exportStar(require("./types"), exports);
var client_1 = require("./client");
Object.defineProperty(exports, "LiteLLMClient", { enumerable: true, get: function () { return client_1.LiteLLMClient; } });
var provisioning_1 = require("./provisioning");
Object.defineProperty(exports, "resolveUserId", { enumerable: true, get: function () { return provisioning_1.resolveUserId; } });
Object.defineProperty(exports, "resolveUserProfile", { enumerable: true, get: function () { return provisioning_1.resolveUserProfile; } });
Object.defineProperty(exports, "toLiteLLMUserId", { enumerable: true, get: function () { return provisioning_1.toLiteLLMUserId; } });
Object.defineProperty(exports, "getOrProvisionUser", { enumerable: true, get: function () { return provisioning_1.getOrProvisionUser; } });
Object.defineProperty(exports, "provisionUser", { enumerable: true, get: function () { return provisioning_1.provisionUser; } });
Object.defineProperty(exports, "readProvisioningDefaults", { enumerable: true, get: function () { return provisioning_1.readProvisioningDefaults; } });
Object.defineProperty(exports, "readRoleConfigs", { enumerable: true, get: function () { return provisioning_1.readRoleConfigs; } });
Object.defineProperty(exports, "applyRoleOverrides", { enumerable: true, get: function () { return provisioning_1.applyRoleOverrides; } });
Object.defineProperty(exports, "resolveUserRole", { enumerable: true, get: function () { return provisioning_1.resolveUserRole; } });
Object.defineProperty(exports, "ProvisioningError", { enumerable: true, get: function () { return provisioning_1.ProvisioningError; } });
var bridge_1 = require("./bridge");
Object.defineProperty(exports, "KeycloakJWTVerifier", { enumerable: true, get: function () { return bridge_1.KeycloakJWTVerifier; } });
Object.defineProperty(exports, "newDefaultVerifier", { enumerable: true, get: function () { return bridge_1.newDefaultVerifier; } });
Object.defineProperty(exports, "readBridgeConfig", { enumerable: true, get: function () { return bridge_1.readBridgeConfig; } });
Object.defineProperty(exports, "resolveBridgeUserId", { enumerable: true, get: function () { return bridge_1.resolveBridgeUserId; } });
Object.defineProperty(exports, "getOrProvisionUserFromClaims", { enumerable: true, get: function () { return bridge_1.getOrProvisionUserFromClaims; } });
Object.defineProperty(exports, "bridgeListKeys", { enumerable: true, get: function () { return bridge_1.bridgeListKeys; } });
Object.defineProperty(exports, "bridgeGenerateKey", { enumerable: true, get: function () { return bridge_1.bridgeGenerateKey; } });
