// @acarmisc/backstage-plugin-litellm-backend/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.litellmPlugin = void 0;
const backend_plugin_api_1 = require("@backstage/backend-plugin-api");
const router_1 = require("./router");
const bridge_1 = require("./bridge");
exports.litellmPlugin = (0, backend_plugin_api_1.createBackendPlugin)({
    pluginId: 'litellm',
    register(reg) {
        reg.registerInit({
            deps: {
                httpRouter: backend_plugin_api_1.coreServices.httpRouter,
                config: backend_plugin_api_1.coreServices.rootConfig,
                logger: backend_plugin_api_1.coreServices.logger,
                auth: backend_plugin_api_1.coreServices.auth,
                discovery: backend_plugin_api_1.coreServices.discovery,
            },
            async init({ httpRouter, config, logger, auth, discovery }) {
                const router = await (0, router_1.createRouter)({ config, logger, auth, discovery });
                httpRouter.use(router);
                // The CLI bridge verifies raw Keycloak JWTs itself (it does NOT use a
                // Backstage-issued token). Exempt its routes from the default auth
                // policy, otherwise the framework rejects the Keycloak Bearer with
                // "Illegal token" before the bridge verifier ever runs.
                if ((0, bridge_1.readBridgeConfig)(config).enabled) {
                    for (const path of [
                        '/bridge/health',
                        '/bridge/keys',
                        '/bridge/models',
                    ]) {
                        httpRouter.addAuthPolicy({ path, allow: 'unauthenticated' });
                    }
                }
            },
        });
    },
});
