// @acarmisc/backstage-plugin-litellm-backend/dist/router.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProvisioningError = void 0;
exports.createRouter = createRouter;
const express_1 = __importStar(require("express"));
const catalog_client_1 = require("@backstage/catalog-client");
const client_1 = require("./client");
const provisioning_1 = require("./provisioning");
Object.defineProperty(exports, "ProvisioningError", { enumerable: true, get: function () { return provisioning_1.ProvisioningError; } });
const bridge_1 = require("./bridge");
async function createRouter(options) {
    const { config, logger, auth, discovery } = options;
    const baseUrl = config.getString('litellm.baseUrl');
    const masterKey = config.getString('litellm.masterKey');
    const userIdDomain = config.getOptionalString('litellm.userIdDomain');
    const client = options.client ?? new client_1.LiteLLMClient({ baseUrl, masterKey });
    const { enabled: provisioningEnabled, defaults: provisioningDefaults } = (0, provisioning_1.readProvisioningDefaults)(config);
    const roleConfigs = (0, provisioning_1.readRoleConfigs)(config);
    const catalogClient = new catalog_client_1.CatalogClient({ discoveryApi: discovery });
    if (provisioningEnabled) {
        logger.info(`LiteLLM auto-provisioning enabled — defaults: budget=$${provisioningDefaults.maxBudget}/${provisioningDefaults.budgetDuration}, models=${provisioningDefaults.models.length
            ? provisioningDefaults.models.join(',')
            : 'all'}, teams=[${provisioningDefaults.teams.join(',')}]`);
    }
    const router = (0, express_1.Router)();
    // JSON body parser. Without this, every POST/PUT endpoint sees an empty
    // req.body. Backstage's httpRouter does not apply a body parser at the
    // plugin-router level, so each plugin must attach its own.
    router.use(express_1.default.json());
    router.get('/health', (_req, res) => {
        res.json({ status: 'ok', provisioning: provisioningEnabled });
    });
    router.get('/user/info', async (req, res) => {
        try {
            const tokenEntityRef = await (0, provisioning_1.resolveUserId)(req, auth);
            const userId = tokenEntityRef
                ? (0, provisioning_1.toLiteLLMUserId)(tokenEntityRef, userIdDomain)
                : req.query.user_id;
            const userInfo = await (0, provisioning_1.getOrProvisionUser)(client, tokenEntityRef, userId, provisioningEnabled, provisioningDefaults, roleConfigs, catalogClient, auth, logger);
            res.json(userInfo);
        }
        catch (error) {
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Failed to fetch user info', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.get('/keys', async (req, res) => {
        try {
            const tokenEntityRef = await (0, provisioning_1.resolveUserId)(req, auth);
            const userId = tokenEntityRef
                ? (0, provisioning_1.toLiteLLMUserId)(tokenEntityRef, userIdDomain)
                : req.query.user_id;
            await (0, provisioning_1.getOrProvisionUser)(client, tokenEntityRef, userId, provisioningEnabled, provisioningDefaults, roleConfigs, catalogClient, auth, logger);
            const keys = await client.listKeys(userId);
            res.json(keys);
        }
        catch (error) {
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Failed to list keys', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.post('/keys/generate', async (req, res) => {
        try {
            // Only alias + max_budget are required. An empty models array is
            // intentional — in LiteLLM `models: []` means "all models the user
            // can access" which is the desired default. Forcing a selection
            // up front is too restrictive for the common case.
            const body = (req.body ?? {});
            const missing = [];
            if (!body.alias?.trim())
                missing.push('alias');
            if (typeof body.max_budget !== 'number' || body.max_budget <= 0) {
                missing.push('max_budget (positive number)');
            }
            if (missing.length) {
                res.status(400).json({
                    error: 'Missing required fields',
                    hint: `Required: ${missing.join(', ')}`,
                });
                return;
            }
            const tokenEntityRef = await (0, provisioning_1.resolveUserId)(req, auth);
            const resolvedUserId = tokenEntityRef
                ? (0, provisioning_1.toLiteLLMUserId)(tokenEntityRef, userIdDomain)
                : undefined;
            if (resolvedUserId) {
                await (0, provisioning_1.getOrProvisionUser)(client, tokenEntityRef, resolvedUserId, provisioningEnabled, provisioningDefaults, roleConfigs, catalogClient, auth, logger);
            }
            // Stamp ownership into LiteLLM key metadata. LiteLLM's native
            // `created_by` column is only populated when the caller authenticates
            // via JWT/SSO; we always call with the master key, so that column
            // stays null. Enriching `metadata` makes the owner identity visible
            // in LiteLLM's UI and queryable via API.
            const profile = tokenEntityRef
                ? await (0, provisioning_1.resolveUserProfile)(tokenEntityRef, catalogClient, auth, logger)
                : {};
            const enrichedMetadata = {
                ...(body.metadata ?? {}),
                created_by_backstage_user: tokenEntityRef ?? 'unknown',
                ...(profile.email && { created_by_email: profile.email }),
                ...(profile.displayName && {
                    created_by_display_name: profile.displayName,
                }),
                created_via: 'backstage',
                created_at_iso: new Date().toISOString(),
            };
            const request = {
                ...body,
                metadata: enrichedMetadata,
                ...(resolvedUserId && { user_id: resolvedUserId }),
            };
            const result = await client.generateKey(request);
            res.json(result);
        }
        catch (error) {
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Failed to generate key', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.post('/keys/:keyId/update', async (req, res) => {
        try {
            const { keyId } = req.params;
            if (!keyId) {
                res.status(400).json({ error: 'keyId is required' });
                return;
            }
            const request = { ...req.body, key: keyId };
            const result = await client.updateKey(request);
            res.json(result);
        }
        catch (error) {
            logger.error('Failed to update key', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.delete('/keys/:keyId', async (req, res) => {
        try {
            const { keyId } = req.params;
            if (!keyId) {
                res.status(400).json({ error: 'keyId is required' });
                return;
            }
            await client.deleteKeys({ keys: [keyId] });
            res.json({ success: true });
        }
        catch (error) {
            logger.error('Failed to delete key', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.get('/models', async (_req, res) => {
        try {
            const models = await client.listModels();
            res.json(models);
        }
        catch (error) {
            logger.error('Failed to list models', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.get('/teams', async (req, res) => {
        try {
            const tokenEntityRef = await (0, provisioning_1.resolveUserId)(req, auth);
            const userId = tokenEntityRef
                ? (0, provisioning_1.toLiteLLMUserId)(tokenEntityRef, userIdDomain)
                : req.query.user_id;
            const userInfo = await (0, provisioning_1.getOrProvisionUser)(client, tokenEntityRef, userId, provisioningEnabled, provisioningDefaults, roleConfigs, catalogClient, auth, logger);
            if (!userInfo?.teams?.length) {
                res.json([]);
                return;
            }
            const teams = await Promise.all(userInfo.teams.map(teamId => client.getTeamInfo(teamId).catch(err => {
                logger.warn(`Failed to fetch team ${teamId}: ${err.message}`);
                return null;
            })));
            res.json(teams.filter(Boolean));
        }
        catch (error) {
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Failed to fetch teams', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.get('/teams/:teamId/usage', async (req, res) => {
        try {
            const { teamId } = req.params;
            const { start_date, end_date } = req.query;
            if (!start_date || !end_date) {
                res.status(400).json({ error: 'start_date and end_date are required' });
                return;
            }
            const usage = await client.getTeamUsage(teamId, start_date, end_date);
            res.json(usage);
        }
        catch (error) {
            logger.error('Failed to fetch team usage', error);
            res.status(500).json({ error: error.message });
        }
    });
    router.get('/usage', async (req, res) => {
        try {
            const { start_date, end_date, group_by } = req.query;
            if (!start_date || !end_date) {
                res.status(400).json({ error: 'start_date and end_date are required' });
                return;
            }
            const tokenEntityRef = await (0, provisioning_1.resolveUserId)(req, auth);
            const userId = tokenEntityRef
                ? (0, provisioning_1.toLiteLLMUserId)(tokenEntityRef, userIdDomain)
                : req.query.user_id;
            if (userId) {
                await (0, provisioning_1.getOrProvisionUser)(client, tokenEntityRef, userId, provisioningEnabled, provisioningDefaults, roleConfigs, catalogClient, auth, logger);
            }
            const usage = await client.getUsage(start_date, end_date, userId, group_by);
            res.json(usage);
        }
        catch (error) {
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Failed to fetch usage', error);
            res.status(500).json({ error: error.message });
        }
    });
    // ---------------------------------------------------------------------------
    // Bridge endpoints (CLI / Abby)
    //
    // Authenticated by a Keycloak access token (JWKS-verified), NOT by Backstage's
    // own auth. Lets CLI clients list/mint virtual keys without holding the master
    // key. Gated by litellm.bridge.enabled; the verifier needs litellm.bridge.issuer.
    const bridgeCfg = (0, bridge_1.readBridgeConfig)(config);
    let tokenVerifier = options.tokenVerifier;
    if (bridgeCfg.enabled && !tokenVerifier) {
        try {
            tokenVerifier = (0, bridge_1.newDefaultVerifier)(bridgeCfg);
        }
        catch (e) {
            logger.error(`LiteLLM bridge enabled but misconfigured: ${e.message}`);
        }
    }
    if (bridgeCfg.enabled && tokenVerifier) {
        const requireClaims = async (req) => {
            const header = req.headers.authorization ?? '';
            const token = header.startsWith('Bearer ') ? header.slice(7) : '';
            if (!token)
                throw new bridge_1.BridgeAuthError('missing Bearer token');
            return tokenVerifier.verify(token);
        };
        const handleBridgeError = (error, res) => {
            if (error instanceof bridge_1.BridgeAuthError) {
                res.status(401).json({ error: 'unauthorized', hint: error.message });
                return;
            }
            if (error instanceof provisioning_1.ProvisioningError) {
                res.status(error.status).json(error.body);
                return;
            }
            logger.error('Bridge request failed', error);
            res.status(500).json({ error: error.message });
        };
        router.get('/bridge/health', (_req, res) => {
            res.json({ status: 'ok', bridge: true, clientId: bridgeCfg.clientId });
        });
        router.get('/bridge/keys', async (req, res) => {
            try {
                const claims = await requireClaims(req);
                const keys = await (0, bridge_1.bridgeListKeys)(client, claims, provisioningEnabled, provisioningDefaults, logger, userIdDomain);
                res.json(keys);
            }
            catch (error) {
                handleBridgeError(error, res);
            }
        });
        router.post('/bridge/keys', async (req, res) => {
            try {
                const claims = await requireClaims(req);
                const result = await (0, bridge_1.bridgeGenerateKey)(client, claims, provisioningEnabled, provisioningDefaults, logger, (req.body ?? {}), userIdDomain);
                res.json(result);
            }
            catch (error) {
                handleBridgeError(error, res);
            }
        });
        router.post('/bridge/keys/regenerate', async (req, res) => {
            try {
                const claims = await requireClaims(req);
                const alias = (req.body ?? {}).alias?.trim();
                if (!alias) {
                    res.status(400).json({ error: 'alias is required' });
                    return;
                }
                const result = await (0, bridge_1.bridgeRegenerateKey)(client, claims, provisioningEnabled, provisioningDefaults, logger, alias, userIdDomain);
                res.json(result);
            }
            catch (error) {
                handleBridgeError(error, res);
            }
        });
        router.get('/bridge/models', async (req, res) => {
            try {
                await requireClaims(req); // authenticate only
                const models = await client.listModels();
                res.json(models);
            }
            catch (error) {
                handleBridgeError(error, res);
            }
        });
    }
    else if (bridgeCfg.enabled) {
        logger.warn('litellm.bridge.enabled is true but no verifier could be built — bridge endpoints not mounted');
    }
    return router;
}
