// @backstage/integration/dist/azureBlobStorage/AzureBlobStorageIntegration.cjs.js
'use strict';

var helpers = require('../helpers.cjs.js');
var config = require('./config.cjs.js');

class AzureBlobStorageIntegration {
  constructor(integrationConfig) {
    this.integrationConfig = integrationConfig;
  }
  static factory = ({
    config: config$1
  }) => {
    const configs = config.readAzureBlobStorageIntegrationConfigs(
      config$1.getOptionalConfigArray("integrations.azureBlobStorage") ?? []
    );
    return helpers.basicIntegrations(
      configs.map((c) => new AzureBlobStorageIntegration(c)),
      (i) => i.config.host
    );
  };
  get type() {
    return "azureBlobStorage";
  }
  get title() {
    return this.integrationConfig.host;
  }
  get config() {
    return this.integrationConfig;
  }
  resolveUrl(options) {
    const resolved = helpers.defaultScmResolveUrl(options);
    return resolved;
  }
  resolveEditUrl(url) {
    return url;
  }
}
const AzureBlobStorageIntergation = AzureBlobStorageIntegration;

exports.AzureBlobStorageIntegration = AzureBlobStorageIntegration;
exports.AzureBlobStorageIntergation = AzureBlobStorageIntergation;
