// @backstage/integration/dist/azureBlobStorage/AzureBlobStorageIntegration.esm.js
import { basicIntegrations, defaultScmResolveUrl } from '../helpers.esm.js';
import { readAzureBlobStorageIntegrationConfigs } from './config.esm.js';

class AzureBlobStorageIntegration {
  constructor(integrationConfig) {
    this.integrationConfig = integrationConfig;
  }
  static factory = ({
    config
  }) => {
    const configs = readAzureBlobStorageIntegrationConfigs(
      config.getOptionalConfigArray("integrations.azureBlobStorage") ?? []
    );
    return basicIntegrations(
      configs.map((c) => new AzureBlobStorageIntegration(c)),
      (i) => i.config.host
    );
  };
  get type() {
    return "azureBlobStorage";
  }
  get title() {
    return this.integrationConfig.host;
  }
  get config() {
    return this.integrationConfig;
  }
  resolveUrl(options) {
    const resolved = defaultScmResolveUrl(options);
    return resolved;
  }
  resolveEditUrl(url) {
    return url;
  }
}
const AzureBlobStorageIntergation = AzureBlobStorageIntegration;

export { AzureBlobStorageIntegration, AzureBlobStorageIntergation };
