// @backstage/module-federation-common/dist/defaults.esm.js
const defaultSharedDependencies = {
  // React
  react: {
    host: {},
    remote: { import: false }
  },
  "react-dom": {
    host: {},
    remote: { import: false }
  },
  // React Router
  "react-router": {
    host: {},
    remote: { import: false }
  },
  "react-router-dom": {
    host: {},
    remote: { import: false }
  },
  // MUI v4
  // not setting import: false for MUI packages as this
  // will break once Backstage moves to BUI
  "@material-ui/core/styles": {
    host: {},
    remote: {}
  },
  "@material-ui/styles": {
    host: {},
    remote: {}
  },
  // MUI v5
  // not setting import: false for MUI packages as this
  // will break once Backstage moves to BUI
  "@mui/material/styles": {
    host: {},
    remote: {}
  },
  "@emotion/react": {
    host: {},
    remote: {}
  }
};
function defaultHostSharedDependencies() {
  return Object.fromEntries(
    Object.entries(defaultSharedDependencies).map(([name, p]) => [
      name,
      {
        eager: true,
        singleton: true,
        requiredVersion: "*",
        ...p.host
      }
    ])
  );
}
function defaultRemoteSharedDependencies() {
  return Object.fromEntries(
    Object.entries(defaultSharedDependencies).map(([name, p]) => [
      name,
      {
        eager: false,
        singleton: true,
        requiredVersion: "*",
        ...p.remote
      }
    ])
  );
}

export { defaultHostSharedDependencies, defaultRemoteSharedDependencies };
