// @backstage/cli-module-build/dist/commands/repo/build.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var chalk = require('chalk');
var cleye = require('cleye');
var node_path = require('node:path');
var packager = require('../../lib/builder/packager.cjs.js');
var cliCommon = require('@backstage/cli-common');
var cliNode = require('@backstage/cli-node');
var buildFrontend = require('../../lib/buildFrontend.cjs.js');
var buildBackend = require('../../lib/buildBackend.cjs.js');
var optionsParser = require('../../lib/optionsParser.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

var build = async ({ args, info }) => {
  const {
    flags: { all, since, minify }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        all: {
          type: Boolean,
          description: "Build all packages, including bundled app and backend packages."
        },
        since: {
          type: String,
          description: "Only build packages and their dev dependents that changed since the specified ref"
        },
        minify: {
          type: Boolean,
          description: "Minify the generated code. Does not apply to app package (app is minified by default)."
        }
      }
    },
    void 0,
    args
  );
  let packages = await cliNode.PackageGraph.listTargetPackages();
  const webpack = process.env.LEGACY_WEBPACK_BUILD ? require("webpack") : void 0;
  if (since) {
    const graph = cliNode.PackageGraph.fromPackages(packages);
    const changedPackages = await graph.listChangedPackages({
      ref: since,
      analyzeLockfile: true
    });
    const withDevDependents = graph.collectPackageNames(
      changedPackages.map((pkg) => pkg.name),
      (pkg) => pkg.localDevDependents.keys()
    );
    packages = Array.from(withDevDependents).map((name) => graph.get(name));
  }
  const apps = new Array();
  const backends = new Array();
  const parseBuildScript = optionsParser.createScriptOptionsParser(["package", "build"], {
    role: { type: "string" },
    minify: { type: "boolean" },
    "skip-build-dependencies": { type: "boolean" },
    stats: { type: "boolean" },
    config: { type: "string", multiple: true },
    "module-federation": { type: "boolean" }
  });
  const options = packages.flatMap((pkg) => {
    const role = pkg.packageJson.backstage?.role ?? cliNode.PackageRoles.detectRoleFromPackage(pkg.packageJson);
    if (!role) {
      console.warn(`Ignored ${pkg.packageJson.name} because it has no role`);
      return [];
    }
    if (role === "frontend") {
      apps.push(pkg);
      return [];
    } else if (role === "backend") {
      backends.push(pkg);
      return [];
    }
    const outputs = packager.getOutputsForRole(role);
    if (outputs.size === 0) {
      console.warn(`Ignored ${pkg.packageJson.name} because it has no output`);
      return [];
    }
    const buildOptions = parseBuildScript(pkg.packageJson.scripts?.build);
    if (!buildOptions) {
      console.warn(
        `Ignored ${pkg.packageJson.name} because it does not have a matching build script`
      );
      return [];
    }
    return {
      targetDir: pkg.dir,
      packageJson: pkg.packageJson,
      outputs,
      logPrefix: `${chalk__default.default.cyan(node_path.relative(cliCommon.targetPaths.rootDir, pkg.dir))}: `,
      workspacePackages: packages,
      minify: minify ?? Boolean(buildOptions.minify)
    };
  });
  console.log("Building packages");
  await packager.buildPackages(options);
  if (all) {
    console.log("Building apps");
    await cliNode.runConcurrentTasks({
      items: apps,
      concurrencyFactor: 1 / 2,
      worker: async (pkg) => {
        const buildOptions = parseBuildScript(pkg.packageJson.scripts?.build);
        if (!buildOptions) {
          console.warn(
            `Ignored ${pkg.packageJson.name} because it does not have a matching build script`
          );
          return;
        }
        const configPaths = buildOptions.config;
        await buildFrontend.buildFrontend({
          targetDir: pkg.dir,
          configPaths: Array.isArray(configPaths) ? configPaths : [],
          writeStats: Boolean(buildOptions.stats),
          webpack
        });
      }
    });
    console.log("Building backends");
    await cliNode.runConcurrentTasks({
      items: backends,
      concurrencyFactor: 1 / 2,
      worker: async (pkg) => {
        const buildOptions = parseBuildScript(pkg.packageJson.scripts?.build);
        if (!buildOptions) {
          console.warn(
            `Ignored ${pkg.packageJson.name} because it does not have a matching build script`
          );
          return;
        }
        await buildBackend.buildBackend({
          targetDir: pkg.dir,
          skipBuildDependencies: true,
          minify: minify ?? Boolean(buildOptions.minify)
        });
      }
    });
  }
};

exports.default = build;
