// @backstage/catalog-model/dist/model/operations/declareRelation.cjs.js
'use strict';

var v3 = require('zod/v3');

const opDeclareRelationV1Schema = v3.z.strictObject({
  op: v3.z.literal("declareRelation.v1"),
  /**
   * The kind that this relation originates from, e.g. "Component".
   */
  fromKind: v3.z.string(),
  /**
   * The technical type of the relation, e.g. "ownedBy".
   */
  type: v3.z.string(),
  /**
   * The kind that this relation points to, e.g. "Group".
   */
  toKind: v3.z.string(),
  /**
   * The properties that apply to this relation.
   */
  properties: v3.z.strictObject({
    /**
     * The technical type of the reverse relation, e.g. "ownerOf".
     */
    reverseType: v3.z.string(),
    /**
     * A human-readable title for the relation type, e.g. "owned by".
     */
    title: v3.z.string(),
    /**
     * A human-readable description of the relation.
     */
    description: v3.z.string()
  })
});
function createDeclareRelationOp(input) {
  return opDeclareRelationV1Schema.parse({
    ...input,
    op: "declareRelation.v1"
  });
}

exports.createDeclareRelationOp = createDeclareRelationOp;
exports.opDeclareRelationV1Schema = opDeclareRelationV1Schema;
