// @backstage/catalog-model/dist/model/operations/declareLabel.cjs.js
'use strict';

var v3 = require('zod/v3');
var zod = require('../jsonSchema/zod.cjs.js');

const opDeclareLabelV1Schema = v3.z.strictObject({
  op: v3.z.literal("declareLabel.v1"),
  /**
   * The name of the label, e.g. "backstage.io/source-location".
   */
  name: v3.z.string(),
  /**
   * The properties that apply to this label.
   */
  properties: v3.z.strictObject({
    /**
     * A human-readable title that can be used for display purposes instead of
     * the technical name.
     */
    title: v3.z.string().optional(),
    /**
     * A human-readable description of the label.
     */
    description: v3.z.string(),
    /**
     * The JSON schema that values of this label must conform to.
     *
     * @remarks
     *
     * If not provided, the label is assumed to be a simple string with no
     * particular schema.
     */
    schema: v3.z.strictObject({
      jsonSchema: zod.jsonSchemaSchema
    }).optional()
  })
});
function createDeclareLabelOp(input) {
  return opDeclareLabelV1Schema.parse({
    ...input,
    op: "declareLabel.v1"
  });
}

exports.createDeclareLabelOp = createDeclareLabelOp;
exports.opDeclareLabelV1Schema = opDeclareLabelV1Schema;
