// @backstage/catalog-model/dist/model/operations/updateKindVersion.cjs.js
'use strict';

var v3 = require('zod/v3');
var zod = require('../jsonSchema/zod.cjs.js');

const relationFieldSchema = v3.z.strictObject({
  /**
   * What field that shall be used to generate relations.
   *
   * @remarks
   *
   * The field value is expected to be a string or string array at runtime.
   */
  selector: v3.z.strictObject({
    /**
     * A dot separated path on the common catalog form, e.g. `spec.owner`.
     */
    path: v3.z.string()
  }),
  /**
   * The relation type that this field generates, e.g. "ownedBy".
   */
  relation: v3.z.string(),
  /**
   * If the given shorthand ref did not have a kind, use this kind as the
   * default. If no default kind is specified, the ref must contain a kind.
   */
  defaultKind: v3.z.string().optional(),
  /**
   * If the given shorthand ref did not have a namespace, either inherit the
   * namespace of the entity itself, or choose the default namespace.
   */
  defaultNamespace: v3.z.enum(["default", "inherit"]).optional(),
  /**
   * Only allow relations to be specified to the given kinds. This list must
   * include the default kind, if any. If no allowed kinds are specified,
   * all kinds are allowed.
   */
  allowedKinds: v3.z.array(v3.z.string()).optional()
});
const opUpdateKindVersionV1Schema = v3.z.strictObject({
  op: v3.z.literal("updateKindVersion.v1"),
  /**
   * The kind that this version belongs to.
   */
  kind: v3.z.string(),
  /**
   * The specific version name, e.g. "v1alpha1". This and the kind group form
   * the full apiVersion.
   */
  name: v3.z.string(),
  /**
   * The spec type that this version applies to, if any.
   *
   * @remarks
   *
   * This can be used to make kinds whose spec effectively are discriminated
   * unions. If you don't specify this, the schema will apply to a spec that has
   * no type given at all, or to those where the type is not among the set of
   * any other known declared spec types.
   */
  specType: v3.z.string().optional(),
  /**
   * The properties that apply to this version update.
   */
  properties: v3.z.strictObject({
    /**
     * A short description of this particular version (and type, where
     * applicable). Specify this if you want to override the default value.
     */
    description: v3.z.string().optional(),
    /**
     * The fields that shall be used to generate relations, if any. Specify this
     * if you want to override the default value.
     */
    relationFields: v3.z.array(relationFieldSchema).optional(),
    /**
     * The JSON schema of the version. Specify this if you want to override the
     * default value.
     *
     * @remarks
     *
     * This schema gets deep merged with the default one for this version. It
     * can therefore be used for both amending and changing existing fields.
     */
    schema: v3.z.strictObject({
      jsonSchema: zod.jsonSchemaSchema
    }).optional()
  })
});
function createUpdateKindVersionOp(input) {
  return opUpdateKindVersionV1Schema.parse({
    ...input,
    op: "updateKindVersion.v1"
  });
}

exports.createUpdateKindVersionOp = createUpdateKindVersionOp;
exports.opUpdateKindVersionV1Schema = opUpdateKindVersionV1Schema;
