// @backstage/catalog-model/dist/model/operations/updateAnnotation.cjs.js
'use strict';

var v3 = require('zod/v3');
var zod = require('../jsonSchema/zod.cjs.js');

const opUpdateAnnotationV1Schema = v3.z.strictObject({
  op: v3.z.literal("updateAnnotation.v1"),
  /**
   * The name of the annotation, e.g. "backstage.io/techdocs-ref".
   */
  name: v3.z.string(),
  /**
   * The properties that apply to this annotation.
   */
  properties: v3.z.strictObject({
    /**
     * A human-readable title that can be used for display purposes instead of
     * the technical name.
     */
    title: v3.z.string().optional(),
    /**
     * A human-readable description of the annotation.
     */
    description: v3.z.string().optional(),
    /**
     * The JSON schema that values of this annotation must conform to.
     *
     * @remarks
     *
     * If not provided, the annotation is assumed to be a simple string with no
     * particular schema.
     */
    schema: v3.z.strictObject({
      jsonSchema: zod.jsonSchemaSchema
    }).optional()
  })
});
function createUpdateAnnotationOp(input) {
  return opUpdateAnnotationV1Schema.parse({
    ...input,
    op: "updateAnnotation.v1"
  });
}

exports.createUpdateAnnotationOp = createUpdateAnnotationOp;
exports.opUpdateAnnotationV1Schema = opUpdateAnnotationV1Schema;
