// @backstage/catalog-model/dist/model/operations/updateRelation.cjs.js
'use strict';

var v3 = require('zod/v3');

const opUpdateRelationV1Schema = v3.z.strictObject({
  op: v3.z.literal("updateRelation.v1"),
  /**
   * The kind that this relation originates from, e.g. "Component".
   */
  fromKind: v3.z.string(),
  /**
   * The technical type of the relation, e.g. "ownedBy".
   */
  type: v3.z.string(),
  /**
   * The kind that this relation points to, e.g. "Group".
   */
  toKind: v3.z.string(),
  /**
   * The properties that apply to this relation.
   */
  properties: v3.z.strictObject({
    /**
     * The technical type of the reverse relation, e.g. "ownerOf".
     */
    reverseType: v3.z.string().optional(),
    /**
     * A human-readable title for the relation type, e.g. "owned by".
     */
    title: v3.z.string().optional(),
    /**
     * A human-readable description of the relation.
     */
    description: v3.z.string().optional()
  })
});
function createUpdateRelationOp(input) {
  return opUpdateRelationV1Schema.parse({
    ...input,
    op: "updateRelation.v1"
  });
}

exports.createUpdateRelationOp = createUpdateRelationOp;
exports.opUpdateRelationV1Schema = opUpdateRelationV1Schema;
