// @backstage/cli-module-actions/dist/package.json.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@backstage/cli-module-actions";
var version = "0.1.2";
var description = "CLI module for executing distributed actions";
var backstage = {
	role: "cli-module"
};
var publishConfig = {
	access: "public",
	main: "dist/index.cjs.js",
	types: "dist/index.d.ts"
};
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "packages/cli-module-actions"
};
var license = "Apache-2.0";
var main = "src/index.ts";
var types = "src/index.ts";
var bin = "bin/backstage-cli-module-actions";
var files = [
	"dist",
	"bin"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/cli-node": "workspace:^",
	"@backstage/errors": "workspace:^",
	chalk: "^4.0.0",
	cleye: "^2.6.0",
	marked: "^15.0.12",
	"marked-terminal": "^7.3.0",
	"strip-ansi": "^7.1.0",
	zod: "^3.25.76 || ^4.0.0"
};
var devDependencies = {
	"@backstage/backend-test-utils": "workspace:^",
	"@backstage/cli": "workspace:^"
};
var packageJson = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	homepage: homepage,
	repository: repository,
	license: license,
	main: main,
	types: types,
	bin: bin,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies
};

exports.backstage = backstage;
exports.bin = bin;
exports.default = packageJson;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.files = files;
exports.homepage = homepage;
exports.license = license;
exports.main = main;
exports.name = name;
exports.publishConfig = publishConfig;
exports.repository = repository;
exports.scripts = scripts;
exports.types = types;
exports.version = version;
