// @backstage/catalog-model/dist/alpha.cjs.js
'use strict';

var AiResourceEntityV1alpha1 = require('./kinds/AiResourceEntityV1alpha1.cjs.js');
var McpServerApiEntity = require('./kinds/McpServerApiEntity.cjs.js');
var compileCatalogModel = require('./model/compileCatalogModel.cjs.js');
var createCatalogModelLayer = require('./model/createCatalogModelLayer.cjs.js');
var createCatalogModelLayerBuilder = require('./model/createCatalogModelLayerBuilder.cjs.js');
var CatalogModelSources = require('./model/sources/CatalogModelSources.cjs.js');
var defaultCatalogEntityModel = require('./model/defaultCatalogEntityModel.cjs.js');



exports.aiResourceEntityModel = AiResourceEntityV1alpha1.aiResourceEntityModel;
exports.aiResourceEntityV1alpha1Validator = AiResourceEntityV1alpha1.aiResourceEntityV1alpha1Validator;
exports.isAiResourceEntity = AiResourceEntityV1alpha1.isAiResourceEntity;
exports.isRuleAiResourceEntity = AiResourceEntityV1alpha1.isRuleAiResourceEntity;
exports.isSkillAiResourceEntity = AiResourceEntityV1alpha1.isSkillAiResourceEntity;
exports.ruleAiResourceEntityV1alpha1Validator = AiResourceEntityV1alpha1.ruleAiResourceEntityV1alpha1Validator;
exports.skillAiResourceEntityV1alpha1Validator = AiResourceEntityV1alpha1.skillAiResourceEntityV1alpha1Validator;
exports.isMcpServerApiEntity = McpServerApiEntity.isMcpServerApiEntity;
exports.mcpServerApiEntityModel = McpServerApiEntity.mcpServerApiEntityModel;
exports.mcpServerApiEntityValidator = McpServerApiEntity.mcpServerApiEntityValidator;
exports.compileCatalogModel = compileCatalogModel.compileCatalogModel;
exports.createCatalogModelLayer = createCatalogModelLayer.createCatalogModelLayer;
exports.createCatalogModelLayerBuilder = createCatalogModelLayerBuilder.createCatalogModelLayerBuilder;
exports.CatalogModelSources = CatalogModelSources.CatalogModelSources;
exports.defaultCatalogEntityModel = defaultCatalogEntityModel.defaultCatalogEntityModel;
