// @backstage/catalog-model/dist/model/createCatalogModelLayerBuilder.esm.js
import { opsFromCatalogModelAnnotation } from './modelActions/addAnnotation.esm.js';
import { opsFromCatalogModelKind } from './modelActions/addKind.esm.js';
import { opsFromCatalogModelAddKindVersion } from './modelActions/addKindVersion.esm.js';
import { opsFromCatalogModelLabel } from './modelActions/addLabel.esm.js';
import { opsFromCatalogModelRelationPair } from './modelActions/addRelationPair.esm.js';
import { opsFromCatalogModelRemoveAnnotation } from './modelActions/removeAnnotation.esm.js';
import { opsFromCatalogModelRemoveKind } from './modelActions/removeKind.esm.js';
import { opsFromCatalogModelRemoveLabel } from './modelActions/removeLabel.esm.js';
import { opsFromCatalogModelRemoveTag } from './modelActions/removeTag.esm.js';
import { opsFromCatalogModelTag } from './modelActions/addTag.esm.js';
import { opsFromCatalogModelUpdateAnnotation } from './modelActions/updateAnnotation.esm.js';
import { opsFromCatalogModelUpdateKind } from './modelActions/updateKind.esm.js';
import { opsFromCatalogModelUpdateLabel } from './modelActions/updateLabel.esm.js';
import { opsFromCatalogModelUpdateRelationPair } from './modelActions/updateRelationPair.esm.js';
import { opsFromCatalogModelUpdateTag } from './modelActions/updateTag.esm.js';
import { OpaqueCatalogModelLayer } from './types.esm.js';

class DefaultCatalogModelLayerBuilder {
  #layerId;
  #ops;
  constructor(options) {
    this.#layerId = options.layerId;
    this.#ops = [];
  }
  addKind(kind) {
    const ops = opsFromCatalogModelKind(kind);
    this.#ops.push(...ops);
  }
  addKindVersion(definition) {
    const ops = opsFromCatalogModelAddKindVersion(definition);
    this.#ops.push(...ops);
  }
  updateKind(kind) {
    const ops = opsFromCatalogModelUpdateKind(kind);
    this.#ops.push(...ops);
  }
  removeKind(kind) {
    const ops = opsFromCatalogModelRemoveKind(kind);
    this.#ops.push(...ops);
  }
  addRelationPair(relation) {
    const ops = opsFromCatalogModelRelationPair(relation);
    this.#ops.push(...ops);
  }
  updateRelationPair(relation) {
    const ops = opsFromCatalogModelUpdateRelationPair(relation);
    this.#ops.push(...ops);
  }
  addAnnotation(annotation) {
    const ops = opsFromCatalogModelAnnotation(annotation);
    this.#ops.push(...ops);
  }
  updateAnnotation(annotation) {
    const ops = opsFromCatalogModelUpdateAnnotation(annotation);
    this.#ops.push(...ops);
  }
  removeAnnotation(annotation) {
    const ops = opsFromCatalogModelRemoveAnnotation(annotation);
    this.#ops.push(...ops);
  }
  addLabel(label) {
    const ops = opsFromCatalogModelLabel(label);
    this.#ops.push(...ops);
  }
  updateLabel(label) {
    const ops = opsFromCatalogModelUpdateLabel(label);
    this.#ops.push(...ops);
  }
  removeLabel(label) {
    const ops = opsFromCatalogModelRemoveLabel(label);
    this.#ops.push(...ops);
  }
  addTag(tag) {
    const ops = opsFromCatalogModelTag(tag);
    this.#ops.push(...ops);
  }
  updateTag(tag) {
    const ops = opsFromCatalogModelUpdateTag(tag);
    this.#ops.push(...ops);
  }
  removeTag(tag) {
    const ops = opsFromCatalogModelRemoveTag(tag);
    this.#ops.push(...ops);
  }
  import(layer) {
    const internal = OpaqueCatalogModelLayer.toInternal(layer);
    this.#ops.push(...internal.ops);
  }
  build() {
    return OpaqueCatalogModelLayer.createInstance("v1", {
      layerId: this.#layerId,
      ops: this.#ops.slice()
    });
  }
}
function createCatalogModelLayerBuilder(options) {
  return new DefaultCatalogModelLayerBuilder(options);
}

export { DefaultCatalogModelLayerBuilder, createCatalogModelLayerBuilder };
