// @backstage/catalog-model/dist/model/createCatalogModelLayerBuilder.cjs.js
'use strict';

var addAnnotation = require('./modelActions/addAnnotation.cjs.js');
var addKind = require('./modelActions/addKind.cjs.js');
var addKindVersion = require('./modelActions/addKindVersion.cjs.js');
var addLabel = require('./modelActions/addLabel.cjs.js');
var addRelationPair = require('./modelActions/addRelationPair.cjs.js');
var removeAnnotation = require('./modelActions/removeAnnotation.cjs.js');
var removeKind = require('./modelActions/removeKind.cjs.js');
var removeLabel = require('./modelActions/removeLabel.cjs.js');
var removeTag = require('./modelActions/removeTag.cjs.js');
var addTag = require('./modelActions/addTag.cjs.js');
var updateAnnotation = require('./modelActions/updateAnnotation.cjs.js');
var updateKind = require('./modelActions/updateKind.cjs.js');
var updateLabel = require('./modelActions/updateLabel.cjs.js');
var updateRelationPair = require('./modelActions/updateRelationPair.cjs.js');
var updateTag = require('./modelActions/updateTag.cjs.js');
var types = require('./types.cjs.js');

class DefaultCatalogModelLayerBuilder {
  #layerId;
  #ops;
  constructor(options) {
    this.#layerId = options.layerId;
    this.#ops = [];
  }
  addKind(kind) {
    const ops = addKind.opsFromCatalogModelKind(kind);
    this.#ops.push(...ops);
  }
  addKindVersion(definition) {
    const ops = addKindVersion.opsFromCatalogModelAddKindVersion(definition);
    this.#ops.push(...ops);
  }
  updateKind(kind) {
    const ops = updateKind.opsFromCatalogModelUpdateKind(kind);
    this.#ops.push(...ops);
  }
  removeKind(kind) {
    const ops = removeKind.opsFromCatalogModelRemoveKind(kind);
    this.#ops.push(...ops);
  }
  addRelationPair(relation) {
    const ops = addRelationPair.opsFromCatalogModelRelationPair(relation);
    this.#ops.push(...ops);
  }
  updateRelationPair(relation) {
    const ops = updateRelationPair.opsFromCatalogModelUpdateRelationPair(relation);
    this.#ops.push(...ops);
  }
  addAnnotation(annotation) {
    const ops = addAnnotation.opsFromCatalogModelAnnotation(annotation);
    this.#ops.push(...ops);
  }
  updateAnnotation(annotation) {
    const ops = updateAnnotation.opsFromCatalogModelUpdateAnnotation(annotation);
    this.#ops.push(...ops);
  }
  removeAnnotation(annotation) {
    const ops = removeAnnotation.opsFromCatalogModelRemoveAnnotation(annotation);
    this.#ops.push(...ops);
  }
  addLabel(label) {
    const ops = addLabel.opsFromCatalogModelLabel(label);
    this.#ops.push(...ops);
  }
  updateLabel(label) {
    const ops = updateLabel.opsFromCatalogModelUpdateLabel(label);
    this.#ops.push(...ops);
  }
  removeLabel(label) {
    const ops = removeLabel.opsFromCatalogModelRemoveLabel(label);
    this.#ops.push(...ops);
  }
  addTag(tag) {
    const ops = addTag.opsFromCatalogModelTag(tag);
    this.#ops.push(...ops);
  }
  updateTag(tag) {
    const ops = updateTag.opsFromCatalogModelUpdateTag(tag);
    this.#ops.push(...ops);
  }
  removeTag(tag) {
    const ops = removeTag.opsFromCatalogModelRemoveTag(tag);
    this.#ops.push(...ops);
  }
  import(layer) {
    const internal = types.OpaqueCatalogModelLayer.toInternal(layer);
    this.#ops.push(...internal.ops);
  }
  build() {
    return types.OpaqueCatalogModelLayer.createInstance("v1", {
      layerId: this.#layerId,
      ops: this.#ops.slice()
    });
  }
}
function createCatalogModelLayerBuilder(options) {
  return new DefaultCatalogModelLayerBuilder(options);
}

exports.DefaultCatalogModelLayerBuilder = DefaultCatalogModelLayerBuilder;
exports.createCatalogModelLayerBuilder = createCatalogModelLayerBuilder;
