// @backstage/cli-module-new/dist/packages/cli-module-new/package.json.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@backstage/cli-module-new";
var version = "0.1.4";
var description = "CLI module for Backstage CLI";
var backstage = {
	role: "cli-module"
};
var publishConfig = {
	access: "public",
	main: "dist/index.cjs.js",
	types: "dist/index.d.ts"
};
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "packages/cli-module-new"
};
var license = "Apache-2.0";
var main = "src/index.ts";
var types = "src/index.ts";
var bin = "bin/backstage-cli-module-new";
var files = [
	"dist",
	"bin",
	"templates"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/cli-common": "workspace:^",
	"@backstage/cli-node": "workspace:^",
	"@backstage/errors": "workspace:^",
	chalk: "^4.0.0",
	cleye: "^2.6.0",
	"fs-extra": "^11.2.0",
	handlebars: "^4.7.3",
	inquirer: "^8.2.0",
	lodash: "^4.17.21",
	ora: "^5.3.0",
	"recursive-readdir": "^2.2.2",
	semver: "^7.5.3",
	yaml: "^2.0.0",
	zod: "^3.25.76",
	"zod-validation-error": "^4.0.2"
};
var devDependencies = {
	"@backstage/backend-test-utils": "workspace:^",
	"@backstage/cli": "workspace:^",
	"@backstage/core-plugin-api": "workspace:^",
	"@backstage/test-utils": "workspace:^",
	"@types/fs-extra": "^11.0.0",
	"@types/inquirer": "^8.1.3",
	"@types/lodash": "^4.14.151",
	"@types/recursive-readdir": "^2.2.0"
};
var packageJson = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	homepage: homepage,
	repository: repository,
	license: license,
	main: main,
	types: types,
	bin: bin,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies
};

exports.backstage = backstage;
exports.bin = bin;
exports.default = packageJson;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.files = files;
exports.homepage = homepage;
exports.license = license;
exports.main = main;
exports.name = name;
exports.publishConfig = publishConfig;
exports.repository = repository;
exports.scripts = scripts;
exports.types = types;
exports.version = version;
