// @backstage/ui/dist/components/Slider/Slider.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { forwardRef, useId, useEffect } from 'react';
import { Slider as Slider$1, SliderOutput, SliderTrack, SliderThumb } from 'react-aria-components';
import clsx from 'clsx';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { SliderDefinition } from './definition.esm.js';

function SliderImpl(props, ref) {
  const { ownProps, restProps } = useDefinition(SliderDefinition, props);
  const { classes, className, label, secondaryLabel, description, isRequired } = ownProps;
  const labelId = useId();
  const descriptionId = useId();
  useEffect(() => {
    if (!label && !restProps["aria-label"] && !restProps["aria-labelledby"]) {
      console.warn(
        "Slider requires either a visible label, aria-label, or aria-labelledby for accessibility"
      );
    }
  }, [label, restProps]);
  const secondaryLabelText = secondaryLabel || (isRequired ? "Required" : null);
  return /* @__PURE__ */ jsxs(
    Slider$1,
    {
      className: clsx(classes.root, className),
      "aria-labelledby": label ? labelId : void 0,
      "aria-describedby": label && description ? descriptionId : void 0,
      ...restProps,
      ref,
      children: [
        label && /* @__PURE__ */ jsxs("div", { className: classes.header, children: [
          /* @__PURE__ */ jsx(
            FieldLabel,
            {
              id: labelId,
              label,
              secondaryLabel: secondaryLabelText,
              description,
              descriptionId: description ? descriptionId : void 0
            }
          ),
          /* @__PURE__ */ jsx(SliderOutput, { className: classes.output, children: ({ state }) => state.values.map((_, i) => state.getThumbValueLabel(i)).join(" \u2013 ") })
        ] }),
        /* @__PURE__ */ jsx(SliderTrack, { className: classes.track, children: ({ state }) => {
          const numThumbs = state.values.length;
          let trackFillStyle;
          if (numThumbs === 1) {
            const percent = state.getThumbPercent(0);
            const isVertical = state.orientation === "vertical";
            trackFillStyle = isVertical ? { bottom: 0, height: `${percent * 100}%` } : { left: 0, width: `${percent * 100}%` };
          } else {
            const start = state.getThumbPercent(0);
            const end = state.getThumbPercent(1);
            const rangePercent = (end - start) * 100;
            const isVertical = state.orientation === "vertical";
            trackFillStyle = isVertical ? { bottom: `${start * 100}%`, height: `${rangePercent}%` } : { left: `${start * 100}%`, width: `${rangePercent}%` };
          }
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx("div", { className: classes.trackFill, style: trackFillStyle }),
            /* @__PURE__ */ jsx(SliderThumb, { index: 0, className: classes.thumb }),
            numThumbs > 1 && /* @__PURE__ */ jsx(SliderThumb, { index: 1, className: classes.thumb })
          ] });
        } }),
        /* @__PURE__ */ jsx(FieldError, {})
      ]
    }
  );
}
const Slider = forwardRef(SliderImpl);
Slider.displayName = "Slider";

export { Slider };
