// @backstage/cli-node/dist/concurrency/runWorkerQueueThreads.cjs.js
'use strict';

var node_worker_threads = require('node:worker_threads');
var concurrency = require('./concurrency.cjs.js');

async function runWorkerQueueThreads(options) {
  const items = Array.from(options.items);
  const workerFactory = options.workerFactory;
  const workerData = options.context;
  const threadCount = Math.min(concurrency.getEnvironmentConcurrency(), items.length);
  const iterator = items[Symbol.iterator]();
  const results = new Array();
  let itemIndex = 0;
  await Promise.all(
    Array(threadCount).fill(0).map(async () => {
      const thread = new node_worker_threads.Worker(`(${workerQueueThread})(${workerFactory})`, {
        eval: true,
        workerData
      });
      return new Promise((resolve, reject) => {
        thread.on("message", (message) => {
          if (message.type === "start" || message.type === "result") {
            if (message.type === "result") {
              results[message.index] = message.result;
            }
            const { value, done } = iterator.next();
            if (done) {
              thread.postMessage({ type: "done" });
            } else {
              thread.postMessage({
                type: "item",
                index: itemIndex,
                item: value
              });
              itemIndex += 1;
            }
          } else if (message.type === "error") {
            const error = new Error(message.error.message);
            error.name = message.error.name;
            error.stack = message.error.stack;
            reject(error);
          }
        });
        thread.on("error", reject);
        thread.on("exit", (code) => {
          if (code !== 0) {
            reject(new Error(`Worker thread exited with code ${code}`));
          } else {
            resolve();
          }
        });
      });
    })
  );
  return { results };
}
function workerQueueThread(workerFuncFactory) {
  const { parentPort, workerData } = require("node:worker_threads");
  Promise.resolve().then(() => workerFuncFactory(workerData)).then(
    (workerFunc) => {
      parentPort.on("message", async (message) => {
        if (message.type === "done") {
          parentPort.close();
          return;
        }
        if (message.type === "item") {
          try {
            const result = await workerFunc(message.item);
            parentPort.postMessage({
              type: "result",
              index: message.index,
              result
            });
          } catch (error) {
            parentPort.postMessage({ type: "error", error });
          }
        }
      });
      parentPort.postMessage({ type: "start" });
    },
    (error) => parentPort.postMessage({ type: "error", error })
  );
}

exports.runWorkerQueueThreads = runWorkerQueueThreads;
