// @backstage/cli-module-auth/dist/cli-module-auth/package.json.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@backstage/cli-module-auth";
var version = "0.1.3";
var description = "CLI module for Backstage CLI";
var backstage = {
	role: "cli-module"
};
var publishConfig = {
	access: "public",
	main: "dist/index.cjs.js",
	types: "dist/index.d.ts"
};
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "packages/cli-module-auth"
};
var license = "Apache-2.0";
var main = "src/index.ts";
var types = "src/index.ts";
var bin = "bin/backstage-cli-module-auth";
var files = [
	"dist",
	"bin"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/cli-node": "workspace:^",
	"@backstage/errors": "workspace:^",
	cleye: "^2.6.0",
	"fs-extra": "^11.2.0",
	glob: "^13.0.0",
	inquirer: "^8.2.0",
	"proper-lockfile": "^4.1.2",
	yaml: "^2.0.0",
	zod: "^3.25.76"
};
var devDependencies = {
	"@backstage/backend-test-utils": "workspace:^",
	"@backstage/cli": "workspace:^",
	"@types/fs-extra": "^11.0.0",
	"@types/proper-lockfile": "^4"
};
var optionalDependencies = {
	keytar: "^7.9.0"
};
var packageJson = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	homepage: homepage,
	repository: repository,
	license: license,
	main: main,
	types: types,
	bin: bin,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	optionalDependencies: optionalDependencies
};

exports.backstage = backstage;
exports.bin = bin;
exports.default = packageJson;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.files = files;
exports.homepage = homepage;
exports.license = license;
exports.main = main;
exports.name = name;
exports.optionalDependencies = optionalDependencies;
exports.publishConfig = publishConfig;
exports.repository = repository;
exports.scripts = scripts;
exports.types = types;
exports.version = version;
