// @backstage/cli-module-build/dist/lib/builder/plugins.cjs.js
'use strict';

var fs = require('fs-extra');
var postcss = require('postcss');
var postcssImport = require('postcss-import');
var node_path = require('node:path');
var rollupPluginutils = require('rollup-pluginutils');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var postcss__default = /*#__PURE__*/_interopDefaultCompat(postcss);
var postcssImport__default = /*#__PURE__*/_interopDefaultCompat(postcssImport);

function forwardFileImports(options) {
  const filter = rollupPluginutils.createFilter(options.include, options.exclude);
  const exportedFiles = /* @__PURE__ */ new Set();
  const generatedFor = /* @__PURE__ */ new Set();
  return {
    name: "forward-file-imports",
    async generateBundle(outputOptions, bundle, isWrite) {
      if (!isWrite) {
        return;
      }
      const dir = outputOptions.dir || node_path.dirname(outputOptions.file);
      if (generatedFor.has(dir)) {
        return;
      }
      for (const output of Object.values(bundle)) {
        if (output.type !== "chunk") {
          continue;
        }
        const chunk = output;
        if (!chunk.facadeModuleId) {
          continue;
        }
        generatedFor.add(dir);
        const srcRoot = node_path.dirname(chunk.facadeModuleId);
        await Promise.all(
          Array.from(exportedFiles).map(async (exportedFile) => {
            const outputPath = node_path.relative(srcRoot, exportedFile);
            const targetFile = node_path.resolve(dir, outputPath);
            await fs__default.default.ensureDir(node_path.dirname(targetFile));
            await fs__default.default.copyFile(exportedFile, targetFile);
          })
        );
        return;
      }
    },
    options(inputOptions) {
      const treeshake = inputOptions.treeshake;
      if (treeshake !== void 0 && typeof treeshake !== "object") {
        throw new Error(
          "Expected treeshake input config to be an object or not set"
        );
      }
      if (treeshake?.moduleSideEffects) {
        throw new Error("treeshake.moduleSideEffects must not be set");
      }
      const moduleSideEffects = (id) => {
        if (filter(id)) {
          return false;
        }
        return true;
      };
      const origExternal = inputOptions.external;
      const external = (id, importer, isResolved) => {
        if (typeof origExternal === "function" && origExternal(id, importer, isResolved)) {
          return true;
        }
        if (Array.isArray(origExternal) && origExternal.includes(id)) {
          return true;
        }
        if (!filter(id)) {
          return false;
        }
        if (!importer) {
          throw new Error(`Unknown importer of file module ${id}`);
        }
        const fullId = isResolved ? id : node_path.resolve(node_path.dirname(importer), id);
        exportedFiles.add(fullId);
        return true;
      };
      return {
        ...inputOptions,
        external,
        treeshake: { ...treeshake, moduleSideEffects }
      };
    }
  };
}
function cssEntryPoints(options) {
  const cssEntries = options.entryPoints.filter((ep) => ep.ext === ".css");
  const generatedFor = /* @__PURE__ */ new Set();
  return {
    name: "backstage-css-entry-points",
    async generateBundle(outputOptions, _bundle, isWrite) {
      if (!isWrite) {
        return;
      }
      const dir = outputOptions.dir || node_path.dirname(outputOptions.file);
      if (generatedFor.has(dir)) {
        return;
      }
      generatedFor.add(dir);
      for (const entryPoint of cssEntries) {
        const sourcePath = node_path.resolve(options.targetDir, entryPoint.path);
        const outputPath = entryPoint.path.replace(/^(\.\/)?src\//, "");
        const source = await fs__default.default.readFile(sourcePath, "utf8");
        const result = await postcss__default.default([postcssImport__default.default()]).process(source, {
          from: sourcePath
        });
        this.emitFile({
          type: "asset",
          fileName: outputPath,
          source: result.css
        });
      }
    }
  };
}

exports.cssEntryPoints = cssEntryPoints;
exports.forwardFileImports = forwardFileImports;
