// @backstage/cli-module-build/dist/lib/buildFrontend.cjs.js
'use strict';

var fs = require('fs-extra');
var node_path = require('node:path');
var bundle = require('./bundler/bundle.cjs.js');
var moduleFederation = require('./bundler/moduleFederation.cjs.js');
require('chalk');
require('react-dev-utils/openBrowser');
require('@rspack/core');
require('@rspack/dev-server');
require('@backstage/cli-common');
var config = require('./config.cjs.js');
require('eslint-rspack-plugin');
require('ts-checker-rspack-plugin');
require('html-webpack-plugin');
require('react-dev-utils/ModuleScopePlugin');
require('@module-federation/enhanced/rspack');
require('lodash/pickBy');
require('yn');
require('@manypkg/get-packages');
require('./bundler/packageDetection.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

async function buildFrontend(options) {
  const { targetDir, writeStats, configPaths, webpack } = options;
  const packageJson = await fs__default.default.readJson(
    node_path.resolve(targetDir, "package.json")
  );
  await bundle.buildBundle({
    targetDir,
    entry: "src/index",
    statsJsonEnabled: writeStats,
    moduleFederationRemote: options.isModuleFederationRemote ? await moduleFederation.getModuleFederationRemoteOptions(
      packageJson,
      node_path.resolve(targetDir)
    ) : void 0,
    ...await config.loadCliConfig({
      args: configPaths,
      fromPackage: packageJson.name
    }),
    webpack
  });
}

exports.buildFrontend = buildFrontend;
