// @backstage/ui/dist/components/Link/Link.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef, useRef } from 'react';
import { useLink, useFocusRing, mergeProps } from 'react-aria';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { useResolvedHref } from '../../hooks/useResolvedHref.esm.js';
import { LinkDefinition } from './definition.esm.js';
import { getNodeText } from '../../analytics/getNodeText.esm.js';

const LinkInternal = forwardRef((props, ref) => {
  const { ownProps, restProps, dataAttributes, analytics } = useDefinition(
    LinkDefinition,
    props
  );
  const { classes, title, children } = ownProps;
  const internalRef = useRef(null);
  const linkRef = ref || internalRef;
  const { linkProps } = useLink(restProps, linkRef);
  const { isFocusVisible, focusProps } = useFocusRing();
  const resolvedHref = useResolvedHref(restProps.href);
  const handleClick = (e) => {
    linkProps.onClick?.(e);
    const text = restProps["aria-label"] ?? getNodeText(children) ?? String(restProps.href ?? "");
    analytics.captureEvent("click", text, {
      attributes: { to: String(restProps.href ?? "") }
    });
  };
  return /* @__PURE__ */ jsx(
    "a",
    {
      ...mergeProps(linkProps, focusProps),
      ...dataAttributes,
      ...restProps,
      href: resolvedHref,
      ref: linkRef,
      title,
      className: classes.root,
      "data-focus-visible": isFocusVisible || void 0,
      onClick: handleClick,
      children
    }
  );
});
LinkInternal.displayName = "LinkInternal";
const Link = forwardRef((props, ref) => {
  return /* @__PURE__ */ jsx(LinkInternal, { ...props, ref });
});
Link.displayName = "Link";

export { Link };
