// @backstage/plugin-catalog-react/dist/components/InspectEntityDialog/components/ColocatedPage.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { ANNOTATION_LOCATION, ANNOTATION_ORIGIN_LOCATION, stringifyEntityRef } from '@backstage/catalog-model';
import { Progress, ResponseErrorPanel } from '@backstage/core-components';
import { useApi } from '@backstage/core-plugin-api';
import { Text, Alert } from '@backstage/ui';
import useAsync from 'react-use/esm/useAsync';
import { catalogApiRef } from '../../../api.esm.js';
import { EntityRefLink } from '../../EntityRefLink/EntityRefLink.esm.js';
import 'react';
import 'lodash/get';
import { makeStyles } from '@material-ui/core/styles';
import { ListSection, ListItemRow } from './common.esm.js';
import { catalogReactTranslationRef } from '../../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles({
  header: {
    paddingLeft: "var(--bui-space-4)",
    marginTop: "var(--bui-space-4)",
    marginBottom: "var(--bui-space-4)"
  },
  headerLabel: {
    margin: 0,
    fontFamily: "monospace",
    fontSize: "var(--bui-font-size-3)",
    fontWeight: "var(--bui-font-weight-regular)"
  },
  entityList: {
    marginTop: "var(--bui-space-4)"
  },
  headerValue: {
    margin: 0,
    marginTop: "var(--bui-space-1)",
    fontFamily: "monospace",
    fontSize: "var(--bui-font-size-3)",
    overflow: "hidden",
    whiteSpace: "nowrap",
    textOverflow: "ellipsis"
  }
});
function useColocated(entity) {
  const catalogApi = useApi(catalogApiRef);
  const currentEntityRef = stringifyEntityRef(entity);
  const location = entity.metadata.annotations?.[ANNOTATION_LOCATION];
  const origin = entity.metadata.annotations?.[ANNOTATION_ORIGIN_LOCATION];
  const { loading, error, value } = useAsync(async () => {
    if (!location && !origin) {
      return [];
    }
    const response = await catalogApi.getEntities({
      filter: [
        ...location ? [{ [`metadata.annotations.${ANNOTATION_LOCATION}`]: location }] : [],
        ...origin ? [
          {
            [`metadata.annotations.${ANNOTATION_ORIGIN_LOCATION}`]: origin
          }
        ] : []
      ]
    });
    return response.items;
  }, [location, origin]);
  return {
    loading,
    error,
    location,
    originLocation: origin,
    colocatedEntities: value?.filter(
      (colocated) => stringifyEntityRef(colocated) !== currentEntityRef
    )
  };
}
function EntityList(props) {
  const classes = useStyles();
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    props.header && /* @__PURE__ */ jsxs("header", { className: classes.header, children: [
      /* @__PURE__ */ jsx("h4", { className: classes.headerLabel, children: props.header[0] }),
      /* @__PURE__ */ jsx("p", { className: classes.headerValue, children: props.header[1] })
    ] }),
    /* @__PURE__ */ jsx(
      ListSection,
      {
        "aria-label": t("inspectEntityDialog.colocatedPage.entityListAriaLabel"),
        className: classes.entityList,
        children: props.entities.map((entity) => /* @__PURE__ */ jsx(ListItemRow, { children: /* @__PURE__ */ jsx(EntityRefLink, { entityRef: entity }) }, stringifyEntityRef(entity)))
      }
    )
  ] });
}
function Contents(props) {
  const { entity } = props;
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const { loading, error, location, originLocation, colocatedEntities } = useColocated(entity);
  if (loading) {
    return /* @__PURE__ */ jsx(Progress, {});
  } else if (error) {
    return /* @__PURE__ */ jsx(ResponseErrorPanel, { error });
  }
  if (!location && !originLocation) {
    return /* @__PURE__ */ jsx(
      Alert,
      {
        status: "warning",
        description: t("inspectEntityDialog.colocatedPage.alertNoLocation"),
        mt: "4"
      }
    );
  } else if (!colocatedEntities?.length) {
    return /* @__PURE__ */ jsx(
      Alert,
      {
        status: "info",
        description: t("inspectEntityDialog.colocatedPage.alertNoEntity"),
        mt: "4"
      }
    );
  }
  if (location === originLocation) {
    return /* @__PURE__ */ jsx(EntityList, { entities: colocatedEntities });
  }
  const atLocation = colocatedEntities.filter(
    (e) => e.metadata.annotations?.[ANNOTATION_LOCATION] === location
  );
  const atOrigin = colocatedEntities.filter(
    (e) => e.metadata.annotations?.[ANNOTATION_ORIGIN_LOCATION] === originLocation
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    atLocation.length > 0 && /* @__PURE__ */ jsx(
      EntityList,
      {
        entities: atLocation,
        header: [
          t("inspectEntityDialog.colocatedPage.locationHeader"),
          location
        ]
      }
    ),
    atOrigin.length > 0 && /* @__PURE__ */ jsx(
      EntityList,
      {
        entities: atOrigin,
        header: [
          t("inspectEntityDialog.colocatedPage.originHeader"),
          originLocation
        ]
      }
    )
  ] });
}
function ColocatedPage(props) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Text, { as: "p", children: t("inspectEntityDialog.colocatedPage.description") }),
    /* @__PURE__ */ jsx(Contents, { entity: props.entity })
  ] });
}

export { ColocatedPage };
