// @backstage/plugin-catalog-react/dist/components/InspectEntityDialog/components/common.esm.js
import { jsx } from 'react/jsx-runtime';
import { makeStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

const useStyles = makeStyles({
  list: {
    listStyle: "none",
    margin: 0,
    padding: 0
  },
  indented: {
    paddingLeft: "var(--bui-space-4)"
  },
  listItem: {
    display: "flex",
    alignItems: "flex-start",
    marginTop: "var(--bui-space-1)",
    paddingLeft: "var(--bui-space-4)",
    fontFamily: "monospace",
    fontSize: "var(--bui-font-size-3)",
    "&:first-child": {
      marginTop: 0
    }
  }
});
function ListSection(props) {
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    "ul",
    {
      className: classNames(
        classes.list,
        props.indent && classes.indented,
        props.className
      ),
      "aria-label": props["aria-label"],
      children: props.children
    }
  );
}
function ListItemRow(props) {
  const classes = useStyles();
  return /* @__PURE__ */ jsx("li", { className: classes.listItem, children: props.children });
}

export { ListItemRow, ListSection };
