// @backstage/plugin-catalog-react/dist/components/InspectEntityDialog/components/OverviewPage.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { Link } from '@backstage/core-components';
import { Flex, TagGroup, Tag, Text, Box, Card, CardHeader, CardBody, ButtonLink, ButtonIcon, VisuallyHidden } from '@backstage/ui';
import { makeStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import { RiQuestionLine, RiCheckLine, RiFileCopyLine } from '@remixicon/react';
import groupBy from 'lodash/groupBy';
import sortBy from 'lodash/sortBy';
import { useState, useRef, useEffect } from 'react';
import { EntityRefLink } from '../../EntityRefLink/EntityRefLink.esm.js';
import { stringifyEntityRef } from '@backstage/catalog-model';
import 'lodash/get';
import { ListSection, ListItemRow } from './common.esm.js';
import { catalogReactTranslationRef } from '../../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles({
  headingWithIcon: {
    display: "flex",
    alignItems: "center"
  },
  definitionList: {
    margin: 0,
    padding: 0
  },
  definitionItem: {
    display: "flex",
    alignItems: "flex-start",
    marginTop: "var(--bui-space-4)",
    paddingLeft: "var(--bui-space-4)",
    fontFamily: "monospace",
    fontSize: "var(--bui-font-size-3)",
    "&:first-child": {
      marginTop: 0
    }
  },
  definitionContent: {
    flex: 1,
    minWidth: 0
  },
  definitionKey: {
    fontWeight: "bold"
  },
  definitionValue: {
    margin: 0,
    marginTop: "var(--bui-space-1)",
    overflow: "hidden",
    whiteSpace: "nowrap",
    textOverflow: "ellipsis"
  },
  copyAction: {
    marginLeft: "var(--bui-space-2)",
    flexShrink: 0
  },
  relationGroup: {
    "& + &": {
      marginTop: "var(--bui-space-4)"
    }
  },
  monospace: {
    fontFamily: "monospace"
  },
  sectionHeading: {
    marginTop: "var(--bui-space-3)"
  },
  metadataList: {
    marginTop: "var(--bui-space-2)"
  },
  relationList: {
    marginTop: "var(--bui-space-2)"
  },
  tagGroup: {
    marginTop: "var(--bui-space-3)",
    paddingLeft: "var(--bui-space-4)"
  }
});
function findLink(value) {
  if (value.match(/^url:https?:\/\//)) {
    return value.slice("url:".length);
  }
  if (value.match(/^https?:\/\//)) {
    return value;
  }
  return void 0;
}
function entriesToItems(entries) {
  return entries.map(([key, value]) => {
    const link = findLink(value);
    return {
      key,
      value: link ? /* @__PURE__ */ jsx(Link, { to: link, children: value }) : value
    };
  });
}
function CopyButton({ text, label }) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const [copied, setCopied] = useState(false);
  const timerRef = useRef();
  useEffect(() => () => clearTimeout(timerRef.current), []);
  const handlePress = async () => {
    try {
      await window.navigator.clipboard.writeText(text);
      setCopied(true);
      timerRef.current = setTimeout(() => setCopied(false), 2e3);
    } catch {
    }
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      ButtonIcon,
      {
        icon: copied ? /* @__PURE__ */ jsx(RiCheckLine, {}) : /* @__PURE__ */ jsx(RiFileCopyLine, {}),
        "aria-label": t("inspectEntityDialog.overviewPage.copyAriaLabel", {
          label
        }),
        variant: "tertiary",
        size: "small",
        onPress: handlePress
      }
    ),
    /* @__PURE__ */ jsx(VisuallyHidden, { role: "status", children: copied ? t("inspectEntityDialog.overviewPage.copiedStatus") : "" })
  ] });
}
function HelpIcon(props) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsx(
    ButtonLink,
    {
      href: props.to,
      target: "_blank",
      rel: "noopener noreferrer",
      variant: "tertiary",
      size: "small",
      iconStart: /* @__PURE__ */ jsx(RiQuestionLine, {}),
      "aria-label": t("inspectEntityDialog.overviewPage.helpLinkAriaLabel")
    }
  );
}
function Container(props) {
  const classes = useStyles();
  return /* @__PURE__ */ jsxs(Card, { children: [
    /* @__PURE__ */ jsx(CardHeader, { children: /* @__PURE__ */ jsxs(
      Text,
      {
        variant: "title-x-small",
        as: "h3",
        className: props.helpLink ? classes.headingWithIcon : void 0,
        children: [
          props.title,
          props.helpLink && /* @__PURE__ */ jsx(HelpIcon, { to: props.helpLink })
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(CardBody, { children: props.children })
  ] });
}
function ListSubheader(props) {
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    Text,
    {
      variant: "body-large",
      as: "h4",
      className: classNames(classes.headingWithIcon, props.className),
      children: props.children
    }
  );
}
function KeyValueList(props) {
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    "dl",
    {
      className: classNames(classes.definitionList, props.className),
      "aria-label": props["aria-label"],
      children: props.items.map((item) => /* @__PURE__ */ jsxs("div", { className: classes.definitionItem, children: [
        /* @__PURE__ */ jsxs("div", { className: classes.definitionContent, children: [
          /* @__PURE__ */ jsx("dt", { className: classes.definitionKey, children: item.key }),
          /* @__PURE__ */ jsx("dd", { className: classes.definitionValue, children: item.value })
        ] }),
        item.copyable && typeof item.value === "string" && /* @__PURE__ */ jsx("div", { className: classes.copyAction, children: /* @__PURE__ */ jsx(CopyButton, { text: item.value, label: item.key }) })
      ] }, item.key))
    }
  );
}
function OverviewPage(props) {
  const classes = useStyles();
  const {
    apiVersion,
    kind,
    metadata,
    spec,
    relations = [],
    status = {}
  } = props.entity;
  const groupedRelations = groupBy(
    sortBy(relations, (r) => r.targetRef),
    "type"
  );
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const entityRef = stringifyEntityRef(props.entity);
  return /* @__PURE__ */ jsxs(Flex, { direction: "column", gap: "4", children: [
    /* @__PURE__ */ jsx(Container, { title: t("inspectEntityDialog.overviewPage.identity.title"), children: /* @__PURE__ */ jsx(
      KeyValueList,
      {
        "aria-label": t("inspectEntityDialog.overviewPage.identity.title"),
        items: [
          { key: "apiVersion", value: apiVersion },
          { key: "kind", value: kind },
          ...spec?.type ? [{ key: "spec.type", value: spec.type.toString() }] : [],
          ...metadata.uid ? [{ key: "uid", value: metadata.uid, copyable: true }] : [],
          ...metadata.etag ? [{ key: "etag", value: metadata.etag, copyable: true }] : [],
          { key: "entityRef", value: entityRef, copyable: true }
        ]
      }
    ) }),
    /* @__PURE__ */ jsxs(Container, { title: t("inspectEntityDialog.overviewPage.metadata.title"), children: [
      !!Object.keys(metadata.annotations || {}).length && /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsxs(ListSubheader, { children: [
          t("inspectEntityDialog.overviewPage.annotations"),
          /* @__PURE__ */ jsx(HelpIcon, { to: "https://backstage.io/docs/features/software-catalog/well-known-annotations" })
        ] }),
        /* @__PURE__ */ jsx(
          KeyValueList,
          {
            items: entriesToItems(Object.entries(metadata.annotations)),
            "aria-label": t("inspectEntityDialog.overviewPage.annotations"),
            className: classes.metadataList
          }
        )
      ] }),
      !!Object.keys(metadata.labels || {}).length && /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(ListSubheader, { className: classes.sectionHeading, children: t("inspectEntityDialog.overviewPage.labels") }),
        /* @__PURE__ */ jsx(
          KeyValueList,
          {
            items: entriesToItems(Object.entries(metadata.labels)),
            "aria-label": t("inspectEntityDialog.overviewPage.labels"),
            className: classes.metadataList
          }
        )
      ] }),
      !!metadata.tags?.length && /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(ListSubheader, { className: classes.sectionHeading, children: t("inspectEntityDialog.overviewPage.tags") }),
        /* @__PURE__ */ jsx(
          TagGroup,
          {
            "aria-label": t("inspectEntityDialog.overviewPage.tags"),
            className: classes.tagGroup,
            children: metadata.tags.map((tag) => /* @__PURE__ */ jsx(Tag, { id: tag, children: tag }, tag))
          }
        )
      ] })
    ] }),
    !!relations.length && /* @__PURE__ */ jsx(
      Container,
      {
        title: t("inspectEntityDialog.overviewPage.relation.title"),
        helpLink: "https://backstage.io/docs/features/software-catalog/well-known-relations",
        children: Object.entries(groupedRelations).map(([type, groupRelations]) => /* @__PURE__ */ jsxs("div", { className: classes.relationGroup, children: [
          /* @__PURE__ */ jsx(ListSubheader, { className: classes.monospace, children: type }),
          /* @__PURE__ */ jsx(ListSection, { "aria-label": type, className: classes.relationList, children: groupRelations.map((group) => /* @__PURE__ */ jsx(ListItemRow, { children: /* @__PURE__ */ jsx(EntityRefLink, { entityRef: group.targetRef }) }, group.targetRef)) })
        ] }, type))
      }
    ),
    !!status.items?.length && /* @__PURE__ */ jsx(
      Container,
      {
        title: t("inspectEntityDialog.overviewPage.status.title"),
        helpLink: "https://backstage.io/docs/features/software-catalog/well-known-statuses",
        children: status.items.map((item, index) => /* @__PURE__ */ jsxs("div", { children: [
          /* @__PURE__ */ jsxs(Text, { children: [
            item.level,
            ": ",
            item.type
          ] }),
          /* @__PURE__ */ jsx(Box, { ml: "4", children: item.message })
        ] }, index))
      }
    )
  ] });
}

export { OverviewPage };
