// @acarmisc/backstage-plugin-litellm/dist/index.esm.js
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/api.ts
import { createApiRef } from "@backstage/core-plugin-api";
var ApiError, liteLlmApiRef, LiteLlmApi;
var init_api = __esm({
  "src/api.ts"() {
    "use strict";
    ApiError = class extends Error {
      constructor(message, status, body) {
        super(message);
        this.status = status;
        this.body = body;
      }
    };
    liteLlmApiRef = createApiRef({
      id: "plugin.litellm.api"
    });
    LiteLlmApi = class {
      constructor(fetchApi, basePath = "/api/litellm") {
        this.fetchApi = fetchApi;
        this.basePath = basePath;
      }
      async throwIfNotOk(response) {
        if (!response.ok) {
          let body;
          try {
            body = await response.json();
          } catch {
            body = await response.text().catch(() => "");
          }
          throw new ApiError(`${response.status} ${response.statusText}`, response.status, body);
        }
      }
      async get(path, params) {
        const url = new URL(`${this.basePath}${path}`, window.location.origin);
        if (params) {
          Object.entries(params).forEach(([key, value]) => url.searchParams.append(key, value));
        }
        const response = await this.fetchApi.fetch(url.toString());
        await this.throwIfNotOk(response);
        return response.json();
      }
      async post(path, body) {
        const response = await this.fetchApi.fetch(`${this.basePath}${path}`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body)
        });
        await this.throwIfNotOk(response);
        return response.json();
      }
      async del(path) {
        const response = await this.fetchApi.fetch(`${this.basePath}${path}`, {
          method: "DELETE",
          headers: { "Content-Type": "application/json" }
        });
        await this.throwIfNotOk(response);
        return response.json();
      }
      // User identity is resolved server-side from the Backstage Bearer token.
      // No user_id param needed on the frontend.
      async getUserInfo() {
        return this.get("/user/info");
      }
      async listKeys() {
        return this.get("/keys");
      }
      async generateKey(request) {
        return this.post("/keys/generate", request);
      }
      async updateKey(keyId, request) {
        return this.post(`/keys/${encodeURIComponent(keyId)}/update`, request);
      }
      async deleteKey(keyId) {
        return this.del(`/keys/${encodeURIComponent(keyId)}`);
      }
      async listModels() {
        return this.get("/models");
      }
      async getTeams() {
        return this.get("/teams");
      }
      async getUsage(startDate, endDate) {
        return this.get("/usage", { start_date: startDate, end_date: endDate });
      }
      async getTeamUsage(teamId, startDate, endDate) {
        return this.get(`/teams/${encodeURIComponent(teamId)}/usage`, {
          start_date: startDate,
          end_date: endDate
        });
      }
    };
  }
});

// src/components/DashboardHeader.tsx
import React from "react";
import { Box, Typography, LinearProgress, Paper, Chip } from "@mui/material";
var DashboardHeader;
var init_DashboardHeader = __esm({
  "src/components/DashboardHeader.tsx"() {
    "use strict";
    DashboardHeader = ({ userInfo, teams, loading }) => {
      if (loading) {
        return /* @__PURE__ */ React.createElement(Paper, { sx: { p: 2, mb: 2 } }, /* @__PURE__ */ React.createElement(LinearProgress, null));
      }
      const displayName = userInfo.user_email ?? userInfo.email ?? userInfo.user_id;
      return /* @__PURE__ */ React.createElement(Paper, { sx: { p: 2, mb: 2 } }, /* @__PURE__ */ React.createElement(Box, { display: "flex", alignItems: "flex-start", gap: 2, flexWrap: "wrap" }, /* @__PURE__ */ React.createElement(Box, { flexGrow: 1 }, /* @__PURE__ */ React.createElement(Typography, { variant: "h6" }, displayName), /* @__PURE__ */ React.createElement(Typography, { variant: "caption", color: "text.secondary" }, userInfo.user_id), teams.length > 0 && /* @__PURE__ */ React.createElement(Box, { display: "flex", gap: 0.75, flexWrap: "wrap", mt: 1 }, teams.map((team) => /* @__PURE__ */ React.createElement(
        Chip,
        {
          key: team.team_id,
          label: team.team_alias || team.team_id,
          size: "small",
          variant: "outlined",
          color: "primary"
        }
      ))))));
    };
  }
});

// src/components/KeysTable.tsx
import React2, { useState } from "react";
import {
  Paper as Paper2,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Button,
  IconButton,
  Box as Box2,
  Typography as Typography2,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  TextField,
  MenuItem,
  Chip as Chip2,
  CircularProgress,
  Autocomplete
} from "@mui/material";
import { ContentCopy, Delete, Add, Edit } from "@mui/icons-material";
var maskKey, shortKeyId, formatDate, emptyForm, keyToEditForm, KeysTable;
var init_KeysTable = __esm({
  "src/components/KeysTable.tsx"() {
    "use strict";
    maskKey = (key) => {
      if (key.length <= 8) return "***";
      return `${key.slice(0, 4)}...${key.slice(-4)}`;
    };
    shortKeyId = (token) => {
      if (!token) return "-";
      if (token.length <= 16) return token;
      return `${token.slice(0, 12)}\u2026`;
    };
    formatDate = (dateStr) => {
      try {
        return new Date(dateStr).toLocaleDateString();
      } catch {
        return dateStr;
      }
    };
    emptyForm = () => ({
      alias: "",
      models: [],
      duration: "30d",
      max_budget: 100,
      tpm_limit: void 0,
      team_id: void 0,
      key_type: "llm_api"
    });
    keyToEditForm = (k) => ({
      key_alias: k.key_alias ?? "",
      models: k.models ?? [],
      max_budget: k.max_budget,
      tpm_limit: k.tpm_limit,
      rpm_limit: k.rpm_limit
    });
    KeysTable = ({
      keys,
      models,
      teams,
      loading,
      onGenerateKey,
      onUpdateKey,
      onDeleteKey
    }) => {
      const [generateModalOpen, setGenerateModalOpen] = useState(false);
      const [newKeyValue, setNewKeyValue] = useState(null);
      const [formData, setFormData] = useState(emptyForm());
      const [submitting, setSubmitting] = useState(false);
      const canGenerate = true;
      const [editingKey, setEditingKey] = useState(null);
      const [editForm, setEditForm] = useState({});
      const [editSubmitting, setEditSubmitting] = useState(false);
      const selectedModels = models.filter((m) => (formData.models || []).includes(m.model_name));
      const selectedTeam = teams.find((t) => t.team_id === formData.team_id) ?? null;
      const editSelectedModels = models.filter((m) => (editForm.models || []).includes(m.model_name));
      const handleGenerate = async () => {
        setSubmitting(true);
        try {
          const response = await onGenerateKey(formData);
          setNewKeyValue(response.key);
          setFormData(emptyForm());
        } catch (error) {
          console.error("Failed to generate key:", error);
        } finally {
          setSubmitting(false);
        }
      };
      const handleCloseModal = () => {
        setGenerateModalOpen(false);
        setNewKeyValue(null);
        setFormData(emptyForm());
      };
      const handleOpenEdit = (k) => {
        setEditingKey(k);
        setEditForm(keyToEditForm(k));
      };
      const handleCloseEdit = () => {
        setEditingKey(null);
        setEditForm({});
      };
      const handleUpdate = async () => {
        if (!editingKey) return;
        setEditSubmitting(true);
        try {
          await onUpdateKey(editingKey.token ?? editingKey.key, editForm);
          handleCloseEdit();
        } catch (error) {
          console.error("Failed to update key:", error);
        } finally {
          setEditSubmitting(false);
        }
      };
      const copyToClipboard = (text) => {
        navigator.clipboard.writeText(text);
      };
      return /* @__PURE__ */ React2.createElement(React2.Fragment, null, /* @__PURE__ */ React2.createElement(Paper2, { sx: { mb: 2 } }, /* @__PURE__ */ React2.createElement(Box2, { display: "flex", justifyContent: "space-between", alignItems: "center", p: 2 }, /* @__PURE__ */ React2.createElement(Typography2, { variant: "h6" }, "Virtual Keys"), /* @__PURE__ */ React2.createElement(
        Button,
        {
          variant: "contained",
          color: "primary",
          startIcon: /* @__PURE__ */ React2.createElement(Add, null),
          onClick: () => setGenerateModalOpen(true)
        },
        "Generate New Key"
      )), /* @__PURE__ */ React2.createElement(TableContainer, null, /* @__PURE__ */ React2.createElement(Table, null, /* @__PURE__ */ React2.createElement(TableHead, null, /* @__PURE__ */ React2.createElement(TableRow, null, /* @__PURE__ */ React2.createElement(TableCell, null, "Alias"), /* @__PURE__ */ React2.createElement(TableCell, null, "Key ID"), /* @__PURE__ */ React2.createElement(TableCell, null, "Created"), /* @__PURE__ */ React2.createElement(TableCell, null, "Spend"), /* @__PURE__ */ React2.createElement(TableCell, null, "Budget"), /* @__PURE__ */ React2.createElement(TableCell, null, "TPM Limit"), /* @__PURE__ */ React2.createElement(TableCell, null, "Models"), /* @__PURE__ */ React2.createElement(TableCell, { align: "right" }, "Actions"))), /* @__PURE__ */ React2.createElement(TableBody, null, loading ? /* @__PURE__ */ React2.createElement(TableRow, null, /* @__PURE__ */ React2.createElement(TableCell, { colSpan: 8, align: "center" }, /* @__PURE__ */ React2.createElement(CircularProgress, { size: 24 }))) : keys.length === 0 ? /* @__PURE__ */ React2.createElement(TableRow, null, /* @__PURE__ */ React2.createElement(TableCell, { colSpan: 8, align: "center" }, /* @__PURE__ */ React2.createElement(Typography2, { color: "text.secondary" }, "No keys found"))) : keys.map((key) => {
        const keyId = key.token ?? key.key;
        return /* @__PURE__ */ React2.createElement(TableRow, { key: keyId }, /* @__PURE__ */ React2.createElement(TableCell, null, key.key_alias || "-"), /* @__PURE__ */ React2.createElement(TableCell, null, /* @__PURE__ */ React2.createElement(Box2, { display: "flex", alignItems: "center", gap: 0.5 }, /* @__PURE__ */ React2.createElement(
          Typography2,
          {
            variant: "body2",
            component: "code",
            color: "text.secondary",
            title: keyId,
            sx: {
              fontFamily: "monospace",
              backgroundColor: "background.default",
              px: 1,
              py: 0.5,
              borderRadius: 1
            }
          },
          shortKeyId(keyId)
        ), /* @__PURE__ */ React2.createElement(IconButton, { size: "small", onClick: () => copyToClipboard(keyId), title: "Copy Key ID" }, /* @__PURE__ */ React2.createElement(ContentCopy, { fontSize: "small" })))), /* @__PURE__ */ React2.createElement(TableCell, null, formatDate(key.created_at)), /* @__PURE__ */ React2.createElement(TableCell, null, "$", key.spend?.toFixed(4) || "0.00"), /* @__PURE__ */ React2.createElement(TableCell, null, key.max_budget ? `$${key.max_budget}` : "-"), /* @__PURE__ */ React2.createElement(TableCell, null, key.tpm_limit || "-"), /* @__PURE__ */ React2.createElement(TableCell, null, /* @__PURE__ */ React2.createElement(Box2, { display: "flex", gap: 0.5, flexWrap: "wrap" }, key.models?.slice(0, 2).map((model) => /* @__PURE__ */ React2.createElement(Chip2, { key: model, label: model, size: "small" })), (key.models?.length || 0) > 2 && /* @__PURE__ */ React2.createElement(Chip2, { label: `+${(key.models?.length || 0) - 2}`, size: "small", variant: "outlined" }))), /* @__PURE__ */ React2.createElement(TableCell, { align: "right" }, /* @__PURE__ */ React2.createElement(IconButton, { onClick: () => handleOpenEdit(key) }, /* @__PURE__ */ React2.createElement(Edit, { fontSize: "small" })), /* @__PURE__ */ React2.createElement(IconButton, { color: "error", onClick: () => onDeleteKey(keyId) }, /* @__PURE__ */ React2.createElement(Delete, null))));
      }))))), /* @__PURE__ */ React2.createElement(Dialog, { open: generateModalOpen, onClose: handleCloseModal, maxWidth: "sm", fullWidth: true }, /* @__PURE__ */ React2.createElement(DialogTitle, null, newKeyValue ? "Key Generated" : "Generate New Key"), /* @__PURE__ */ React2.createElement(DialogContent, null, newKeyValue ? /* @__PURE__ */ React2.createElement(Box2, null, /* @__PURE__ */ React2.createElement(Typography2, { variant: "body2", color: "text.secondary", gutterBottom: true }, "Copy this key now. You won't be able to see it again."), /* @__PURE__ */ React2.createElement(
        Box2,
        {
          display: "flex",
          alignItems: "center",
          gap: 1,
          mt: 2,
          p: 2,
          sx: {
            backgroundColor: "action.hover",
            border: "1px solid",
            borderColor: "divider",
            borderRadius: 1
          }
        },
        /* @__PURE__ */ React2.createElement(
          Typography2,
          {
            component: "code",
            color: "text.primary",
            sx: { fontFamily: "monospace", wordBreak: "break-all", flex: 1 }
          },
          newKeyValue
        ),
        /* @__PURE__ */ React2.createElement(IconButton, { onClick: () => copyToClipboard(newKeyValue) }, /* @__PURE__ */ React2.createElement(ContentCopy, null))
      )) : /* @__PURE__ */ React2.createElement(Box2, { display: "flex", flexDirection: "column", gap: 2, mt: 1 }, /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "Alias",
          value: formData.alias || "",
          onChange: (e) => setFormData({ ...formData, alias: e.target.value }),
          required: true,
          fullWidth: true
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          select: true,
          label: "Duration",
          value: formData.duration || "30d",
          onChange: (e) => setFormData({ ...formData, duration: e.target.value }),
          fullWidth: true
        },
        /* @__PURE__ */ React2.createElement(MenuItem, { value: "1d" }, "1 Day"),
        /* @__PURE__ */ React2.createElement(MenuItem, { value: "7d" }, "7 Days"),
        /* @__PURE__ */ React2.createElement(MenuItem, { value: "30d" }, "30 Days"),
        /* @__PURE__ */ React2.createElement(MenuItem, { value: "90d" }, "90 Days"),
        /* @__PURE__ */ React2.createElement(MenuItem, { value: "1y" }, "1 Year")
      ), teams.length > 0 && /* @__PURE__ */ React2.createElement(
        Autocomplete,
        {
          options: teams,
          getOptionLabel: (t) => t.team_alias || t.team_id,
          value: selectedTeam,
          onChange: (_e, team) => setFormData({ ...formData, team_id: team?.team_id }),
          renderInput: (params) => /* @__PURE__ */ React2.createElement(
            TextField,
            {
              ...params,
              label: "Team",
              helperText: "Optional: bind this key to a specific team for scoped access",
              fullWidth: true
            }
          )
        }
      ), models.length > 0 && /* @__PURE__ */ React2.createElement(
        Autocomplete,
        {
          multiple: true,
          options: models,
          groupBy: (m) => m.mode || "other",
          getOptionLabel: (m) => m.model_name,
          value: selectedModels,
          onChange: (_e, selected) => setFormData({ ...formData, models: selected.map((m) => m.model_name) }),
          renderOption: (props, m) => /* @__PURE__ */ React2.createElement("li", { ...props }, m.model_name, m.supports_function_calling && " \u{1F527}", m.supports_vision && " \u{1F441}\uFE0F"),
          renderInput: (params) => /* @__PURE__ */ React2.createElement(TextField, { ...params, label: "Models", helperText: "Leave empty to allow all models", fullWidth: true })
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "Max Budget (USD)",
          type: "number",
          value: formData.max_budget ?? "",
          onChange: (e) => setFormData({ ...formData, max_budget: e.target.value ? Number(e.target.value) : void 0 }),
          required: true,
          fullWidth: true
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "TPM Limit",
          type: "number",
          value: formData.tpm_limit ?? "",
          onChange: (e) => setFormData({ ...formData, tpm_limit: e.target.value ? Number(e.target.value) : void 0 }),
          fullWidth: true
        }
      ))), /* @__PURE__ */ React2.createElement(DialogActions, null, /* @__PURE__ */ React2.createElement(Button, { onClick: handleCloseModal }, newKeyValue ? "Done" : "Cancel"), !newKeyValue && /* @__PURE__ */ React2.createElement(
        Button,
        {
          onClick: handleGenerate,
          variant: "contained",
          color: "primary",
          disabled: submitting || !canGenerate
        },
        submitting ? /* @__PURE__ */ React2.createElement(CircularProgress, { size: 24 }) : "Generate"
      ), newKeyValue && /* @__PURE__ */ React2.createElement(Button, { onClick: handleCloseModal, variant: "contained", color: "success" }, "Done"))), /* @__PURE__ */ React2.createElement(Dialog, { open: !!editingKey, onClose: handleCloseEdit, maxWidth: "sm", fullWidth: true }, /* @__PURE__ */ React2.createElement(DialogTitle, null, "Edit Key"), /* @__PURE__ */ React2.createElement(DialogContent, null, editingKey && /* @__PURE__ */ React2.createElement(Box2, { display: "flex", flexDirection: "column", gap: 2, mt: 1 }, /* @__PURE__ */ React2.createElement(Typography2, { variant: "body2", color: "text.secondary" }, /* @__PURE__ */ React2.createElement("code", { style: { fontFamily: "monospace", color: "inherit" } }, maskKey(editingKey.key))), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "Alias",
          value: editForm.key_alias || "",
          onChange: (e) => setEditForm({ ...editForm, key_alias: e.target.value }),
          fullWidth: true
        }
      ), models.length > 0 && /* @__PURE__ */ React2.createElement(
        Autocomplete,
        {
          multiple: true,
          options: models,
          groupBy: (m) => m.mode || "other",
          getOptionLabel: (m) => m.model_name,
          value: editSelectedModels,
          onChange: (_e, selected) => setEditForm({ ...editForm, models: selected.map((m) => m.model_name) }),
          renderInput: (params) => /* @__PURE__ */ React2.createElement(TextField, { ...params, label: "Models", fullWidth: true })
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "Max Budget (USD)",
          type: "number",
          value: editForm.max_budget ?? "",
          onChange: (e) => setEditForm({ ...editForm, max_budget: e.target.value ? Number(e.target.value) : void 0 }),
          fullWidth: true
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "TPM Limit",
          type: "number",
          value: editForm.tpm_limit ?? "",
          onChange: (e) => setEditForm({ ...editForm, tpm_limit: e.target.value ? Number(e.target.value) : void 0 }),
          fullWidth: true
        }
      ), /* @__PURE__ */ React2.createElement(
        TextField,
        {
          label: "RPM Limit",
          type: "number",
          value: editForm.rpm_limit ?? "",
          onChange: (e) => setEditForm({ ...editForm, rpm_limit: e.target.value ? Number(e.target.value) : void 0 }),
          fullWidth: true
        }
      ))), /* @__PURE__ */ React2.createElement(DialogActions, null, /* @__PURE__ */ React2.createElement(Button, { onClick: handleCloseEdit }, "Cancel"), /* @__PURE__ */ React2.createElement(Button, { onClick: handleUpdate, variant: "contained", color: "primary", disabled: editSubmitting }, editSubmitting ? /* @__PURE__ */ React2.createElement(CircularProgress, { size: 24 }) : "Save"))));
    };
  }
});

// src/components/UsageStats.tsx
import React3, { useMemo, useState as useState2 } from "react";
import {
  Paper as Paper3,
  Box as Box3,
  Typography as Typography3,
  FormControl,
  InputLabel,
  Select,
  MenuItem as MenuItem2,
  Grid,
  Tabs,
  Tab,
  Table as Table2,
  TableBody as TableBody2,
  TableCell as TableCell2,
  TableContainer as TableContainer2,
  TableHead as TableHead2,
  TableRow as TableRow2,
  Chip as Chip3,
  LinearProgress as LinearProgress2,
  Skeleton
} from "@mui/material";
import {
  AreaChart,
  Area,
  BarChart,
  Bar,
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  ResponsiveContainer,
  Legend,
  ReferenceLine
} from "recharts";
function modelColor(model) {
  let h = 5381;
  for (let i = 0; i < model.length; i++) h = (h << 5) + h + model.charCodeAt(i) >>> 0;
  return MODEL_COLORS[h % MODEL_COLORS.length];
}
var TOP_N_MODELS, PERIOD_LS_KEY, MODEL_COLORS, fmtUsd, fmtInt, fmtPct, KpiCard, ChartSkeleton, EmptyChart, UsageStats;
var init_UsageStats = __esm({
  "src/components/UsageStats.tsx"() {
    "use strict";
    TOP_N_MODELS = 6;
    PERIOD_LS_KEY = "litellm_usage_period";
    MODEL_COLORS = [
      "#8884d8",
      "#82ca9d",
      "#ffc658",
      "#ff7300",
      "#0088fe",
      "#00C49F",
      "#FFBB28",
      "#FF8042",
      "#a4de6c",
      "#d0ed57"
    ];
    fmtUsd = (n) => `$${(n ?? 0).toFixed(n < 1 ? 4 : 2)}`;
    fmtInt = (n) => (n ?? 0).toLocaleString();
    fmtPct = (n) => `${(n * 100).toFixed(1)}%`;
    KpiCard = ({ label, value, hint }) => /* @__PURE__ */ React3.createElement(Paper3, { variant: "outlined", sx: { p: 2, height: "100%" } }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "caption", color: "text.secondary" }, label), /* @__PURE__ */ React3.createElement(Typography3, { variant: "h5", sx: { mt: 0.5 } }, value), hint ? /* @__PURE__ */ React3.createElement(Typography3, { variant: "caption", color: "text.secondary" }, hint) : null);
    ChartSkeleton = ({ height = 260 }) => /* @__PURE__ */ React3.createElement(Skeleton, { variant: "rectangular", height, sx: { borderRadius: 1 } });
    EmptyChart = ({
      height = 260,
      message = "No data for this period"
    }) => /* @__PURE__ */ React3.createElement(Box3, { height, display: "flex", alignItems: "center", justifyContent: "center" }, /* @__PURE__ */ React3.createElement(Typography3, { color: "text.secondary", variant: "body2" }, message));
    UsageStats = ({
      usage,
      models,
      dateRange,
      onDateRangeChange,
      loading,
      userInfo
    }) => {
      const [selectedModel, setSelectedModel] = useState2("all");
      const [tab, setTab] = useState2("costs");
      const selectedPreset = useMemo(() => {
        if (dateRange.start.toDateString() === dateRange.end.toDateString()) return "today";
        const diffMs = dateRange.end.getTime() - dateRange.start.getTime();
        const diffHours = diffMs / (1e3 * 60 * 60);
        if (diffHours <= 26) return "24h";
        const diffDays = Math.ceil(diffMs / (1e3 * 60 * 60 * 24));
        if (diffDays <= 7) return "7d";
        return "30d";
      }, [dateRange]);
      const handlePresetChange = (preset) => {
        const end = /* @__PURE__ */ new Date();
        const start = /* @__PURE__ */ new Date();
        if (preset === "today") start.setHours(0, 0, 0, 0);
        else if (preset === "24h") start.setHours(start.getHours() - 24);
        else if (preset === "7d") start.setDate(start.getDate() - 7);
        else if (preset === "30d") start.setDate(start.getDate() - 30);
        try {
          localStorage.setItem(PERIOD_LS_KEY, preset);
        } catch {
        }
        onDateRangeChange({ start, end });
      };
      const dailyData = useMemo(
        () => (usage?.daily_usage ?? []).map((d) => ({
          date: d.date,
          spend: d.spend,
          promptTokens: d.prompt_tokens,
          completionTokens: d.completion_tokens,
          totalTokens: d.total_tokens,
          apiRequests: d.api_requests,
          successfulRequests: d.successful_requests,
          failedRequests: d.failed_requests
        })),
        [usage]
      );
      const cumulativeData = useMemo(() => {
        let cum = 0;
        return dailyData.map((d) => {
          cum += d.spend;
          return { date: d.date, cumulative: cum };
        });
      }, [dailyData]);
      const successRateData = useMemo(
        () => dailyData.filter((d) => d.apiRequests > 0).map((d) => ({
          date: d.date,
          successRate: parseFloat((d.successfulRequests / d.apiRequests * 100).toFixed(1))
        })),
        [dailyData]
      );
      const { modelSpendByDate, topModels: topSpendModels } = useMemo(() => {
        const rows = usage?.daily_by_model ?? [];
        const modelTotals = {};
        for (const r of rows) modelTotals[r.model] = (modelTotals[r.model] ?? 0) + r.spend;
        const top = Object.entries(modelTotals).sort(([, a], [, b]) => b - a).slice(0, TOP_N_MODELS).map(([m]) => m);
        const dateMap = {};
        for (const r of rows) {
          if (!dateMap[r.date]) dateMap[r.date] = {};
          const bucket = top.includes(r.model) ? r.model : "Other";
          dateMap[r.date][bucket] = (dateMap[r.date][bucket] ?? 0) + r.spend;
        }
        const sorted = Object.entries(dateMap).sort(([a], [b]) => a.localeCompare(b)).map(([date, m]) => ({ date, ...m }));
        const hasOther = sorted.some((r) => r.Other > 0);
        return { modelSpendByDate: sorted, topModels: hasOther ? [...top, "Other"] : top };
      }, [usage]);
      const modelRows = useMemo(() => {
        const entries = Object.entries(usage?.usage_by_model ?? {}).map(([model, d]) => ({
          model,
          spend: d.total_spend,
          promptTokens: d.prompt_tokens,
          completionTokens: d.completion_tokens,
          totalTokens: d.total_tokens,
          apiRequests: d.api_requests,
          successfulRequests: d.successful_requests,
          failedRequests: d.failed_requests,
          successRate: d.api_requests > 0 ? d.successful_requests / d.api_requests : 0
        }));
        return selectedModel === "all" ? entries : entries.filter((e) => e.model === selectedModel);
      }, [usage, selectedModel]);
      const topModelSpendBars = useMemo(
        () => [...modelRows].sort((a, b) => b.spend - a.spend).slice(0, 10),
        [modelRows]
      );
      const keyRows = useMemo(
        () => Object.entries(usage?.usage_by_key ?? {}).map(([keyHash, d]) => ({
          keyHash,
          keyAlias: d.key_alias ?? keyHash.slice(0, 8),
          teamId: d.team_id ?? null,
          models: d.models,
          spend: d.total_spend,
          totalTokens: d.total_tokens,
          promptTokens: d.prompt_tokens,
          completionTokens: d.completion_tokens,
          apiRequests: d.api_requests,
          successfulRequests: d.successful_requests,
          failedRequests: d.failed_requests,
          successRate: d.api_requests > 0 ? d.successful_requests / d.api_requests : 0
        })),
        [usage]
      );
      const topKeySpendBars = useMemo(
        () => [...keyRows].sort((a, b) => b.spend - a.spend).slice(0, 10),
        [keyRows]
      );
      const totalRequests = usage?.api_requests ?? 0;
      const overallSuccessRate = totalRequests > 0 ? (usage?.successful_requests ?? 0) / totalRequests : 0;
      const maxBudget = userInfo?.max_budget ?? 0;
      const totalCumSpend = cumulativeData[cumulativeData.length - 1]?.cumulative ?? 0;
      return /* @__PURE__ */ React3.createElement(Paper3, { sx: { p: 2 } }, /* @__PURE__ */ React3.createElement(Box3, { display: "flex", justifyContent: "space-between", alignItems: "center", mb: 2, flexWrap: "wrap", gap: 2 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "h6" }, "Usage Analytics"), /* @__PURE__ */ React3.createElement(Box3, { display: "flex", gap: 2, alignItems: "center", flexWrap: "wrap" }, /* @__PURE__ */ React3.createElement(FormControl, { size: "small", sx: { minWidth: 140 } }, /* @__PURE__ */ React3.createElement(InputLabel, null, "Period"), /* @__PURE__ */ React3.createElement(
        Select,
        {
          value: selectedPreset,
          label: "Period",
          onChange: (e) => handlePresetChange(e.target.value)
        },
        /* @__PURE__ */ React3.createElement(MenuItem2, { value: "today" }, "Today"),
        /* @__PURE__ */ React3.createElement(MenuItem2, { value: "24h" }, "Last 24h"),
        /* @__PURE__ */ React3.createElement(MenuItem2, { value: "7d" }, "Last 7 days"),
        /* @__PURE__ */ React3.createElement(MenuItem2, { value: "30d" }, "Last 30 days")
      )), tab === "models" && /* @__PURE__ */ React3.createElement(FormControl, { size: "small", sx: { minWidth: 180 } }, /* @__PURE__ */ React3.createElement(InputLabel, null, "Model"), /* @__PURE__ */ React3.createElement(
        Select,
        {
          value: selectedModel,
          label: "Model",
          onChange: (e) => setSelectedModel(e.target.value)
        },
        /* @__PURE__ */ React3.createElement(MenuItem2, { value: "all" }, "All Models"),
        models.map((m) => /* @__PURE__ */ React3.createElement(MenuItem2, { key: m.model_name, value: m.model_name }, m.model_name))
      )))), /* @__PURE__ */ React3.createElement(Grid, { container: true, spacing: 2, sx: { mb: 2 } }, /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 6, sm: 3 }, loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, { height: 80 }) : /* @__PURE__ */ React3.createElement(KpiCard, { label: "Total Spend", value: fmtUsd(usage?.total_spend ?? 0) })), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 6, sm: 3 }, loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, { height: 80 }) : /* @__PURE__ */ React3.createElement(KpiCard, { label: "Total Requests", value: fmtInt(totalRequests), hint: `${fmtInt(usage?.failed_requests ?? 0)} failed` })), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 6, sm: 3 }, loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, { height: 80 }) : /* @__PURE__ */ React3.createElement(KpiCard, { label: "Success Rate", value: totalRequests > 0 ? fmtPct(overallSuccessRate) : "\u2014" })), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 6, sm: 3 }, loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, { height: 80 }) : /* @__PURE__ */ React3.createElement(
        KpiCard,
        {
          label: "Total Tokens",
          value: fmtInt(usage?.total_tokens ?? 0),
          hint: `${fmtInt(usage?.prompt_tokens ?? 0)} in \xB7 ${fmtInt(usage?.completion_tokens ?? 0)} out`
        }
      ))), /* @__PURE__ */ React3.createElement(Tabs, { value: tab, onChange: (_, v) => setTab(v), sx: { mb: 2 } }, /* @__PURE__ */ React3.createElement(Tab, { value: "costs", label: "Costs" }), /* @__PURE__ */ React3.createElement(Tab, { value: "models", label: "Model Activity" }), /* @__PURE__ */ React3.createElement(Tab, { value: "keys", label: "Key Activity" })), tab === "costs" && /* @__PURE__ */ React3.createElement(Grid, { container: true, spacing: 3 }, /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12, md: 6 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Daily Spend by Model"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : modelSpendByDate.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: 260 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(AreaChart, { data: modelSpendByDate }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "date", tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(YAxis, { tick: { fontSize: 12 }, tickFormatter: (v) => `$${v.toFixed(2)}` }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [`$${v.toFixed(4)}`, void 0] }), /* @__PURE__ */ React3.createElement(Legend, null), topSpendModels.map((m) => /* @__PURE__ */ React3.createElement(
        Area,
        {
          key: m,
          type: "monotone",
          dataKey: m,
          stackId: "spend",
          stroke: modelColor(m),
          fill: modelColor(m),
          fillOpacity: 0.6
        }
      )))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12, md: 6 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Daily Token Usage"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : dailyData.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: 260 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(AreaChart, { data: dailyData }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "date", tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(YAxis, { tick: { fontSize: 12 }, tickFormatter: fmtInt }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [fmtInt(v), void 0] }), /* @__PURE__ */ React3.createElement(Legend, null), /* @__PURE__ */ React3.createElement(Area, { type: "monotone", dataKey: "promptTokens", name: "Input (prompt)", stackId: "tok", stroke: "#8884d8", fill: "#8884d8", fillOpacity: 0.5 }), /* @__PURE__ */ React3.createElement(Area, { type: "monotone", dataKey: "completionTokens", name: "Output (completion)", stackId: "tok", stroke: "#82ca9d", fill: "#82ca9d", fillOpacity: 0.5 }))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12, md: 6 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Daily Requests"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : dailyData.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: 260 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(BarChart, { data: dailyData }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "date", tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(YAxis, { tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(Tooltip, null), /* @__PURE__ */ React3.createElement(Legend, null), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "successfulRequests", name: "Successful", fill: "#82ca9d", stackId: "r" }), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "failedRequests", name: "Failed", fill: "#e57373", stackId: "r" }))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12, md: 6 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Daily Success Rate"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : successRateData.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: 260 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(LineChart, { data: successRateData }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "date", tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(YAxis, { domain: [0, 100], tick: { fontSize: 12 }, tickFormatter: (v) => `${v}%` }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [`${v}%`, "Success rate"] }), /* @__PURE__ */ React3.createElement(ReferenceLine, { y: 100, stroke: "#82ca9d", strokeDasharray: "4 2" }), /* @__PURE__ */ React3.createElement(Line, { type: "monotone", dataKey: "successRate", name: "Success rate", stroke: "#8884d8", dot: { r: 3 } }))))), (maxBudget > 0 || cumulativeData.length > 0) && /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Cumulative Spend", maxBudget > 0 ? ` vs Budget (${fmtUsd(maxBudget)})` : "", maxBudget > 0 && /* @__PURE__ */ React3.createElement(Typography3, { component: "span", variant: "caption", color: "text.secondary", sx: { ml: 1 } }, fmtUsd(totalCumSpend), " used \xB7 ", fmtUsd(Math.max(0, maxBudget - totalCumSpend)), " remaining")), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, { height: 180 }) : cumulativeData.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, { height: 180 }) : /* @__PURE__ */ React3.createElement(Box3, { height: 180 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(AreaChart, { data: cumulativeData }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "date", tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(YAxis, { tick: { fontSize: 12 }, tickFormatter: (v) => `$${v.toFixed(2)}` }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [`$${v.toFixed(4)}`, "Cumulative spend"] }), maxBudget > 0 && /* @__PURE__ */ React3.createElement(ReferenceLine, { y: maxBudget, stroke: "#e57373", strokeDasharray: "6 3", label: { value: `Budget ${fmtUsd(maxBudget)}`, position: "insideTopRight", fontSize: 11 } }), /* @__PURE__ */ React3.createElement(Area, { type: "monotone", dataKey: "cumulative", name: "Cumulative spend", stroke: "#ffc658", fill: "#ffc658", fillOpacity: 0.3 })))))), tab === "models" && /* @__PURE__ */ React3.createElement(Grid, { container: true, spacing: 3 }, /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Spend by Model"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : topModelSpendBars.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: Math.max(200, topModelSpendBars.length * 36) }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(BarChart, { data: topModelSpendBars, layout: "vertical", margin: { left: 20 } }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { type: "number", tick: { fontSize: 12 }, tickFormatter: (v) => `$${v.toFixed(2)}` }), /* @__PURE__ */ React3.createElement(YAxis, { type: "category", dataKey: "model", tick: { fontSize: 11 }, width: 200 }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [fmtUsd(v), "Spend"] }), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "spend", name: "Spend", radius: [0, 4, 4, 0] }, topModelSpendBars.map((r) => /* @__PURE__ */ React3.createElement("rect", { key: r.model, fill: modelColor(r.model) }))))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Tokens by Model"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : modelRows.length === 0 ? /* @__PURE__ */ React3.createElement(EmptyChart, null) : /* @__PURE__ */ React3.createElement(Box3, { height: 260 }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(BarChart, { data: modelRows }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { dataKey: "model", tick: { fontSize: 10 }, interval: 0, angle: -15, textAnchor: "end", height: 60 }), /* @__PURE__ */ React3.createElement(YAxis, { tick: { fontSize: 12 } }), /* @__PURE__ */ React3.createElement(Tooltip, null), /* @__PURE__ */ React3.createElement(Legend, null), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "promptTokens", name: "Prompt", fill: "#8884d8", stackId: "t" }), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "completionTokens", name: "Completion", fill: "#82ca9d", stackId: "t" }))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(TableContainer2, { component: Paper3, variant: "outlined" }, /* @__PURE__ */ React3.createElement(Table2, { size: "small" }, /* @__PURE__ */ React3.createElement(TableHead2, null, /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, null, "Model"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Spend"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Requests"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Success"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Failed"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Prompt"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Completion"), /* @__PURE__ */ React3.createElement(TableCell2, { sx: { minWidth: 120 } }, "Success rate"))), /* @__PURE__ */ React3.createElement(TableBody2, null, loading ? /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, { colSpan: 8 }, /* @__PURE__ */ React3.createElement(LinearProgress2, null))) : modelRows.length === 0 ? /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, { colSpan: 8, align: "center" }, "No model activity")) : [...modelRows].sort((a, b) => b.spend - a.spend || b.totalTokens - a.totalTokens).map((r) => /* @__PURE__ */ React3.createElement(TableRow2, { key: r.model }, /* @__PURE__ */ React3.createElement(TableCell2, null, /* @__PURE__ */ React3.createElement(Box3, { display: "flex", alignItems: "center", gap: 0.75 }, /* @__PURE__ */ React3.createElement(Box3, { sx: { width: 10, height: 10, borderRadius: "50%", bgcolor: modelColor(r.model), flexShrink: 0 } }), r.model)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtUsd(r.spend)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.apiRequests)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.successfulRequests)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.failedRequests)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.promptTokens)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.completionTokens)), /* @__PURE__ */ React3.createElement(TableCell2, null, r.apiRequests > 0 ? /* @__PURE__ */ React3.createElement(Box3, { display: "flex", alignItems: "center", gap: 1 }, /* @__PURE__ */ React3.createElement(
        LinearProgress2,
        {
          variant: "determinate",
          value: r.successRate * 100,
          sx: { flex: 1, height: 6, borderRadius: 3 }
        }
      ), /* @__PURE__ */ React3.createElement(Typography3, { variant: "caption" }, fmtPct(r.successRate))) : "\u2014")))))))), tab === "keys" && /* @__PURE__ */ React3.createElement(Grid, { container: true, spacing: 3 }, (loading || topKeySpendBars.length > 0) && /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(Typography3, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Spend by Key"), loading ? /* @__PURE__ */ React3.createElement(ChartSkeleton, null) : /* @__PURE__ */ React3.createElement(Box3, { height: Math.max(160, topKeySpendBars.length * 36) }, /* @__PURE__ */ React3.createElement(ResponsiveContainer, { width: "100%", height: "100%" }, /* @__PURE__ */ React3.createElement(BarChart, { data: topKeySpendBars, layout: "vertical", margin: { left: 20 } }, /* @__PURE__ */ React3.createElement(CartesianGrid, { strokeDasharray: "3 3" }), /* @__PURE__ */ React3.createElement(XAxis, { type: "number", tick: { fontSize: 12 }, tickFormatter: (v) => `$${v.toFixed(2)}` }), /* @__PURE__ */ React3.createElement(YAxis, { type: "category", dataKey: "keyAlias", tick: { fontSize: 11 }, width: 160 }), /* @__PURE__ */ React3.createElement(Tooltip, { formatter: (v) => [fmtUsd(v), "Spend"] }), /* @__PURE__ */ React3.createElement(Bar, { dataKey: "spend", name: "Spend", fill: "#8884d8", radius: [0, 4, 4, 0] }))))), /* @__PURE__ */ React3.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React3.createElement(TableContainer2, { component: Paper3, variant: "outlined" }, /* @__PURE__ */ React3.createElement(Table2, { size: "small" }, /* @__PURE__ */ React3.createElement(TableHead2, null, /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, null, "Key"), /* @__PURE__ */ React3.createElement(TableCell2, null, "Models"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Spend"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Requests"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Tokens"), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, "Failed"), /* @__PURE__ */ React3.createElement(TableCell2, { sx: { minWidth: 120 } }, "Success rate"))), /* @__PURE__ */ React3.createElement(TableBody2, null, loading ? /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, { colSpan: 7 }, /* @__PURE__ */ React3.createElement(LinearProgress2, null))) : keyRows.length === 0 ? /* @__PURE__ */ React3.createElement(TableRow2, null, /* @__PURE__ */ React3.createElement(TableCell2, { colSpan: 7, align: "center" }, "No key activity")) : [...keyRows].sort((a, b) => b.spend - a.spend || b.apiRequests - a.apiRequests).map((r) => /* @__PURE__ */ React3.createElement(TableRow2, { key: r.keyHash }, /* @__PURE__ */ React3.createElement(TableCell2, null, /* @__PURE__ */ React3.createElement(Typography3, { variant: "body2" }, r.keyAlias), r.teamId ? /* @__PURE__ */ React3.createElement(Typography3, { variant: "caption", color: "text.secondary" }, "team: ", r.teamId) : null), /* @__PURE__ */ React3.createElement(TableCell2, null, /* @__PURE__ */ React3.createElement(Box3, { display: "flex", gap: 0.5, flexWrap: "wrap" }, r.models.map((m) => /* @__PURE__ */ React3.createElement(Chip3, { key: m, label: m, size: "small", variant: "outlined" })))), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtUsd(r.spend)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.apiRequests)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.totalTokens)), /* @__PURE__ */ React3.createElement(TableCell2, { align: "right" }, fmtInt(r.failedRequests)), /* @__PURE__ */ React3.createElement(TableCell2, null, r.apiRequests > 0 ? /* @__PURE__ */ React3.createElement(Box3, { display: "flex", alignItems: "center", gap: 1 }, /* @__PURE__ */ React3.createElement(
        LinearProgress2,
        {
          variant: "determinate",
          value: r.successRate * 100,
          sx: { flex: 1, height: 6, borderRadius: 3 }
        }
      ), /* @__PURE__ */ React3.createElement(Typography3, { variant: "caption" }, fmtPct(r.successRate))) : "\u2014")))))))));
    };
  }
});

// src/components/TeamUsage.tsx
import React4, { useState as useState3 } from "react";
import {
  Paper as Paper4,
  Box as Box4,
  Typography as Typography4,
  LinearProgress as LinearProgress3,
  Chip as Chip4,
  Table as Table3,
  TableBody as TableBody3,
  TableCell as TableCell3,
  TableHead as TableHead3,
  TableRow as TableRow3,
  TableContainer as TableContainer3,
  Collapse,
  IconButton as IconButton2,
  CircularProgress as CircularProgress2
} from "@mui/material";
import { ExpandMore, ExpandLess, Group } from "@mui/icons-material";
import {
  AreaChart as AreaChart2,
  Area as Area2,
  XAxis as XAxis2,
  YAxis as YAxis2,
  CartesianGrid as CartesianGrid2,
  Tooltip as Tooltip2,
  ResponsiveContainer as ResponsiveContainer2
} from "recharts";
var TeamCard, TeamUsage;
var init_TeamUsage = __esm({
  "src/components/TeamUsage.tsx"() {
    "use strict";
    TeamCard = ({ team, usage, usageLoading }) => {
      const [expanded, setExpanded] = useState3(false);
      const budget = team.max_budget ?? 0;
      const spend = team.spend ?? 0;
      const budgetPct = budget > 0 ? Math.min(spend / budget * 100, 100) : 0;
      const isOver = budget > 0 && spend >= budget;
      const isNear = budget > 0 && spend >= budget * 0.8 && !isOver;
      const dailyData = usage?.daily_usage?.map((d) => ({ date: d.date, spend: d.spend })) ?? [];
      return /* @__PURE__ */ React4.createElement(Paper4, { variant: "outlined", sx: { p: 2, mb: 2 } }, /* @__PURE__ */ React4.createElement(Box4, { display: "flex", alignItems: "center", gap: 1 }, /* @__PURE__ */ React4.createElement(Group, { color: "action" }), /* @__PURE__ */ React4.createElement(Box4, { flexGrow: 1 }, /* @__PURE__ */ React4.createElement(Typography4, { variant: "subtitle1", fontWeight: 600 }, team.team_alias ?? team.team_id), team.models?.length ? /* @__PURE__ */ React4.createElement(Box4, { display: "flex", gap: 0.5, flexWrap: "wrap", mt: 0.5 }, team.models.map((m) => /* @__PURE__ */ React4.createElement(Chip4, { key: m, label: m, size: "small", variant: "outlined" }))) : /* @__PURE__ */ React4.createElement(Typography4, { variant: "caption", color: "text.secondary" }, "All models")), /* @__PURE__ */ React4.createElement(IconButton2, { size: "small", onClick: () => setExpanded((e) => !e) }, expanded ? /* @__PURE__ */ React4.createElement(ExpandLess, null) : /* @__PURE__ */ React4.createElement(ExpandMore, null))), budget > 0 && /* @__PURE__ */ React4.createElement(Box4, { mt: 1.5 }, /* @__PURE__ */ React4.createElement(Box4, { display: "flex", justifyContent: "space-between", mb: 0.5 }, /* @__PURE__ */ React4.createElement(Typography4, { variant: "body2" }, "$", spend.toFixed(2), " / $", budget.toFixed(2)), isOver && /* @__PURE__ */ React4.createElement(Chip4, { label: "Over Budget", size: "small", color: "error" }), isNear && /* @__PURE__ */ React4.createElement(Chip4, { label: "Near Limit", size: "small", color: "warning" })), /* @__PURE__ */ React4.createElement(
        LinearProgress3,
        {
          variant: "determinate",
          value: budgetPct,
          color: isOver ? "error" : isNear ? "warning" : "primary",
          sx: { height: 6, borderRadius: 1 }
        }
      )), /* @__PURE__ */ React4.createElement(Collapse, { in: expanded }, /* @__PURE__ */ React4.createElement(Box4, { mt: 2 }, usageLoading ? /* @__PURE__ */ React4.createElement(Box4, { display: "flex", justifyContent: "center", p: 2 }, /* @__PURE__ */ React4.createElement(CircularProgress2, { size: 24 })) : dailyData.length > 0 ? /* @__PURE__ */ React4.createElement(React4.Fragment, null, /* @__PURE__ */ React4.createElement(Typography4, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Daily Spend"), /* @__PURE__ */ React4.createElement(Box4, { height: 160 }, /* @__PURE__ */ React4.createElement(ResponsiveContainer2, { width: "100%", height: "100%" }, /* @__PURE__ */ React4.createElement(AreaChart2, { data: dailyData }, /* @__PURE__ */ React4.createElement(CartesianGrid2, { strokeDasharray: "3 3" }), /* @__PURE__ */ React4.createElement(XAxis2, { dataKey: "date", tick: { fontSize: 11 } }), /* @__PURE__ */ React4.createElement(YAxis2, { tick: { fontSize: 11 }, tickFormatter: (v) => `$${v.toFixed(2)}` }), /* @__PURE__ */ React4.createElement(Tooltip2, { formatter: (v) => [`$${v.toFixed(4)}`, "Spend"] }), /* @__PURE__ */ React4.createElement(Area2, { type: "monotone", dataKey: "spend", stroke: "#8884d8", fill: "#8884d8", fillOpacity: 0.3 }))))) : null, team.members_with_roles?.length ? /* @__PURE__ */ React4.createElement(Box4, { mt: 2 }, /* @__PURE__ */ React4.createElement(Typography4, { variant: "subtitle2", color: "text.secondary", gutterBottom: true }, "Members"), /* @__PURE__ */ React4.createElement(TableContainer3, null, /* @__PURE__ */ React4.createElement(Table3, { size: "small" }, /* @__PURE__ */ React4.createElement(TableHead3, null, /* @__PURE__ */ React4.createElement(TableRow3, null, /* @__PURE__ */ React4.createElement(TableCell3, null, "User"), /* @__PURE__ */ React4.createElement(TableCell3, null, "Role"))), /* @__PURE__ */ React4.createElement(TableBody3, null, team.members_with_roles.map((m) => /* @__PURE__ */ React4.createElement(TableRow3, { key: m.user_id }, /* @__PURE__ */ React4.createElement(TableCell3, { sx: { fontFamily: "monospace", fontSize: 12 } }, m.user_id), /* @__PURE__ */ React4.createElement(TableCell3, null, /* @__PURE__ */ React4.createElement(
        Chip4,
        {
          label: m.role,
          size: "small",
          color: m.role === "admin" ? "primary" : "default"
        }
      )))))))) : null)));
    };
    TeamUsage = ({
      teams,
      loading,
      getTeamUsage,
      getTeamUsageLoading
    }) => {
      if (loading) {
        return /* @__PURE__ */ React4.createElement(Paper4, { sx: { p: 2 } }, /* @__PURE__ */ React4.createElement(LinearProgress3, null));
      }
      if (!teams.length) {
        return /* @__PURE__ */ React4.createElement(Paper4, { sx: { p: 2 } }, /* @__PURE__ */ React4.createElement(Box4, { display: "flex", alignItems: "center", gap: 1 }, /* @__PURE__ */ React4.createElement(Group, { color: "disabled" }), /* @__PURE__ */ React4.createElement(Typography4, { color: "text.secondary", variant: "body2" }, "No team membership found in LiteLLM for this account.")));
      }
      return /* @__PURE__ */ React4.createElement(Box4, null, /* @__PURE__ */ React4.createElement(Typography4, { variant: "h6", mb: 1 }, "Teams"), teams.map((team) => /* @__PURE__ */ React4.createElement(
        TeamCard,
        {
          key: team.team_id,
          team,
          usage: getTeamUsage(team.team_id),
          usageLoading: getTeamUsageLoading(team.team_id)
        }
      )));
    };
  }
});

// src/components/LiteLLMPage.tsx
var LiteLLMPage_exports = {};
__export(LiteLLMPage_exports, {
  LiteLLMPage: () => LiteLLMPage
});
import React5, { useState as useState4, useCallback, useMemo as useMemo2 } from "react";
import { Grid as Grid2, Box as Box5, Snackbar, Alert, CircularProgress as CircularProgress3, Typography as Typography5, Paper as Paper5 } from "@mui/material";
import { useAsync, useAsyncRetry } from "react-use";
import { useApi } from "@backstage/core-plugin-api";
function initDateRange() {
  let preset = "7d";
  try {
    preset = localStorage.getItem(PERIOD_LS_KEY2) ?? "7d";
  } catch {
  }
  const end = /* @__PURE__ */ new Date();
  const start = /* @__PURE__ */ new Date();
  if (preset === "today") start.setHours(0, 0, 0, 0);
  else if (preset === "24h") start.setHours(start.getHours() - 24);
  else if (preset === "30d") start.setDate(start.getDate() - 30);
  else start.setDate(start.getDate() - 7);
  return { start, end };
}
var PERIOD_LS_KEY2, LiteLLMPage;
var init_LiteLLMPage = __esm({
  "src/components/LiteLLMPage.tsx"() {
    "use strict";
    init_DashboardHeader();
    init_KeysTable();
    init_UsageStats();
    init_TeamUsage();
    init_api();
    PERIOD_LS_KEY2 = "litellm_usage_period";
    LiteLLMPage = () => {
      const api = useApi(liteLlmApiRef);
      const [dateRange, setDateRange] = useState4(initDateRange);
      const [snackbar, setSnackbar] = useState4(null);
      const [teamUsageCache, setTeamUsageCache] = useState4({});
      const [teamUsageLoading, setTeamUsageLoading] = useState4({});
      const { value: userInfo, loading: userLoading, error: userError } = useAsync(
        () => api.getUserInfo(),
        [api]
      );
      const { value: keys, loading: keysLoading, retry: refreshKeys } = useAsyncRetry(
        async () => {
          try {
            return await api.listKeys();
          } catch (e) {
            setSnackbar({ message: `Failed to load keys: ${e.message}`, severity: "error" });
            return [];
          }
        },
        [api]
      );
      const { value: allModels, loading: modelsLoading } = useAsync(
        () => api.listModels().catch(() => []),
        [api]
      );
      const { value: allTeams, loading: teamsLoading } = useAsync(
        () => api.getTeams().catch(() => []),
        [api]
      );
      const teams = useMemo2(() => {
        if (!allTeams?.length) return [];
        if (!userInfo) return allTeams;
        const userId = userInfo.user_id;
        if (userInfo.teams?.length) {
          return allTeams.filter((t) => userInfo.teams.includes(t.team_id));
        }
        const byMembership = allTeams.filter(
          (t) => t.members_with_roles?.some((m) => m.user_id === userId)
        );
        return byMembership.length > 0 ? byMembership : allTeams;
      }, [allTeams, userInfo]);
      const { value: usage, loading: usageLoading } = useAsync(async () => {
        const startDate = dateRange.start.toISOString().split("T")[0];
        const endDate = dateRange.end.toISOString().split("T")[0];
        return api.getUsage(startDate, endDate);
      }, [api, dateRange]);
      const loadTeamUsage = useCallback(async (teamId) => {
        if (teamUsageCache[teamId] !== void 0 || teamUsageLoading[teamId]) return;
        setTeamUsageLoading((prev) => ({ ...prev, [teamId]: true }));
        try {
          const startDate = dateRange.start.toISOString().split("T")[0];
          const endDate = dateRange.end.toISOString().split("T")[0];
          const data = await api.getTeamUsage(teamId, startDate, endDate);
          setTeamUsageCache((prev) => ({ ...prev, [teamId]: data }));
        } catch {
          setTeamUsageCache((prev) => ({ ...prev, [teamId]: null }));
        } finally {
          setTeamUsageLoading((prev) => ({ ...prev, [teamId]: false }));
        }
      }, [api, dateRange, teamUsageCache, teamUsageLoading]);
      const allowedModels = useMemo2(() => {
        if (!allModels?.length) return [];
        const userModels = userInfo?.models;
        const teamModels = teams?.flatMap((t) => t.models ?? []);
        const hasUserRestriction = userModels && userModels.length > 0;
        const hasTeamRestriction = teamModels && teamModels.length > 0;
        if (!hasUserRestriction && !hasTeamRestriction) return allModels;
        const allowed = /* @__PURE__ */ new Set([
          ...hasUserRestriction ? userModels : allModels.map((m) => m.model_name),
          ...hasTeamRestriction ? teamModels : []
        ]);
        return allModels.filter((m) => allowed.has(m.model_name));
      }, [allModels, userInfo, teams]);
      const handleGenerateKey = useCallback(
        async (request) => {
          const response = await api.generateKey(request);
          setSnackbar({ message: "Key generated successfully", severity: "success" });
          refreshKeys();
          return response;
        },
        [api, refreshKeys]
      );
      const handleUpdateKey = useCallback(
        async (keyId, request) => {
          try {
            await api.updateKey(keyId, request);
            setSnackbar({ message: "Key updated successfully", severity: "success" });
            refreshKeys();
          } catch (e) {
            setSnackbar({ message: `Failed to update key: ${e.message}`, severity: "error" });
            throw e;
          }
        },
        [api, refreshKeys]
      );
      const handleDeleteKey = useCallback(
        async (keyId) => {
          try {
            await api.deleteKey(keyId);
            setSnackbar({
              message: "Key revoked successfully",
              severity: "success"
            });
            refreshKeys();
          } catch (e) {
            if (e.body?.success && (e.body?.message?.includes("already deleted") || e.body?.message?.includes("never existed"))) {
              setSnackbar({
                message: "Key was already deleted",
                severity: "warning"
              });
              refreshKeys();
              return;
            }
            setSnackbar({
              message: `Failed to revoke key: ${e.message}`,
              severity: "error"
            });
          } finally {
            refreshKeys();
          }
        },
        [api, refreshKeys]
      );
      const isInitialLoading = userLoading && !userInfo;
      if (isInitialLoading) {
        return /* @__PURE__ */ React5.createElement(Box5, { display: "flex", justifyContent: "center", alignItems: "center", minHeight: "50vh" }, /* @__PURE__ */ React5.createElement(CircularProgress3, null));
      }
      if (userError || !userInfo) {
        const isProvisioningEnabled = userError?.body?.provisioning === true;
        const hint = userError?.body?.hint;
        return /* @__PURE__ */ React5.createElement(Box5, { p: 3 }, /* @__PURE__ */ React5.createElement(Paper5, { sx: { p: 3 } }, /* @__PURE__ */ React5.createElement(Typography5, { variant: "h6", gutterBottom: true }, "Account not provisioned"), /* @__PURE__ */ React5.createElement(Typography5, { color: "text.secondary", paragraph: true }, "Your Backstage account is not linked to a LiteLLM user."), hint ? /* @__PURE__ */ React5.createElement(Typography5, { variant: "body2", color: "text.secondary" }, hint) : /* @__PURE__ */ React5.createElement(Typography5, { variant: "body2", color: "text.secondary" }, isProvisioningEnabled ? "Auto-provisioning is enabled but failed. Check the backend logs." : "Set litellm.provisioning.enabled: true in app-config.yaml to enable auto-provisioning, or ask your administrator to create the account manually.")));
      }
      return /* @__PURE__ */ React5.createElement(Box5, { p: 3 }, /* @__PURE__ */ React5.createElement(Grid2, { container: true, spacing: 2 }, /* @__PURE__ */ React5.createElement(Grid2, { item: true, xs: 12 }, /* @__PURE__ */ React5.createElement(DashboardHeader, { userInfo, teams: teams ?? [], loading: userLoading || teamsLoading })), /* @__PURE__ */ React5.createElement(Grid2, { item: true, xs: 12 }, /* @__PURE__ */ React5.createElement(
        TeamUsage,
        {
          teams: teams ?? [],
          loading: teamsLoading,
          dateRange,
          getTeamUsage: (teamId) => {
            if (teamUsageCache[teamId] === void 0) loadTeamUsage(teamId);
            return teamUsageCache[teamId] ?? null;
          },
          getTeamUsageLoading: (teamId) => teamUsageLoading[teamId] ?? false
        }
      )), /* @__PURE__ */ React5.createElement(Grid2, { item: true, xs: 12 }, /* @__PURE__ */ React5.createElement(
        KeysTable,
        {
          keys: keys ?? [],
          models: allowedModels,
          teams: teams ?? [],
          loading: keysLoading || modelsLoading,
          onGenerateKey: handleGenerateKey,
          onUpdateKey: handleUpdateKey,
          onDeleteKey: handleDeleteKey
        }
      )), /* @__PURE__ */ React5.createElement(Grid2, { item: true, xs: 12 }, /* @__PURE__ */ React5.createElement(
        UsageStats,
        {
          usage: usage ?? null,
          models: allModels ?? [],
          dateRange,
          onDateRangeChange: setDateRange,
          loading: usageLoading,
          userInfo
        }
      ))), /* @__PURE__ */ React5.createElement(
        Snackbar,
        {
          open: !!snackbar,
          autoHideDuration: 5e3,
          onClose: () => setSnackbar(null),
          anchorOrigin: { vertical: "bottom", horizontal: "right" }
        },
        snackbar ? /* @__PURE__ */ React5.createElement(Alert, { severity: snackbar.severity, onClose: () => setSnackbar(null) }, snackbar.message) : void 0
      ));
    };
  }
});

// src/plugin.tsx
init_api();
import React6 from "react";
import { TrendingUp as TrendingUpIcon } from "@mui/icons-material";
import {
  createFrontendPlugin,
  ApiBlueprint,
  PageBlueprint,
  fetchApiRef
} from "@backstage/frontend-plugin-api";
var liteLlmApi = ApiBlueprint.make({
  params: (defineParams) => defineParams({
    api: liteLlmApiRef,
    deps: { fetchApi: fetchApiRef },
    factory: ({ fetchApi }) => new LiteLlmApi(fetchApi)
  })
});
var liteLlmPage = PageBlueprint.make({
  params: {
    path: "/litellm",
    title: "LiteLLM",
    icon: /* @__PURE__ */ React6.createElement(TrendingUpIcon, null),
    loader: async () => {
      const { LiteLLMPage: LiteLLMPage2 } = await Promise.resolve().then(() => (init_LiteLLMPage(), LiteLLMPage_exports));
      return /* @__PURE__ */ React6.createElement(LiteLLMPage2, null);
    }
  }
});
var litellmPlugin = createFrontendPlugin({
  pluginId: "litellm",
  extensions: [liteLlmApi, liteLlmPage]
});

// src/index.ts
init_LiteLLMPage();
init_DashboardHeader();
init_KeysTable();
init_UsageStats();
init_TeamUsage();

// src/components/LiteLLMHomeWidget.tsx
init_api();
import React7, { useState as useState5, useEffect } from "react";
import {
  Paper as Paper6,
  Box as Box6,
  Typography as Typography6,
  FormControl as FormControl2,
  Select as Select2,
  MenuItem as MenuItem3,
  Grid as Grid3,
  CircularProgress as CircularProgress4,
  Alert as Alert2
} from "@mui/material";
import { AreaChart as AreaChart3, Area as Area3, ResponsiveContainer as ResponsiveContainer3 } from "recharts";
import { useApi as useApi2 } from "@backstage/core-plugin-api";
var fmtUsd2 = (n) => `$${(n ?? 0).toFixed(n < 1 ? 4 : 2)}`;
var fmtInt2 = (n) => (n ?? 0).toLocaleString();
function presetToDateRange(preset) {
  const end = /* @__PURE__ */ new Date();
  const start = /* @__PURE__ */ new Date();
  if (preset === "today") {
    start.setHours(0, 0, 0, 0);
  } else if (preset === "7d") {
    start.setDate(start.getDate() - 7);
  } else {
    start.setDate(start.getDate() - 30);
  }
  return { start, end };
}
var Kpi = ({ label, value }) => /* @__PURE__ */ React7.createElement(Box6, null, /* @__PURE__ */ React7.createElement(Typography6, { variant: "caption", color: "text.secondary", display: "block" }, label), /* @__PURE__ */ React7.createElement(Typography6, { variant: "subtitle1", fontWeight: 600 }, value));
var LiteLLMHomeWidget = ({
  defaultPeriod = "7d",
  title = "LiteLLM Usage"
}) => {
  const api = useApi2(liteLlmApiRef);
  const [period, setPeriod] = useState5(defaultPeriod);
  const [loading, setLoading] = useState5(true);
  const [error, setError] = useState5(null);
  const [usage, setUsage] = useState5(null);
  const [keys, setKeys] = useState5([]);
  useEffect(() => {
    let cancelled = false;
    setLoading(true);
    setError(null);
    const { start, end } = presetToDateRange(period);
    const startDate = start.toISOString().split("T")[0];
    const endDate = end.toISOString().split("T")[0];
    Promise.all([api.getUsage(startDate, endDate), api.listKeys()]).then(([usageData, keysData]) => {
      if (!cancelled) {
        setUsage(usageData);
        setKeys(keysData);
        setLoading(false);
      }
    }).catch((err) => {
      if (!cancelled) {
        setError(err.message ?? "Failed to load usage data");
        setLoading(false);
      }
    });
    return () => {
      cancelled = true;
    };
  }, [api, period]);
  const dailyData = (usage?.daily_usage ?? []).map((d) => ({
    date: d.date,
    spend: d.spend
  }));
  const hasSparkline = dailyData.length > 0;
  return /* @__PURE__ */ React7.createElement(Paper6, { sx: { p: 2 } }, /* @__PURE__ */ React7.createElement(Box6, { display: "flex", justifyContent: "space-between", alignItems: "center", mb: 1.5 }, /* @__PURE__ */ React7.createElement(Typography6, { variant: "h6" }, title), /* @__PURE__ */ React7.createElement(FormControl2, { size: "small", sx: { minWidth: 90 } }, /* @__PURE__ */ React7.createElement(
    Select2,
    {
      value: period,
      onChange: (e) => setPeriod(e.target.value),
      displayEmpty: true
    },
    /* @__PURE__ */ React7.createElement(MenuItem3, { value: "today" }, "Today"),
    /* @__PURE__ */ React7.createElement(MenuItem3, { value: "7d" }, "7d"),
    /* @__PURE__ */ React7.createElement(MenuItem3, { value: "30d" }, "30d")
  ))), loading && /* @__PURE__ */ React7.createElement(Box6, { display: "flex", justifyContent: "center", alignItems: "center", minHeight: 120 }, /* @__PURE__ */ React7.createElement(CircularProgress4, { size: 32 })), !loading && error && /* @__PURE__ */ React7.createElement(Alert2, { severity: "error", sx: { mt: 1 } }, error), !loading && !error && /* @__PURE__ */ React7.createElement(React7.Fragment, null, /* @__PURE__ */ React7.createElement(Grid3, { container: true, spacing: 2, sx: { mb: hasSparkline ? 1.5 : 0 } }, /* @__PURE__ */ React7.createElement(Grid3, { item: true, xs: 6 }, /* @__PURE__ */ React7.createElement(Kpi, { label: "USD Spent", value: fmtUsd2(usage?.total_spend ?? 0) })), /* @__PURE__ */ React7.createElement(Grid3, { item: true, xs: 6 }, /* @__PURE__ */ React7.createElement(Kpi, { label: "Tokens In", value: fmtInt2(usage?.prompt_tokens ?? 0) })), /* @__PURE__ */ React7.createElement(Grid3, { item: true, xs: 6 }, /* @__PURE__ */ React7.createElement(Kpi, { label: "Tokens Out", value: fmtInt2(usage?.completion_tokens ?? 0) })), /* @__PURE__ */ React7.createElement(Grid3, { item: true, xs: 6 }, /* @__PURE__ */ React7.createElement(Kpi, { label: "Keys", value: fmtInt2(keys.length) }))), hasSparkline && /* @__PURE__ */ React7.createElement(Box6, { height: 120 }, /* @__PURE__ */ React7.createElement(ResponsiveContainer3, { width: "100%", height: "100%" }, /* @__PURE__ */ React7.createElement(AreaChart3, { data: dailyData, margin: { top: 4, right: 0, bottom: 0, left: 0 } }, /* @__PURE__ */ React7.createElement(
    Area3,
    {
      type: "monotone",
      dataKey: "spend",
      stroke: "#8884d8",
      fill: "#8884d8",
      fillOpacity: 0.3,
      dot: false,
      isAnimationActive: false
    }
  ))))));
};

// src/index.ts
init_api();
export {
  DashboardHeader,
  KeysTable,
  LiteLLMHomeWidget,
  LiteLLMPage,
  LiteLlmApi,
  TeamUsage,
  UsageStats,
  liteLlmApiRef,
  litellmPlugin
};
