// @backstage/cli-module-build/dist/lib/bundler/config.cjs.js
'use strict';

var node_path = require('node:path');
var core = require('@rspack/core');
var ESLintRspackPlugin = require('eslint-rspack-plugin');
var tsCheckerRspackPlugin = require('ts-checker-rspack-plugin');
var HtmlWebpackPlugin = require('html-webpack-plugin');
var ModuleScopePlugin = require('react-dev-utils/ModuleScopePlugin');
var rspack = require('@module-federation/enhanced/rspack');
var fs = require('fs-extra');
var optimization = require('./optimization.cjs.js');
var pickBy = require('lodash/pickBy');
var cliCommon = require('@backstage/cli-common');
var transforms = require('./transforms.cjs.js');
var _package = require('../../package.json.cjs.js');
var yn = require('yn');
var hasReactDomClient = require('./hasReactDomClient.cjs.js');
var linkWorkspaces = require('./linkWorkspaces.cjs.js');
var ConfigInjectingHtmlWebpackPlugin = require('./ConfigInjectingHtmlWebpackPlugin.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var ESLintRspackPlugin__default = /*#__PURE__*/_interopDefaultCompat(ESLintRspackPlugin);
var HtmlWebpackPlugin__default = /*#__PURE__*/_interopDefaultCompat(HtmlWebpackPlugin);
var ModuleScopePlugin__default = /*#__PURE__*/_interopDefaultCompat(ModuleScopePlugin);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var pickBy__default = /*#__PURE__*/_interopDefaultCompat(pickBy);
var yn__default = /*#__PURE__*/_interopDefaultCompat(yn);

function resolveBaseUrl(config, moduleFederationRemote) {
  const baseUrl = config.getOptionalString("app.baseUrl");
  const defaultBaseUrl = moduleFederationRemote ? `http://localhost:${process.env.PORT ?? "3000"}` : "http://localhost:3000";
  try {
    return new URL(baseUrl ?? "/", defaultBaseUrl);
  } catch (error) {
    throw new Error(`Invalid app.baseUrl, ${error}`);
  }
}
function resolveEndpoint(config, moduleFederationRemote) {
  const url = resolveBaseUrl(config, moduleFederationRemote);
  return {
    host: config.getOptionalString("app.listen.host") ?? url.hostname,
    port: config.getOptionalNumber("app.listen.port") ?? Number(url.port) ?? (url.protocol === "https:" ? 443 : 80)
  };
}
async function readBuildInfo() {
  const timestamp = Date.now();
  let commit;
  try {
    commit = await cliCommon.runOutput(["git", "rev-parse", "HEAD"]);
  } catch (error) {
  }
  let gitVersion;
  try {
    gitVersion = await cliCommon.runOutput(["git", "describe", "--always"]);
  } catch (error) {
  }
  if (commit === void 0 || gitVersion === void 0) {
    console.info(
      "NOTE: Did not compute git version or commit hash, could not execute the git command line utility"
    );
  }
  const { version: packageVersion } = await fs__default.default.readJson(
    cliCommon.targetPaths.resolve("package.json")
  );
  return {
    cliVersion: _package.version,
    gitVersion: gitVersion ?? "unknown",
    packageVersion,
    timestamp,
    commit: commit ?? "unknown"
  };
}
async function createConfig(paths, options) {
  const {
    checksEnabled,
    isDev,
    frontendConfig,
    moduleFederationRemote,
    publicSubPath = "",
    webpack
  } = options;
  const { plugins, loaders } = transforms.transforms(options);
  const validBaseUrl = resolveBaseUrl(frontendConfig, moduleFederationRemote);
  let publicPath = validBaseUrl.pathname.replace(/\/$/, "");
  if (publicSubPath) {
    publicPath = `${publicPath}${publicSubPath}`.replace("//", "/");
  }
  if (isDev) {
    const { host, port } = resolveEndpoint(
      options.frontendConfig,
      options.moduleFederationRemote
    );
    const refreshOptions = {
      overlay: {
        sockProtocol: "ws",
        sockHost: host,
        sockPort: port
      }
    };
    if (webpack) {
      const ReactRefreshPlugin = require("@pmmmwh/react-refresh-webpack-plugin");
      plugins.push(new ReactRefreshPlugin(refreshOptions));
    } else {
      const RspackReactRefreshPlugin = require("@rspack/plugin-react-refresh");
      plugins.push(new RspackReactRefreshPlugin(refreshOptions));
    }
  }
  if (checksEnabled) {
    const TsCheckerPlugin = webpack ? require("fork-ts-checker-webpack-plugin") : tsCheckerRspackPlugin.TsCheckerRspackPlugin;
    const ESLintPlugin = webpack ? require("eslint-webpack-plugin") : ESLintRspackPlugin__default.default;
    plugins.push(
      new TsCheckerPlugin({
        typescript: { configFile: paths.targetTsConfig, memoryLimit: 8192 }
      }),
      new ESLintPlugin({
        cache: false,
        // Cache seems broken
        context: paths.targetPath,
        files: ["**/*.(ts|tsx|mts|cts|js|jsx|mjs|cjs)"]
      })
    );
  }
  const bundler = webpack ? webpack : core.rspack;
  plugins.push(
    new bundler.ProvidePlugin({
      process: require.resolve("process/browser"),
      Buffer: ["buffer", "Buffer"]
    })
  );
  if (!options.moduleFederationRemote) {
    const templateOptions = {
      meta: {
        "backstage-app-mode": options?.appMode ?? "public"
      },
      template: paths.targetHtml,
      templateParameters: {
        publicPath,
        config: frontendConfig
      }
    };
    if (webpack) {
      plugins.push(new HtmlWebpackPlugin__default.default(templateOptions));
    } else {
      plugins.push(
        new ConfigInjectingHtmlWebpackPlugin.ConfigInjectingHtmlWebpackPlugin(
          templateOptions,
          options.getFrontendAppConfigs
        )
      );
    }
    plugins.push(
      new HtmlWebpackPlugin__default.default({
        meta: {
          "backstage-app-mode": options?.appMode ?? "public",
          // This is added to be written in the later step, and finally read by the extra entry point
          "backstage-public-path": "<%= publicPath %>/"
        },
        minify: false,
        publicPath: "<%= publicPath %>",
        filename: "index.html.tmpl",
        template: `${require.resolve("raw-loader")}!${paths.targetHtml}`
      })
    );
  }
  if (options.moduleFederationRemote) {
    const AdaptedModuleFederationPlugin = webpack ? require("@module-federation/enhanced/webpack").ModuleFederationPlugin : rspack.ModuleFederationPlugin;
    const exposes = options.moduleFederationRemote.exposes ? Object.fromEntries(
      Object.entries(options.moduleFederationRemote?.exposes).map(
        ([k, v]) => [k, node_path.resolve(paths.targetPath, v)]
      )
    ) : {
      ".": paths.targetEntry
    };
    plugins.push(
      new AdaptedModuleFederationPlugin({
        filename: "remoteEntry.js",
        exposes,
        name: options.moduleFederationRemote.name,
        runtime: false,
        shared: options.moduleFederationRemote.sharedDependencies
      })
    );
  }
  const buildInfo = await readBuildInfo();
  plugins.push(
    webpack ? new webpack.DefinePlugin({
      "process.env.BUILD_INFO": JSON.stringify(buildInfo),
      "process.env.APP_CONFIG": webpack.DefinePlugin.runtimeValue(
        () => JSON.stringify(options.getFrontendAppConfigs()),
        true
      ),
      // This allows for conditional imports of react-dom/client, since there's no way
      // to check for presence of it in source code without module resolution errors.
      "process.env.HAS_REACT_DOM_CLIENT": JSON.stringify(
        hasReactDomClient.hasReactDomClient()
      )
    }) : new bundler.DefinePlugin({
      "process.env.BUILD_INFO": JSON.stringify(buildInfo),
      "process.env.APP_CONFIG": JSON.stringify([]),
      // Inject via index.html instead
      // This allows for conditional imports of react-dom/client, since there's no way
      // to check for presence of it in source code without module resolution errors.
      "process.env.HAS_REACT_DOM_CLIENT": JSON.stringify(
        hasReactDomClient.hasReactDomClient()
      )
    })
  );
  if (options.linkedWorkspace) {
    plugins.push(
      ...await linkWorkspaces.createWorkspaceLinkingPlugins(
        bundler,
        options.linkedWorkspace
      )
    );
  }
  const reactRefreshFiles = webpack ? [
    require.resolve(
      "@pmmmwh/react-refresh-webpack-plugin/lib/runtime/RefreshUtils.js"
    ),
    require.resolve(
      "@pmmmwh/react-refresh-webpack-plugin/overlay/index.js"
    ),
    require.resolve("react-refresh")
  ] : [];
  const mode = isDev ? "development" : "production";
  const optimization$1 = optimization.optimization(options);
  return {
    mode,
    profile: false,
    ...isDev ? {
      watchOptions: {
        ignored: /node_modules\/(?!__backstage-autodetected-plugins__)/
      }
    } : {},
    optimization: optimization$1,
    bail: false,
    performance: {
      hints: false
      // we check the gzip size instead
    },
    devtool: isDev ? "eval-cheap-module-source-map" : "source-map",
    context: paths.targetPath,
    entry: [
      /* eslint-disable-next-line no-restricted-syntax */
      cliCommon.findOwnPaths(__dirname).resolve("config/webpack-public-path"),
      ...options.additionalEntryPoints ?? [],
      paths.targetEntry
    ],
    resolve: {
      extensions: [".ts", ".tsx", ".mjs", ".js", ".jsx", ".json", ".wasm"],
      mainFields: ["browser", "module", "main"],
      fallback: {
        ...pickBy__default.default(require("node-stdlib-browser")),
        module: false,
        dgram: false,
        dns: false,
        fs: false,
        http2: false,
        net: false,
        tls: false,
        child_process: false,
        /* new ignores */
        path: false,
        https: false,
        http: false,
        util: require.resolve("util/")
      },
      // FIXME: see also https://github.com/web-infra-dev/rspack/issues/3408
      ...webpack && {
        plugins: [
          new ModuleScopePlugin__default.default(
            [paths.targetSrc, paths.targetDev],
            [paths.targetPackageJson, ...reactRefreshFiles]
          )
        ]
      }
    },
    module: {
      rules: loaders
    },
    output: {
      uniqueName: options.moduleFederationRemote?.name,
      path: paths.targetDist,
      publicPath: options.moduleFederationRemote ? "auto" : `${publicPath}/`,
      filename: isDev ? "[name].js" : "static/[name].[contenthash:8].js",
      chunkFilename: isDev ? "[name].chunk.js" : "static/[name].[contenthash:8].chunk.js",
      ...isDev ? {
        devtoolModuleFilenameTemplate: (info) => `file:///${node_path.resolve(info.absoluteResourcePath).replace(
          /\\/g,
          "/"
        )}`
      } : {}
    },
    experiments: {
      lazyCompilation: yn__default.default(process.env.EXPERIMENTAL_LAZY_COMPILATION),
      ...!webpack && {
        // We're still using `style-loader` for custom `insert` option
        css: false
      }
    },
    plugins,
    ignoreWarnings: [
      // @protobufjs/inquire uses require(moduleName) with a dynamic argument.
      // Since protobufjs >=7.5.9 this code path is never exercised (the
      // bundler-safe optional module lookups backport replaced all call sites),
      // but webpack/rspack still statically analyzes the source and emits a
      // "Critical dependency" warning. Safe to suppress.
      // See https://github.com/protobufjs/protobuf.js/issues/2057
      {
        module: /@protobufjs[\\/]inquire/,
        message: /Critical dependency: the request of a dependency is an expression/
      },
      // TODO: remove this warning skipping as soon as the corresponding bundler limitation
      // described in issue https://github.com/web-infra-dev/rspack/issues/13635 is fixed
      // when PR: https://github.com/web-infra-dev/rspack/pull/13636 is merged.
      ...options.moduleFederationRemote ? [
        {
          message: /No version specified and unable to automatically determine one\. No version in description file/
        }
      ] : []
    ]
  };
}

exports.createConfig = createConfig;
exports.resolveBaseUrl = resolveBaseUrl;
exports.resolveEndpoint = resolveEndpoint;
