// @backstage/cli-module-build/dist/lib/bundler/paths.cjs.js
'use strict';

var fs = require('fs-extra');
var node_path = require('node:path');
var cliCommon = require('@backstage/cli-common');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

function resolveBundlingPaths(options) {
  const { entry, targetDir = cliCommon.targetPaths.dir } = options;
  const resolveTargetModule = (pathString) => {
    for (const ext of ["mjs", "js", "ts", "tsx", "jsx"]) {
      const filePath = node_path.resolve(targetDir, `${pathString}.${ext}`);
      if (fs__default.default.pathExistsSync(filePath)) {
        return filePath;
      }
    }
    return node_path.resolve(targetDir, `${pathString}.js`);
  };
  let targetPublic = void 0;
  let targetHtml = node_path.resolve(targetDir, "public/index.html");
  if (fs__default.default.pathExistsSync(targetHtml)) {
    targetPublic = node_path.resolve(targetDir, "public");
  } else {
    targetHtml = node_path.resolve(targetDir, `${entry}.html`);
    if (!fs__default.default.pathExistsSync(targetHtml)) {
      targetHtml = cliCommon.findOwnPaths(__dirname).resolve(
        "templates/serve_index.html"
      );
    }
  }
  const targetRunFile = node_path.resolve(targetDir, "src/run.ts");
  const runFileExists = fs__default.default.pathExistsSync(targetRunFile);
  return {
    targetHtml,
    targetPublic,
    targetPath: node_path.resolve(targetDir, "."),
    targetRunFile: runFileExists ? targetRunFile : void 0,
    targetDist: node_path.resolve(targetDir, options.dist ?? "dist"),
    targetAssets: node_path.resolve(targetDir, "assets"),
    targetSrc: node_path.resolve(targetDir, "src"),
    targetDev: node_path.resolve(targetDir, "dev"),
    targetEntry: resolveTargetModule(entry),
    targetTsConfig: cliCommon.targetPaths.resolveRoot("tsconfig.json"),
    targetPackageJson: node_path.resolve(targetDir, "package.json"),
    rootNodeModules: cliCommon.targetPaths.resolveRoot("node_modules"),
    root: cliCommon.targetPaths.rootDir
  };
}
async function resolveOptionalBundlingPaths(options) {
  const resolvedPaths = resolveBundlingPaths(options);
  if (await fs__default.default.pathExists(resolvedPaths.targetEntry)) {
    return resolvedPaths;
  }
  return void 0;
}

exports.resolveBundlingPaths = resolveBundlingPaths;
exports.resolveOptionalBundlingPaths = resolveOptionalBundlingPaths;
