// @backstage/cli-module-build/dist/lib/config.cjs.js
'use strict';

var configLoader = require('@backstage/config-loader');
var config = require('@backstage/config');
var cliCommon = require('@backstage/cli-common');
var getPackages = require('@manypkg/get-packages');
var cliNode = require('@backstage/cli-node');
var node_path = require('node:path');

async function loadCliConfig(options) {
  const targetDir = options.targetDir ?? cliCommon.targetPaths.dir;
  const { packages } = await getPackages.getPackages(targetDir);
  let localPackageNames;
  if (options.fromPackage) {
    if (packages.length) {
      const graph = cliNode.PackageGraph.fromPackages(packages);
      localPackageNames = Array.from(
        graph.collectPackageNames([options.fromPackage], (node) => {
          if (node.name === "@backstage/cli") {
            return void 0;
          }
          return node.localDependencies.keys();
        })
      );
    } else {
      localPackageNames = [options.fromPackage];
    }
  } else {
    localPackageNames = packages.map((p) => p.packageJson.name);
  }
  const schema = await configLoader.loadConfigSchema({
    dependencies: localPackageNames,
    packagePaths: [cliCommon.targetPaths.resolveRoot("package.json")]
  });
  const source = configLoader.ConfigSources.default({
    allowMissingDefaultConfig: true,
    watch: Boolean(options.watch),
    rootDir: cliCommon.targetPaths.rootDir,
    argv: options.args.flatMap((t) => ["--config", node_path.resolve(targetDir, t)])
  });
  const appConfigs = await new Promise((resolve, reject) => {
    async function loadConfigReaderLoop() {
      let loaded = false;
      try {
        const abortController = new AbortController();
        for await (const { configs } of source.readConfigData({
          signal: abortController.signal
        })) {
          if (loaded) {
            const newFrontendAppConfigs = schema.process(configs, {
              visibility: ["frontend"],
              withFilteredKeys: options.withFilteredKeys,
              ignoreSchemaErrors: true
            });
            options.watch?.(newFrontendAppConfigs);
          } else {
            resolve(configs);
            loaded = true;
            if (!options.watch) {
              abortController.abort();
            }
          }
        }
      } catch (error) {
        if (loaded) {
          console.error(`Failed to reload configuration, ${error}`);
        } else {
          reject(error);
        }
      }
    }
    loadConfigReaderLoop();
  });
  const configurationLoadedMessage = appConfigs.length ? `Loaded config from ${appConfigs.map((c) => c.context).join(", ")}` : `No configuration files found, running without config`;
  process.stderr.write(`${configurationLoadedMessage}
`);
  const frontendAppConfigs = schema.process(appConfigs, {
    visibility: ["frontend"],
    withFilteredKeys: options.withFilteredKeys,
    ignoreSchemaErrors: true
  });
  const frontendConfig = config.ConfigReader.fromConfigs(frontendAppConfigs);
  const fullConfig = config.ConfigReader.fromConfigs(appConfigs);
  return {
    schema,
    appConfigs,
    frontendConfig,
    frontendAppConfigs,
    fullConfig
  };
}

exports.loadCliConfig = loadCliConfig;
