// @backstage/catalog-model/dist/model/compileCatalogModel.esm.js
import { InputError } from '@backstage/errors';
import lodash from 'lodash';
import { mergeJsonSchemas } from './jsonSchema/mergeJsonSchemas.esm.js';
import { ops } from './operations/util.esm.js';
import { OpaqueCatalogModelLayer } from './types.esm.js';

function sortOps(input) {
  return lodash.sortBy(input, (op) => ops[op.op].order);
}
function applyDeclareKind(kinds, op) {
  if (kinds.has(op.kind)) {
    throw new InputError(`Kind "${op.kind}" is declared more than once`);
  }
  kinds.set(op.kind, {
    group: op.group,
    singular: op.properties.singular,
    plural: op.properties.plural,
    description: op.properties.description,
    versions: /* @__PURE__ */ new Map()
  });
}
function applyDeclareKindVersion(kinds, op) {
  const kind = kinds.get(op.kind);
  if (!kind) {
    throw new InputError(
      `Cannot declare version "${op.name}" for unknown kind "${op.kind}"`
    );
  }
  let version = kind.versions.get(op.name);
  if (!version) {
    version = {
      name: op.name,
      apiVersion: `${kind.group}/${op.name}`,
      specTypes: /* @__PURE__ */ new Map()
    };
    kind.versions.set(op.name, version);
  }
  if (version.specTypes.has(op.specType)) {
    const label = op.specType ? `spec type "${op.specType}"` : "default spec type";
    throw new InputError(
      `Version "${op.name}" of kind "${op.kind}" already has ${label} declared`
    );
  }
  version.specTypes.set(op.specType, {
    description: op.properties.description,
    relationFields: op.properties.relationFields,
    jsonSchema: op.properties.schema.jsonSchema
  });
}
function applyDeclareRelation(relations, op) {
  const existing = relations.get(op.type);
  if (existing) {
    existing.fromKinds.add(op.fromKind);
    existing.toKinds.add(op.toKind);
  } else {
    relations.set(op.type, {
      fromKinds: /* @__PURE__ */ new Set([op.fromKind]),
      toKinds: /* @__PURE__ */ new Set([op.toKind]),
      description: op.properties.description,
      forward: {
        type: op.type,
        title: op.properties.title
      },
      reverse: {
        type: op.properties.reverseType,
        title: op.properties.title
      }
    });
  }
}
function applyUpdateKind(kinds, op) {
  const kind = kinds.get(op.kind);
  if (!kind) {
    throw new InputError(`Cannot update unknown kind "${op.kind}"`);
  }
  if (op.properties.singular !== void 0) {
    kind.singular = op.properties.singular;
  }
  if (op.properties.plural !== void 0) {
    kind.plural = op.properties.plural;
  }
  if (op.properties.description !== void 0) {
    kind.description = op.properties.description;
  }
}
function applyUpdateKindVersion(kinds, op) {
  const kind = kinds.get(op.kind);
  if (!kind) {
    throw new InputError(
      `Cannot update version "${op.name}" for unknown kind "${op.kind}"`
    );
  }
  const version = kind.versions.get(op.name);
  if (!version) {
    throw new InputError(
      `Cannot update unknown version "${op.name}" of kind "${op.kind}"`
    );
  }
  const specType = version.specTypes.get(op.specType);
  if (!specType) {
    const label = op.specType ? `spec type "${op.specType}"` : "default spec type";
    throw new InputError(
      `Cannot update undeclared ${label} on version "${op.name}" of kind "${op.kind}"`
    );
  }
  if (op.properties.description !== void 0) {
    specType.description = op.properties.description;
  }
  if (op.properties.relationFields !== void 0) {
    specType.relationFields = op.properties.relationFields;
  }
  if (op.properties.schema !== void 0) {
    specType.jsonSchema = mergeJsonSchemas(
      specType.jsonSchema,
      op.properties.schema.jsonSchema
    );
  }
}
function applyUpdateRelation(relations, op) {
  const relation = relations.get(op.type);
  if (!relation) {
    throw new InputError(`Cannot update undeclared relation "${op.type}"`);
  }
  relation.fromKinds.add(op.fromKind);
  relation.toKinds.add(op.toKind);
  if (op.properties.reverseType !== void 0) {
    relation.reverse.type = op.properties.reverseType;
  }
  if (op.properties.title !== void 0) {
    relation.forward.title = op.properties.title;
    relation.reverse.title = op.properties.title;
  }
  if (op.properties.description !== void 0) {
    relation.description = op.properties.description;
  }
}
function applyRemoveKind(kinds, op) {
  if (!kinds.has(op.kind)) {
    throw new InputError(`Cannot remove unknown kind "${op.kind}"`);
  }
  kinds.delete(op.kind);
}
function applyDeclareAnnotation(annotations, op) {
  if (annotations.has(op.name)) {
    throw new InputError(`Annotation "${op.name}" is declared more than once`);
  }
  annotations.set(op.name, {
    title: op.properties.title,
    description: op.properties.description,
    schema: op.properties.schema
  });
}
function applyDeclareLabel(labels, op) {
  if (labels.has(op.name)) {
    throw new InputError(`Label "${op.name}" is declared more than once`);
  }
  labels.set(op.name, {
    title: op.properties.title,
    description: op.properties.description,
    schema: op.properties.schema
  });
}
function applyDeclareTag(tags, op) {
  if (tags.has(op.name)) {
    throw new InputError(`Tag "${op.name}" is declared more than once`);
  }
  tags.set(op.name, {
    title: op.properties.title,
    description: op.properties.description
  });
}
function applyUpdateAnnotation(annotations, op) {
  const annotation = annotations.get(op.name);
  if (!annotation) {
    throw new InputError(`Cannot update undeclared annotation "${op.name}"`);
  }
  if (op.properties.title !== void 0) {
    annotation.title = op.properties.title;
  }
  if (op.properties.description !== void 0) {
    annotation.description = op.properties.description;
  }
  if (op.properties.schema !== void 0) {
    annotation.schema = op.properties.schema;
  }
}
function applyUpdateLabel(labels, op) {
  const label = labels.get(op.name);
  if (!label) {
    throw new InputError(`Cannot update undeclared label "${op.name}"`);
  }
  if (op.properties.title !== void 0) {
    label.title = op.properties.title;
  }
  if (op.properties.description !== void 0) {
    label.description = op.properties.description;
  }
  if (op.properties.schema !== void 0) {
    label.schema = op.properties.schema;
  }
}
function applyUpdateTag(tags, op) {
  const tag = tags.get(op.name);
  if (!tag) {
    throw new InputError(`Cannot update undeclared tag "${op.name}"`);
  }
  if (op.properties.title !== void 0) {
    tag.title = op.properties.title;
  }
  if (op.properties.description !== void 0) {
    tag.description = op.properties.description;
  }
}
function applyRemoveAnnotation(annotations, op) {
  if (!annotations.has(op.name)) {
    throw new InputError(`Cannot remove unknown annotation "${op.name}"`);
  }
  annotations.delete(op.name);
}
function applyRemoveLabel(labels, op) {
  if (!labels.has(op.name)) {
    throw new InputError(`Cannot remove unknown label "${op.name}"`);
  }
  labels.delete(op.name);
}
function applyRemoveTag(tags, op) {
  if (!tags.has(op.name)) {
    throw new InputError(`Cannot remove unknown tag "${op.name}"`);
  }
  tags.delete(op.name);
}
function buildFullSchema(options) {
  const annotationProperties = {};
  for (const [name, state] of options.annotations) {
    annotationProperties[name] = state.schema?.jsonSchema ?? { type: "string" };
  }
  const labelProperties = {};
  for (const [name, state] of options.labels) {
    labelProperties[name] = state.schema?.jsonSchema ?? { type: "string" };
  }
  const metadataSchema = {
    type: "object",
    required: ["name"],
    additionalProperties: true,
    properties: {
      uid: {
        type: "string",
        description: "A globally unique ID for the entity.",
        minLength: 1
      },
      etag: {
        type: "string",
        description: "An opaque string that changes for each update operation to any part of the entity, including metadata.",
        minLength: 1
      },
      name: {
        type: "string",
        description: "The name of the entity. Must be unique within the catalog at any given point in time, for any given namespace + kind pair.",
        minLength: 1
      },
      namespace: {
        type: "string",
        description: "The namespace that the entity belongs to.",
        default: "default",
        minLength: 1
      },
      title: {
        type: "string",
        description: "A display name of the entity, to be presented in user interfaces instead of the name property, when available.",
        minLength: 1
      },
      description: {
        type: "string",
        description: "A short (typically relatively few words, on one line) description of the entity."
      },
      annotations: {
        type: "object",
        description: "Key/value pairs of non-identifying auxiliary information attached to the entity.",
        additionalProperties: { type: "string" },
        properties: annotationProperties
      },
      labels: {
        type: "object",
        description: "Key/value pairs of identifying information attached to the entity.",
        additionalProperties: { type: "string" },
        properties: labelProperties
      },
      tags: {
        type: "array",
        description: "A list of single-valued strings, to for example classify catalog entities in various ways.",
        items: { type: "string", minLength: 1 }
      },
      links: {
        type: "array",
        description: "A list of external hyperlinks related to the entity.",
        items: {
          type: "object",
          required: ["url"],
          properties: {
            url: { type: "string", minLength: 1 },
            title: { type: "string", minLength: 1 },
            icon: { type: "string", minLength: 1 },
            type: { type: "string", minLength: 1 }
          }
        }
      }
    }
  };
  const kindRequired = Array.isArray(options.kindSchema.required) ? options.kindSchema.required : [];
  const generatedSchema = {
    type: "object",
    required: [.../* @__PURE__ */ new Set([...kindRequired, "apiVersion", "kind", "metadata"])],
    additionalProperties: false,
    properties: {
      apiVersion: { const: options.apiVersion },
      kind: { const: options.kind },
      metadata: metadataSchema
    }
  };
  return mergeJsonSchemas(options.kindSchema, generatedSchema);
}
function compileCatalogModel(inputs) {
  let allOps = [];
  for (const input of inputs) {
    const internal = OpaqueCatalogModelLayer.toInternal(input);
    allOps = allOps.concat(internal.ops);
  }
  const sortedOps = sortOps(allOps);
  const annotations = /* @__PURE__ */ new Map();
  const labels = /* @__PURE__ */ new Map();
  const tags = /* @__PURE__ */ new Map();
  const kinds = /* @__PURE__ */ new Map();
  const relations = /* @__PURE__ */ new Map();
  for (const op of sortedOps) {
    switch (op.op) {
      case "declareAnnotation.v1":
        applyDeclareAnnotation(annotations, op);
        break;
      case "declareLabel.v1":
        applyDeclareLabel(labels, op);
        break;
      case "declareTag.v1":
        applyDeclareTag(tags, op);
        break;
      case "declareKind.v1":
        applyDeclareKind(kinds, op);
        break;
      case "declareKindVersion.v1":
        applyDeclareKindVersion(kinds, op);
        break;
      case "declareRelation.v1":
        applyDeclareRelation(relations, op);
        break;
      case "updateAnnotation.v1":
        applyUpdateAnnotation(annotations, op);
        break;
      case "updateLabel.v1":
        applyUpdateLabel(labels, op);
        break;
      case "updateTag.v1":
        applyUpdateTag(tags, op);
        break;
      case "updateKind.v1":
        applyUpdateKind(kinds, op);
        break;
      case "updateKindVersion.v1":
        applyUpdateKindVersion(kinds, op);
        break;
      case "updateRelation.v1":
        applyUpdateRelation(relations, op);
        break;
      case "removeAnnotation.v1":
        applyRemoveAnnotation(annotations, op);
        break;
      case "removeLabel.v1":
        applyRemoveLabel(labels, op);
        break;
      case "removeTag.v1":
        applyRemoveTag(tags, op);
        break;
      case "removeKind.v1":
        applyRemoveKind(kinds, op);
        break;
      default:
        throw new InputError(`Unknown op type "${op.op}"`);
    }
  }
  const compiledKinds = /* @__PURE__ */ new Map();
  for (const [kindName, kindState] of kinds) {
    for (const version of kindState.versions.values()) {
      for (const [specType, specificKind] of version.specTypes) {
        const key = `${kindName}\0${version.apiVersion}\0${specType ?? ""}`;
        compiledKinds.set(key, {
          description: specificKind.description ?? kindState.description,
          apiVersions: [version.apiVersion],
          names: {
            kind: kindName,
            singular: kindState.singular,
            plural: kindState.plural
          },
          relationFields: (specificKind.relationFields ?? []).map((f) => ({
            path: f.selector.path,
            relation: f.relation,
            defaultKind: f.defaultKind,
            defaultNamespace: f.defaultNamespace,
            allowedKinds: f.allowedKinds
          })),
          jsonSchema: buildFullSchema({
            kind: kindName,
            apiVersion: version.apiVersion,
            kindSchema: specificKind.jsonSchema,
            annotations,
            labels})
        });
      }
    }
  }
  const compiledRelations = /* @__PURE__ */ new Map();
  for (const kindName of kinds.keys()) {
    compiledRelations.set(
      kindName,
      [...relations.values()].filter((r) => r.fromKinds.has(kindName)).map((r) => {
        const reverseEntry = relations.get(r.reverse.type);
        return {
          fromKind: [...r.fromKinds],
          toKind: [...r.toKinds],
          description: r.description,
          forward: r.forward,
          reverse: {
            type: r.reverse.type,
            title: reverseEntry?.forward.title ?? r.reverse.title
          }
        };
      })
    );
  }
  const kindSummaries = [...kinds.entries()].map(
    ([kindName, kindState]) => ({
      description: kindState.description,
      names: {
        kind: kindName,
        singular: kindState.singular,
        plural: kindState.plural
      },
      versions: [...kindState.versions.values()].flatMap(
        (version) => [...version.specTypes.keys()].map((specType) => ({
          apiVersion: version.apiVersion,
          ...specType !== void 0 ? { specType } : void 0
        }))
      )
    })
  );
  const annotationSummaries = [
    ...annotations.entries()
  ].map(([name, state]) => ({
    name,
    ...state.title !== void 0 ? { title: state.title } : void 0,
    description: state.description
  }));
  const labelSummaries = [...labels.entries()].map(
    ([name, state]) => ({
      name,
      ...state.title !== void 0 ? { title: state.title } : void 0,
      description: state.description
    })
  );
  const tagSummaries = [...tags.entries()].map(
    ([name, state]) => ({
      name,
      ...state.title !== void 0 ? { title: state.title } : void 0,
      description: state.description
    })
  );
  const relationSummaries = [
    ...relations.values()
  ].map((r) => {
    const reverseEntry = relations.get(r.reverse.type);
    return {
      fromKind: [...r.fromKinds],
      toKind: [...r.toKinds],
      description: r.description,
      forward: r.forward,
      reverse: {
        type: r.reverse.type,
        title: reverseEntry?.forward.title ?? r.reverse.title
      }
    };
  });
  return {
    listKinds() {
      return kindSummaries;
    },
    listRelations() {
      return relationSummaries;
    },
    getMetadata() {
      return {
        annotations: annotationSummaries,
        labels: labelSummaries,
        tags: tagSummaries
      };
    },
    getKind(options) {
      const type = options.spec?.type;
      if (!kinds.has(options.kind)) {
        return void 0;
      }
      const kindState = kinds.get(options.kind);
      const version = [...kindState.versions.values()].find(
        (v) => v.apiVersion === options.apiVersion
      );
      if (!version) {
        throw new TypeError(
          `Kind "${options.kind}" exists, but has no version matching apiVersion "${options.apiVersion}"`
        );
      }
      const key = `${options.kind}\0${version.apiVersion}\0${type ?? ""}`;
      const result = compiledKinds.get(key);
      if (result) {
        return result;
      }
      if (type !== void 0) {
        const fallback = compiledKinds.get(
          `${options.kind}\0${version.apiVersion}\0`
        );
        if (fallback) {
          return fallback;
        }
      }
      throw new TypeError(
        `Kind "${options.kind}" version "${version.name}" exists, but has no matching spec type`
      );
    },
    getRelations(options) {
      return compiledRelations.get(options.kind);
    }
  };
}

export { compileCatalogModel };
