// @backstage/cli-module-build/dist/package.json.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@backstage/cli-module-build";
var version = "0.1.4";
var description = "CLI module for Backstage CLI";
var backstage = {
	role: "cli-module"
};
var publishConfig = {
	access: "public",
	main: "dist/index.cjs.js",
	types: "dist/index.d.ts"
};
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "packages/cli-module-build"
};
var license = "Apache-2.0";
var main = "src/index.ts";
var types = "src/index.ts";
var bin = "bin/backstage-cli-module-build";
var files = [
	"dist",
	"bin",
	"config",
	"templates"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	test: "backstage-cli package test"
};
var jest = {
	coveragePathIgnorePatterns: [
		"/__fixtures__/"
	]
};
var dependencies = {
	"@backstage/cli-common": "workspace:^",
	"@backstage/cli-node": "workspace:^",
	"@backstage/config": "workspace:^",
	"@backstage/config-loader": "workspace:^",
	"@backstage/errors": "workspace:^",
	"@backstage/module-federation-common": "workspace:^",
	"@manypkg/get-packages": "^1.1.3",
	"@module-federation/enhanced": "^2.3.3",
	"@pmmmwh/react-refresh-webpack-plugin": "^0.6.0",
	"@rollup/plugin-commonjs": "^26.0.0",
	"@rollup/plugin-json": "^6.0.0",
	"@rollup/plugin-node-resolve": "^15.0.0",
	"@rollup/plugin-yaml": "^4.0.0",
	"@rspack/core": "^1.4.11",
	"@rspack/dev-server": "^1.1.4",
	"@rspack/plugin-react-refresh": "^1.4.3",
	"@swc/core": "^1.15.6",
	bfj: "^9.0.2",
	buffer: "^6.0.3",
	chalk: "^4.0.0",
	chokidar: "^3.3.1",
	cleye: "^2.6.0",
	"cross-spawn": "^7.0.3",
	"css-loader": "^6.5.1",
	"ctrlc-windows": "^2.1.0",
	"esbuild-loader": "^4.0.0",
	"eslint-rspack-plugin": "^4.2.1",
	"eslint-webpack-plugin": "^4.2.0",
	"fork-ts-checker-webpack-plugin": "^9.0.0",
	"fs-extra": "^11.2.0",
	glob: "^13.0.0",
	"html-webpack-plugin": "^5.6.3",
	lodash: "^4.17.21",
	"mini-css-extract-plugin": "^2.4.2",
	"node-stdlib-browser": "^1.3.1",
	"npm-packlist": "^5.0.0",
	"p-queue": "^6.6.2",
	portfinder: "^1.0.32",
	postcss: "^8.1.0",
	"postcss-import": "^16.1.0",
	process: "^0.11.10",
	"raw-loader": "^4.0.2",
	"react-dev-utils": "^12.0.0-next.60",
	"react-refresh": "^0.18.0",
	rollup: "^4.59.0",
	"rollup-plugin-dts": "^6.1.0",
	"rollup-plugin-esbuild": "^6.1.1",
	"rollup-plugin-postcss": "^4.0.0",
	"rollup-pluginutils": "^2.8.2",
	"shell-quote": "^1.8.1",
	"style-loader": "^3.3.1",
	"swc-loader": "^0.2.3",
	tar: "^7.5.6",
	"ts-checker-rspack-plugin": "^1.1.5",
	"ts-morph": "^24.0.0",
	util: "^0.12.3",
	webpack: "~5.105.0",
	"webpack-dev-server": "^5.0.0",
	"yml-loader": "^2.1.0",
	yn: "^4.0.0"
};
var devDependencies = {
	"@backstage/backend-test-utils": "workspace:^",
	"@backstage/cli": "workspace:^",
	"@types/fs-extra": "^11.0.0",
	"@types/lodash": "^4.14.151",
	"@types/npm-packlist": "^3.0.0",
	"@types/shell-quote": "^1.7.5",
	"embedded-postgres": "18.3.0-beta.17"
};
var peerDependencies = {
	"embedded-postgres": "^18.3.0-beta.16"
};
var peerDependenciesMeta = {
	"embedded-postgres": {
		optional: true
	}
};
var packageJson = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	homepage: homepage,
	repository: repository,
	license: license,
	main: main,
	types: types,
	bin: bin,
	files: files,
	scripts: scripts,
	jest: jest,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta
};

exports.backstage = backstage;
exports.bin = bin;
exports.default = packageJson;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.files = files;
exports.homepage = homepage;
exports.jest = jest;
exports.license = license;
exports.main = main;
exports.name = name;
exports.peerDependencies = peerDependencies;
exports.peerDependenciesMeta = peerDependenciesMeta;
exports.publishConfig = publishConfig;
exports.repository = repository;
exports.scripts = scripts;
exports.types = types;
exports.version = version;
