// @backstage/cli-module-migrate/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cliNode = require('@backstage/cli-node');
var _package = require('./package.json.cjs.js');

var index = cliNode.createCliModule({
  packageJson: _package.default,
  init: async (reg) => {
    reg.addCommand({
      path: ["versions:migrate"],
      description: "Migrate any plugins that have been moved to the @backstage-community namespace automatically",
      execute: { loader: () => import('./commands/versions/migrate.cjs.js') }
    });
    reg.addCommand({
      path: ["versions:bump"],
      description: "Bump Backstage packages to the latest versions",
      execute: { loader: () => import('./commands/versions/bump.cjs.js') }
    });
    reg.addCommand({
      path: ["migrate", "package-roles"],
      description: `Add package role field to packages that don't have it`,
      execute: {
        loader: () => import('./commands/packageRole.cjs.js')
      }
    });
    reg.addCommand({
      path: ["migrate", "package-scripts"],
      description: "Set package scripts according to each package role",
      execute: {
        loader: () => import('./commands/packageScripts.cjs.js')
      }
    });
    reg.addCommand({
      path: ["migrate", "package-exports"],
      description: "Synchronize package subpath export definitions",
      execute: {
        loader: () => import('./commands/packageExports.cjs.js')
      }
    });
    reg.addCommand({
      path: ["migrate", "package-lint-configs"],
      description: "Migrates all packages to use @backstage/cli/config/eslint-factory",
      execute: {
        loader: () => import('./commands/packageLintConfigs.cjs.js')
      }
    });
    reg.addCommand({
      path: ["migrate", "react-router-deps"],
      description: "Migrates the react-router dependencies for all packages to be peer dependencies",
      execute: {
        loader: () => import('./commands/reactRouterDeps.cjs.js')
      }
    });
  }
});

exports.default = index;
