// @backstage/plugin-catalog-react/dist/alpha/blueprints/EntityContentLayoutBlueprint.esm.js
import { createExtensionDataRef, createExtensionBlueprint, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { entityFilterExpressionDataRef, entityFilterFunctionDataRef } from './extensionData.esm.js';
import { createZodV4FilterPredicateSchema } from '@backstage/filter-predicates';
import { z } from 'zod/v4';
import { resolveEntityFilterData } from './resolveEntityFilterData.esm.js';

const entityCardLayoutComponentDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-layout.component"
});
const EntityContentLayoutBlueprint = createExtensionBlueprint({
  kind: "entity-content-layout",
  attachTo: { id: "entity-content:catalog/overview", input: "layouts" },
  output: [
    entityFilterFunctionDataRef.optional(),
    entityFilterExpressionDataRef.optional(),
    entityCardLayoutComponentDataRef
  ],
  dataRefs: {
    filterFunction: entityFilterFunctionDataRef,
    filterExpression: entityFilterExpressionDataRef,
    component: entityCardLayoutComponentDataRef
  },
  configSchema: {
    type: z.string().optional(),
    filter: z.union([z.string(), createZodV4FilterPredicateSchema()]).optional()
  },
  *factory({
    loader,
    filter
  }, { node, config }) {
    yield* resolveEntityFilterData(filter, config, node);
    yield entityCardLayoutComponentDataRef(
      ExtensionBoundary.lazyComponent(node, loader)
    );
  }
});

export { EntityContentLayoutBlueprint };
