// @backstage/ui/dist/components/DateRangePicker/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './DateRangePicker.module.css.esm.js';

const DateRangePickerDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DateRangePicker"
  },
  propDefs: {
    size: { dataAttribute: true, default: "small" },
    className: {},
    label: {},
    description: {},
    secondaryLabel: {}
  }
});
const DateRangePickerGroupDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DateRangePickerGroup",
    dateFields: "bui-DateRangePickerDateFields",
    dateInput: "bui-DateRangePickerDateInput",
    segment: "bui-DateRangePickerSegment",
    separator: "bui-DateRangePickerSeparator",
    button: "bui-DateRangePickerButton"
  },
  bg: "consumer",
  propDefs: {}
});
const DateRangePickerCalendarDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DateRangePickerCalendar",
    header: "bui-DateRangePickerCalendarHeader",
    heading: "bui-DateRangePickerCalendarHeading",
    navButton: "bui-DateRangePickerCalendarNavButton",
    grid: "bui-DateRangePickerCalendarGrid",
    gridHeader: "bui-DateRangePickerCalendarGridHeader",
    headerCell: "bui-DateRangePickerCalendarHeaderCell",
    gridBody: "bui-DateRangePickerCalendarGridBody",
    cell: "bui-DateRangePickerCalendarCell"
  },
  propDefs: {}
});

export { DateRangePickerCalendarDefinition, DateRangePickerDefinition, DateRangePickerGroupDefinition };
