// @backstage/cli-module-test-jest/dist/package.json.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var name = "@backstage/cli-module-test-jest";
var version = "0.1.3";
var description = "CLI module for Backstage CLI";
var backstage = {
	role: "cli-module"
};
var publishConfig = {
	access: "public",
	main: "dist/index.cjs.js",
	types: "dist/index.d.ts"
};
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "packages/cli-module-test-jest"
};
var license = "Apache-2.0";
var main = "src/index.ts";
var types = "src/index.ts";
var bin = "bin/backstage-cli-module-test-jest";
var files = [
	"dist",
	"bin",
	"config"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/cli-common": "workspace:^",
	"@backstage/cli-node": "workspace:^",
	"@swc/core": "^1.15.6",
	"@swc/jest": "^0.2.39",
	cleye: "^2.6.0",
	"cross-fetch": "^4.0.0",
	"fs-extra": "^11.2.0",
	glob: "^13.0.0",
	"jest-css-modules": "^2.1.0",
	sucrase: "^3.20.2",
	yargs: "^16.2.0"
};
var devDependencies = {
	"@backstage/cli": "workspace:^"
};
var peerDependencies = {
	"@jest/environment-jsdom-abstract": "^30.0.0",
	jest: "^29.0.0 || ^30.0.0",
	"jest-environment-jsdom": "*",
	jsdom: "^27.1.0"
};
var peerDependenciesMeta = {
	"@jest/environment-jsdom-abstract": {
		optional: true
	},
	"jest-environment-jsdom": {
		optional: true
	},
	jsdom: {
		optional: true
	}
};
var packageJson = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	homepage: homepage,
	repository: repository,
	license: license,
	main: main,
	types: types,
	bin: bin,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta
};

exports.backstage = backstage;
exports.bin = bin;
exports.default = packageJson;
exports.dependencies = dependencies;
exports.description = description;
exports.devDependencies = devDependencies;
exports.files = files;
exports.homepage = homepage;
exports.license = license;
exports.main = main;
exports.name = name;
exports.peerDependencies = peerDependencies;
exports.peerDependenciesMeta = peerDependenciesMeta;
exports.publishConfig = publishConfig;
exports.repository = repository;
exports.scripts = scripts;
exports.types = types;
exports.version = version;
