// @backstage/cli-module-actions/dist/commands/execute.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var ActionsClient = require('../lib/ActionsClient.cjs.js');
var schemaToFlags = require('../lib/schemaToFlags.cjs.js');
var resolveAuth = require('../lib/resolveAuth.cjs.js');
var format = require('../lib/format.cjs.js');

function parseArgs(args) {
  const instanceIdx = args.indexOf("--instance");
  const instanceFlag = instanceIdx !== -1 ? args[instanceIdx + 1] : void 0;
  const skipIndices = /* @__PURE__ */ new Set();
  if (instanceIdx !== -1) {
    skipIndices.add(instanceIdx);
    skipIndices.add(instanceIdx + 1);
  }
  let actionId;
  let actionIdIdx = -1;
  for (let i = 0; i < args.length; i++) {
    if (!skipIndices.has(i) && !args[i].startsWith("-")) {
      actionId = args[i];
      actionIdIdx = i;
      break;
    }
  }
  const wantsHelp = args.includes("--help") || args.includes("-h");
  return { instanceFlag, actionId, actionIdIdx, wantsHelp };
}
function showGenericHelp(info, args) {
  cleye.cli(
    {
      name: info.usage,
      parameters: ["<action-id>"],
      flags: {
        instance: {
          type: String,
          description: "Name of the instance to use"
        }
      }
    },
    void 0,
    args
  );
}
async function showActionHelp(info, actionId, instanceFlag) {
  try {
    const { accessToken, baseUrl } = await resolveAuth.resolveAuth(instanceFlag);
    const client = new ActionsClient.ActionsClient(baseUrl, accessToken);
    const actions = await client.listForPlugin(actionId);
    const action = actions.find((a) => a.id === actionId);
    if (!action) {
      return false;
    }
    const { flags: flagDefs } = schemaToFlags.schemaToFlags(action.schema.input);
    const flags = format.flagDefsToFlagInfo(flagDefs);
    flags.push({
      name: "instance",
      type: "string",
      description: "Name of the instance to use"
    });
    process.stdout.write(
      await format.formatActionHelp({
        action,
        usage: `${info.usage ?? "backstage actions execute"} ${actionId}`,
        flags
      })
    );
    return true;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    process.stderr.write(
      `Unable to retrieve action schema: ${msg}
Showing generic help.
`
    );
    return false;
  }
}
var execute = async ({ args, info }) => {
  const { instanceFlag, actionId, actionIdIdx, wantsHelp } = parseArgs(args);
  if (wantsHelp) {
    if (!actionId || !await showActionHelp(info, actionId, instanceFlag)) {
      showGenericHelp(info, args);
    }
    return;
  }
  if (!actionId) {
    showGenericHelp(info, ["--help", ...args]);
    throw new Error("Action ID is required");
  }
  const { accessToken, baseUrl } = await resolveAuth.resolveAuth(instanceFlag);
  const client = new ActionsClient.ActionsClient(baseUrl, accessToken);
  const actions = await client.listForPlugin(actionId);
  const action = actions.find((a) => a.id === actionId);
  if (!action) {
    throw new Error(
      `Action "${actionId}" not found. Run "actions list" to see available actions.`
    );
  }
  const inputSchema = action.schema.input;
  const { flags: schemaFlags, complexKeys } = schemaToFlags.schemaToFlags(inputSchema);
  const flagArgs = args.filter((_, i) => i !== actionIdIdx);
  const { flags } = cleye.cli(
    {
      name: info.usage,
      flags: {
        ...schemaFlags,
        instance: {
          type: String,
          description: "Name of the instance to use"
        }
      }
    },
    void 0,
    flagArgs
  );
  const allFlags = flags;
  const input = {};
  for (const [key, value] of Object.entries(allFlags)) {
    if (key === "instance" || value === void 0) {
      continue;
    }
    if (complexKeys.has(key) && typeof value === "string") {
      try {
        input[key] = JSON.parse(value);
      } catch {
        throw new Error(`Invalid JSON for --${key}. Expected a JSON string.`);
      }
    } else {
      input[key] = value;
    }
  }
  const output = await client.execute(actionId, input);
  process.stdout.write(`${JSON.stringify(output, null, 2)}
`);
};

exports.default = execute;
