// @backstage/cli-module-build/dist/commands/package/start/command.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var startPackage = require('./startPackage.cjs.js');
var resolveLinkedWorkspace = require('./resolveLinkedWorkspace.cjs.js');
var role = require('../../../lib/role.cjs.js');
var cliCommon = require('@backstage/cli-common');

var command = async ({ args, info }) => {
  const {
    flags: {
      config,
      role: role$1,
      check,
      require: requirePath,
      link,
      entrypoint,
      inspect,
      inspectBrk
    }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        config: {
          type: [String],
          description: "Config files to load instead of app-config.yaml",
          default: []
        },
        role: {
          type: String,
          description: "Run the command with an explicit package role"
        },
        check: {
          type: Boolean,
          description: "Enable type checking and linting if available"
        },
        require: {
          type: String,
          description: "Add a --require argument to the node process"
        },
        link: {
          type: String,
          description: "Link an external workspace for module resolution"
        },
        entrypoint: {
          type: String,
          description: 'The entrypoint to start from, relative to the package root. Can point to either a file (without extension) or a directory (in which case the index file in that directory is used). Defaults to "dev"'
        },
        inspect: {
          type: String,
          description: "Enable the Node.js inspector, optionally at a specific host:port"
        },
        inspectBrk: {
          type: String,
          description: "Enable the Node.js inspector and break before user code starts"
        }
      }
    },
    void 0,
    args
  );
  await startPackage.startPackage({
    role: await role.findRoleFromCommand({ role: role$1 }),
    entrypoint,
    targetDir: cliCommon.targetPaths.dir,
    configPaths: config,
    checksEnabled: Boolean(check),
    linkedWorkspace: await resolveLinkedWorkspace.resolveLinkedWorkspace(link),
    inspectEnabled: inspect || (inspect === "" ? true : void 0),
    inspectBrkEnabled: inspectBrk || (inspectBrk === "" ? true : void 0),
    require: requirePath
  });
};

exports.default = command;
