// @backstage/cli-node/dist/cli-module/CommandGraph.cjs.js
'use strict';

var InternalCliModule = require('../cli-internal/src/InternalCliModule.cjs.js');
var InternalCommandNode = require('../cli-internal/src/InternalCommandNode.cjs.js');
require('node:fs');
require('node:os');
require('node:path');
require('../cli-internal/src/knownPluginPackages.cjs.js');

class CommandGraph {
  #roots = [];
  get roots() {
    return this.#roots;
  }
  add(command, module) {
    const { path } = command;
    let current = this.#roots;
    for (let i = 0; i < path.length - 1; i++) {
      const name2 = path[i];
      let next = current.find((node) => getNodeName(node) === name2);
      if (!next) {
        next = InternalCommandNode.OpaqueCommandTreeNode.createInstance("v1", {
          name: name2,
          children: []
        });
        current.push(next);
      } else if (InternalCommandNode.OpaqueCommandLeafNode.isType(next)) {
        throw new Error(
          formatConflictError(
            path,
            module,
            InternalCommandNode.OpaqueCommandLeafNode.toInternal(next).module
          )
        );
      }
      current = InternalCommandNode.OpaqueCommandTreeNode.toInternal(next).children;
    }
    const name = path[path.length - 1];
    const existing = current.find((node) => getNodeName(node) === name);
    if (existing) {
      throw new Error(
        formatConflictError(path, module, findNodeModule(existing))
      );
    }
    current.push(
      InternalCommandNode.OpaqueCommandLeafNode.createInstance("v1", {
        name,
        command,
        module
      })
    );
  }
}
function getNodeName(node) {
  if (InternalCommandNode.OpaqueCommandTreeNode.isType(node)) {
    return InternalCommandNode.OpaqueCommandTreeNode.toInternal(node).name;
  }
  return InternalCommandNode.OpaqueCommandLeafNode.toInternal(node).name;
}
function findNodeModule(node) {
  if (InternalCommandNode.OpaqueCommandLeafNode.isType(node)) {
    return InternalCommandNode.OpaqueCommandLeafNode.toInternal(node).module;
  }
  for (const child of InternalCommandNode.OpaqueCommandTreeNode.toInternal(node).children) {
    const module = findNodeModule(child);
    if (module) {
      return module;
    }
  }
  return void 0;
}
function getModuleName(module) {
  if (module && InternalCliModule.OpaqueCliModule.isType(module)) {
    return InternalCliModule.OpaqueCliModule.toInternal(module).packageName;
  }
  return void 0;
}
function formatConflictError(path, newModule, existingModule) {
  const command = path.join(" ");
  const newPackage = getModuleName(newModule);
  const existingPackage = getModuleName(existingModule);
  if (newPackage && existingPackage) {
    return `Command "${command}" from "${newPackage}" conflicts with an existing command from "${existingPackage}"`;
  }
  if (newPackage) {
    return `Command "${command}" from "${newPackage}" conflicts with an existing command`;
  }
  if (existingPackage) {
    return `Command "${command}" conflicts with an existing command from "${existingPackage}"`;
  }
  return `Command "${command}" conflicts with an existing command`;
}

exports.CommandGraph = CommandGraph;
