// @backstage/cli-node/dist/cli-module/runCli.cjs.js
'use strict';

var InternalCliModule = require('../cli-internal/src/InternalCliModule.cjs.js');
var InternalCommandNode = require('../cli-internal/src/InternalCommandNode.cjs.js');
require('node:fs');
require('node:os');
require('node:path');
require('../cli-internal/src/knownPluginPackages.cjs.js');
var errors = require('@backstage/errors');
var chalk = require('chalk');
var cleye = require('cleye');
var CommandGraph = require('./CommandGraph.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

function exit(message, code = 1) {
  process.stderr.write(`
${chalk__default.default.red(message)}

`);
  process.exit(code);
}
function exitWithError(error) {
  if (!errors.isError(error)) {
    return exit(errors.stringifyError(error));
  }
  switch (error.name) {
    case "InputError":
      return exit(
        error.message,
        74
        /* input/output error */
      );
    case "NotFoundError":
      return exit(
        error.message,
        127
        /* command not found */
      );
    case "NotImplementedError":
      return exit(
        error.message,
        64
        /* command line usage error */
      );
    case "AuthenticationError":
    case "NotAllowedError":
      return exit(
        error.message,
        77
        /* permission denied */
      );
    case "ExitCodeError":
      return exit(
        error.message,
        "code" in error && typeof error.code === "number" ? error.code : 1
      );
    default:
      return exit(
        errors.stringifyError(error),
        "code" in error && typeof error.code === "number" ? error.code : 1
      );
  }
}
function getNodeName(node) {
  if (InternalCommandNode.OpaqueCommandTreeNode.isType(node)) {
    return InternalCommandNode.OpaqueCommandTreeNode.toInternal(node).name;
  }
  return InternalCommandNode.OpaqueCommandLeafNode.toInternal(node).name;
}
function getNodeDescription(node) {
  if (InternalCommandNode.OpaqueCommandTreeNode.isType(node)) {
    return InternalCommandNode.OpaqueCommandTreeNode.toInternal(node).name;
  }
  return InternalCommandNode.OpaqueCommandLeafNode.toInternal(node).command.description;
}
function createHelpOptions(options) {
  const { nodes, includeHelpCommand, version } = options;
  const visibleNodes = nodes.filter((node) => !InternalCommandNode.isCommandNodeHidden(node));
  return {
    version,
    render(nodesToRender, renderers) {
      const commandsNode = nodesToRender.find((node) => node.id === "commands");
      if (commandsNode) {
        const commandRows = visibleNodes.map((node) => [
          getNodeName(node),
          getNodeDescription(node)
        ]);
        if (includeHelpCommand) {
          commandRows.push(["help", "Display help for command"]);
        }
        commandsNode.data.body.data.tableData = commandRows;
      }
      return renderers.render(nodesToRender);
    }
  };
}
async function executeCommand(commandToExecute, args, programName) {
  const context = {
    args,
    info: {
      usage: [programName, ...commandToExecute.path].join(" "),
      name: commandToExecute.path.join(" ")
    }
  };
  if (typeof commandToExecute.execute === "function") {
    await commandToExecute.execute(context);
  } else {
    const mod = await commandToExecute.execute.loader();
    const fn = typeof mod.default === "function" ? mod.default : mod.default.default;
    await fn(context);
  }
}
async function runCommandLevel(options) {
  const { nodes, argv, programName, commandPath = [], version } = options;
  const name = [programName, ...commandPath].join(" ");
  const commandArgs = argv.slice(1);
  const commands = nodes.map(
    (node) => cleye.command(
      {
        name: getNodeName(node),
        help: false
      },
      async () => {
        try {
          if (InternalCommandNode.OpaqueCommandTreeNode.isType(node)) {
            await runCommandLevel({
              nodes: InternalCommandNode.OpaqueCommandTreeNode.toInternal(node).children,
              argv: commandArgs,
              programName,
              commandPath: [...commandPath, getNodeName(node)],
              version
            });
          } else {
            await executeCommand(
              InternalCommandNode.OpaqueCommandLeafNode.toInternal(node).command,
              commandArgs,
              programName
            );
            process.exit(0);
          }
        } catch (error) {
          exitWithError(error);
        }
      }
    )
  );
  const includeHelpCommand = !nodes.some((node) => getNodeName(node) === "help");
  if (includeHelpCommand && nodes.length > 0) {
    commands.push(
      cleye.command(
        {
          name: "help",
          parameters: ["[command...]"],
          help: false
        },
        async () => {
          await runCommandLevel({
            nodes,
            argv: [...commandArgs, "--help"],
            programName,
            commandPath,
            version
          });
        }
      )
    );
  }
  await cleye.cli(
    {
      name,
      flags: version ? {
        version: {
          type: Boolean,
          alias: "V",
          description: "Show version"
        }
      } : void 0,
      commands,
      help: createHelpOptions({
        nodes,
        includeHelpCommand: includeHelpCommand && nodes.length > 0,
        version
      })
    },
    (parsed) => {
      if (version && parsed.flags.version) {
        console.log(version);
        process.exit(0);
        return;
      }
      if (argv.length === 0) {
        return;
      }
      console.log();
      console.log(
        chalk__default.default.red(`Invalid command: ${[...commandPath, ...argv].join(" ")}`)
      );
      console.log();
      parsed.showHelp();
      process.exit(1);
    },
    [...argv]
  );
}
function handleUnhandledRejection(rejection) {
  exitWithError(new errors.ForwardedError("Unhandled rejection", rejection));
}
function hasVersionFlag(args) {
  const separatorIndex = args.indexOf("--");
  const options = separatorIndex === -1 ? args : args.slice(0, separatorIndex);
  return options.includes("-V") || options.includes("--version");
}
async function runCli(options) {
  const { modules, name, version } = options;
  const graph = new CommandGraph.CommandGraph();
  for (const module of modules) {
    if (!InternalCliModule.OpaqueCliModule.isType(module)) {
      throw new Error(
        `Invalid CLI module: expected a module created with createCliModule`
      );
    }
    for (const commandToAdd of await InternalCliModule.OpaqueCliModule.toInternal(module).commands) {
      graph.add(commandToAdd, module);
    }
  }
  if (!process.listeners("unhandledRejection").includes(handleUnhandledRejection)) {
    process.on("unhandledRejection", handleUnhandledRejection);
  }
  const args = process.argv.slice(2);
  if (version && hasVersionFlag(args)) {
    console.log(version);
    process.exit(0);
    return;
  }
  await runCommandLevel({
    nodes: graph.roots,
    argv: args,
    programName: name,
    version
  });
}

exports.runCli = runCli;
