// @backstage/plugin-catalog-react/dist/hooks/useEntityListProvider.esm.js
import { jsx } from 'react/jsx-runtime';
import { useApi } from '@backstage/core-plugin-api';
import { createVersionedContext, createVersionedValueMap, useVersionedContext } from '@backstage/version-bridge';
import { compact, isEqual } from 'lodash';
import qs from 'qs';
import { createContext, useState, useMemo, useRef, useEffect, useCallback, useContext } from 'react';
import { useLocation } from 'react-router-dom';
import useAsyncFn from 'react-use/esm/useAsyncFn';
import useDebounce from 'react-use/esm/useDebounce';
import useMountedState from 'react-use/esm/useMountedState';
import { catalogApiRef } from '../api.esm.js';
import { reduceCatalogFilters, reduceBackendCatalogFilters, reduceEntityFilters } from '../utils/filters.esm.js';

const NewEntityListContext = createVersionedContext("entity-list-context");
const OldEntityListContext = createContext(void 0);
const EntityListProvider = (props) => {
  const isMounted = useMountedState();
  const catalogApi = useApi(catalogApiRef);
  const [requestedFilters, setRequestedFilters] = useState(
    {}
  );
  const location = useLocation();
  let paginationMode = "none";
  if (props.pagination === true) {
    paginationMode = "cursor";
  } else if (typeof props.pagination === "object") {
    paginationMode = props.pagination.mode ?? "cursor";
  }
  const paginationLimit = typeof props.pagination === "object" ? props.pagination.limit ?? 20 : 20;
  const {
    queryParameters,
    cursor: initialCursor,
    offset: initialOffset,
    limit: initialLimit
  } = useMemo(() => {
    const parsed = qs.parse(location.search, {
      ignoreQueryPrefix: true,
      arrayLimit: 1e4
    });
    let limit2 = paginationLimit;
    if (typeof parsed.limit === "string") {
      const queryLimit = Number.parseInt(parsed.limit, 10);
      if (!isNaN(queryLimit)) {
        limit2 = queryLimit;
      }
    }
    const offset2 = typeof parsed.offset === "string" && paginationMode === "offset" ? Number.parseInt(parsed.offset, 10) : void 0;
    return {
      queryParameters: parsed.filters ?? {},
      cursor: typeof parsed.cursor === "string" && paginationMode === "cursor" ? parsed.cursor : void 0,
      offset: paginationMode === "offset" && offset2 && !isNaN(offset2) ? offset2 : void 0,
      limit: limit2
    };
  }, [paginationMode, location.search, paginationLimit]);
  const [cursor, setCursor] = useState(initialCursor);
  const [offset, setOffset] = useState(initialOffset);
  const [limit, setLimit] = useState(initialLimit);
  const [backendState, setBackendState] = useState({
    backendEntities: []
  });
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();
  const [refreshToken, setRefreshToken] = useState(0);
  const lastFetchParamsRef = useRef(void 0);
  const fetchGenRef = useRef(0);
  const adjustedFilters = useMemo(() => {
    const kindValue = requestedFilters.kind?.value?.toLocaleLowerCase("en-US");
    return kindValue === "user" || kindValue === "group" ? { ...requestedFilters, owners: void 0 } : requestedFilters;
  }, [requestedFilters]);
  const refresh = async () => {
    const compacted = compact(Object.values(adjustedFilters));
    let fetchParams;
    let doFetch;
    if (paginationMode !== "none") {
      if (cursor) {
        fetchParams = { cursor, limit };
        doFetch = async () => {
          const response = await catalogApi.queryEntities({
            cursor,
            limit,
            totalItems: "exclude"
          });
          return {
            backendEntities: response.items,
            pageInfo: response.pageInfo
          };
        };
      } else {
        const backendFilter = reduceCatalogFilters(compacted);
        fetchParams = { ...backendFilter, limit, offset };
        doFetch = async () => {
          const response = await catalogApi.queryEntities({
            ...backendFilter,
            limit,
            offset,
            totalItems: "exclude"
          });
          return {
            backendEntities: response.items,
            pageInfo: response.pageInfo
          };
        };
      }
    } else {
      const backendFilter = reduceBackendCatalogFilters(compacted);
      const { orderFields } = reduceCatalogFilters(compacted);
      fetchParams = { filter: backendFilter, order: orderFields };
      doFetch = async () => {
        const response = await catalogApi.getEntities({
          filter: backendFilter,
          order: orderFields
        });
        return { backendEntities: response.items };
      };
    }
    if (compacted.length === 0 && lastFetchParamsRef.current === void 0 && !cursor) {
      setLoading(false);
      return;
    }
    if (isEqual(fetchParams, lastFetchParamsRef.current)) {
      return;
    }
    lastFetchParamsRef.current = fetchParams;
    const gen = ++fetchGenRef.current;
    setLoading(true);
    setError(void 0);
    try {
      const result = await doFetch();
      if (gen === fetchGenRef.current) {
        setBackendState(result);
      }
    } catch (e) {
      if (gen === fetchGenRef.current) {
        lastFetchParamsRef.current = void 0;
        setError(e);
      }
    } finally {
      if (gen === fetchGenRef.current) {
        setLoading(false);
      }
    }
  };
  useDebounce(refresh, 10, [
    adjustedFilters,
    cursor,
    limit,
    offset,
    refreshToken
  ]);
  const [{ value: totalItems, loading: totalItemsLoading }, refreshCount] = useAsyncFn(async () => {
    if (paginationMode === "none") {
      return void 0;
    }
    const compacted = compact(Object.values(adjustedFilters));
    if (compacted.length === 0) {
      return void 0;
    }
    const backendFilter = reduceCatalogFilters(compacted);
    try {
      const response = await catalogApi.queryEntities({
        ...backendFilter,
        limit: 0
      });
      return response.totalItems;
    } catch {
      return void 0;
    }
  }, [catalogApi, paginationMode, adjustedFilters]);
  useDebounce(refreshCount, 10, [adjustedFilters, refreshToken]);
  const entities = useMemo(() => {
    const compacted = compact(Object.values(adjustedFilters));
    const entityFilter = reduceEntityFilters(compacted);
    return backendState.backendEntities.filter(entityFilter);
  }, [adjustedFilters, backendState.backendEntities]);
  useEffect(() => {
    if (!isMounted() || Object.keys(requestedFilters).length === 0) {
      return;
    }
    const queryParams = Object.keys(requestedFilters).reduce((params, key) => {
      const filter = requestedFilters[key];
      if (filter?.toQueryValue) {
        params[key] = filter.toQueryValue();
      }
      return params;
    }, {});
    const oldParams = qs.parse(location.search, {
      ignoreQueryPrefix: true,
      arrayLimit: 1e4
    });
    const newParams = qs.stringify(
      {
        ...oldParams,
        filters: queryParams,
        ...paginationMode === "none" ? {} : { cursor, limit, offset }
      },
      { addQueryPrefix: true, arrayFormat: "repeat" }
    );
    const newUrl = `${window.location.pathname}${newParams}`;
    window.history?.replaceState(null, document.title, newUrl);
  }, [
    cursor,
    isMounted,
    limit,
    location.search,
    offset,
    requestedFilters,
    paginationMode
  ]);
  const updateFilters = useCallback(
    (update) => {
      if (paginationMode === "cursor") {
        setCursor(void 0);
      } else if (paginationMode === "offset") {
        setOffset(0);
      }
      setRequestedFilters((prevFilters) => {
        const newFilters = typeof update === "function" ? update(prevFilters) : update;
        return { ...prevFilters, ...newFilters };
      });
    },
    [paginationMode]
  );
  const triggerRefresh = useCallback(() => {
    lastFetchParamsRef.current = void 0;
    setRefreshToken((t) => t + 1);
  }, []);
  const pageInfo = useMemo(() => {
    if (paginationMode !== "cursor") {
      return void 0;
    }
    const prevCursor = backendState.pageInfo?.prevCursor;
    const nextCursor = backendState.pageInfo?.nextCursor;
    return {
      prev: prevCursor ? () => setCursor(prevCursor) : void 0,
      next: nextCursor ? () => setCursor(nextCursor) : void 0
    };
  }, [paginationMode, backendState.pageInfo]);
  const value = useMemo(
    () => ({
      filters: requestedFilters,
      entities,
      backendEntities: backendState.backendEntities,
      updateFilters,
      queryParameters,
      loading,
      error,
      pageInfo,
      totalItems: paginationMode === "none" ? entities.length : totalItems ?? backendState.totalItems,
      totalItemsLoading: paginationMode !== "none" && totalItemsLoading,
      limit,
      offset,
      setLimit,
      setOffset,
      paginationMode,
      refresh: triggerRefresh
    }),
    [
      requestedFilters,
      entities,
      backendState,
      totalItems,
      totalItemsLoading,
      updateFilters,
      queryParameters,
      loading,
      error,
      pageInfo,
      paginationMode,
      limit,
      offset,
      setLimit,
      setOffset,
      triggerRefresh
    ]
  );
  return /* @__PURE__ */ jsx(OldEntityListContext.Provider, { value, children: /* @__PURE__ */ jsx(
    NewEntityListContext.Provider,
    {
      value: createVersionedValueMap({ 1: value }),
      children: props.children
    }
  ) });
};
function useEntityList() {
  const versionedHolder = useVersionedContext("entity-list-context");
  const oldContext = useContext(OldEntityListContext);
  if (versionedHolder) {
    const value = versionedHolder.atVersion(1);
    if (!value) {
      throw new Error("EntityListContext v1 not available");
    }
    return value;
  }
  if (oldContext) {
    return oldContext;
  }
  throw new Error("useEntityList must be used within EntityListProvider");
}

export { EntityListProvider, NewEntityListContext, OldEntityListContext, useEntityList };
