// @backstage/cli-module-auth/dist/lib/storage.cjs.js
'use strict';

var errors = require('@backstage/errors');
var fs = require('fs-extra');
var os = require('node:os');
var path = require('node:path');
var lockfile = require('proper-lockfile');
var YAML = require('yaml');
var v3 = require('zod/v3');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var os__default = /*#__PURE__*/_interopDefaultCompat(os);
var path__default = /*#__PURE__*/_interopDefaultCompat(path);
var lockfile__default = /*#__PURE__*/_interopDefaultCompat(lockfile);
var YAML__default = /*#__PURE__*/_interopDefaultCompat(YAML);

const METADATA_FILE = "auth-instances.yaml";
const INSTANCE_NAME_PATTERN = /^[a-zA-Z0-9._:@-]+$/;
const storedInstanceSchema = v3.z.object({
  name: v3.z.string().min(1).regex(INSTANCE_NAME_PATTERN, "Instance name contains invalid characters"),
  baseUrl: v3.z.string().url(),
  clientId: v3.z.string().min(1),
  issuedAt: v3.z.number().int().nonnegative(),
  accessTokenExpiresAt: v3.z.number().int().nonnegative(),
  selected: v3.z.boolean().optional(),
  metadata: v3.z.record(v3.z.string(), v3.z.unknown()).optional()
});
const authYamlSchema = v3.z.object({
  instances: v3.z.array(storedInstanceSchema).default([])
});
function getMetadataFilePath() {
  const root = process.env.XDG_CONFIG_HOME || (process.platform === "win32" ? process.env.APPDATA || path__default.default.join(os__default.default.homedir(), "AppData", "Roaming") : path__default.default.join(os__default.default.homedir(), ".config"));
  return path__default.default.join(root, "backstage-cli", METADATA_FILE);
}
async function readAll() {
  const file = getMetadataFilePath();
  if (!await fs__default.default.pathExists(file)) {
    return { instances: [] };
  }
  const text = await fs__default.default.readFile(file, "utf8");
  if (!text.trim()) {
    return { instances: [] };
  }
  try {
    const doc = YAML__default.default.parse(text);
    const parsed = authYamlSchema.safeParse(doc);
    if (parsed.success) {
      return parsed.data;
    }
    return { instances: [] };
  } catch {
    return { instances: [] };
  }
}
async function writeAll(data) {
  const file = getMetadataFilePath();
  await fs__default.default.ensureDir(path__default.default.dirname(file));
  const yaml = YAML__default.default.stringify(authYamlSchema.parse(data), { indentSeq: false });
  await fs__default.default.writeFile(file, yaml, { encoding: "utf8", mode: 384 });
}
async function getAllInstances() {
  const { instances } = await readAll();
  if (instances.length === 0) {
    return { instances: [], selected: void 0 };
  }
  const selected = instances.find((i) => i.selected) ?? instances[0];
  return {
    instances: instances.map((i) => ({
      ...i,
      selected: i.name === selected.name
    })),
    selected
  };
}
async function getInstanceByName(name) {
  const { instances } = await readAll();
  const instance = instances.find((i) => i.name === name);
  if (!instance) {
    throw new errors.NotFoundError(`Instance '${name}' not found`);
  }
  return instance;
}
async function upsertInstance(instance) {
  const data = await readAll();
  const idx = data.instances.findIndex((i) => i.name === instance.name);
  if (idx === -1) {
    data.instances.push(instance);
  } else {
    data.instances[idx] = instance;
  }
  await writeAll(data);
}
async function removeInstance(name) {
  const data = await readAll();
  const next = data.instances.filter((i) => i.name !== name);
  if (next.length !== data.instances.length) {
    await writeAll({ instances: next });
  }
}
async function setSelectedInstance(name) {
  return withMetadataLock(async () => {
    const data = await readAll();
    let found = false;
    data.instances = data.instances.map((i) => {
      if (i.name === name) {
        found = true;
        return { ...i, selected: true };
      }
      const { selected, ...rest } = i;
      return { ...rest, selected: false };
    });
    if (!found) {
      throw new Error(`Unknown instance '${name}'`);
    }
    await writeAll(data);
  });
}
async function withMetadataLock(fn) {
  const file = getMetadataFilePath();
  await fs__default.default.ensureDir(path__default.default.dirname(file));
  if (!await fs__default.default.pathExists(file)) {
    await fs__default.default.writeFile(file, "", { encoding: "utf8", mode: 384 });
  }
  const release = await lockfile__default.default.lock(file, {
    retries: { retries: 5, factor: 1.5, minTimeout: 100, maxTimeout: 1e3 }
  });
  try {
    return await fn();
  } finally {
    await release();
  }
}

exports.getAllInstances = getAllInstances;
exports.getInstanceByName = getInstanceByName;
exports.removeInstance = removeInstance;
exports.setSelectedInstance = setSelectedInstance;
exports.upsertInstance = upsertInstance;
exports.withMetadataLock = withMetadataLock;
