// @backstage/cli-node/dist/auth/CliAuth.cjs.js
'use strict';

var storage = require('./storage.cjs.js');
var secretStore = require('./secretStore.cjs.js');
var authIdentifiers = require('./authIdentifiers.cjs.js');
var httpJson = require('./httpJson.cjs.js');
var v3 = require('zod/v3');

const TokenResponseSchema = v3.z.object({
  access_token: v3.z.string().min(1),
  token_type: v3.z.string().min(1),
  expires_in: v3.z.number().positive().finite(),
  refresh_token: v3.z.string().min(1).optional()
});
class CliAuth {
  #secretStore;
  #instance;
  /**
   * Resolve the current auth instance and return a ready-to-use
   * {@link CliAuth} object. Throws when no instance can be found.
   */
  static async create(options) {
    const instance = await storage.getSelectedInstance(options?.instanceName);
    const secretStore$1 = await secretStore.getSecretStore();
    return new CliAuth(instance, secretStore$1);
  }
  constructor(instance, secretStore) {
    this.#instance = instance;
    this.#secretStore = secretStore;
  }
  /** Returns the name of the resolved auth instance. */
  getInstanceName() {
    return this.#instance.name;
  }
  /** Returns the base URL of the resolved auth instance. */
  getBaseUrl() {
    return this.#instance.baseUrl;
  }
  /**
   * Returns a valid access token, refreshing it first if the current
   * token is expired or about to expire.
   */
  async getAccessToken() {
    if (storage.accessTokenNeedsRefresh(this.#instance)) {
      await this.#refreshAccessToken();
    }
    const service = authIdentifiers.getAuthInstanceService(this.#instance.name);
    const token = await this.#secretStore.get(service, "accessToken");
    if (!token) {
      throw new Error(
        'No access token found. Run "auth login" to authenticate.'
      );
    }
    return token;
  }
  /**
   * Reads a per-instance metadata value previously stored by the
   * auth module (e.g. `pluginSources`).
   */
  async getMetadata(key) {
    return storage.getInstanceMetadata(this.#instance.name, key);
  }
  /**
   * Writes a per-instance metadata value to the on-disk instance store.
   */
  async setMetadata(key, value) {
    return storage.updateInstanceMetadata(this.#instance.name, key, value);
  }
  async #refreshAccessToken() {
    const service = authIdentifiers.getAuthInstanceService(this.#instance.name);
    const refreshToken = await this.#secretStore.get(service, "refreshToken") ?? "";
    if (!refreshToken) {
      throw new Error(
        "Access token is expired and no refresh token is available"
      );
    }
    const response = await httpJson.httpJson(
      `${this.#instance.baseUrl}/api/auth/v1/token`,
      {
        method: "POST",
        body: {
          grant_type: "refresh_token",
          refresh_token: refreshToken
        },
        signal: AbortSignal.timeout(3e4)
      }
    );
    const parsed = TokenResponseSchema.safeParse(response);
    if (!parsed.success) {
      throw new Error(`Invalid token response: ${parsed.error.message}`);
    }
    const token = parsed.data;
    await this.#secretStore.set(service, "accessToken", token.access_token);
    if (token.refresh_token) {
      await this.#secretStore.set(service, "refreshToken", token.refresh_token);
    }
    const issuedAt = Date.now();
    const accessTokenExpiresAt = Date.now() + token.expires_in * 1e3;
    this.#instance = { ...this.#instance, issuedAt, accessTokenExpiresAt };
    await storage.updateInstance(this.#instance.name, {
      issuedAt,
      accessTokenExpiresAt
    });
  }
}

exports.CliAuth = CliAuth;
