// @backstage/catalog-model/dist/model/operations/declareLabel.esm.js
import { z } from 'zod/v3';
import { jsonSchemaSchema } from '../jsonSchema/zod.esm.js';

const opDeclareLabelV1Schema = z.strictObject({
  op: z.literal("declareLabel.v1"),
  /**
   * The name of the label, e.g. "backstage.io/source-location".
   */
  name: z.string(),
  /**
   * The properties that apply to this label.
   */
  properties: z.strictObject({
    /**
     * A human-readable title that can be used for display purposes instead of
     * the technical name.
     */
    title: z.string().optional(),
    /**
     * A human-readable description of the label.
     */
    description: z.string(),
    /**
     * The JSON schema that values of this label must conform to.
     *
     * @remarks
     *
     * If not provided, the label is assumed to be a simple string with no
     * particular schema.
     */
    schema: z.strictObject({
      jsonSchema: jsonSchemaSchema
    }).optional()
  })
});
function createDeclareLabelOp(input) {
  return opDeclareLabelV1Schema.parse({
    ...input,
    op: "declareLabel.v1"
  });
}

export { createDeclareLabelOp, opDeclareLabelV1Schema };
