// @backstage/catalog-model/dist/model/operations/updateRelation.esm.js
import { z } from 'zod/v3';

const opUpdateRelationV1Schema = z.strictObject({
  op: z.literal("updateRelation.v1"),
  /**
   * The kind that this relation originates from, e.g. "Component".
   */
  fromKind: z.string(),
  /**
   * The technical type of the relation, e.g. "ownedBy".
   */
  type: z.string(),
  /**
   * The kind that this relation points to, e.g. "Group".
   */
  toKind: z.string(),
  /**
   * The properties that apply to this relation.
   */
  properties: z.strictObject({
    /**
     * The technical type of the reverse relation, e.g. "ownerOf".
     */
    reverseType: z.string().optional(),
    /**
     * A human-readable title for the relation type, e.g. "owned by".
     */
    title: z.string().optional(),
    /**
     * A human-readable description of the relation.
     */
    description: z.string().optional()
  })
});
function createUpdateRelationOp(input) {
  return opUpdateRelationV1Schema.parse({
    ...input,
    op: "updateRelation.v1"
  });
}

export { createUpdateRelationOp, opUpdateRelationV1Schema };
