// @backstage/catalog-model/dist/model/operations/updateLabel.cjs.js
'use strict';

var v3 = require('zod/v3');
var zod = require('../jsonSchema/zod.cjs.js');

const opUpdateLabelV1Schema = v3.z.strictObject({
  op: v3.z.literal("updateLabel.v1"),
  /**
   * The name of the label, e.g. "backstage.io/source-location".
   */
  name: v3.z.string(),
  /**
   * The properties that apply to this label.
   */
  properties: v3.z.strictObject({
    /**
     * A human-readable title that can be used for display purposes instead of
     * the technical name.
     */
    title: v3.z.string().optional(),
    /**
     * A human-readable description of the label.
     */
    description: v3.z.string().optional(),
    /**
     * The JSON schema that values of this label must conform to.
     *
     * @remarks
     *
     * If not provided, the label is assumed to be a simple string with no
     * particular schema.
     */
    schema: v3.z.strictObject({
      jsonSchema: zod.jsonSchemaSchema
    }).optional()
  })
});
function createUpdateLabelOp(input) {
  return opUpdateLabelV1Schema.parse({
    ...input,
    op: "updateLabel.v1"
  });
}

exports.createUpdateLabelOp = createUpdateLabelOp;
exports.opUpdateLabelV1Schema = opUpdateLabelV1Schema;
