// @backstage/catalog-model/dist/model/operations/updateAnnotation.esm.js
import { z } from 'zod/v3';
import { jsonSchemaSchema } from '../jsonSchema/zod.esm.js';

const opUpdateAnnotationV1Schema = z.strictObject({
  op: z.literal("updateAnnotation.v1"),
  /**
   * The name of the annotation, e.g. "backstage.io/techdocs-ref".
   */
  name: z.string(),
  /**
   * The properties that apply to this annotation.
   */
  properties: z.strictObject({
    /**
     * A human-readable title that can be used for display purposes instead of
     * the technical name.
     */
    title: z.string().optional(),
    /**
     * A human-readable description of the annotation.
     */
    description: z.string().optional(),
    /**
     * The JSON schema that values of this annotation must conform to.
     *
     * @remarks
     *
     * If not provided, the annotation is assumed to be a simple string with no
     * particular schema.
     */
    schema: z.strictObject({
      jsonSchema: jsonSchemaSchema
    }).optional()
  })
});
function createUpdateAnnotationOp(input) {
  return opUpdateAnnotationV1Schema.parse({
    ...input,
    op: "updateAnnotation.v1"
  });
}

export { createUpdateAnnotationOp, opUpdateAnnotationV1Schema };
