// @backstage/cli-module-lint/dist/commands/package/lint.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('fs-extra');
var cleye = require('cleye');
var cliCommon = require('@backstage/cli-common');
var eslint = require('eslint');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

var lint = async ({ args, info }) => {
  const {
    flags: { fix, format, outputFile, maxWarnings },
    _: directories
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      parameters: ["[directories...]"],
      flags: {
        fix: {
          type: Boolean,
          description: "Attempt to automatically fix violations"
        },
        format: {
          type: String,
          description: "Lint report output format",
          default: "eslint-formatter-friendly"
        },
        outputFile: {
          type: String,
          description: "Write the lint report to a file instead of stdout"
        },
        maxWarnings: {
          type: String,
          description: "Fail if more than this number of warnings. -1 allows warnings. (default: -1)"
        }
      }
    },
    void 0,
    args
  );
  const eslint$1 = new eslint.ESLint({
    cwd: cliCommon.targetPaths.dir,
    fix,
    extensions: ["js", "jsx", "ts", "tsx", "mjs", "cjs"]
  });
  const results = await eslint$1.lintFiles(
    directories.length ? directories : ["."]
  );
  const maxWarningsNum = maxWarnings ? +maxWarnings : -1;
  const ignoreWarnings = maxWarningsNum === -1;
  const failed = results.some((r) => r.errorCount > 0) || !ignoreWarnings && results.reduce((current, next) => current + next.warningCount, 0) > maxWarningsNum;
  if (fix) {
    await eslint.ESLint.outputFixes(results);
  }
  const formatter = await eslint$1.loadFormatter(format);
  if (format === "eslint-formatter-friendly") {
    process.chdir(cliCommon.targetPaths.rootDir);
  }
  const resultText = await formatter.format(results);
  if (resultText) {
    if (outputFile) {
      await fs__default.default.writeFile(cliCommon.targetPaths.resolve(outputFile), resultText);
    } else {
      console.log(resultText);
    }
  }
  if (failed) {
    process.exit(1);
  }
};

exports.default = lint;
