// @backstage/cli-node/dist/auth/secretStore.cjs.js
'use strict';

var node_fs = require('node:fs');
var os = require('node:os');
var path = require('node:path');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefaultCompat(os);
var path__default = /*#__PURE__*/_interopDefaultCompat(path);

async function loadKeytar() {
  try {
    const keytar = require("keytar");
    if (keytar && typeof keytar.getPassword === "function") {
      return keytar;
    }
  } catch {
  }
  return void 0;
}
class KeytarSecretStore {
  keytar;
  constructor(keytar) {
    this.keytar = keytar;
  }
  async get(service, account) {
    const result = await this.keytar.getPassword(service, account);
    return result ?? void 0;
  }
  async set(service, account, secret) {
    await this.keytar.setPassword(service, account, secret);
  }
  async delete(service, account) {
    await this.keytar.deletePassword(service, account);
  }
}
async function pathExists(p) {
  try {
    await node_fs.promises.stat(p);
    return true;
  } catch {
    return false;
  }
}
class FileSecretStore {
  baseDir;
  constructor() {
    const root = process.env.XDG_DATA_HOME || (process.platform === "win32" ? process.env.APPDATA || path__default.default.join(os__default.default.homedir(), "AppData", "Roaming") : path__default.default.join(os__default.default.homedir(), ".local", "share"));
    this.baseDir = path__default.default.join(root, "backstage-cli", "auth-secrets");
  }
  filePath(service, account) {
    return path__default.default.join(
      this.baseDir,
      encodeURIComponent(service),
      `${encodeURIComponent(account)}.secret`
    );
  }
  async get(service, account) {
    const file = this.filePath(service, account);
    if (!await pathExists(file)) {
      return void 0;
    }
    return await node_fs.promises.readFile(file, "utf8");
  }
  async set(service, account, secret) {
    const file = this.filePath(service, account);
    await node_fs.promises.mkdir(path__default.default.dirname(file), { recursive: true });
    await node_fs.promises.writeFile(file, secret, { encoding: "utf8", mode: 384 });
  }
  async delete(service, account) {
    const file = this.filePath(service, account);
    try {
      await node_fs.promises.unlink(file);
    } catch (err) {
      if (err.code !== "ENOENT") {
        throw err;
      }
    }
  }
}
let singleton;
async function getSecretStore() {
  if (!singleton) {
    const keytar = await loadKeytar();
    if (keytar) {
      singleton = new KeytarSecretStore(keytar);
    } else {
      singleton = new FileSecretStore();
    }
  }
  return singleton;
}

exports.getSecretStore = getSecretStore;
