// @backstage/cli-node/dist/auth/storage.cjs.js
'use strict';

var errors = require('@backstage/errors');
var node_fs = require('node:fs');
var os = require('node:os');
var path = require('node:path');
var lockfile = require('proper-lockfile');
var YAML = require('yaml');
var v3 = require('zod/v3');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefaultCompat(os);
var path__default = /*#__PURE__*/_interopDefaultCompat(path);
var lockfile__default = /*#__PURE__*/_interopDefaultCompat(lockfile);
var YAML__default = /*#__PURE__*/_interopDefaultCompat(YAML);

const METADATA_FILE = "auth-instances.yaml";
const INSTANCE_NAME_PATTERN = /^[a-zA-Z0-9._:@-]+$/;
const storedInstanceSchema = v3.z.object({
  name: v3.z.string().min(1).regex(INSTANCE_NAME_PATTERN, "Instance name contains invalid characters"),
  baseUrl: v3.z.string().url(),
  clientId: v3.z.string().min(1),
  issuedAt: v3.z.number().int().nonnegative(),
  accessTokenExpiresAt: v3.z.number().int().nonnegative(),
  selected: v3.z.boolean().optional(),
  metadata: v3.z.record(v3.z.string(), v3.z.unknown()).optional()
});
const authYamlSchema = v3.z.object({
  instances: v3.z.array(storedInstanceSchema).default([])
});
async function pathExists(p) {
  try {
    await node_fs.promises.stat(p);
    return true;
  } catch {
    return false;
  }
}
function getMetadataFilePath() {
  const root = process.env.XDG_CONFIG_HOME || (process.platform === "win32" ? process.env.APPDATA || path__default.default.join(os__default.default.homedir(), "AppData", "Roaming") : path__default.default.join(os__default.default.homedir(), ".config"));
  return path__default.default.join(root, "backstage-cli", METADATA_FILE);
}
async function readAll() {
  const file = getMetadataFilePath();
  if (!await pathExists(file)) {
    return { instances: [] };
  }
  const text = await node_fs.promises.readFile(file, "utf8");
  if (!text.trim()) {
    return { instances: [] };
  }
  try {
    const doc = YAML__default.default.parse(text);
    const parsed = authYamlSchema.safeParse(doc);
    if (parsed.success) {
      return parsed.data;
    }
    return { instances: [] };
  } catch {
    return { instances: [] };
  }
}
async function writeAll(data) {
  const file = getMetadataFilePath();
  await node_fs.promises.mkdir(path__default.default.dirname(file), { recursive: true });
  const yaml = YAML__default.default.stringify(authYamlSchema.parse(data), { indentSeq: false });
  await node_fs.promises.writeFile(file, yaml, { encoding: "utf8", mode: 384 });
}
async function withMetadataLock(fn) {
  const file = getMetadataFilePath();
  await node_fs.promises.mkdir(path__default.default.dirname(file), { recursive: true });
  if (!await pathExists(file)) {
    await node_fs.promises.writeFile(file, "", { encoding: "utf8", mode: 384 });
  }
  const release = await lockfile__default.default.lock(file, {
    retries: { retries: 5, factor: 1.5, minTimeout: 100, maxTimeout: 1e3 }
  });
  try {
    return await fn();
  } finally {
    await release();
  }
}
async function getAllInstances() {
  const { instances } = await readAll();
  if (instances.length === 0) {
    return { instances: [], selected: void 0 };
  }
  const selected = instances.find((i) => i.selected) ?? instances[0];
  return {
    instances: instances.map((i) => ({
      ...i,
      selected: i.name === selected.name
    })),
    selected
  };
}
async function getSelectedInstance(instanceName) {
  if (instanceName) {
    return await getInstanceByName(instanceName);
  }
  const { selected } = await getAllInstances();
  if (!selected) {
    throw new Error(
      'No instances found. Run "auth login" to authenticate first.'
    );
  }
  return selected;
}
async function getInstanceByName(name) {
  const { instances } = await readAll();
  const instance = instances.find((i) => i.name === name);
  if (!instance) {
    throw new errors.NotFoundError(`Instance '${name}' not found`);
  }
  return instance;
}
async function getInstanceMetadata(instanceName, key) {
  const instance = await getInstanceByName(instanceName);
  return instance.metadata?.[key];
}
async function updateInstanceMetadata(instanceName, key, value) {
  return withMetadataLock(async () => {
    const data = await readAll();
    const idx = data.instances.findIndex((i) => i.name === instanceName);
    if (idx === -1) {
      throw new errors.NotFoundError(`Instance '${instanceName}' not found`);
    }
    data.instances[idx] = {
      ...data.instances[idx],
      metadata: { ...data.instances[idx].metadata, [key]: value }
    };
    await writeAll(data);
  });
}
async function updateInstance(instanceName, updates) {
  return withMetadataLock(async () => {
    const data = await readAll();
    const idx = data.instances.findIndex((i) => i.name === instanceName);
    if (idx === -1) {
      throw new errors.NotFoundError(`Instance '${instanceName}' not found`);
    }
    data.instances[idx] = { ...data.instances[idx], ...updates };
    await writeAll(data);
  });
}
function accessTokenNeedsRefresh(instance) {
  return instance.accessTokenExpiresAt <= Date.now() + 2 * 6e4;
}

exports.accessTokenNeedsRefresh = accessTokenNeedsRefresh;
exports.getAllInstances = getAllInstances;
exports.getInstanceByName = getInstanceByName;
exports.getInstanceMetadata = getInstanceMetadata;
exports.getMetadataFilePath = getMetadataFilePath;
exports.getSelectedInstance = getSelectedInstance;
exports.readAll = readAll;
exports.updateInstance = updateInstance;
exports.updateInstanceMetadata = updateInstanceMetadata;
