// @backstage/catalog-model/dist/model/modelActions/addKindVersion.cjs.js
'use strict';

var reduceKindSchema = require('../jsonSchema/reduceKindSchema.cjs.js');
var validateKindRootSchemaSemantics = require('../jsonSchema/validateKindRootSchemaSemantics.cjs.js');
var validateMetaSchema = require('../jsonSchema/validateMetaSchema.cjs.js');
var declareKindVersion = require('../operations/declareKindVersion.cjs.js');

function opsFromCatalogModelAddKindVersion(definition) {
  const ops = [];
  for (const version of definition.versions) {
    const jsonSchema = reduceKindSchema.reduceKindSchema(version.schema.jsonSchema);
    validateMetaSchema.validateMetaSchema(jsonSchema);
    validateKindRootSchemaSemantics.validateKindRootSchemaSemantics(jsonSchema);
    const names = Array.isArray(version.name) ? version.name : [version.name];
    for (const name of names) {
      const specTypes = version.specType ? [version.specType].flat() : [void 0];
      for (const specType of specTypes) {
        ops.push(
          declareKindVersion.createDeclareKindVersionOp({
            kind: definition.kind,
            name,
            specType,
            properties: {
              description: version.description,
              relationFields: version.relationFields,
              schema: {
                jsonSchema
              }
            }
          })
        );
      }
    }
  }
  return ops;
}

exports.opsFromCatalogModelAddKindVersion = opsFromCatalogModelAddKindVersion;
