// @backstage/catalog-model/dist/model/modelActions/addKindVersion.esm.js
import { reduceKindSchema } from '../jsonSchema/reduceKindSchema.esm.js';
import { validateKindRootSchemaSemantics } from '../jsonSchema/validateKindRootSchemaSemantics.esm.js';
import { validateMetaSchema } from '../jsonSchema/validateMetaSchema.esm.js';
import { createDeclareKindVersionOp } from '../operations/declareKindVersion.esm.js';

function opsFromCatalogModelAddKindVersion(definition) {
  const ops = [];
  for (const version of definition.versions) {
    const jsonSchema = reduceKindSchema(version.schema.jsonSchema);
    validateMetaSchema(jsonSchema);
    validateKindRootSchemaSemantics(jsonSchema);
    const names = Array.isArray(version.name) ? version.name : [version.name];
    for (const name of names) {
      const specTypes = version.specType ? [version.specType].flat() : [void 0];
      for (const specType of specTypes) {
        ops.push(
          createDeclareKindVersionOp({
            kind: definition.kind,
            name,
            specType,
            properties: {
              description: version.description,
              relationFields: version.relationFields,
              schema: {
                jsonSchema
              }
            }
          })
        );
      }
    }
  }
  return ops;
}

export { opsFromCatalogModelAddKindVersion };
