// @backstage/catalog-model/dist/model/modelActions/addRelationPair.esm.js
import { createDeclareRelationOp } from '../operations/declareRelation.esm.js';

function opsFromCatalogModelRelationPair(relationPair) {
  const ops = [];
  for (const firstKind of [relationPair.fromKind].flat()) {
    for (const secondKind of [relationPair.toKind].flat()) {
      ops.push(
        createDeclareRelationOp({
          fromKind: firstKind,
          type: relationPair.forward.type,
          toKind: secondKind,
          properties: {
            reverseType: relationPair.reverse.type,
            title: relationPair.forward.title,
            description: relationPair.description
          }
        })
      );
      ops.push(
        createDeclareRelationOp({
          fromKind: secondKind,
          type: relationPair.reverse.type,
          toKind: firstKind,
          properties: {
            reverseType: relationPair.forward.type,
            title: relationPair.reverse.title,
            description: relationPair.description
          }
        })
      );
    }
  }
  return ops;
}

export { opsFromCatalogModelRelationPair };
