// @backstage/catalog-model/dist/model/modelActions/updateKind.esm.js
import { reduceKindSchema } from '../jsonSchema/reduceKindSchema.esm.js';
import { validateMetaSchema } from '../jsonSchema/validateMetaSchema.esm.js';
import { createUpdateKindOp } from '../operations/updateKind.esm.js';
import { createUpdateKindVersionOp } from '../operations/updateKindVersion.esm.js';

function opsFromCatalogModelUpdateKind(kind) {
  const ops = [];
  if (kind.names.singular || kind.names.plural || kind.description) {
    ops.push(
      createUpdateKindOp({
        kind: kind.names.kind,
        properties: {
          singular: kind.names.singular,
          plural: kind.names.plural,
          description: kind.description
        }
      })
    );
  }
  for (const version of kind.versions ?? []) {
    const jsonSchema = version.schema ? reduceKindSchema(version.schema.jsonSchema) : void 0;
    if (jsonSchema) {
      validateMetaSchema(jsonSchema);
    }
    const names = Array.isArray(version.name) ? version.name : [version.name];
    for (const name of names) {
      const specTypes = version.specType?.length ? [version.specType].flat() : [void 0];
      for (const specType of specTypes) {
        ops.push(
          createUpdateKindVersionOp({
            kind: kind.names.kind,
            name,
            specType,
            properties: {
              description: version.description,
              relationFields: version.relationFields,
              schema: jsonSchema ? { jsonSchema } : void 0
            }
          })
        );
      }
    }
  }
  return ops;
}

export { opsFromCatalogModelUpdateKind };
