// @backstage/ui/dist/utils/selectableCollection.esm.js
function normalizeOption(option) {
  if (option.id !== void 0) {
    return option;
  }
  return {
    id: option.value,
    label: option.label,
    disabled: option.disabled
  };
}
function normalizeOptions(options) {
  return options.map((option) => {
    if ("options" in option) {
      return {
        title: option.title,
        options: option.options.map(normalizeOption)
      };
    }
    return normalizeOption(option);
  });
}
function isAsyncListSource(source) {
  return typeof source === "object" && source !== null && "items" in source && "filterText" in source && "setFilterText" in source && "loadingState" in source && "loadMore" in source;
}
function flattenOptions(options) {
  return options.flatMap((option) => {
    if ("options" in option) {
      return option.options;
    }
    return option;
  });
}
function getItemKeys(items) {
  return new Set(Array.from(items, (item) => item.id));
}
function resolveCollectionSource({
  options,
  items
}) {
  const plainOptions = Array.isArray(options) ? options : void 0;
  const normalizedOptions = plainOptions ? normalizeOptions(plainOptions) : void 0;
  const flatOptions = normalizedOptions ? flattenOptions(normalizedOptions) : void 0;
  const asyncOptions = options && !Array.isArray(options) ? options : void 0;
  if (items !== void 0) {
    return {
      options: plainOptions,
      flatOptions,
      source: items,
      rendersItems: true
    };
  }
  if (asyncOptions !== void 0) {
    return {
      options: plainOptions,
      flatOptions,
      source: asyncOptions,
      rendersItems: true
    };
  }
  return {
    options: plainOptions,
    flatOptions,
    source: flatOptions ?? [],
    rendersItems: false
  };
}
function toSelection(value) {
  if (value == null) {
    return /* @__PURE__ */ new Set();
  }
  return new Set(Array.isArray(value) ? value : [value]);
}

export { flattenOptions, getItemKeys, isAsyncListSource, normalizeOption, normalizeOptions, resolveCollectionSource, toSelection };
