// @backstage/integration/dist/index.cjs.js
'use strict';

var AwsS3Integration = require('./awsS3/AwsS3Integration.cjs.js');
var config$2 = require('./awsS3/config.cjs.js');
var AwsCodeCommitIntegration = require('./awsCodeCommit/AwsCodeCommitIntegration.cjs.js');
var config$1 = require('./awsCodeCommit/config.cjs.js');
var AzureBlobStorageIntegration = require('./azureBlobStorage/AzureBlobStorageIntegration.cjs.js');
var config$3 = require('./azureBlobStorage/config.cjs.js');
var DefaultAzureCredentialsProvider = require('./azureBlobStorage/DefaultAzureCredentialsProvider.cjs.js');
var AzureIntegration = require('./azure/AzureIntegration.cjs.js');
var config$4 = require('./azure/config.cjs.js');
var core$1 = require('./azure/core.cjs.js');
var DefaultAzureDevOpsCredentialsProvider = require('./azure/DefaultAzureDevOpsCredentialsProvider.cjs.js');
var BitbucketCloudIntegration = require('./bitbucketCloud/BitbucketCloudIntegration.cjs.js');
var config$5 = require('./bitbucketCloud/config.cjs.js');
var core$2 = require('./bitbucketCloud/core.cjs.js');
var BitbucketServerIntegration = require('./bitbucketServer/BitbucketServerIntegration.cjs.js');
var config$6 = require('./bitbucketServer/config.cjs.js');
var core$3 = require('./bitbucketServer/core.cjs.js');
var GerritIntegration = require('./gerrit/GerritIntegration.cjs.js');
var config$7 = require('./gerrit/config.cjs.js');
var core = require('./gerrit/core.cjs.js');
var GiteaIntegration = require('./gitea/GiteaIntegration.cjs.js');
var core$5 = require('./gitea/core.cjs.js');
var config$8 = require('./gitea/config.cjs.js');
var config$9 = require('./github/config.cjs.js');
var core$6 = require('./github/core.cjs.js');
var DefaultGithubCredentialsProvider = require('./github/DefaultGithubCredentialsProvider.cjs.js');
var SingleInstanceGithubCredentialsProvider = require('./github/SingleInstanceGithubCredentialsProvider.cjs.js');
var GithubIntegration = require('./github/GithubIntegration.cjs.js');
var config = require('./gitlab/config.cjs.js');
var core$4 = require('./gitlab/core.cjs.js');
var GitLabIntegration = require('./gitlab/GitLabIntegration.cjs.js');
var DefaultGitlabCredentialsProvider = require('./gitlab/DefaultGitlabCredentialsProvider.cjs.js');
var config$a = require('./googleGcs/config.cjs.js');
var GoogleGcsIntegration = require('./googleGcs/GoogleGcsIntegration.cjs.js');
var HarnessIntegration = require('./harness/HarnessIntegration.cjs.js');
var core$7 = require('./harness/core.cjs.js');
var config$b = require('./harness/config.cjs.js');
var helpers = require('./helpers.cjs.js');
var ScmIntegrations = require('./ScmIntegrations.cjs.js');



exports.AwsS3Integration = AwsS3Integration.AwsS3Integration;
exports.readAwsS3IntegrationConfig = config$2.readAwsS3IntegrationConfig;
exports.readAwsS3IntegrationConfigs = config$2.readAwsS3IntegrationConfigs;
exports.AwsCodeCommitIntegration = AwsCodeCommitIntegration.AwsCodeCommitIntegration;
exports.readAwsCodeCommitIntegrationConfig = config$1.readAwsCodeCommitIntegrationConfig;
exports.readAwsCodeCommitIntegrationConfigs = config$1.readAwsCodeCommitIntegrationConfigs;
exports.AzureBlobStorageIntegration = AzureBlobStorageIntegration.AzureBlobStorageIntegration;
exports.AzureBlobStorageIntergation = AzureBlobStorageIntegration.AzureBlobStorageIntergation;
exports.readAzureBlobStorageIntegrationConfig = config$3.readAzureBlobStorageIntegrationConfig;
exports.readAzureBlobStorageIntegrationConfigs = config$3.readAzureBlobStorageIntegrationConfigs;
exports.DefaultAzureCredentialsManager = DefaultAzureCredentialsProvider.DefaultAzureCredentialsManager;
exports.AzureIntegration = AzureIntegration.AzureIntegration;
exports.readAzureIntegrationConfig = config$4.readAzureIntegrationConfig;
exports.readAzureIntegrationConfigs = config$4.readAzureIntegrationConfigs;
exports.getAzureCommitsUrl = core$1.getAzureCommitsUrl;
exports.getAzureDownloadUrl = core$1.getAzureDownloadUrl;
exports.getAzureFileFetchUrl = core$1.getAzureFileFetchUrl;
exports.DefaultAzureDevOpsCredentialsProvider = DefaultAzureDevOpsCredentialsProvider.DefaultAzureDevOpsCredentialsProvider;
exports.BitbucketCloudIntegration = BitbucketCloudIntegration.BitbucketCloudIntegration;
exports.readBitbucketCloudIntegrationConfig = config$5.readBitbucketCloudIntegrationConfig;
exports.readBitbucketCloudIntegrationConfigs = config$5.readBitbucketCloudIntegrationConfigs;
exports.getBitbucketCloudDefaultBranch = core$2.getBitbucketCloudDefaultBranch;
exports.getBitbucketCloudDownloadUrl = core$2.getBitbucketCloudDownloadUrl;
exports.getBitbucketCloudFileFetchUrl = core$2.getBitbucketCloudFileFetchUrl;
exports.getBitbucketCloudOAuthToken = core$2.getBitbucketCloudOAuthToken;
exports.getBitbucketCloudRequestOptions = core$2.getBitbucketCloudRequestOptions;
exports.BitbucketServerIntegration = BitbucketServerIntegration.BitbucketServerIntegration;
exports.readBitbucketServerIntegrationConfig = config$6.readBitbucketServerIntegrationConfig;
exports.readBitbucketServerIntegrationConfigs = config$6.readBitbucketServerIntegrationConfigs;
exports.getBitbucketServerDefaultBranch = core$3.getBitbucketServerDefaultBranch;
exports.getBitbucketServerDownloadUrl = core$3.getBitbucketServerDownloadUrl;
exports.getBitbucketServerFileFetchUrl = core$3.getBitbucketServerFileFetchUrl;
exports.getBitbucketServerRequestOptions = core$3.getBitbucketServerRequestOptions;
exports.GerritIntegration = GerritIntegration.GerritIntegration;
exports.readGerritIntegrationConfig = config$7.readGerritIntegrationConfig;
exports.readGerritIntegrationConfigs = config$7.readGerritIntegrationConfigs;
exports.buildGerritGitilesArchiveUrlFromLocation = core.buildGerritGitilesArchiveUrlFromLocation;
exports.getGerritBranchApiUrl = core.getGerritBranchApiUrl;
exports.getGerritCloneRepoUrl = core.getGerritCloneRepoUrl;
exports.getGerritFileContentsApiUrl = core.getGerritFileContentsApiUrl;
exports.getGerritProjectsApiUrl = core.getGerritProjectsApiUrl;
exports.getGerritRequestOptions = core.getGerritRequestOptions;
exports.getGitilesAuthenticationUrl = core.getGitilesAuthenticationUrl;
exports.parseGerritJsonResponse = core.parseGerritJsonResponse;
exports.parseGitilesUrlRef = core.parseGitilesUrlRef;
exports.GiteaIntegration = GiteaIntegration.GiteaIntegration;
exports.getGiteaArchiveUrl = core$5.getGiteaArchiveUrl;
exports.getGiteaEditContentsUrl = core$5.getGiteaEditContentsUrl;
exports.getGiteaFileContentsUrl = core$5.getGiteaFileContentsUrl;
exports.getGiteaLatestCommitUrl = core$5.getGiteaLatestCommitUrl;
exports.getGiteaRequestOptions = core$5.getGiteaRequestOptions;
exports.parseGiteaUrl = core$5.parseGiteaUrl;
exports.readGiteaConfig = config$8.readGiteaConfig;
exports.readGithubIntegrationConfig = config$9.readGithubIntegrationConfig;
exports.readGithubIntegrationConfigs = config$9.readGithubIntegrationConfigs;
exports.getGithubFileFetchUrl = core$6.getGithubFileFetchUrl;
exports.DefaultGithubCredentialsProvider = DefaultGithubCredentialsProvider.DefaultGithubCredentialsProvider;
exports.GithubAppCredentialsMux = SingleInstanceGithubCredentialsProvider.GithubAppCredentialsMux;
exports.SingleInstanceGithubCredentialsProvider = SingleInstanceGithubCredentialsProvider.SingleInstanceGithubCredentialsProvider;
exports.GithubIntegration = GithubIntegration.GithubIntegration;
exports.replaceGithubUrlType = GithubIntegration.replaceGithubUrlType;
exports.getGitLabIntegrationRelativePath = config.getGitLabIntegrationRelativePath;
exports.readGitLabIntegrationConfig = config.readGitLabIntegrationConfig;
exports.readGitLabIntegrationConfigs = config.readGitLabIntegrationConfigs;
exports.getGitLabFileFetchUrl = core$4.getGitLabFileFetchUrl;
exports.getGitLabRequestOptions = core$4.getGitLabRequestOptions;
exports.GitLabIntegration = GitLabIntegration.GitLabIntegration;
exports.replaceGitLabUrlType = GitLabIntegration.replaceGitLabUrlType;
exports.DefaultGitlabCredentialsProvider = DefaultGitlabCredentialsProvider.DefaultGitlabCredentialsProvider;
exports.readGoogleGcsIntegrationConfig = config$a.readGoogleGcsIntegrationConfig;
exports.GoogleGcsIntegration = GoogleGcsIntegration.GoogleGcsIntegration;
exports.HarnessIntegration = HarnessIntegration.HarnessIntegration;
exports.getHarnessArchiveUrl = core$7.getHarnessArchiveUrl;
exports.getHarnessFileContentsUrl = core$7.getHarnessFileContentsUrl;
exports.getHarnessLatestCommitUrl = core$7.getHarnessLatestCommitUrl;
exports.getHarnessRequestOptions = core$7.getHarnessRequestOptions;
exports.parseHarnessUrl = core$7.parseHarnessUrl;
exports.readHarnessConfig = config$b.readHarnessConfig;
exports.defaultScmResolveUrl = helpers.defaultScmResolveUrl;
exports.ScmIntegrations = ScmIntegrations.ScmIntegrations;
