// @backstage/cli-module-new/dist/commands/new.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var createNewPackage = require('../lib/createNewPackage.cjs.js');

var _new = async ({ args, info }) => {
  for (const flag of ["skipInstall", "npmRegistry", "baseVersion"]) {
    if (args.some((a) => a === `--${flag}` || a.startsWith(`--${flag}=`))) {
      process.stderr.write(
        `DEPRECATION WARNING: --${flag} is deprecated, use the kebab-case form instead
`
      );
    }
  }
  const {
    flags: {
      select,
      option: rawArgOptions,
      skipInstall,
      scope,
      npmRegistry,
      baseVersion,
      license,
      private: isPrivate
    }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        select: {
          type: String,
          description: "Select the thing you want to be creating upfront"
        },
        option: {
          type: [String],
          description: "Pre-fill options for the creation process",
          default: []
        },
        skipInstall: {
          type: Boolean,
          description: `Skips running 'yarn install' and 'yarn lint --fix'`
        },
        scope: {
          type: String,
          description: "The scope to use for new packages"
        },
        npmRegistry: {
          type: String,
          description: "The package registry to use for new packages"
        },
        baseVersion: {
          type: String,
          description: "The version to use for any new packages (default: 0.1.0)"
        },
        license: {
          type: String,
          description: "The license to use for any new packages (default: Apache-2.0)"
        },
        private: {
          type: Boolean,
          description: "Mark new packages as private",
          default: true
        }
      }
    },
    void 0,
    args
  );
  const prefilledParams = parseParams(rawArgOptions);
  let pluginInfix = void 0;
  let packagePrefix = void 0;
  if (scope) {
    const normalizedScope = scope.startsWith("@") ? scope : `@${scope}`;
    packagePrefix = normalizedScope.includes("/") ? normalizedScope : `${normalizedScope}/`;
    pluginInfix = scope.includes("backstage") ? "plugin-" : "backstage-plugin-";
  }
  if (isPrivate === false || [npmRegistry, baseVersion, license].filter(Boolean).length !== 0) {
    console.warn(
      `Global template configuration via CLI flags is deprecated, see https://backstage.io/docs/cli/new for information on how to configure package templating`
    );
  }
  await createNewPackage.createNewPackage({
    prefilledParams,
    preselectedTemplateId: select,
    configOverrides: {
      license,
      version: baseVersion,
      private: isPrivate,
      publishRegistry: npmRegistry,
      packageNamePrefix: packagePrefix,
      packageNamePluginInfix: pluginInfix
    },
    skipInstall: Boolean(skipInstall)
  });
};
function parseParams(optionStrings) {
  const options = {};
  for (const str of optionStrings) {
    const [key] = str.split("=", 1);
    const value = str.slice(key.length + 1);
    if (!key || str[key.length] !== "=") {
      throw new Error(
        `Invalid option '${str}', must be of the format <key>=<value>`
      );
    }
    options[key] = value;
  }
  return options;
}

exports.default = _new;
