// @backstage/cli-module-config/dist/commands/print.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var yaml = require('yaml');
var config$1 = require('@backstage/config');
var config = require('../lib/config.cjs.js');

var print = async ({ args, info }) => {
  const {
    flags: { config: config$1, lax, frontend, withSecrets, format, package: pkg }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        package: { type: String, description: "Package to print config for" },
        lax: {
          type: Boolean,
          description: "Do not require environment variables to be set"
        },
        frontend: { type: Boolean, description: "Only print frontend config" },
        withSecrets: {
          type: Boolean,
          description: "Include secrets in the output"
        },
        format: { type: String, description: "Output format (yaml or json)" },
        config: {
          type: [String],
          description: "Config files to load instead of app-config.yaml"
        }
      }
    },
    void 0,
    args
  );
  const { schema, appConfigs } = await config.loadCliConfig({
    args: config$1,
    fromPackage: pkg,
    mockEnv: lax,
    fullVisibility: !frontend
  });
  const visibility = getVisibilityOption(frontend, withSecrets);
  const data = serializeConfigData(appConfigs, schema, visibility);
  if (format === "json") {
    process.stdout.write(`${JSON.stringify(data, null, 2)}
`);
  } else {
    process.stdout.write(`${yaml.stringify(data)}
`);
  }
};
function getVisibilityOption(frontend, withSecrets) {
  if (frontend && withSecrets) {
    throw new Error("Not allowed to combine frontend and secret config");
  }
  if (frontend) {
    return "frontend";
  } else if (withSecrets) {
    return "secret";
  }
  return "backend";
}
function serializeConfigData(appConfigs, schema, visibility) {
  if (visibility === "frontend") {
    const frontendConfigs = schema.process(appConfigs, {
      visibility: ["frontend"]
    });
    return config$1.ConfigReader.fromConfigs(frontendConfigs).get();
  } else if (visibility === "secret") {
    return config$1.ConfigReader.fromConfigs(appConfigs).get();
  }
  const sanitizedConfigs = schema.process(appConfigs, {
    valueTransform: (value, context) => context.visibility === "secret" ? "<secret>" : value
  });
  return config$1.ConfigReader.fromConfigs(sanitizedConfigs).get();
}

exports.default = print;
