// @backstage/catalog-model/dist/kinds/AiResourceEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
require('lodash');
require('ajv');
var entityKindSchemaValidator = require('../validation/entityKindSchemaValidator.cjs.js');
require('../validation/makeValidator.cjs.js');
var AiResource_v1alpha1_schema = require('../schema/kinds/AiResource.v1alpha1.schema.json.cjs.js');
var AiResource_v1alpha1_skill_schema = require('../schema/kinds/AiResource.v1alpha1.skill.schema.json.cjs.js');
var AiResource_v1alpha1_rule_schema = require('../schema/kinds/AiResource.v1alpha1.rule.schema.json.cjs.js');

const defaultValidator = entityKindSchemaValidator.entityKindSchemaValidator(AiResource_v1alpha1_schema.default);
const aiResourceEntityV1alpha1Validator = {
  async check(data) {
    return defaultValidator(data) === data;
  }
};
const skillValidator = entityKindSchemaValidator.entityKindSchemaValidator(AiResource_v1alpha1_skill_schema.default);
const skillAiResourceEntityV1alpha1Validator = {
  async check(data) {
    return skillValidator(data) === data;
  }
};
const isAiResourceEntity = (entity) => entity.apiVersion === "backstage.io/v1alpha1" && entity.kind === "AiResource";
const isSkillAiResourceEntity = (entity) => isAiResourceEntity(entity) && entity.spec?.type === "skill";
const isRuleAiResourceEntity = (entity) => isAiResourceEntity(entity) && entity.spec?.type === "rule";
const ruleValidator = entityKindSchemaValidator.entityKindSchemaValidator(AiResource_v1alpha1_rule_schema.default);
const ruleAiResourceEntityV1alpha1Validator = {
  async check(data) {
    return ruleValidator(data) === data;
  }
};
const baseRelationFields = [
  {
    selector: { path: "spec.owner" },
    relation: "ownedBy",
    defaultKind: "Group",
    defaultNamespace: "inherit",
    allowedKinds: ["Group", "User"]
  },
  {
    selector: { path: "spec.system" },
    relation: "partOf",
    defaultKind: "System",
    defaultNamespace: "inherit"
  }
];
const aiResourceEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-ai-resource",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "AiResource",
        singular: "airesource",
        plural: "airesources"
      },
      description: "An AI resource represents contextual information consumed by AI coding tools, such as skills and rules.",
      versions: [
        {
          name: "v1alpha1",
          relationFields: baseRelationFields,
          schema: {
            jsonSchema: AiResource_v1alpha1_schema.default
          }
        },
        {
          name: "v1alpha1",
          specType: "skill",
          relationFields: [
            ...baseRelationFields,
            {
              selector: { path: "spec.dependsOn" },
              relation: "dependsOn",
              defaultKind: "AiResource",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: AiResource_v1alpha1_skill_schema.default
          }
        },
        {
          name: "v1alpha1",
          specType: "rule",
          relationFields: baseRelationFields,
          schema: {
            jsonSchema: AiResource_v1alpha1_rule_schema.default
          }
        }
      ]
    });
  }
});

exports.aiResourceEntityModel = aiResourceEntityModel;
exports.aiResourceEntityV1alpha1Validator = aiResourceEntityV1alpha1Validator;
exports.isAiResourceEntity = isAiResourceEntity;
exports.isRuleAiResourceEntity = isRuleAiResourceEntity;
exports.isSkillAiResourceEntity = isSkillAiResourceEntity;
exports.ruleAiResourceEntityV1alpha1Validator = ruleAiResourceEntityV1alpha1Validator;
exports.skillAiResourceEntityV1alpha1Validator = skillAiResourceEntityV1alpha1Validator;
