// @backstage/catalog-model/dist/kinds/ApiEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var API_v1alpha1_schema = require('../schema/kinds/API.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const apiEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(API_v1alpha1_schema.default);
const apiEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-api",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "API",
        singular: "api",
        plural: "apis"
      },
      description: "An API describes an interface that can be exposed by a component.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: { jsonSchema: API_v1alpha1_schema.default }
        }
      ]
    });
  }
});

exports.apiEntityModel = apiEntityModel;
exports.apiEntityV1alpha1Validator = apiEntityV1alpha1Validator;
