// @backstage/catalog-model/dist/kinds/ApiEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/API.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const apiEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const apiEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-api",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "API",
        singular: "api",
        plural: "apis"
      },
      description: "An API describes an interface that can be exposed by a component.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: { jsonSchema }
        }
      ]
    });
  }
});

export { apiEntityModel, apiEntityV1alpha1Validator };
