// @backstage/catalog-model/dist/kinds/ComponentEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var Component_v1alpha1_schema = require('../schema/kinds/Component.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const componentEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(Component_v1alpha1_schema.default);
const componentEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-component",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Component",
        singular: "component",
        plural: "components"
      },
      description: "A Component describes a software component, usually with a distinct deployable or linkable artifact.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.subcomponentOf" },
              relation: "partOf",
              defaultKind: "Component",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.providesApis" },
              relation: "providesApi",
              defaultKind: "API",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.consumesApis" },
              relation: "consumesApi",
              defaultKind: "API",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.dependsOn" },
              relation: "dependsOn",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.dependencyOf" },
              relation: "dependencyOf",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: Component_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.componentEntityModel = componentEntityModel;
exports.componentEntityV1alpha1Validator = componentEntityV1alpha1Validator;
