// @backstage/catalog-model/dist/kinds/DomainEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var Domain_v1alpha1_schema = require('../schema/kinds/Domain.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const domainEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(Domain_v1alpha1_schema.default);
const domainEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-domain",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Domain",
        singular: "domain",
        plural: "domains"
      },
      description: "A Domain groups a collection of systems that share terminology, domain models, business purpose, or documentation.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.subdomainOf" },
              relation: "partOf",
              defaultKind: "Domain",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: Domain_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.domainEntityModel = domainEntityModel;
exports.domainEntityV1alpha1Validator = domainEntityV1alpha1Validator;
