// @backstage/catalog-model/dist/kinds/DomainEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/Domain.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const domainEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const domainEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-domain",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Domain",
        singular: "domain",
        plural: "domains"
      },
      description: "A Domain groups a collection of systems that share terminology, domain models, business purpose, or documentation.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.subdomainOf" },
              relation: "partOf",
              defaultKind: "Domain",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema
          }
        }
      ]
    });
  }
});

export { domainEntityModel, domainEntityV1alpha1Validator };
