// @backstage/catalog-model/dist/kinds/GroupEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/Group.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const groupEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const groupEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-group",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Group",
        singular: "group",
        plural: "groups"
      },
      description: "A Group describes an organizational entity, such as a team, a business unit, or a loose collection of people.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.parent" },
              relation: "childOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            },
            {
              selector: { path: "spec.children" },
              relation: "parentOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            },
            {
              selector: { path: "spec.members" },
              relation: "hasMember",
              defaultKind: "User",
              defaultNamespace: "inherit",
              allowedKinds: ["User"]
            }
          ],
          schema: {
            jsonSchema
          }
        }
      ]
    });
  }
});

export { groupEntityModel, groupEntityV1alpha1Validator };
