// @backstage/catalog-model/dist/kinds/relations.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';

const RELATION_OWNED_BY = "ownedBy";
const RELATION_OWNER_OF = "ownerOf";
const RELATION_CONSUMES_API = "consumesApi";
const RELATION_API_CONSUMED_BY = "apiConsumedBy";
const RELATION_PROVIDES_API = "providesApi";
const RELATION_API_PROVIDED_BY = "apiProvidedBy";
const RELATION_DEPENDS_ON = "dependsOn";
const RELATION_DEPENDENCY_OF = "dependencyOf";
const RELATION_PARENT_OF = "parentOf";
const RELATION_CHILD_OF = "childOf";
const RELATION_MEMBER_OF = "memberOf";
const RELATION_HAS_MEMBER = "hasMember";
const RELATION_PART_OF = "partOf";
const RELATION_HAS_PART = "hasPart";
const wellKnownRelationsModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/well-known-relations",
  builder: (model) => {
    model.addRelationPair({
      fromKind: [
        "API",
        "Component",
        "Domain",
        "Group",
        "Location",
        "Resource",
        "System",
        "User"
      ],
      toKind: ["Group", "User"],
      description: "An ownership relation where the owner is usually an organizational entity (user or group), and the other entity can be anything.",
      forward: { type: "ownedBy", title: "owned by" },
      reverse: { type: "ownerOf", title: "owner of" }
    });
    model.addRelationPair({
      fromKind: "Component",
      toKind: "API",
      description: "A relation from a component to an API it provides for consumption by others.",
      forward: { type: "providesApi", title: "provides API" },
      reverse: { type: "apiProvidedBy", title: "API provided by" }
    });
    model.addRelationPair({
      fromKind: "Component",
      toKind: "API",
      description: "A relation from a component to an API it consumes.",
      forward: { type: "consumesApi", title: "consumes API" },
      reverse: { type: "apiConsumedBy", title: "API consumed by" }
    });
    model.addRelationPair({
      fromKind: ["Component", "Resource"],
      toKind: ["Component", "Resource"],
      description: "A dependency relation expressing that an entity needs another entity to function.",
      forward: { type: "dependsOn", title: "depends on" },
      reverse: { type: "dependencyOf", title: "dependency of" }
    });
    model.addRelationPair({
      fromKind: "Group",
      toKind: "Group",
      description: "A parent/child relation to build up a tree, used for example to describe the organizational structure between groups.",
      forward: { type: "parentOf", title: "parent of" },
      reverse: { type: "childOf", title: "child of" }
    });
    model.addRelationPair({
      fromKind: "User",
      toKind: "Group",
      description: "A membership relation, typically for users in a group.",
      forward: { type: "memberOf", title: "member of" },
      reverse: { type: "hasMember", title: "has member" }
    });
    model.addRelationPair({
      fromKind: ["Component", "API", "Resource"],
      toKind: ["Component", "System"],
      description: "A part/whole relation where a component, API, or resource belongs to a system or a component is a subcomponent of another.",
      forward: { type: "partOf", title: "part of" },
      reverse: { type: "hasPart", title: "has part" }
    });
    model.addRelationPair({
      fromKind: "System",
      toKind: "Domain",
      description: "A part/whole relation where a system belongs to a domain.",
      forward: { type: "partOf", title: "part of" },
      reverse: { type: "hasPart", title: "has part" }
    });
    model.addRelationPair({
      fromKind: "Domain",
      toKind: "Domain",
      description: "A part/whole relation where a domain is a subdomain of another domain.",
      forward: { type: "partOf", title: "part of" },
      reverse: { type: "hasPart", title: "has part" }
    });
  }
});

export { RELATION_API_CONSUMED_BY, RELATION_API_PROVIDED_BY, RELATION_CHILD_OF, RELATION_CONSUMES_API, RELATION_DEPENDENCY_OF, RELATION_DEPENDS_ON, RELATION_HAS_MEMBER, RELATION_HAS_PART, RELATION_MEMBER_OF, RELATION_OWNED_BY, RELATION_OWNER_OF, RELATION_PARENT_OF, RELATION_PART_OF, RELATION_PROVIDES_API, wellKnownRelationsModel };
