// @backstage/catalog-model/dist/kinds/ResourceEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/Resource.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const resourceEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const resourceEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-resource",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Resource",
        singular: "resource",
        plural: "resources"
      },
      description: "A Resource describes the infrastructure a system needs to operate, like databases, topics, or buckets.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.dependsOn" },
              relation: "dependsOn",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.dependencyOf" },
              relation: "dependencyOf",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema
          }
        }
      ]
    });
  }
});

export { resourceEntityModel, resourceEntityV1alpha1Validator };
