// @backstage/catalog-model/dist/kinds/SystemEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/System.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const systemEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const systemEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-system",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "System",
        singular: "system",
        plural: "systems"
      },
      description: "A System is a collection of resources and components that exposes one or several APIs.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.domain" },
              relation: "partOf",
              defaultKind: "Domain",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema
          }
        }
      ]
    });
  }
});

export { systemEntityModel, systemEntityV1alpha1Validator };
