// @backstage/catalog-model/dist/kinds/UserEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var User_v1alpha1_schema = require('../schema/kinds/User.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const userEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(User_v1alpha1_schema.default);
const userEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-user",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "User",
        singular: "user",
        plural: "users"
      },
      description: "A User describes a person, such as an employee or a contractor.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.memberOf" },
              relation: "memberOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            }
          ],
          schema: {
            jsonSchema: User_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.userEntityModel = userEntityModel;
exports.userEntityV1alpha1Validator = userEntityV1alpha1Validator;
