// @backstage/ui/dist/components/DatePicker/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './DatePicker.module.css.esm.js';

const DatePickerDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DatePicker"
  },
  propDefs: {
    size: { dataAttribute: true, default: "small" },
    className: {},
    label: {},
    description: {},
    secondaryLabel: {}
  }
});
const DatePickerGroupDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DatePickerGroup",
    dateInput: "bui-DatePickerDateInput",
    segment: "bui-DatePickerSegment",
    button: "bui-DatePickerButton"
  },
  bg: "consumer",
  propDefs: {}
});
const DatePickerCalendarDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DatePickerCalendar",
    header: "bui-DatePickerCalendarHeader",
    heading: "bui-DatePickerCalendarHeading",
    navButton: "bui-DatePickerCalendarNavButton",
    grid: "bui-DatePickerCalendarGrid",
    gridHeader: "bui-DatePickerCalendarGridHeader",
    headerCell: "bui-DatePickerCalendarHeaderCell",
    gridBody: "bui-DatePickerCalendarGridBody",
    cell: "bui-DatePickerCalendarCell"
  },
  propDefs: {}
});

export { DatePickerCalendarDefinition, DatePickerDefinition, DatePickerGroupDefinition };
