// @backstage/cli-module-build/dist/commands/buildWorkspace.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('fs-extra');
var cleye = require('cleye');
var createDistWorkspace = require('../lib/packager/createDistWorkspace.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

var buildWorkspace = async ({ args, info }) => {
  const normalizedArgs = args.map((a) => {
    if (a === "--alwaysYarnPack") {
      return "--always-pack";
    }
    if (a.startsWith("--alwaysYarnPack=")) {
      return `--always-pack${a.substring("--alwaysYarnPack".length)}`;
    }
    return a;
  });
  const {
    flags: { alwaysPack },
    _: positionals
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      parameters: ["<workspace-dir>", "[packages...]"],
      flags: {
        alwaysPack: {
          type: Boolean,
          description: "Force workspace output to be a result of running `yarn pack` on each package (warning: very slow)"
        }
      }
    },
    void 0,
    normalizedArgs
  );
  const [dir, ...packages] = positionals;
  if (!await fs__default.default.pathExists(dir)) {
    throw new Error(`Target workspace directory doesn't exist, '${dir}'`);
  }
  await createDistWorkspace.createDistWorkspace(packages, {
    targetDir: dir,
    alwaysPack,
    enableFeatureDetection: true
  });
};

exports.default = buildWorkspace;
