// @backstage/ui/dist/components/Header/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Header.module.css.esm.js';

const HeaderDefinition = defineComponent()({
  styles,
  bg: "consumer",
  classNames: {
    headerTop: "bui-HeaderTop",
    stickySentinel: "bui-HeaderStickySentinel",
    content: "bui-HeaderContent",
    headerBottom: "bui-HeaderBottom",
    breadcrumbs: "bui-HeaderBreadcrumbs",
    breadcrumbsSmall: "bui-HeaderBreadcrumbsSmall",
    breadcrumbLink: "bui-HeaderBreadcrumbLink",
    breadcrumbLinkSmall: "bui-HeaderBreadcrumbLinkSmall",
    breadcrumbSeparator: "bui-HeaderBreadcrumbSeparator",
    titleStack: "bui-HeaderTitleStack",
    title: "bui-HeaderTitle",
    titleSmall: "bui-HeaderTitleSmall",
    tabsWrapper: "bui-HeaderTabsWrapper",
    controls: "bui-HeaderControls",
    tags: "bui-HeaderTags",
    tag: "bui-HeaderTag",
    description: "bui-HeaderDescription",
    metaRow: "bui-HeaderMetaRow",
    metaItem: "bui-HeaderMetaItem"
  },
  propDefs: {
    title: {},
    customActions: {},
    tabs: {},
    activeTabId: {},
    breadcrumbs: {},
    description: {},
    tags: {},
    metadata: {},
    className: {},
    sticky: {}
  }
});
const HeaderPageDefinition = HeaderDefinition;

export { HeaderDefinition, HeaderPageDefinition };
