// @backstage/ui/dist/components/Header/HeaderNav.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useRef, useState, useMemo, useCallback } from 'react';
import { useFocusVisible, useHover, useLink } from 'react-aria';
import { useInRouterContext, useResolvedPath, useLocation, resolvePath, matchRoutes } from 'react-router-dom';
import { Button } from 'react-aria-components';
import { RiArrowDownSLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { HeaderNavDefinition, HeaderNavGroupDefinition, HeaderNavItemDefinition } from './HeaderNavDefinition.esm.js';
import { HeaderNavIndicators } from './HeaderNavIndicators.esm.js';
import { MenuTrigger, Menu, MenuItem } from '../Menu/Menu.esm.js';
import '../Menu/definition.esm.js';

function isTabGroup(tab) {
  return "items" in tab;
}
function HeaderNavLink(props) {
  const { ownProps, analytics } = useDefinition(HeaderNavItemDefinition, props);
  const { id, label, href, active, registerRef, onHighlight } = ownProps;
  const linkRef = useRef(null);
  const { linkProps } = useLink({ href }, linkRef);
  const { hoverProps } = useHover({
    onHoverStart: () => onHighlight(id),
    onHoverEnd: () => onHighlight(null)
  });
  const handleClick = (e) => {
    linkProps.onClick?.(e);
    analytics.captureEvent("click", label, {
      attributes: { to: href }
    });
  };
  return /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
    "a",
    {
      ...linkProps,
      ...hoverProps,
      ref: (el) => {
        linkRef.current = el;
        registerRef(id, el);
      },
      className: ownProps.classes.root,
      "aria-current": active ? "page" : void 0,
      onClick: handleClick,
      onFocus: () => onHighlight(id),
      onBlur: () => onHighlight(null),
      children: label
    }
  ) });
}
function HeaderNavGroupItem(props) {
  const { group, active, activeChildId, registerRef, onHighlight } = props;
  const { ownProps } = useDefinition(HeaderNavGroupDefinition, {});
  const { hoverProps } = useHover({
    onHoverStart: () => onHighlight(group.id),
    onHoverEnd: () => onHighlight(null)
  });
  return /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsxs(MenuTrigger, { children: [
    /* @__PURE__ */ jsxs(
      Button,
      {
        ref: (el) => {
          registerRef(group.id, el);
        },
        className: ownProps.classes.root,
        "aria-current": active ? "page" : void 0,
        ...hoverProps,
        onFocus: () => onHighlight(group.id),
        onBlur: () => onHighlight(null),
        children: [
          group.label,
          /* @__PURE__ */ jsx(RiArrowDownSLine, { size: 16 })
        ]
      }
    ),
    /* @__PURE__ */ jsx(
      Menu,
      {
        selectionMode: "single",
        selectedKeys: new Set(activeChildId ? [activeChildId] : []),
        children: group.items.map((item) => /* @__PURE__ */ jsx(MenuItem, { id: item.id, href: item.href, children: item.label }, item.id))
      }
    )
  ] }) });
}
function useAutoActiveTabId(tabs) {
  const basePath = useResolvedPath(".").pathname;
  const { pathname } = useLocation();
  return useMemo(() => {
    const allTabs = tabs.flatMap((tab) => isTabGroup(tab) ? tab.items : [tab]);
    const routeObjects = allTabs.map((tab) => ({
      path: `${resolvePath(tab.href, basePath).pathname}/*`,
      id: tab.id
    }));
    const matches = matchRoutes(routeObjects, pathname);
    return matches?.[0]?.route.id;
  }, [tabs, basePath, pathname]);
}
function HeaderNavAutoDetect(props) {
  const activeTabId = useAutoActiveTabId(props.tabs);
  return /* @__PURE__ */ jsx(HeaderNavInner, { tabs: props.tabs, activeTabId });
}
function HeaderNavInner(props) {
  const { tabs, activeTabId } = props;
  const { ownProps } = useDefinition(HeaderNavDefinition, {
    tabs,
    activeTabId
  });
  const { classes } = ownProps;
  const { isFocusVisible } = useFocusVisible();
  const navRef = useRef(null);
  const itemRefs = useRef(/* @__PURE__ */ new Map());
  const [highlightedKey, setHighlightedKey] = useState(null);
  const { activeKey, activeChildId } = useMemo(() => {
    if (!activeTabId) return { activeKey: void 0, activeChildId: void 0 };
    for (const item of tabs) {
      if (isTabGroup(item)) {
        const child = item.items.find((c) => c.id === activeTabId);
        if (child) {
          return { activeKey: item.id, activeChildId: child.id };
        }
      } else if (item.id === activeTabId) {
        return { activeKey: item.id, activeChildId: void 0 };
      }
    }
    return { activeKey: void 0, activeChildId: void 0 };
  }, [activeTabId, tabs]);
  const registerRef = useCallback((key, el) => {
    if (el) {
      itemRefs.current.set(key, el);
    } else {
      itemRefs.current.delete(key);
    }
  }, []);
  return /* @__PURE__ */ jsxs(
    "nav",
    {
      ref: navRef,
      "aria-label": "Content navigation",
      className: classes.root,
      "data-focus-visible": isFocusVisible || void 0,
      children: [
        /* @__PURE__ */ jsx("ul", { role: "list", className: classes.list, children: tabs.map(
          (item) => isTabGroup(item) ? /* @__PURE__ */ jsx(
            HeaderNavGroupItem,
            {
              group: item,
              active: activeKey === item.id,
              activeChildId,
              registerRef,
              onHighlight: setHighlightedKey
            },
            item.id
          ) : /* @__PURE__ */ jsx(
            HeaderNavLink,
            {
              id: item.id,
              label: item.label,
              href: item.href,
              active: activeKey === item.id,
              registerRef,
              onHighlight: setHighlightedKey
            },
            item.id
          )
        ) }),
        /* @__PURE__ */ jsx(
          HeaderNavIndicators,
          {
            navRef,
            itemRefs,
            activeKey,
            highlightedKey,
            classes: { active: classes.active, hovered: classes.hovered }
          }
        )
      ]
    }
  );
}
function HeaderNav(props) {
  const inRouter = useInRouterContext();
  if (props.activeTabId === void 0 && inRouter) {
    return /* @__PURE__ */ jsx(HeaderNavAutoDetect, { tabs: props.tabs });
  }
  return /* @__PURE__ */ jsx(HeaderNavInner, { tabs: props.tabs, activeTabId: props.activeTabId });
}

export { HeaderNav };
