// @backstage/ui/dist/components/Header/Header.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { Text } from '../Text/Text.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../Text/Text.module.css.esm.js';
import { RiArrowRightSLine } from '@remixicon/react';
import { HeaderNav } from './HeaderNav.esm.js';
import { HeaderDefinition } from './definition.esm.js';
import { sanitizeUrl } from '@braintree/sanitize-url';
import { Lexer } from 'marked';
import { Link } from '../Link/Link.esm.js';
import '../Link/Link.module.css.esm.js';
import { useMemo, useRef, useState, useEffect, Fragment as Fragment$1 } from 'react';
import { Container } from '../Container/Container.esm.js';
import '../Container/Container.module.css.esm.js';

const getScrollParent = (element) => {
  let parent = element?.parentElement;
  while (parent) {
    const { overflow, overflowX, overflowY } = window.getComputedStyle(parent);
    if (/(auto|scroll|overlay|hidden)/.test(`${overflow}${overflowX}${overflowY}`)) {
      return parent;
    }
    parent = parent.parentElement;
  }
  return null;
};
const isStickySentinelOutOfView = (sentinel, root) => {
  const sentinelRect = sentinel.getBoundingClientRect();
  const rootTop = root ? root.getBoundingClientRect().top : 0;
  return sentinelRect.bottom <= rootTop;
};
function renderInlineMarkdown(text) {
  return Lexer.lexInline(text).map((token, i) => {
    if (token.type === "link") {
      const href = sanitizeUrl(token.href);
      if (href === "about:blank") return token.text;
      return /* @__PURE__ */ jsx(Link, { href, standalone: true, children: token.text }, i);
    }
    return token.raw;
  });
}
const Header = (props) => {
  const { ownProps, dataAttributes } = useDefinition(HeaderDefinition, props, {
    classNameTarget: "content"
  });
  const {
    classes,
    title,
    tabs,
    activeTabId,
    customActions,
    breadcrumbs,
    description,
    tags,
    metadata,
    sticky
  } = ownProps;
  const descriptionNodes = useMemo(
    () => description ? renderInlineMarkdown(description) : null,
    [description]
  );
  const stickySentinelRef = useRef(null);
  const [isStuck, setIsStuck] = useState(false);
  useEffect(() => {
    if (!sticky) {
      setIsStuck(false);
      return;
    }
    const sentinel = stickySentinelRef.current;
    if (!sentinel) {
      return;
    }
    const root = getScrollParent(sentinel);
    if (typeof IntersectionObserver === "undefined") {
      const updateStuckState = () => {
        setIsStuck(isStickySentinelOutOfView(sentinel, root));
      };
      const scrollTarget = root ?? window;
      updateStuckState();
      scrollTarget.addEventListener("scroll", updateStuckState, {
        passive: true
      });
      window.addEventListener("resize", updateStuckState);
      return () => {
        scrollTarget.removeEventListener("scroll", updateStuckState);
        window.removeEventListener("resize", updateStuckState);
      };
    }
    const observer = new IntersectionObserver(
      ([entry]) => {
        setIsStuck(!entry.isIntersecting);
      },
      { root, threshold: 0 }
    );
    observer.observe(sentinel);
    return () => {
      observer.disconnect();
    };
  }, [sticky]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Container,
      {
        className: classes.headerTop,
        "data-sticky": sticky || void 0,
        children: tags && tags.length > 0 && /* @__PURE__ */ jsx("ul", { className: classes.tags, children: tags.map((tag, i) => /* @__PURE__ */ jsx(
          "li",
          {
            className: classes.tag,
            children: tag.href ? /* @__PURE__ */ jsx(
              Link,
              {
                href: tag.href,
                variant: "body-medium",
                color: "secondary",
                standalone: true,
                children: tag.label
              }
            ) : /* @__PURE__ */ jsx(Text, { variant: "body-medium", color: "secondary", children: tag.label })
          },
          `${i}:${tag.label}:${tag.href ?? ""}`
        )) })
      }
    ),
    sticky && /* @__PURE__ */ jsx(
      "div",
      {
        ref: stickySentinelRef,
        className: classes.stickySentinel,
        "data-sticky": "",
        "aria-hidden": "true"
      }
    ),
    /* @__PURE__ */ jsxs(
      Container,
      {
        className: classes.content,
        "data-sticky": sticky || void 0,
        "data-stuck": isStuck || void 0,
        "data-has-tags": tags && tags.length > 0 ? "" : void 0,
        ...dataAttributes,
        children: [
          /* @__PURE__ */ jsx("div", { className: classes.titleStack, children: isStuck ? /* @__PURE__ */ jsxs("div", { className: classes.breadcrumbsSmall, children: [
            breadcrumbs && breadcrumbs.map((breadcrumb) => /* @__PURE__ */ jsxs(Fragment$1, { children: [
              /* @__PURE__ */ jsx(
                Link,
                {
                  href: breadcrumb.href,
                  color: "secondary",
                  className: classes.breadcrumbLinkSmall,
                  standalone: true,
                  children: breadcrumb.label
                }
              ),
              /* @__PURE__ */ jsx(
                RiArrowRightSLine,
                {
                  className: classes.breadcrumbSeparator,
                  size: 16,
                  color: "var(--bui-fg-secondary)"
                }
              )
            ] }, breadcrumb.label)),
            /* @__PURE__ */ jsx("h2", { className: classes.titleSmall, children: title })
          ] }) : /* @__PURE__ */ jsxs("div", { className: classes.breadcrumbs, children: [
            breadcrumbs && breadcrumbs.map((breadcrumb) => /* @__PURE__ */ jsxs(Fragment$1, { children: [
              /* @__PURE__ */ jsx(
                Link,
                {
                  href: breadcrumb.href,
                  color: "secondary",
                  className: classes.breadcrumbLink,
                  standalone: true,
                  children: breadcrumb.label
                }
              ),
              /* @__PURE__ */ jsx(
                RiArrowRightSLine,
                {
                  className: classes.breadcrumbSeparator,
                  size: 16,
                  color: "var(--bui-fg-secondary)"
                }
              )
            ] }, breadcrumb.label)),
            /* @__PURE__ */ jsx("h2", { className: classes.title, children: title })
          ] }) }),
          /* @__PURE__ */ jsx("div", { className: classes.controls, children: customActions })
        ]
      }
    ),
    /* @__PURE__ */ jsxs(
      Container,
      {
        className: classes.headerBottom,
        "data-sticky": sticky || void 0,
        children: [
          description && /* @__PURE__ */ jsx(
            Text,
            {
              variant: "body-medium",
              color: "secondary",
              className: classes.description,
              children: descriptionNodes
            }
          ),
          metadata && metadata.length > 0 && /* @__PURE__ */ jsx("dl", { className: classes.metaRow, children: metadata.map((item, i) => /* @__PURE__ */ jsxs("div", { className: classes.metaItem, children: [
            /* @__PURE__ */ jsx("dt", { children: /* @__PURE__ */ jsx(Text, { variant: "body-medium", color: "secondary", children: item.label }) }),
            /* @__PURE__ */ jsx("dd", { children: typeof item.value === "string" ? /* @__PURE__ */ jsx(Text, { variant: "body-medium", children: item.value }) : item.value })
          ] }, `${i}:${item.label}`)) }),
          tabs && /* @__PURE__ */ jsx("div", { className: classes.tabsWrapper, children: /* @__PURE__ */ jsx(HeaderNav, { tabs, activeTabId }) })
        ]
      }
    )
  ] });
};
const HeaderPage = Header;

export { Header, HeaderPage };
