// @backstage/ui/dist/components/Header/HeaderNavDefinition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './HeaderNav.module.css.esm.js';

const HeaderNavDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-HeaderNav",
    list: "bui-HeaderNavList",
    active: "bui-HeaderNavActive",
    hovered: "bui-HeaderNavHovered"
  },
  propDefs: {
    tabs: {},
    activeTabId: {},
    children: {},
    className: {}
  }
});
const HeaderNavItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-HeaderNavItem"
  },
  analytics: true,
  propDefs: {
    noTrack: {},
    id: {},
    label: {},
    href: {},
    active: {},
    registerRef: {},
    onHighlight: {},
    className: {}
  }
});
const HeaderNavGroupDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-HeaderNavGroup"
  },
  propDefs: {
    className: {}
  }
});

export { HeaderNavDefinition, HeaderNavGroupDefinition, HeaderNavItemDefinition };
