// @backstage/ui/dist/components/Header/HeaderMetadataUsers.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { Avatar } from '../Avatar/Avatar.esm.js';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import 'react-router-dom';
import '../Avatar/Avatar.module.css.esm.js';
import { TooltipTrigger, Tooltip } from '../Tooltip/Tooltip.esm.js';
import '../Tooltip/Tooltip.module.css.esm.js';
import { Text } from '../Text/Text.esm.js';
import '../Text/Text.module.css.esm.js';
import { Link } from '../Link/Link.esm.js';
import '../Link/Link.module.css.esm.js';
import { Pressable } from 'react-aria';
import styles from './HeaderMetadataUsers.module.css.esm.js';

const HeaderMetadataUsers = ({
  users
}) => {
  if (users.length === 0) return null;
  if (users.length === 1) {
    const user = users[0];
    if (user.href) {
      return /* @__PURE__ */ jsxs(
        Link,
        {
          href: user.href,
          variant: "body-medium",
          standalone: true,
          className: styles.single,
          children: [
            /* @__PURE__ */ jsx(
              Avatar,
              {
                src: user.src ?? "data:,",
                name: user.name,
                size: "small",
                purpose: "decoration"
              }
            ),
            user.name
          ]
        }
      );
    }
    return /* @__PURE__ */ jsxs("div", { className: styles.single, children: [
      /* @__PURE__ */ jsx(
        Avatar,
        {
          src: user.src ?? "data:,",
          name: user.name,
          size: "small",
          purpose: "decoration"
        }
      ),
      /* @__PURE__ */ jsx(Text, { variant: "body-medium", children: user.name })
    ] });
  }
  return /* @__PURE__ */ jsx("ul", { className: styles.stack, children: users.map((user, i) => /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsxs(TooltipTrigger, { children: [
    user.href ? /* @__PURE__ */ jsx(
      Link,
      {
        href: user.href,
        "aria-label": user.name,
        className: styles.avatarLink,
        children: /* @__PURE__ */ jsx(
          Avatar,
          {
            src: user.src ?? "data:,",
            name: user.name,
            size: "small",
            purpose: "decoration"
          }
        )
      }
    ) : /* @__PURE__ */ jsx(Pressable, { children: /* @__PURE__ */ jsx(
      Avatar,
      {
        src: user.src ?? "data:,",
        name: user.name,
        size: "small",
        purpose: "informative"
      }
    ) }),
    /* @__PURE__ */ jsx(Tooltip, { children: user.name })
  ] }) }, user.href ?? `${i}:${user.name}`)) });
};

export { HeaderMetadataUsers };
