// @backstage/cli-module-auth/dist/commands/login.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var localServer = require('../lib/localServer.cjs.js');
var node_child_process = require('node:child_process');
var pkce = require('../lib/pkce.cjs.js');
var http = require('../lib/http.cjs.js');
var storage = require('../lib/storage.cjs.js');
require('../cli-internal/src/InternalCliModule.cjs.js');
require('../cli-internal/src/InternalCommandNode.cjs.js');
var authIdentifiers = require('../cli-internal/src/authIdentifiers.cjs.js');
var secretStore = require('../cli-internal/src/secretStore.cjs.js');
require('../cli-internal/src/knownPluginPackages.cjs.js');
var crypto = require('node:crypto');
var fs = require('fs-extra');
var path = require('node:path');
var glob = require('glob');
var YAML = require('yaml');
var inquirer = require('inquirer');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var crypto__default = /*#__PURE__*/_interopDefaultCompat(crypto);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var path__default = /*#__PURE__*/_interopDefaultCompat(path);
var YAML__default = /*#__PURE__*/_interopDefaultCompat(YAML);
var inquirer__default = /*#__PURE__*/_interopDefaultCompat(inquirer);

const TOKEN_EXCHANGE_TIMEOUT_MS = 3e4;
var login = async ({ args, info }) => {
  const {
    flags: { backendUrl, noBrowser, instance: instanceFlag }
  } = cleye.cli(
    {
      name: info.usage,
      flags: {
        backendUrl: { type: String, description: "Backend base URL" },
        noBrowser: {
          type: Boolean,
          description: "Do not open browser automatically"
        },
        instance: {
          type: String,
          description: "Name for this instance (used by other auth commands)"
        }
      }
    },
    void 0,
    args
  );
  const { instances, selected } = await storage.getAllInstances();
  let backendBaseUrl;
  let instanceName;
  if (instanceFlag) {
    instanceName = instanceFlag;
    const targetInstance = instances.find((i) => i.name === instanceFlag);
    if (targetInstance) {
      backendBaseUrl = normalizeUrl(backendUrl) ?? targetInstance.baseUrl;
    } else {
      backendBaseUrl = normalizeUrl(backendUrl) ?? await pickBaseUrl();
    }
  } else if (backendUrl) {
    backendBaseUrl = normalizeUrl(backendUrl);
    instanceName = deriveInstanceName(backendBaseUrl);
  } else if (instances.length > 0) {
    const choice = await promptForInstance(instances, selected);
    if (choice === "__new__") {
      backendBaseUrl = await pickBaseUrl();
      instanceName = deriveInstanceName(backendBaseUrl);
    } else {
      const targetInstance = instances.find((i) => i.name === choice);
      if (!targetInstance) {
        throw new Error("Instance not found");
      }
      backendBaseUrl = targetInstance.baseUrl;
      instanceName = targetInstance.name;
    }
  } else {
    backendBaseUrl = await pickBaseUrl();
    instanceName = deriveInstanceName(backendBaseUrl);
  }
  const authBaseUrl = `${backendBaseUrl}/api/auth`;
  const clientId = `${authBaseUrl}/.well-known/oauth-client/cli.json`;
  const metadataResponse = await fetch(clientId, {
    signal: AbortSignal.timeout(3e4)
  });
  if (!metadataResponse.ok) {
    throw new Error(
      `Server does not support CLI authentication. Ensure CIMD is enabled on the backend.`
    );
  }
  const { verifier, challenge, state } = createPkceState();
  const callback = await localServer.startCallbackServer({ state });
  try {
    const authorizeUrl = buildAuthorizeUrl({
      authBaseUrl,
      clientId,
      redirectUri: callback.url,
      state,
      challenge
    });
    await openBrowserOrPrint(authorizeUrl, noBrowser);
    const code = await waitForAuthorizationCode(callback, state);
    const token = await exchangeAuthorizationCode({
      authBaseUrl,
      code,
      redirectUri: callback.url,
      verifier
    });
    await persistInstance({
      instanceName,
      backendBaseUrl,
      clientId,
      token
    });
    process.stdout.write("Login successful\n");
  } finally {
    await callback.close();
  }
};
async function promptForInstance(instances, selected) {
  const choices = instances.map((i) => ({
    name: `${i.name === selected?.name ? "* " : "  "}${i.name} (${i.baseUrl})`,
    value: i.name
  }));
  choices.push({
    name: "Add new instance...",
    value: "__new__"
  });
  const { choice } = await inquirer__default.default.prompt([
    {
      type: "list",
      name: "choice",
      message: "Select instance to authenticate:",
      choices,
      default: selected?.name ?? "__new__"
    }
  ]);
  return choice;
}
async function pickBaseUrl() {
  const cwd = process.cwd();
  const candidates = [];
  const patterns = [
    "app-config.yaml",
    "app-config.*.yaml",
    "packages/*/app-config.yaml",
    "packages/*/app-config.*.yaml"
  ];
  const files = patterns.flatMap((p) => glob.globSync(p, { cwd, nodir: true }));
  for (const file of files) {
    try {
      const content = await fs__default.default.readFile(path__default.default.resolve(cwd, file), "utf8");
      const doc = YAML__default.default.parse(content);
      const url = doc?.backend?.baseUrl;
      if (url) {
        candidates.push({ url: normalizeUrl(url), file });
      }
    } catch {
    }
  }
  const list = [...new Map(candidates.map((c) => [c.url, c])).values()];
  if (list.length === 0) {
    const { manual } = await inquirer__default.default.prompt([
      { type: "input", name: "manual", message: "Enter backend base URL" }
    ]);
    return normalizeUrl(manual);
  }
  if (list.length === 1) {
    return list[0].url;
  }
  const { picked } = await inquirer__default.default.prompt([
    {
      type: "list",
      name: "picked",
      message: "Select backend base URL",
      choices: [
        ...list.map((e) => ({ name: `${e.url} (${e.file})`, value: e.url })),
        { name: "Enter manually", value: "__manual__" }
      ]
    }
  ]);
  if (picked === "__manual__") {
    const { manual } = await inquirer__default.default.prompt([
      { type: "input", name: "manual", message: "Enter backend base URL" }
    ]);
    return normalizeUrl(manual);
  }
  return picked;
}
function normalizeUrl(u) {
  if (u === void 0) {
    return void 0;
  }
  try {
    const url = new URL(u);
    return url.toString().replace(/\/$/, "");
  } catch {
    throw new Error(`'${u}' is not a valid URL`);
  }
}
function deriveInstanceName(url) {
  return new URL(url).host;
}
function createPkceState() {
  const verifier = pkce.generateVerifier();
  const challenge = pkce.challengeFromVerifier(verifier);
  const state = cryptoRandom();
  return { verifier, challenge, state };
}
function buildAuthorizeUrl(options) {
  const { authBaseUrl, clientId, redirectUri, state, challenge } = options;
  const authorize = new URL(`${authBaseUrl}/v1/authorize`);
  authorize.searchParams.set("client_id", clientId);
  authorize.searchParams.set("redirect_uri", redirectUri);
  authorize.searchParams.set("response_type", "code");
  authorize.searchParams.set("scope", "openid offline_access");
  authorize.searchParams.set("state", state);
  authorize.searchParams.set("code_challenge", challenge);
  authorize.searchParams.set("code_challenge_method", "S256");
  return authorize.toString();
}
async function openBrowserOrPrint(url, noBrowser) {
  if (noBrowser) {
    process.stdout.write(`Open this URL to continue: ${url}
`);
  } else {
    process.stdout.write(`Opening the following URL: ${url}
`);
    openInBrowser(url);
  }
}
async function waitForAuthorizationCode(callback, expectedState) {
  const { code, state } = await callback.waitForCode();
  if (state !== expectedState) {
    throw new Error("State mismatch");
  }
  return code;
}
async function exchangeAuthorizationCode(options) {
  const { authBaseUrl, code, redirectUri, verifier } = options;
  return await http.httpJson(`${authBaseUrl}/v1/token`, {
    method: "POST",
    body: {
      grant_type: "authorization_code",
      code,
      redirect_uri: redirectUri,
      code_verifier: verifier
    },
    signal: AbortSignal.timeout(TOKEN_EXCHANGE_TIMEOUT_MS)
  });
}
async function persistInstance(options) {
  const { instanceName, backendBaseUrl, clientId, token } = options;
  const secretStore$1 = await secretStore.getSecretStore();
  await storage.withMetadataLock(async () => {
    const service = authIdentifiers.getAuthInstanceService(instanceName);
    await secretStore$1.set(service, "accessToken", token.access_token);
    if (token.refresh_token) {
      await secretStore$1.set(service, "refreshToken", token.refresh_token);
    } else {
      process.stderr.write(
        "Warning: No refresh token received. You will need to re-authenticate when the access token expires.\n"
      );
    }
    let existing;
    try {
      existing = await storage.getInstanceByName(instanceName);
    } catch {
    }
    await storage.upsertInstance({
      name: instanceName,
      baseUrl: backendBaseUrl,
      clientId,
      issuedAt: Date.now(),
      accessTokenExpiresAt: Date.now() + token.expires_in * 1e3,
      selected: existing?.selected,
      metadata: existing?.metadata
    });
  });
}
function cryptoRandom() {
  return crypto__default.default.randomBytes(32).toString("hex");
}
function openInBrowser(url) {
  const handleError = (error) => {
    const message = error instanceof Error ? error.message : "Unknown error";
    process.stderr.write(
      `Warning: Failed to open browser automatically: ${message}
`
    );
    process.stderr.write(`Please open this URL manually: ${url}
`);
  };
  const spawnOpts = { detached: true, stdio: "ignore" };
  let child;
  try {
    if (process.platform === "darwin") {
      child = node_child_process.spawn("open", [url], spawnOpts);
    } else if (process.platform === "win32") {
      child = node_child_process.spawn(
        "powershell",
        ["-Command", `Start-Process '${url.replace(/'/g, "''")}'`],
        spawnOpts
      );
    } else {
      child = node_child_process.spawn("xdg-open", [url], spawnOpts);
    }
    child.unref();
    child.on("error", handleError);
  } catch (error) {
    handleError(error);
  }
}

exports.default = login;
exports.openInBrowser = openInBrowser;
