// @backstage/cli-module-build/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cliNode = require('@backstage/cli-node');
var _package = require('./package.json.cjs.js');

var index = cliNode.createCliModule({
  packageJson: _package.default,
  init: async (reg) => {
    reg.addCommand({
      path: ["package", "build"],
      description: "Build a package for production deployment or publishing",
      execute: { loader: () => import('./commands/package/build/index.cjs.js') }
    });
    reg.addCommand({
      path: ["repo", "build"],
      description: "Build packages in the project, excluding bundled app and backend packages.",
      execute: { loader: () => import('./commands/repo/build.cjs.js') }
    });
    reg.addCommand({
      path: ["package", "bundle"],
      description: "Bundle a plugin for dynamic loading. Creates a self-contained plugin package that can be deployed and loaded dynamically by a Backstage application. Supports both backend and frontend plugins. Experimental.",
      experimental: true,
      execute: { loader: () => import('./commands/package/bundle/index.cjs.js') }
    });
    reg.addCommand({
      path: ["package", "start"],
      description: "Start a package for local development",
      execute: { loader: () => import('./commands/package/start/index.cjs.js') }
    });
    reg.addCommand({
      path: ["repo", "start"],
      description: "Starts packages in the repo for local development",
      execute: { loader: () => import('./commands/repo/start.cjs.js') }
    });
    reg.addCommand({
      path: ["package", "clean"],
      description: "Delete cache directories",
      execute: {
        loader: () => import('./commands/package/clean.cjs.js')
      }
    });
    reg.addCommand({
      path: ["package", "prepack"],
      description: "Prepares a package for packaging before publishing",
      execute: {
        loader: () => import('./commands/package/prepack.cjs.js')
      }
    });
    reg.addCommand({
      path: ["package", "postpack"],
      description: "Restores the changes made by the prepack command",
      execute: {
        loader: () => import('./commands/package/postpack.cjs.js')
      }
    });
    reg.addCommand({
      path: ["repo", "clean"],
      description: "Delete cache and output directories",
      execute: {
        loader: () => import('./commands/repo/clean.cjs.js')
      }
    });
    reg.addCommand({
      path: ["build-workspace"],
      description: "Builds a temporary dist workspace from the provided packages",
      execute: { loader: () => import('./commands/buildWorkspace.cjs.js') }
    });
  }
});

exports.default = index;
