// @backstage/cli-module-new/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cliNode = require('@backstage/cli-node');
var errors = require('@backstage/errors');
var _package = require('./packages/cli-module-new/package.json.cjs.js');

var index = cliNode.createCliModule({
  packageJson: _package.default,
  init: async (reg) => {
    reg.addCommand({
      path: ["new"],
      description: "Open up an interactive guide to creating new things in your app",
      execute: { loader: () => import('./commands/new.cjs.js') }
    });
    reg.addCommand({
      path: ["create"],
      description: "Create a new Backstage app",
      deprecated: true,
      execute: async () => {
        throw new errors.NotImplementedError(
          `This command has been removed, use 'backstage-cli new' instead`
        );
      }
    });
    reg.addCommand({
      path: ["create-plugin"],
      description: "Create a new Backstage plugin",
      deprecated: true,
      execute: async () => {
        throw new errors.NotImplementedError(
          `This command has been removed, use 'backstage-cli new' instead`
        );
      }
    });
  }
});

exports.default = index;
