// @backstage/cli-module-actions/dist/lib/ActionsClient.cjs.js
'use strict';

var httpJson = require('./httpJson.cjs.js');

function extractPluginId(actionId) {
  const colonIndex = actionId.indexOf(":");
  if (colonIndex === -1) {
    throw new Error(
      `Invalid action ID "${actionId}". Expected format "pluginId:actionName".`
    );
  }
  return actionId.substring(0, colonIndex);
}
function pluginActionsUrl(baseUrl, pluginId) {
  return new URL(
    `/api/${encodeURIComponent(pluginId)}/.backstage/actions/v1/actions`,
    baseUrl
  ).toString();
}
class ActionsClient {
  constructor(baseUrl, accessToken) {
    this.baseUrl = baseUrl;
    this.accessToken = accessToken;
  }
  async list(pluginSources) {
    return Promise.all(
      pluginSources.map(async (pluginId) => {
        const url = pluginActionsUrl(this.baseUrl, pluginId);
        const response = await httpJson.httpJson(url, {
          headers: { Authorization: `Bearer ${this.accessToken}` },
          signal: AbortSignal.timeout(3e4)
        });
        return { pluginId, actions: response.actions };
      })
    );
  }
  async listForPlugin(actionId) {
    const pluginId = extractPluginId(actionId);
    const grouped = await this.list([pluginId]);
    return grouped.flatMap((g) => g.actions);
  }
  async execute(actionId, input) {
    const pluginId = extractPluginId(actionId);
    const url = `${pluginActionsUrl(
      this.baseUrl,
      pluginId
    )}/${encodeURIComponent(actionId)}/invoke`;
    const response = await httpJson.httpJson(url, {
      method: "POST",
      headers: { Authorization: `Bearer ${this.accessToken}` },
      body: input ?? {},
      signal: AbortSignal.timeout(3e4)
    });
    return response.output;
  }
}

exports.ActionsClient = ActionsClient;
