// @backstage/cli-module-actions/dist/lib/format.cjs.js
'use strict';

var chalk = require('chalk');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

async function renderMarkdown(text) {
  const { Marked } = await import('marked');
  const { markedTerminal } = await import('marked-terminal');
  const stripAnsi = (await import('strip-ansi')).default;
  const instance = new Marked(markedTerminal());
  const sanitized = stripAnsi(text);
  return instance.parse(sanitized);
}
function dedent(text) {
  const trimmed = text.replace(/^\n+/, "").replace(/\n+$/, "");
  const lines = trimmed.split("\n");
  const nonEmptyLines = lines.filter((l) => l.trim().length > 0);
  if (nonEmptyLines.length === 0) return trimmed;
  const minIndent = Math.min(
    ...nonEmptyLines.map((l) => l.match(/^(\s*)/)[0].length)
  );
  if (minIndent === 0) return trimmed;
  return lines.map((l) => l.slice(minIndent)).join("\n");
}
function terminalWidth() {
  return process.stdout.columns || 80;
}
function formatActionList(grouped) {
  const width = terminalWidth();
  const lines = [];
  for (let i = 0; i < grouped.length; i++) {
    const { pluginId, actions } = grouped[i];
    if (i > 0) {
      lines.push("");
    }
    const header = `\u2500\u2500 ${pluginId} `;
    const remaining = Math.max(0, width - header.length);
    lines.push(chalk__default.default.bold(`${header}${"\u2500".repeat(remaining)}`));
    lines.push("");
    if (actions.length === 0) {
      lines.push(`  ${chalk__default.default.dim("(no actions)")}`);
      continue;
    }
    const maxIdLen = Math.max(...actions.map((a) => a.id.length));
    const idColWidth = maxIdLen + 4;
    for (const action of actions) {
      const paddedId = action.id.padEnd(idColWidth);
      if (!action.title) {
        lines.push(`  ${chalk__default.default.cyan(paddedId.trimEnd())}`);
        continue;
      }
      lines.push(`  ${chalk__default.default.cyan(paddedId)}${chalk__default.default.dim(action.title)}`);
    }
  }
  return lines.join("\n");
}
const typeHintNames = {
  String: "string",
  Number: "number",
  Boolean: ""
};
function flagDefsToFlagInfo(defs) {
  return Object.entries(defs).map(([name, def]) => ({
    name,
    type: typeHintNames[def.type.name] ?? "string",
    description: def.description
  }));
}
async function formatActionHelp(options) {
  const { action, usage, flags } = options;
  const lines = [];
  lines.push(chalk__default.default.bold.cyan(action.id));
  if (action.title) {
    lines.push(`  ${action.title}`);
  }
  if (action.description) {
    lines.push("");
    const dedented = dedent(action.description);
    const rendered = await renderMarkdown(dedented);
    lines.push(rendered.trimEnd());
  }
  lines.push("");
  lines.push(chalk__default.default.bold("Usage:"));
  lines.push(`  ${usage} ${chalk__default.default.dim("[flags]")}`);
  if (flags.length > 0) {
    lines.push("");
    lines.push(chalk__default.default.bold("Flags:"));
    const maxFlagLen = Math.max(
      ...flags.map((f) => {
        const typeHint = f.type ? ` <${f.type}>` : "";
        return `  --${f.name}${typeHint}`.length;
      })
    );
    const colWidth = maxFlagLen + 4;
    for (const flag of flags) {
      const typeHint = flag.type ? ` <${flag.type}>` : "";
      const left = `  --${flag.name}${typeHint}`.padEnd(colWidth);
      const desc = flag.description ? chalk__default.default.dim(flag.description) : "";
      lines.push(`${left}${desc}`);
    }
  }
  lines.push("");
  return lines.join("\n");
}

exports.flagDefsToFlagInfo = flagDefsToFlagInfo;
exports.formatActionHelp = formatActionHelp;
exports.formatActionList = formatActionList;
