// @backstage/cli-module-lint/dist/commands/repo/lint.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var chalk = require('chalk');
var fs = require('fs-extra');
var cleye = require('cleye');
var node_crypto = require('node:crypto');
var node_path = require('node:path');
var cliNode = require('@backstage/cli-node');
var cliCommon = require('@backstage/cli-common');
var optionsParser = require('../../lib/optionsParser.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

function depCount(pkg) {
  const deps = pkg.dependencies ? Object.keys(pkg.dependencies).length : 0;
  const devDeps = pkg.devDependencies ? Object.keys(pkg.devDependencies).length : 0;
  return deps + devDeps;
}
var lint = async ({ args, info }) => {
  for (const flag of [
    "outputFile",
    "successCache",
    "successCacheDir",
    "maxWarnings"
  ]) {
    if (args.some((a) => a === `--${flag}` || a.startsWith(`--${flag}=`))) {
      process.stderr.write(
        `DEPRECATION WARNING: --${flag} is deprecated, use the kebab-case form instead
`
      );
    }
  }
  const {
    flags: {
      fix,
      format,
      outputFile,
      successCache: useSuccessCache,
      successCacheDir,
      since,
      maxWarnings
    }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        fix: {
          type: Boolean,
          description: "Attempt to automatically fix violations"
        },
        format: {
          type: String,
          description: "Lint report output format",
          default: "eslint-formatter-friendly"
        },
        outputFile: {
          type: String,
          description: "Write the lint report to a file instead of stdout"
        },
        successCache: {
          type: Boolean,
          description: "Enable success caching, which skips running tests for unchanged packages that were successful in the previous run"
        },
        successCacheDir: {
          type: String,
          description: "Set the success cache location, (default: node_modules/.cache/backstage-cli)"
        },
        since: {
          type: String,
          description: "Only lint packages that changed since the specified ref"
        },
        maxWarnings: {
          type: String,
          description: "Fail if more than this number of warnings. -1 allows warnings. (default: -1)"
        }
      }
    },
    void 0,
    args
  );
  let packages = await cliNode.PackageGraph.listTargetPackages();
  const cache = cliNode.SuccessCache.create({
    name: "lint",
    basePath: successCacheDir
  });
  const cacheContext = useSuccessCache ? {
    entries: await cache.read(),
    lockfile: await cliNode.Lockfile.load(cliCommon.targetPaths.resolveRoot("yarn.lock"))
  } : void 0;
  if (since) {
    const graph = cliNode.PackageGraph.fromPackages(packages);
    packages = await graph.listChangedPackages({
      ref: since,
      analyzeLockfile: true
    });
  }
  packages.sort((a, b) => depCount(b.packageJson) - depCount(a.packageJson));
  if (format === "eslint-formatter-friendly") {
    process.chdir(cliCommon.targetPaths.rootDir);
  }
  if (!process.env.FORCE_COLOR) {
    process.env.FORCE_COLOR = "1";
  }
  const parseLintScript = optionsParser.createScriptOptionsParser(["package", "lint"], {
    fix: { type: "boolean" },
    format: { type: "string" },
    "output-file": { type: "string" },
    "max-warnings": { type: "string" }
  });
  const items = await Promise.all(
    packages.map(async (pkg) => {
      const lintOptions = parseLintScript(pkg.packageJson.scripts?.lint);
      const base = {
        fullDir: pkg.dir,
        relativeDir: node_path.relative(cliCommon.targetPaths.rootDir, pkg.dir),
        lintOptions,
        parentHash: void 0
      };
      if (!cacheContext) {
        return base;
      }
      const hash = node_crypto.createHash("sha1");
      hash.update(
        cacheContext.lockfile.getDependencyTreeHash(pkg.packageJson.name)
      );
      hash.update("\0");
      hash.update(JSON.stringify(lintOptions ?? {}));
      hash.update("\0");
      hash.update(process.version);
      hash.update("\0");
      hash.update("v1");
      return {
        ...base,
        parentHash: hash.digest("hex")
      };
    })
  );
  const { results: resultsList } = await cliNode.runWorkerQueueThreads({
    items: items.filter((item) => item.lintOptions),
    // Filter out packages without lint script
    context: {
      fix: Boolean(fix),
      format,
      shouldCache: Boolean(cacheContext),
      maxWarnings: maxWarnings ?? "-1",
      successCache: cacheContext?.entries,
      rootDir: cliCommon.targetPaths.rootDir
    },
    workerFactory: async ({
      fix: workerFix,
      format: workerFormat,
      shouldCache,
      successCache,
      rootDir,
      maxWarnings: workerMaxWarnings
    }) => {
      const { ESLint } = require("eslint");
      const crypto = require("node:crypto");
      const globby = require("globby");
      const { readFile } = require("node:fs/promises");
      const workerPath = require("node:path");
      return async ({
        fullDir,
        relativeDir,
        parentHash
      }) => {
        process.cwd = () => fullDir;
        const start = Date.now();
        const eslint = new ESLint({
          cwd: fullDir,
          fix: workerFix,
          extensions: ["js", "jsx", "ts", "tsx", "mjs", "cjs"]
        });
        let sha = void 0;
        if (shouldCache) {
          const result = await globby(relativeDir, {
            gitignore: true,
            onlyFiles: true,
            cwd: rootDir
          });
          const hash = crypto.createHash("sha1");
          hash.update(parentHash);
          hash.update("\0");
          for (const path of result.sort()) {
            const absPath = workerPath.resolve(rootDir, path);
            const pathInPackage = workerPath.relative(fullDir, absPath);
            if (await eslint.isPathIgnored(pathInPackage)) {
              continue;
            }
            hash.update(pathInPackage);
            hash.update("\0");
            hash.update(await readFile(absPath));
            hash.update("\0");
            hash.update(
              JSON.stringify(
                await eslint.calculateConfigForFile(pathInPackage)
              ).replaceAll(rootDir, "")
            );
            hash.update("\0");
          }
          sha = await hash.digest("hex");
          if (successCache?.has(sha)) {
            console.log(`Skipped ${relativeDir} due to cache hit`);
            return { relativeDir, sha, failed: false };
          }
        }
        const formatter = await eslint.loadFormatter(workerFormat);
        const results = await eslint.lintFiles(["."]);
        const count = String(results.length).padStart(3);
        const time = ((Date.now() - start) / 1e3).toFixed(2);
        console.log(`Checked ${count} files in ${relativeDir} ${time}s`);
        if (workerFix) {
          await ESLint.outputFixes(results);
        }
        const ignoreWarnings = +workerMaxWarnings === -1;
        const resultText = formatter.format(results);
        const failed2 = results.some((r) => r.errorCount > 0) || !ignoreWarnings && results.reduce((current, next) => current + next.warningCount, 0) > +workerMaxWarnings;
        return {
          relativeDir,
          resultText,
          failed: failed2,
          sha
        };
      };
    }
  });
  const outputSuccessCache = [];
  const jsonResults = [];
  let errorOutput = "";
  let failed = false;
  for (const {
    relativeDir,
    resultText,
    failed: runFailed,
    sha
  } of resultsList) {
    if (runFailed) {
      console.log(chalk__default.default.red(`Lint failed in ${relativeDir}`));
      failed = true;
      if (resultText) {
        if (outputFile) {
          if (format === "json") {
            jsonResults.push(resultText);
          } else {
            errorOutput += `${resultText}
`;
          }
        } else {
          console.log();
          console.log(resultText.trimStart());
        }
      }
    } else if (sha) {
      outputSuccessCache.push(sha);
    }
  }
  if (format === "json") {
    let mergedJsonResults = [];
    for (const jsonResult of jsonResults) {
      mergedJsonResults = mergedJsonResults.concat(JSON.parse(jsonResult));
    }
    errorOutput = JSON.stringify(mergedJsonResults, null, 2);
  }
  if (outputFile && errorOutput) {
    await fs__default.default.writeFile(cliCommon.targetPaths.resolveRoot(outputFile), errorOutput);
  }
  if (cacheContext) {
    await cache.write(outputSuccessCache);
  }
  if (failed) {
    process.exit(1);
  }
};

exports.default = lint;
