// @backstage/cli-module-config/dist/commands/docs.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var configLoader = require('@backstage/config-loader');
var openBrowser = require('react-dev-utils/openBrowser');
var chalk = require('chalk');
var config = require('../lib/config.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var openBrowser__default = /*#__PURE__*/_interopDefaultCompat(openBrowser);
var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

const DOCS_URL = "https://config.backstage.io";
var docs = async ({ args, info }) => {
  const {
    flags: { package: pkg }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        package: {
          type: String,
          description: "Only include the schema that applies to the given package"
        }
      }
    },
    void 0,
    args
  );
  const { schema: appSchemas } = await config.loadCliConfig({
    args: [],
    fromPackage: pkg,
    mockEnv: true
  });
  const schema = configLoader.mergeConfigSchemas(
    appSchemas.serialize().schemas.map(
      (_) => _.value
    )
  );
  const url = `${DOCS_URL}#schema=${JSON.stringify(schema)}`;
  console.log();
  console.log(
    chalk__default.default.cyan(
      "Opening configuration reference documentation in your browser..."
    )
  );
  console.log(`  ${chalk__default.default.cyan(url)}`);
  console.log();
  const opened = openBrowser__default.default(url);
  if (!opened) {
    console.log(
      chalk__default.default.yellow("\u26A0\uFE0F  WARNING: Unable to open browser automatically.")
    );
    console.log(chalk__default.default.yellow("Please open the URL manually in your browser."));
  }
};

exports.default = docs;
