// @backstage/cli-module-config/dist/commands/schema.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var yaml = require('yaml');
var config = require('../lib/config.cjs.js');
var configLoader = require('@backstage/config-loader');

var schema = async ({ args, info }) => {
  const {
    flags: { merge, format, package: pkg }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        package: { type: String, description: "Package to print schema for" },
        format: { type: String, description: "Output format (yaml or json)" },
        merge: {
          type: Boolean,
          description: "Merge all schemas into a single schema"
        }
      }
    },
    void 0,
    args
  );
  const { schema } = await config.loadCliConfig({
    args: [],
    fromPackage: pkg,
    mockEnv: true
  });
  let configSchema;
  if (merge) {
    configSchema = configLoader.mergeConfigSchemas(
      schema.serialize().schemas.map(
        (_) => _.value
      )
    );
    configSchema.title = "Application Configuration Schema";
    configSchema.description = "This is the schema describing the structure of the app-config.yaml configuration file.";
  } else {
    configSchema = schema.serialize();
  }
  if (format === "json") {
    process.stdout.write(`${JSON.stringify(configSchema, null, 2)}
`);
  } else {
    process.stdout.write(`${yaml.stringify(configSchema)}
`);
  }
};

exports.default = schema;
