// @backstage/cli-module-config/dist/commands/validate.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var config = require('../lib/config.cjs.js');

var validate = async ({ args, info }) => {
  const {
    flags: { config: config$1, lax, frontend, deprecated, strict, package: pkg }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        package: {
          type: String,
          description: "Package to validate config for"
        },
        lax: {
          type: Boolean,
          description: "Do not require environment variables to be set"
        },
        frontend: {
          type: Boolean,
          description: "Only validate frontend config"
        },
        deprecated: { type: Boolean, description: "Output deprecated keys" },
        strict: { type: Boolean, description: "Enable strict validation" },
        config: {
          type: [String],
          description: "Config files to load instead of app-config.yaml"
        }
      }
    },
    void 0,
    args
  );
  await config.loadCliConfig({
    args: config$1,
    fromPackage: pkg,
    mockEnv: lax,
    fullVisibility: !frontend,
    withDeprecatedKeys: deprecated,
    strict
  });
};

exports.default = validate;
