// @backstage/cli-module-new/dist/lib/preparation/collectPortableTemplateInput.cjs.js
'use strict';

var inquirer = require('inquirer');
var codeowners = require('../codeowners/codeowners.cjs.js');
var cliCommon = require('@backstage/cli-common');
require('../../packages/cli-internal/src/InternalCliModule.cjs.js');
require('../../packages/cli-internal/src/InternalCommandNode.cjs.js');
require('node:fs');
require('node:os');
require('node:path');
var knownPluginPackages = require('../../packages/cli-internal/src/knownPluginPackages.cjs.js');
var resolvePackageParams = require('./resolvePackageParams.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var inquirer__default = /*#__PURE__*/_interopDefaultCompat(inquirer);

async function collectPortableTemplateInput(options) {
  const { config, template, prefilledParams } = options;
  const codeOwnersFilePath = await codeowners.getCodeownersFilePath(cliCommon.targetPaths.rootDir);
  const prompts = getPromptsForRole(template.role);
  if (codeOwnersFilePath) {
    prompts.push(ownerPrompt());
  }
  const deprecatedParams = {};
  if (config.isUsingDefaultTemplates && prefilledParams.id) {
    console.warn(
      `DEPRECATION WARNING: The 'id' parameter is deprecated, use 'pluginId' instead`
    );
    deprecatedParams.pluginId = prefilledParams.id;
  }
  const parameters = {
    ...template.values,
    ...prefilledParams,
    ...deprecatedParams
  };
  if (template.role === "backend-plugin-module" || template.role === "frontend-plugin-module") {
    const pluginPkg = parameters.pluginPackage?.trim();
    if (!pluginPkg || !isValidNpmPackageName(pluginPkg)) {
      delete parameters.pluginPackage;
    } else {
      parameters.pluginPackage = pluginPkg;
    }
    if (parameters.pluginId && !parameters.pluginPackage) {
      const knownPackages = template.role === "backend-plugin-module" ? knownPluginPackages.knownBackendPluginPackageNameByPluginId : knownPluginPackages.knownFrontendPluginPackageNameByPluginId;
      const knownPackage = knownPackages[parameters.pluginId];
      if (knownPackage) {
        parameters.pluginPackage = knownPackage;
      }
    }
  }
  const needsAnswer = [];
  const prefilledAnswers = {};
  for (const prompt of prompts) {
    if (prompt.name && parameters[prompt.name] !== void 0) {
      prefilledAnswers[prompt.name] = parameters[prompt.name];
    } else {
      needsAnswer.push(prompt);
    }
  }
  const promptAnswers = await inquirer__default.default.prompt(
    needsAnswer
  );
  const answers = {
    ...parameters,
    ...prefilledAnswers,
    ...promptAnswers
  };
  let pluginPackage;
  if (template.role === "backend-plugin-module" || template.role === "frontend-plugin-module") {
    const knownPackages = template.role === "backend-plugin-module" ? knownPluginPackages.knownBackendPluginPackageNameByPluginId : knownPluginPackages.knownFrontendPluginPackageNameByPluginId;
    pluginPackage = answers.pluginPackage ?? knownPackages[answers.pluginId];
  }
  const roleParams = {
    role: template.role,
    name: answers.name,
    pluginId: answers.pluginId,
    moduleId: answers.moduleId,
    pluginPackage
  };
  const packageParams = resolvePackageParams.resolvePackageParams({
    roleParams,
    packagePrefix: config.packageNamePrefix,
    pluginInfix: config.packageNamePluginInfix
  });
  return {
    roleParams,
    owner: answers.owner,
    license: config.license,
    version: config.version,
    private: config.private,
    publishRegistry: config.publishRegistry,
    packageName: packageParams.packageName,
    packagePath: packageParams.packagePath
  };
}
function namePrompt() {
  return {
    type: "input",
    name: "name",
    message: "Enter the name of the package, without scope [required]",
    validate: (value) => {
      if (!value) {
        return "Please enter the name of the package";
      } else if (!/^[a-z0-9]+(-[a-z0-9]+)*$/.test(value)) {
        return "Package names must be lowercase and contain only letters, digits, and dashes.";
      }
      return true;
    }
  };
}
function pluginIdPrompt() {
  return {
    type: "input",
    name: "pluginId",
    message: "Enter the ID of the plugin [required]",
    validate: (value) => {
      if (!value) {
        return "Please enter the ID of the plugin";
      } else if (!/^[a-z0-9]+(-[a-z0-9]+)*$/.test(value)) {
        return "Plugin IDs must be lowercase and contain only letters, digits, and dashes.";
      }
      return true;
    }
  };
}
function moduleIdIdPrompt() {
  return {
    type: "input",
    name: "moduleId",
    message: "Enter the ID of the module [required]",
    validate: (value) => {
      if (!value) {
        return "Please enter the ID of the module";
      } else if (!/^[a-z0-9]+(-[a-z0-9]+)*$/.test(value)) {
        return "Module IDs must be lowercase and contain only letters, digits, and dashes.";
      }
      return true;
    }
  };
}
function pluginPackagePrompt(role) {
  const knownPackages = role === "backend-plugin-module" ? knownPluginPackages.knownBackendPluginPackageNameByPluginId : knownPluginPackages.knownFrontendPluginPackageNameByPluginId;
  const examplePackage = role === "backend-plugin-module" ? "@backstage/plugin-catalog-backend" : "@backstage/plugin-catalog";
  return {
    type: "input",
    name: "pluginPackage",
    filter: (value) => value.trim(),
    message: `Enter the package name of the plugin this module extends (e.g. ${examplePackage}) [required]`,
    validate: (value) => {
      if (!value) {
        return "Please enter the package name of the plugin";
      }
      if (!isValidNpmPackageName(value)) {
        return `Please enter a valid npm package name (e.g. ${examplePackage} or my-plugin)`;
      }
      return true;
    },
    when: (answers) => !knownPackages[answers.pluginId]
  };
}
function getPromptsForRole(role) {
  switch (role) {
    case "web-library":
    case "node-library":
    case "common-library":
    case "cli-module":
      return [namePrompt()];
    case "plugin-web-library":
    case "plugin-node-library":
    case "plugin-common-library":
    case "frontend-plugin":
    case "backend-plugin":
      return [pluginIdPrompt()];
    case "frontend-plugin-module":
    case "backend-plugin-module":
      return [pluginIdPrompt(), moduleIdIdPrompt(), pluginPackagePrompt(role)];
    default:
      return [];
  }
}
function ownerPrompt() {
  return {
    type: "input",
    name: "owner",
    message: "Enter an owner to add to CODEOWNERS [optional]",
    validate: (value) => {
      if (!value) {
        return true;
      }
      const ownerIds = codeowners.parseOwnerIds(value);
      if (!ownerIds) {
        return "The owner must be a space separated list of team names (e.g. @org/team-name), usernames (e.g. @username), or the email addresses (e.g. user@example.com).";
      }
      return true;
    }
  };
}
const packageNamePattern = /^[a-z0-9][a-z0-9._-]*$/;
const scopedPackageNamePattern = /^@[a-z0-9][a-z0-9._-]*\/[a-z0-9][a-z0-9._-]*$/;
function isValidNpmPackageName(name) {
  return packageNamePattern.test(name) || scopedPackageNamePattern.test(name);
}

exports.collectPortableTemplateInput = collectPortableTemplateInput;
exports.getPromptsForRole = getPromptsForRole;
exports.moduleIdIdPrompt = moduleIdIdPrompt;
exports.namePrompt = namePrompt;
exports.ownerPrompt = ownerPrompt;
exports.pluginIdPrompt = pluginIdPrompt;
exports.pluginPackagePrompt = pluginPackagePrompt;
