// @backstage/cli-module-new/dist/lib/preparation/loadPortableTemplateConfig.cjs.js
'use strict';

var fs = require('fs-extra');
var path = require('node:path');
var cliCommon = require('@backstage/cli-common');
var defaultTemplates = require('../defaultTemplates.cjs.js');
var types = require('../types.cjs.js');
var yaml = require('yaml');
var v3$1 = require('zod/v3');
var v3 = require('zod-validation-error/v3');
var errors = require('@backstage/errors');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

async function detectFrontendSystem(basePath) {
  const appPkgPath = path.join(basePath, "packages", "app", "package.json");
  try {
    const appPkgJson = await fs__default.default.readJson(appPkgPath);
    const deps = {
      ...appPkgJson.dependencies,
      ...appPkgJson.devDependencies
    };
    if (deps["@backstage/frontend-defaults"] || deps["@backstage/frontend-app-api"]) {
      return "new";
    }
    if (deps["@backstage/app-defaults"] || deps["@backstage/core-app-api"]) {
      return "legacy";
    }
  } catch {
  }
  return "unknown";
}
const newFrontendTemplates = [
  "@backstage/cli-module-new/templates/frontend-plugin",
  "@backstage/cli-module-new/templates/frontend-plugin-module"
];
const legacyFrontendTemplates = [
  "@backstage/cli-module-new/templates/legacy-frontend-plugin"
];
function filterTemplateEntriesForFrontendSystem(entries, frontendSystem) {
  if (frontendSystem === "unknown") {
    return entries;
  }
  if (frontendSystem === "new") {
    return entries.filter((e) => !legacyFrontendTemplates.includes(e.rawPointer));
  }
  return entries.filter((e) => !newFrontendTemplates.includes(e.rawPointer));
}
const defaults = {
  license: "Apache-2.0",
  version: "0.1.0",
  private: true,
  publishRegistry: void 0,
  packageNamePrefix: "@internal/",
  packageNamePluginInfix: "plugin-"
};
const newConfigSchema = v3$1.z.object({
  templates: v3$1.z.array(v3$1.z.string()).optional(),
  globals: v3$1.z.object({
    license: v3$1.z.string().optional(),
    version: v3$1.z.string().optional(),
    private: v3$1.z.boolean().optional(),
    publishRegistry: v3$1.z.string().optional(),
    namePrefix: v3$1.z.string().optional(),
    namePluginInfix: v3$1.z.string().optional()
  }).optional()
}).strict();
const pkgJsonWithNewConfigSchema = v3$1.z.object({
  backstage: v3$1.z.object({
    cli: v3$1.z.object({
      new: newConfigSchema.optional()
    }).optional()
  }).optional()
});
function computePackageNamePluginInfix(packageNamePrefix, namePluginInfix) {
  const packageNamePluginInfix = namePluginInfix ?? (packageNamePrefix.includes("backstage") ? defaults.packageNamePluginInfix : "backstage-plugin-");
  return {
    packageNamePluginInfix
  };
}
async function loadPortableTemplateConfig(options = {}) {
  const { overrides = {} } = options;
  const pkgPath = options.packagePath ?? cliCommon.targetPaths.resolveRoot("package.json");
  const pkgJson = await fs__default.default.readJson(pkgPath);
  const parsed = pkgJsonWithNewConfigSchema.safeParse(pkgJson);
  if (!parsed.success) {
    throw new errors.ForwardedError(
      `Failed to load templating configuration from '${pkgPath}'`,
      v3.fromZodError(parsed.error)
    );
  }
  const config = parsed.data.backstage?.cli?.new;
  const basePath = path.dirname(pkgPath);
  const isUsingDefaultTemplates = !config?.templates;
  let templatePointerEntries = await Promise.all(
    (config?.templates ?? defaultTemplates.defaultTemplates).map(async (rawPointer) => {
      try {
        const templatePath = resolveLocalTemplatePath(rawPointer, basePath);
        const pointer = await peekLocalTemplateDefinition(templatePath);
        return { pointer, rawPointer };
      } catch (error) {
        throw new errors.ForwardedError(
          `Failed to load template definition '${rawPointer}'`,
          error
        );
      }
    })
  );
  if (isUsingDefaultTemplates) {
    const frontendSystem = await detectFrontendSystem(basePath);
    templatePointerEntries = filterTemplateEntriesForFrontendSystem(
      templatePointerEntries,
      frontendSystem
    );
  }
  const templateNameConflicts = /* @__PURE__ */ new Map();
  for (const { pointer, rawPointer } of templatePointerEntries) {
    const conflict = templateNameConflicts.get(pointer.name);
    if (conflict) {
      throw new Error(
        `Invalid template configuration, received conflicting template name '${pointer.name}' from '${conflict}' and '${rawPointer}'`
      );
    }
    templateNameConflicts.set(pointer.name, rawPointer);
  }
  const packageNamePrefix = overrides.packageNamePrefix ?? config?.globals?.namePrefix ?? defaults.packageNamePrefix;
  const { packageNamePluginInfix } = computePackageNamePluginInfix(
    packageNamePrefix,
    overrides.packageNamePluginInfix ?? config?.globals?.namePluginInfix
  );
  return {
    isUsingDefaultTemplates,
    templatePointers: templatePointerEntries.map(({ pointer }) => pointer),
    license: overrides.license ?? config?.globals?.license ?? defaults.license,
    version: overrides.version ?? config?.globals?.version ?? defaults.version,
    private: overrides.private ?? config?.globals?.private ?? defaults.private,
    publishRegistry: overrides.publishRegistry ?? config?.globals?.publishRegistry ?? defaults.publishRegistry,
    packageNamePrefix,
    packageNamePluginInfix
  };
}
const CLI_TEMPLATE_PREFIX = "@backstage/cli/templates/";
const CLI_MODULE_NEW_TEMPLATE_PREFIX = "@backstage/cli-module-new/templates/";
function resolveLocalTemplatePath(pointer, basePath) {
  if (path.isAbsolute(pointer)) {
    throw new Error(`Template target may not be an absolute path`);
  }
  if (pointer.startsWith(".")) {
    return path.resolve(basePath, pointer, types.TEMPLATE_FILE_NAME);
  }
  const resolvedPointer = pointer.startsWith(CLI_TEMPLATE_PREFIX) ? `${CLI_MODULE_NEW_TEMPLATE_PREFIX}${pointer.slice(
    CLI_TEMPLATE_PREFIX.length
  )}` : pointer;
  return require.resolve(`${resolvedPointer}/${types.TEMPLATE_FILE_NAME}`, {
    paths: [basePath]
  });
}
const partialTemplateDefinitionSchema = v3$1.z.object({
  name: v3$1.z.string(),
  description: v3$1.z.string().optional()
});
async function peekLocalTemplateDefinition(target) {
  const content = await fs__default.default.readFile(target, "utf8");
  const rawTemplate = yaml.parse(content);
  const parsed = partialTemplateDefinitionSchema.safeParse(rawTemplate);
  if (!parsed.success) {
    throw v3.fromZodError(parsed.error);
  }
  return {
    name: parsed.data.name,
    description: parsed.data.description,
    target
  };
}

exports.loadPortableTemplateConfig = loadPortableTemplateConfig;
