// @backstage/catalog-model/dist/kinds/AiResourceEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import 'lodash';
import 'ajv';
import { entityKindSchemaValidator } from '../validation/entityKindSchemaValidator.esm.js';
import '../validation/makeValidator.esm.js';
import defaultJsonSchema from '../schema/kinds/AiResource.v1alpha1.schema.json.esm.js';
import skillJsonSchema from '../schema/kinds/AiResource.v1alpha1.skill.schema.json.esm.js';
import ruleJsonSchema from '../schema/kinds/AiResource.v1alpha1.rule.schema.json.esm.js';

const defaultValidator = entityKindSchemaValidator(defaultJsonSchema);
const aiResourceEntityV1alpha1Validator = {
  async check(data) {
    return defaultValidator(data) === data;
  }
};
const skillValidator = entityKindSchemaValidator(skillJsonSchema);
const skillAiResourceEntityV1alpha1Validator = {
  async check(data) {
    return skillValidator(data) === data;
  }
};
const isAiResourceEntity = (entity) => entity.apiVersion === "backstage.io/v1alpha1" && entity.kind === "AiResource";
const isSkillAiResourceEntity = (entity) => isAiResourceEntity(entity) && entity.spec?.type === "skill";
const isRuleAiResourceEntity = (entity) => isAiResourceEntity(entity) && entity.spec?.type === "rule";
const ruleValidator = entityKindSchemaValidator(ruleJsonSchema);
const ruleAiResourceEntityV1alpha1Validator = {
  async check(data) {
    return ruleValidator(data) === data;
  }
};
const baseRelationFields = [
  {
    selector: { path: "spec.owner" },
    relation: "ownedBy",
    defaultKind: "Group",
    defaultNamespace: "inherit",
    allowedKinds: ["Group", "User"]
  },
  {
    selector: { path: "spec.system" },
    relation: "partOf",
    defaultKind: "System",
    defaultNamespace: "inherit"
  }
];
const aiResourceEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-ai-resource",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "AiResource",
        singular: "airesource",
        plural: "airesources"
      },
      description: "An AI resource represents contextual information consumed by AI coding tools, such as skills and rules.",
      versions: [
        {
          name: "v1alpha1",
          relationFields: baseRelationFields,
          schema: {
            jsonSchema: defaultJsonSchema
          }
        },
        {
          name: "v1alpha1",
          specType: "skill",
          relationFields: [
            ...baseRelationFields,
            {
              selector: { path: "spec.dependsOn" },
              relation: "dependsOn",
              defaultKind: "AiResource",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: skillJsonSchema
          }
        },
        {
          name: "v1alpha1",
          specType: "rule",
          relationFields: baseRelationFields,
          schema: {
            jsonSchema: ruleJsonSchema
          }
        }
      ]
    });
  }
});

export { aiResourceEntityModel, aiResourceEntityV1alpha1Validator, isAiResourceEntity, isRuleAiResourceEntity, isSkillAiResourceEntity, ruleAiResourceEntityV1alpha1Validator, skillAiResourceEntityV1alpha1Validator };
