// @backstage/catalog-model/dist/kinds/GroupEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var Group_v1alpha1_schema = require('../schema/kinds/Group.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const groupEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(Group_v1alpha1_schema.default);
const groupEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-group",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Group",
        singular: "group",
        plural: "groups"
      },
      description: "A Group describes an organizational entity, such as a team, a business unit, or a loose collection of people.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.parent" },
              relation: "childOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            },
            {
              selector: { path: "spec.children" },
              relation: "parentOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            },
            {
              selector: { path: "spec.members" },
              relation: "hasMember",
              defaultKind: "User",
              defaultNamespace: "inherit",
              allowedKinds: ["User"]
            }
          ],
          schema: {
            jsonSchema: Group_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.groupEntityModel = groupEntityModel;
exports.groupEntityV1alpha1Validator = groupEntityV1alpha1Validator;
