// @backstage/catalog-model/dist/kinds/LocationEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var Location_v1alpha1_schema = require('../schema/kinds/Location.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const locationEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(Location_v1alpha1_schema.default);
const locationEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-location",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Location",
        singular: "location",
        plural: "locations"
      },
      description: "A Location is a marker that references other places to look for catalog data.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          schema: {
            jsonSchema: Location_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.locationEntityModel = locationEntityModel;
exports.locationEntityV1alpha1Validator = locationEntityV1alpha1Validator;
