// @backstage/catalog-model/dist/kinds/McpServerApiEntity.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import mcpServerSchema from '../schema/kinds/API.v1alpha1.mcp-server.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const mcpServerApiEntityValidator = ajvCompiledJsonSchemaValidator(mcpServerSchema);
function isMcpServerApiEntity(entity) {
  return entity.spec.type === "mcp-server";
}
const mcpServerApiEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-api-mcp-server",
  builder: (model) => {
    model.addKindVersion({
      kind: "API",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          specType: "mcp-server",
          description: "An MCP (Model Context Protocol) server exposed as an API entity.",
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: { jsonSchema: mcpServerSchema }
        }
      ]
    });
  }
});

export { isMcpServerApiEntity, mcpServerApiEntityModel, mcpServerApiEntityValidator };
