// @backstage/catalog-model/dist/kinds/McpServerApiEntity.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var API_v1alpha1_mcpServer_schema = require('../schema/kinds/API.v1alpha1.mcp-server.schema.json.cjs.js');
var util = require('./util.cjs.js');

const mcpServerApiEntityValidator = util.ajvCompiledJsonSchemaValidator(API_v1alpha1_mcpServer_schema.default);
function isMcpServerApiEntity(entity) {
  return entity.spec.type === "mcp-server";
}
const mcpServerApiEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-api-mcp-server",
  builder: (model) => {
    model.addKindVersion({
      kind: "API",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          specType: "mcp-server",
          description: "An MCP (Model Context Protocol) server exposed as an API entity.",
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: { jsonSchema: API_v1alpha1_mcpServer_schema.default }
        }
      ]
    });
  }
});

exports.isMcpServerApiEntity = isMcpServerApiEntity;
exports.mcpServerApiEntityModel = mcpServerApiEntityModel;
exports.mcpServerApiEntityValidator = mcpServerApiEntityValidator;
