// @backstage/catalog-model/dist/kinds/ResourceEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var Resource_v1alpha1_schema = require('../schema/kinds/Resource.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const resourceEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(Resource_v1alpha1_schema.default);
const resourceEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-resource",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "Resource",
        singular: "resource",
        plural: "resources"
      },
      description: "A Resource describes the infrastructure a system needs to operate, like databases, topics, or buckets.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.dependsOn" },
              relation: "dependsOn",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.dependencyOf" },
              relation: "dependencyOf",
              defaultNamespace: "inherit"
            },
            {
              selector: { path: "spec.system" },
              relation: "partOf",
              defaultKind: "System",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: Resource_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.resourceEntityModel = resourceEntityModel;
exports.resourceEntityV1alpha1Validator = resourceEntityV1alpha1Validator;
