// @backstage/catalog-model/dist/kinds/SystemEntityV1alpha1.cjs.js
'use strict';

var createCatalogModelLayer = require('../model/createCatalogModelLayer.cjs.js');
var System_v1alpha1_schema = require('../schema/kinds/System.v1alpha1.schema.json.cjs.js');
var util = require('./util.cjs.js');

const systemEntityV1alpha1Validator = util.ajvCompiledJsonSchemaValidator(System_v1alpha1_schema.default);
const systemEntityModel = createCatalogModelLayer.createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-system",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "System",
        singular: "system",
        plural: "systems"
      },
      description: "A System is a collection of resources and components that exposes one or several APIs.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.owner" },
              relation: "ownedBy",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group", "User"]
            },
            {
              selector: { path: "spec.domain" },
              relation: "partOf",
              defaultKind: "Domain",
              defaultNamespace: "inherit"
            }
          ],
          schema: {
            jsonSchema: System_v1alpha1_schema.default
          }
        }
      ]
    });
  }
});

exports.systemEntityModel = systemEntityModel;
exports.systemEntityV1alpha1Validator = systemEntityV1alpha1Validator;
