// @backstage/catalog-model/dist/kinds/UserEntityV1alpha1.esm.js
import { createCatalogModelLayer } from '../model/createCatalogModelLayer.esm.js';
import jsonSchema from '../schema/kinds/User.v1alpha1.schema.json.esm.js';
import { ajvCompiledJsonSchemaValidator } from './util.esm.js';

const userEntityV1alpha1Validator = ajvCompiledJsonSchemaValidator(jsonSchema);
const userEntityModel = createCatalogModelLayer({
  layerId: "catalog.backstage.io/kind-user",
  builder: (model) => {
    model.addKind({
      group: "backstage.io",
      names: {
        kind: "User",
        singular: "user",
        plural: "users"
      },
      description: "A User describes a person, such as an employee or a contractor.",
      versions: [
        {
          name: ["v1alpha1", "v1beta1"],
          relationFields: [
            {
              selector: { path: "spec.memberOf" },
              relation: "memberOf",
              defaultKind: "Group",
              defaultNamespace: "inherit",
              allowedKinds: ["Group"]
            }
          ],
          schema: {
            jsonSchema
          }
        }
      ]
    });
  }
});

export { userEntityModel, userEntityV1alpha1Validator };
