// @backstage/cli-module-build/dist/commands/package/build/command.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var cleye = require('cleye');
var fs = require('fs-extra');
var packager = require('../../../lib/builder/packager.cjs.js');
var types = require('../../../lib/builder/types.cjs.js');
var role = require('../../../lib/role.cjs.js');
var cliNode = require('@backstage/cli-node');
var cliCommon = require('@backstage/cli-common');
var buildFrontend = require('../../../lib/buildFrontend.cjs.js');
var buildBackend = require('../../../lib/buildBackend.cjs.js');
var urls = require('../../../lib/urls.cjs.js');
var chalk = require('chalk');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

var command = async ({ args, info }) => {
  const {
    flags: {
      role: role$1,
      minify,
      skipBuildDependencies,
      stats,
      config,
      moduleFederation
    }
  } = cleye.cli(
    {
      name: info.usage,
      booleanFlagNegation: true,
      flags: {
        role: {
          type: String,
          description: "Run the command with an explicit package role"
        },
        minify: {
          type: Boolean,
          description: "Minify the generated code. Does not apply to app package (app is minified by default)."
        },
        skipBuildDependencies: {
          type: Boolean,
          description: "Skip the automatic building of local dependencies. Applies to backend packages only."
        },
        stats: {
          type: Boolean,
          description: "If bundle stats are available, write them to the output directory. Applies to app packages only."
        },
        config: {
          type: [String],
          description: "Config files to load instead of app-config.yaml. Applies to app packages only.",
          default: []
        },
        moduleFederation: {
          type: Boolean,
          description: "Build a package as a module federation remote. Applies to frontend plugin packages only."
        }
      }
    },
    void 0,
    args
  );
  const webpack = process.env.LEGACY_WEBPACK_BUILD ? require("webpack") : void 0;
  const resolvedRole = await role.findRoleFromCommand({ role: role$1 });
  if (resolvedRole === "frontend" || resolvedRole === "backend") {
    const configPaths = config.map((arg) => {
      if (urls.isValidUrl(arg)) {
        return arg;
      }
      return cliCommon.targetPaths.resolve(arg);
    });
    if (resolvedRole === "frontend") {
      return buildFrontend.buildFrontend({
        targetDir: cliCommon.targetPaths.dir,
        configPaths,
        writeStats: Boolean(stats),
        webpack
      });
    }
    return buildBackend.buildBackend({
      targetDir: cliCommon.targetPaths.dir,
      configPaths,
      skipBuildDependencies: Boolean(skipBuildDependencies),
      minify: Boolean(minify)
    });
  }
  let isModuleFederationRemote = void 0;
  if (resolvedRole === "frontend-dynamic-container") {
    console.log(
      chalk__default.default.yellow(
        `\u26A0\uFE0F  WARNING: The 'frontend-dynamic-container' package role is experimental and will receive immediate breaking changes in the future.`
      )
    );
    isModuleFederationRemote = true;
  }
  if (moduleFederation) {
    isModuleFederationRemote = true;
  }
  if (isModuleFederationRemote) {
    console.log("Building package as a module federation remote");
    return buildFrontend.buildFrontend({
      targetDir: cliCommon.targetPaths.dir,
      configPaths: [],
      writeStats: Boolean(stats),
      isModuleFederationRemote,
      webpack
    });
  }
  const roleInfo = cliNode.PackageRoles.getRoleInfo(resolvedRole);
  const outputs = /* @__PURE__ */ new Set();
  if (roleInfo.output.includes("cjs")) {
    outputs.add(types.Output.cjs);
  }
  if (roleInfo.output.includes("esm")) {
    outputs.add(types.Output.esm);
  }
  if (roleInfo.output.includes("types")) {
    outputs.add(types.Output.types);
  }
  const packageJson = await fs__default.default.readJson(
    cliCommon.targetPaths.resolve("package.json")
  );
  return packager.buildPackage({
    outputs,
    packageJson,
    minify: Boolean(minify),
    workspacePackages: await cliNode.PackageGraph.listTargetPackages()
  });
};

exports.default = command;
