// @backstage/cli-module-build/dist/lib/runner/startEmbeddedDb.cjs.js
'use strict';

var os = require('node:os');
var fs = require('fs-extra');
var node_path = require('node:path');
var portfinder = require('portfinder');
var errors = require('@backstage/errors');
var configLoader = require('@backstage/config-loader');
var cliCommon = require('@backstage/cli-common');
var chalk = require('chalk');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefaultCompat(os);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

const TEMP_DIR_PREFIX = "backstage-dev-db-";
const PID_FILE = "backstage.pid";
function isProcessAlive(pid) {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}
async function cleanStaleDatabases() {
  const tmpBase = os__default.default.tmpdir();
  const entries = (await fs__default.default.readdir(tmpBase)).filter(
    (d) => d.startsWith(TEMP_DIR_PREFIX)
  );
  await Promise.all(
    entries.map(async (d) => {
      const dir = node_path.resolve(tmpBase, d);
      const raw = await fs__default.default.readFile(node_path.resolve(dir, PID_FILE), "utf8").catch(() => void 0);
      const pid = raw ? Number(raw.trim()) : NaN;
      if (!pid || !isProcessAlive(pid)) {
        await fs__default.default.remove(dir);
      }
    })
  );
}
async function startEmbeddedDb(options) {
  const dbConfig = await readDatabaseConfig(
    options.configPaths,
    options.targetDir
  );
  if (dbConfig?.client !== "embedded-postgres") {
    return void 0;
  }
  return startEmbeddedDbInternal(dbConfig.connection);
}
async function startEmbeddedDbInternal(userConfig) {
  console.warn(
    chalk__default.default.yellow(
      "WARNING: Using embedded-postgres for local development is experimental and subject to change"
    )
  );
  const { default: EmbeddedPostgres } = await import('embedded-postgres').catch(
    (error) => {
      throw new errors.ForwardedError(
        `Failed to load 'embedded-postgres' which is required when using 'embedded-postgres' as the database client. It must be installed as an explicit dependency in your project`,
        error
      );
    }
  );
  await cleanStaleDatabases();
  const host = userConfig?.host ?? "localhost";
  const user = userConfig?.user ?? "postgres";
  const password = userConfig?.password ?? "postgres";
  const port = userConfig?.port ?? await portfinder.getPortPromise();
  const tmpDir = await fs__default.default.mkdtemp(node_path.resolve(os__default.default.tmpdir(), TEMP_DIR_PREFIX));
  const pg = new EmbeddedPostgres({
    databaseDir: tmpDir,
    user,
    password,
    port,
    persistent: false,
    onError(messageOrError) {
      console.error(`[embedded-postgres]`, messageOrError);
    },
    onLog() {
    }
  });
  try {
    await pg.initialise();
    await fs__default.default.writeFile(node_path.resolve(tmpDir, PID_FILE), String(process.pid));
    await pg.start();
  } catch (error) {
    await pg.stop().catch(() => {
    });
    await fs__default.default.remove(tmpDir).catch(() => {
    });
    throw error;
  }
  const configOverride = {
    client: "pg"
  };
  const defaultedConnection = {};
  if (!userConfig?.host) {
    defaultedConnection.host = host;
  }
  if (!userConfig?.user) {
    defaultedConnection.user = user;
  }
  if (!userConfig?.password) {
    defaultedConnection.password = password;
  }
  if (!userConfig?.port) {
    defaultedConnection.port = port;
  }
  if (Object.keys(defaultedConnection).length > 0) {
    configOverride.connection = defaultedConnection;
  }
  return {
    configOverride,
    async close() {
      await pg.stop();
      await fs__default.default.remove(tmpDir);
    }
  };
}
async function readDatabaseConfig(configPaths, targetDir) {
  const rootDir = cliCommon.targetPaths.rootDir;
  const configBaseDir = targetDir ?? rootDir;
  const source = configLoader.ConfigSources.default({
    rootDir,
    allowMissingDefaultConfig: true,
    argv: (configPaths ?? []).flatMap((p) => [
      "--config",
      node_path.isAbsolute(p) ? p : node_path.resolve(configBaseDir, p)
    ])
  });
  const config = await configLoader.ConfigSources.toConfig(source);
  try {
    const client = config.getOptionalString("backend.database.client");
    if (!client) {
      return void 0;
    }
    const rawConnection = config.getOptional("backend.database.connection");
    if (typeof rawConnection === "string" || rawConnection === void 0) {
      return { client };
    }
    const host = config.getOptionalString("backend.database.connection.host");
    const port = config.getOptionalNumber("backend.database.connection.port");
    const user = config.getOptionalString("backend.database.connection.user");
    const password = config.getOptionalString(
      "backend.database.connection.password"
    );
    const connection = host !== void 0 || port !== void 0 || user !== void 0 || password !== void 0 ? { host, port, user, password } : void 0;
    return { client, connection };
  } finally {
    config.close();
  }
}

exports.startEmbeddedDb = startEmbeddedDb;
