// @backstage/frontend-plugin-api/dist/wiring/createFrontendPlugin.esm.js
import { resolveExtensionDefinitions, resolveExtensionDefinition } from './resolveExtensionDefinition.esm.js';
import { ID_PATTERN } from './constants.esm.js';
import { OpaqueFrontendPlugin } from '../frontend-internal/src/wiring/InternalFrontendPlugin.esm.js';
import '../frontend-internal/src/wiring/InternalExtensionDefinition.esm.js';
import '../frontend-internal/src/wiring/InternalExtensionInput.esm.js';
import '../frontend-internal/src/wiring/InternalSwappableComponentRef.esm.js';

function createFrontendPlugin(options) {
  const pluginId = options.pluginId;
  if (!ID_PATTERN.test(pluginId)) {
    console.warn(
      `WARNING: The pluginId '${pluginId}' will be invalid soon, please change it to match the pattern ${ID_PATTERN} (letters, digits, and dashes only, starting with a letter)`
    );
  }
  const { extensions, extensionDefinitionsById } = resolveExtensionDefinitions(
    options.extensions ?? [],
    {
      namespace: pluginId,
      featureType: "Plugin"
    }
  );
  return OpaqueFrontendPlugin.createInstance("v1", {
    pluginId,
    id: pluginId,
    title: options.title,
    icon: options.icon,
    routes: options.routes ?? {},
    externalRoutes: options.externalRoutes ?? {},
    featureFlags: options.featureFlags ?? [],
    if: options.if,
    extensions,
    infoOptions: options.info,
    // This method is overridden when the plugin instance is installed in an app
    async info() {
      throw new Error(
        `Attempted to load plugin info for plugin '${pluginId}', but the plugin instance is not installed in an app`
      );
    },
    getExtension(id) {
      const ext = extensionDefinitionsById.get(id);
      if (!ext) {
        throw new Error(
          `Attempted to get non-existent extension '${id}' from plugin '${pluginId}'`
        );
      }
      return ext;
    },
    toString() {
      return `Plugin{id=${pluginId}}`;
    },
    withOverrides(overrides) {
      let ifPredicate = options.if;
      if ("if" in overrides) {
        ifPredicate = overrides.if;
      }
      const overrideExtensions = overrides.extensions ?? [];
      const overriddenExtensionIds = new Set(
        overrideExtensions.map(
          (e) => resolveExtensionDefinition(e, { namespace: pluginId }).id
        )
      );
      const nonOverriddenExtensions = (options.extensions ?? []).filter(
        (e) => !overriddenExtensionIds.has(
          resolveExtensionDefinition(e, { namespace: pluginId }).id
        )
      );
      return createFrontendPlugin({
        ...options,
        pluginId,
        if: ifPredicate,
        title: overrides.title ?? options.title,
        icon: overrides.icon ?? options.icon,
        extensions: [...nonOverriddenExtensions, ...overrideExtensions],
        info: {
          ...options.info,
          ...overrides.info
        }
      });
    }
  });
}

export { createFrontendPlugin };
