// @backstage/cli-module-build/dist/lib/packager/createDistWorkspace.cjs.js
'use strict';

var chalk = require('chalk');
var fs = require('fs-extra');
var node_path = require('node:path');
var os = require('node:os');
var tar = require('tar');
var partition = require('lodash/partition');
var cliCommon = require('@backstage/cli-common');
var computeTopologicalLayers = require('./computeTopologicalLayers.cjs.js');
var _package = require('../../package.json.cjs.js');
var packager = require('../builder/packager.cjs.js');
var types = require('../builder/types.cjs.js');
var productionPack = require('./productionPack.cjs.js');
var cliNode = require('@backstage/cli-node');
var typeDistProject = require('../typeDistProject.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

function _interopNamespaceCompat(e) {
  if (e && typeof e === 'object' && 'default' in e) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var tar__namespace = /*#__PURE__*/_interopNamespaceCompat(tar);
var partition__default = /*#__PURE__*/_interopDefaultCompat(partition);

const UNSAFE_PACKAGES = [
  ...Object.keys(_package.dependencies),
  ...Object.keys(_package.devDependencies)
];
function prefixLogFunc(prefix, logFn) {
  return (data) => {
    for (const line of data.toString("utf8").split(/\r?\n/)) {
      if (line) logFn(`${prefix}${line}`);
    }
  };
}
async function createDistWorkspace(packageNames, options = {}) {
  const logger = options.logger ?? { log: console.log, warn: console.warn };
  const targetDir = options.targetDir ?? await fs__default.default.mkdtemp(node_path.resolve(os.tmpdir(), "dist-workspace"));
  const packages = await cliNode.PackageGraph.listTargetPackages();
  const targets = await resolveLocalDependencies(packageNames, packages);
  if (options.buildDependencies) {
    const exclude = options.buildExcludes ?? [];
    const configPaths = options.configPaths ?? [];
    const toBuild = new Set(
      targets.map((_) => _.name).filter((name) => !exclude.includes(name))
    );
    const standardBuilds = new Array();
    const customBuild = new Array();
    for (const pkg of packages) {
      if (!toBuild.has(pkg.packageJson.name)) {
        continue;
      }
      const role = pkg.packageJson.backstage?.role;
      if (!role) {
        logger.warn(
          `Building ${pkg.packageJson.name} separately because it has no role`
        );
        customBuild.push({ dir: pkg.dir, name: pkg.packageJson.name });
        continue;
      }
      const buildScript = pkg.packageJson.scripts?.build;
      if (!buildScript) {
        customBuild.push({ dir: pkg.dir, name: pkg.packageJson.name });
        continue;
      }
      if (!buildScript.startsWith("backstage-cli package build")) {
        logger.warn(
          `Building ${pkg.packageJson.name} separately because it has a custom build script, '${buildScript}'`
        );
        customBuild.push({ dir: pkg.dir, name: pkg.packageJson.name });
        continue;
      }
      if (cliNode.PackageRoles.getRoleInfo(role).output.includes("bundle")) {
        logger.warn(
          `Building ${pkg.packageJson.name} separately because it is a bundled package`
        );
        const args = buildScript.includes("--config") ? [] : configPaths.map((p) => ["--config", p]).flat();
        customBuild.push({ dir: pkg.dir, name: pkg.packageJson.name, args });
        continue;
      }
      const outputs = packager.getOutputsForRole(role);
      outputs.delete(types.Output.types);
      if (outputs.size > 0) {
        standardBuilds.push({
          targetDir: pkg.dir,
          packageJson: pkg.packageJson,
          outputs,
          logPrefix: `${chalk__default.default.cyan(
            node_path.relative(cliCommon.targetPaths.rootDir, pkg.dir)
          )}: `,
          minify: options.minify,
          workspacePackages: packages
        });
      }
    }
    await packager.buildPackages(standardBuilds);
    if (customBuild.length > 0) {
      await cliNode.runConcurrentTasks({
        items: customBuild,
        worker: async ({ name, dir, args }) => {
          await cliCommon.run(["yarn", "run", "build", ...args || []], {
            cwd: dir,
            onStdout: prefixLogFunc(`${name}: `, logger.log),
            onStderr: prefixLogFunc(`${name}: `, logger.warn)
          }).waitForExit();
        }
      });
    }
  }
  await moveToDistWorkspace(
    targetDir,
    targets,
    Boolean(options.alwaysPack),
    Boolean(options.enableFeatureDetection),
    logger
  );
  const files = options.files ?? ["yarn.lock", "package.json"];
  for (const file of files) {
    const src = typeof file === "string" ? file : file.src;
    const dest = typeof file === "string" ? file : file.dest;
    await fs__default.default.copy(cliCommon.targetPaths.resolveRoot(src), node_path.resolve(targetDir, dest));
  }
  if (options.skeleton) {
    const skeletonFiles = targets.map((target) => {
      const dir = node_path.relative(cliCommon.targetPaths.rootDir, target.dir);
      return node_path.join(dir, "package.json");
    }).sort();
    await tar__namespace.create(
      {
        file: node_path.resolve(targetDir, options.skeleton),
        cwd: targetDir,
        portable: true,
        noMtime: true,
        gzip: options.skeleton.endsWith(".gz")
      },
      skeletonFiles
    );
  }
  return { targetDir, targets };
}
const FAST_PACK_SCRIPTS = [
  void 0,
  "backstage-cli prepack",
  "backstage-cli package prepack"
];
async function packToDirectory(options) {
  const { packageDir, packageName, targetDir, logger } = options;
  const archivePath = options.archivePath ?? node_path.resolve(targetDir, "temp-archive.tgz");
  const prefix = `${packageName} [pack]: `;
  await fs__default.default.ensureDir(targetDir);
  await cliCommon.run(["yarn", "pack", "--filename", archivePath], {
    cwd: packageDir,
    onStdout: prefixLogFunc(prefix, logger.log),
    onStderr: prefixLogFunc(prefix, logger.warn)
  }).waitForExit();
  await tar__namespace.extract({ file: archivePath, cwd: targetDir, strip: 1 });
  await fs__default.default.remove(archivePath);
}
async function moveToDistWorkspace(workspaceDir, localPackages, alwaysPack, enableFeatureDetection, logger) {
  const [fastPackPackages, slowPackPackages] = partition__default.default(
    localPackages,
    (pkg) => !alwaysPack && FAST_PACK_SCRIPTS.includes(pkg.packageJson.scripts?.prepack)
  );
  const featureDetectionProject = fastPackPackages.length > 0 && enableFeatureDetection ? await typeDistProject.createTypeDistProject() : void 0;
  await Promise.all(
    fastPackPackages.map(async (target) => {
      logger.log(`Moving ${target.name} into dist workspace`);
      const outputDir = node_path.relative(cliCommon.targetPaths.rootDir, target.dir);
      const absoluteOutputPath = node_path.resolve(workspaceDir, outputDir);
      await productionPack.productionPack({
        packageDir: target.dir,
        targetDir: absoluteOutputPath,
        featureDetectionProject
      });
    })
  );
  let archiveIndex = 0;
  async function pack(target, archive = `temp-package-${archiveIndex++}.tgz`) {
    logger.log(`Repacking ${target.name} into dist workspace`);
    const absoluteOutputPath = node_path.resolve(
      workspaceDir,
      node_path.relative(cliCommon.targetPaths.rootDir, target.dir)
    );
    await packToDirectory({
      packageDir: target.dir,
      packageName: target.name,
      targetDir: absoluteOutputPath,
      archivePath: node_path.resolve(workspaceDir, archive),
      logger
    });
    if (target.packageJson.bundled) {
      const pkgJson = await fs__default.default.readJson(
        node_path.resolve(absoluteOutputPath, "package.json")
      );
      delete pkgJson.dependencies;
      delete pkgJson.devDependencies;
      delete pkgJson.peerDependencies;
      delete pkgJson.optionalDependencies;
      await fs__default.default.writeJson(
        node_path.resolve(absoluteOutputPath, "package.json"),
        pkgJson,
        {
          spaces: 2
        }
      );
    }
  }
  const [unsafePackages, safePackages] = partition__default.default(
    slowPackPackages,
    (p) => UNSAFE_PACKAGES.includes(p.name)
  );
  for (const target of unsafePackages) {
    await pack(target, `temp-package.tgz`);
  }
  const layers = computeTopologicalLayers.computeTopologicalLayers(safePackages);
  for (const layer of layers) {
    await cliNode.runConcurrentTasks({
      items: layer,
      worker: pack
    });
  }
}
async function resolveLocalDependencies(packageNames, packages) {
  const packageGraph = cliNode.PackageGraph.fromPackages(packages);
  const targetNames = packageGraph.collectPackageNames(packageNames, (node) => {
    if (node.packageJson.bundled) {
      return void 0;
    }
    return node.publishedLocalDependencies.keys();
  });
  return Array.from(targetNames).map((name) => packageGraph.get(name));
}

exports.createDistWorkspace = createDistWorkspace;
exports.packToDirectory = packToDirectory;
exports.resolveLocalDependencies = resolveLocalDependencies;
